{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SARS-CoV-2 data processing\n",
    "\n",
    "This notebook contains the following analysis:\n",
    "\n",
    "1. [Data preprocessing, selecting regions, and filtering data](#preprocess)  \n",
    "2. [Finding groups of linked mutations and classifying mutations as synonymous or nonsynonymous](#linked)   \n",
    "3. [Finding frequency trajectories of SNVs](#trajectories)  \n",
    "4. [Inferring selection coefficients](#inference)  \n",
    "5. [Error analysis](#errors)  \n",
    "    a. [Bootstrapping sequence data to find errors](#bootstrap)   \n",
    "    b. [Subsampling regions to find errors](#section_subsample_regions)   \n",
    "6. [Early detection](#early_detection)  \n",
    "    a. [Finding the null distribution and analying early detections for groups of SNVs accross all regions](#null_distribution)  \n",
    "    b. [Making a table of number of detections for major variants](#variant_detection_table)  \n",
    "    c. [Inferring selection for variants before they actually appeared](#selection_on_date)   \n",
    "    d. [Early detection of Alpha, Delta, and Omicron](#variant_detection)  \n",
    "    e. [Early detection of Alpha, Delta, and Omicron trimming data by submission date](#submission_date)  \n",
    "7. [Calculating statistics](#statistics) \n",
    "8. [Nextstrain version](#nextstrain_version)\n",
    "8. [Analyzing the effect of travel on inferred coeffficients for EU1 in the UK](#EU1_travel)  \n",
    "9. [Analyzing the effect on inferred coefficients of masking out Alpha mutations](#alphamask)  \n",
    "10. [Removing probematic sequences and check how inference is affected](#remove_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version 3.8.3 (default, Jul  2 2020, 11:26:31) \n",
      "[Clang 10.0.0 ]\n",
      "numpy version 1.18.5\n",
      "pandas version 1.0.5\n",
      "matplotlib version 3.5.1\n",
      "scipy version 1.5.0\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# import modules and global variables\n",
    "import os, sys\n",
    "from imp import reload\n",
    "import datetime as dt\n",
    "import shutil\n",
    "import posixpath as ppath\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "print('matplotlib version %s' % matplotlib.__version__)\n",
    "\n",
    "import scipy\n",
    "from scipy import linalg\n",
    "print('scipy version %s' % scipy.__version__)\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "# sequence related variables\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "\n",
    "# local directories\n",
    "SARS_DIR     = '/Users/brianlee/SARS-CoV-2-Data'\n",
    "INF_DIR      = os.path.join('/Users/brianlee/Python/MPL', '6-29-20-epidemiological')\n",
    "DATA_DATE    = '2022-06-09'\n",
    "DATA_DIR     = os.path.join(SARS_DIR, DATA_DATE)\n",
    "SCRIPT_DIR   = os.path.join(SARS_DIR, 'Processing-new')\n",
    "\n",
    "# cluster directories\n",
    "USER_NAME    = 'blee098'\n",
    "SSH_HOME     = 'blee098@cluster.hpcc.ucr.edu:'\n",
    "SSH_DATA     = ppath.join('/rhome', USER_NAME, 'bigdata', 'SARS-CoV-2-Data', DATA_DATE)\n",
    "SCRATCH      = ppath.join(SSH_DATA, 'scratch')\n",
    "DATABASE_DIR = ppath.join(SSH_DATA, f'lmdb-{DATA_DATE}')\n",
    "metadata_new = ppath.join(SSH_DATA, MERGED_FINAL)\n",
    "\n",
    "# metadata variables\n",
    "METADATA_COLS    = [   'accession', 'virus_name',            'date', 'location',             'location_additional', 'submission_date']\n",
    "METADATA_OCOLS   = ['Accession ID', 'Virus name', 'Collection date', 'Location', 'Additional location information', 'Submission date']\n",
    "METADATA_XFORM   = [           str,    str.lower,               str,  str.lower,                         str.lower,               str]\n",
    "METADATA_COMP    = os.path.join(SARS_DIR, f'metadata_tsv_2022_06_09.tar.xz')\n",
    "METADATA_FILE    = os.path.join(SARS_DIR, f'metadata-{DATA_DATE}', 'metadata.tsv')\n",
    "METADATA_DIR     = os.path.join(SARS_DIR, f'metadata-{DATA_DATE}')\n",
    "MSA_NAME         = 'msa_0608'\n",
    "METADATA_INTERIM = 'merged-interim.csv'\n",
    "MERGED_FINAL     = f'metadata-{DATA_DATE}.csv'\n",
    "REF_TAG          = 'EPI_ISL_402125'\n",
    "IHME_DATA        = 'ihme-popsize-2022_09_10'\n",
    "REF_FILE         = f'ref-index-{DATA_DATE}'\n",
    "MIN_REGIONAL_SEQS = 1000\n",
    "\n",
    "#os.chdir(SCRIPT_DIR)\n",
    "import data_processing as dp\n",
    "#os.chdir(SARS_DIR)\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transfer files to the cluster\n",
    "print('scp %s/%s %s &&' % (SCRIPT_DIR, 'data_processing.py', SSH_HOME))\n",
    "\n",
    "file_paths = os.path.join(SCRIPT_DIR, '*')\n",
    "print(f'scp -r {file_paths} {SSH_HOME}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make files on directory\n",
    "print(f'mkdir -p {SSH_DATA} &&')\n",
    "print('mkdir -p ./MPL/out &&')\n",
    "dir_list = [\n",
    "    'regions-times',\n",
    "    'msa-reg', 'genome-data', 'genome-trimmed', 'genome-filtered', 'max-freqs', \n",
    "    'freqs', 'freqs-tv', 'freqs-regional', 'freqs-dt-10', 'freqs-dt-20', 'freqs-0.5pct', \n",
    "    'freqs-5pct', 'freqs-bootstrap', 'counts',\n",
    "    'trajectories', 'trajectories-nosmooth', 'trajectories-uk',\n",
    "    'infer-bs', 'infer-subsample', 'infer-tv', 'infer',\n",
    "    'genome-filtered-south-africa', 'genome-filtered-london', 'genome-filtered-uk',\n",
    "    'trajectories-south-africa', 'trajectories-london', 'trajectories-uk',\n",
    "    #'freqs-tv-london', 'freqs-tv-south-africa', 'freqs-tv-uk', \n",
    "    #'infer-tv-uk', 'infer-tv-london', 'infer-tv-south-africa',\n",
    "    'genome-filtered-variants',\n",
    "    'freqs-tv-variants', 'freqs-regional-variants', 'infer-tv-variants',\n",
    "    'freqs-uk-sd', 'freqs-london-sd', 'freqs-south-africa-sd',\n",
    "    'infer-south-africa-sd', 'infer-london-sd', 'infer-uk-sd'\n",
    "]\n",
    "for name in dir_list:\n",
    "    dir_path = ppath.join(SSH_DATA, name)\n",
    "    if name!=dir_list[-1]:\n",
    "        print(f'mkdir {dir_path} &&')\n",
    "    else:\n",
    "        print(f'mkdir {dir_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variant definitions\n",
    "\n",
    "epsilon        = np.unique(['NSP13-260-0-T', 'S-13-1-T', 'S-152-2-T', 'NC-28271-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', \n",
    "                  'S-452-1-G', 'N-205-1-T', 'ORF3a-57-2-T', 'NSP9-65-0-G']) # Done\n",
    "\n",
    "alpha          = np.unique(['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "                  'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "                  'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G'])    # Done\n",
    "\n",
    "# NEED TO ADD S:D138Y\n",
    "gamma          = np.unique(['NSP9-31-2-T', 'NSP1-156-2-C', 'NSP3-10-2-T', 'NSP3-370-1-T', 'NSP3-977-0-C', 'NSP3-1200-2-G', 'NSP3-1298-2-G', \n",
    "                  'NSP12-140-2-T', 'NSP13-341-2-T', 'S-20-1-A', 'S-417-1-C', 'S-1027-1-T', 'ORF3a-253-0-C', 'ORF8-92-0-A', 'N-80-1-G', \n",
    "                  'S-190-2-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-501-0-T', 'NSP6-106-0--', 'NSP6-106-1--', \n",
    "                  'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'N-202-0-T', 'N-202-1-C', 'S-18-0-T', 'S-26-0-T', 'S-655-0-T',\n",
    "                  'S-1176-0-T'])    # Done\n",
    "\n",
    "twentyE_EU1    = np.unique(['NSP16-199-2-C', 'NSP1-60-2-C', 'NSP3-1189-2-T', 'M-93-2-G', 'N-220-1-T', 'ORF10-30-0-T', 'S-222-1-T',\n",
    "                  'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G'])    # Done\n",
    "# This is Pango lineage B.1.177\n",
    "\n",
    "delta          = np.unique(['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', \n",
    "                  'S-158-0--', 'S-478-1-A', 'S-681-1-G', 'S-950-0-A', 'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', \n",
    "                  'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'N-63-1-G', 'N-203-1-T', \n",
    "                  'N-377-0-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-452-1-G', 'NC-28270--', 'NSP4-492-1-T'])    # Done\n",
    "\n",
    "beta           = np.unique(['S-80-1-C', 'NSP3-837-2-T', 'S-215-1-G', 'E-71-1-T', 'S-240-2--', 'S-241-0--', 'S-242-1--', 'S-242-2--', 'S-243-0--', \n",
    "                  'S-240-1--', 'S-241-1--', 'S-241-2--', 'S-242-0--', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-501-0-T',\n",
    "                  'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', \n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-417-2-T', 'S-484-0-A', 'S-701-1-T', 'ORF3a-57-2-T', 'NSP2-85-1-T', 'NSP5-90-1-G',\n",
    "                  'N-205-1-T', 'NC-173-T', 'ORF8-120-2-T'])    # Done\n",
    "\n",
    "lambda_new     = np.unique(['S-246-2--', 'NSP3-1569-0-G', 'S-247-0--', 'S-247-1--', 'S-247-2--', 'S-248-0--', 'S-248-1--', 'S-248-2--', 'S-249-0--', \n",
    "                  'S-249-1--', 'S-249-2--', 'S-250-0--', 'S-250-1--', 'S-250-2--', 'S-251-0--', 'S-251-1--', 'S-251-2--', 'S-252-0--', \n",
    "                  'S-246-1--', 'S-252-1--', 'S-252-2--', 'S-253-0--', 'N-214-0-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', \n",
    "                  'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', \n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-75-1-T', 'S-76-1-T', 'S-452-1-A', 'S-490-1-C', 'S-859-1-A', 'NSP3-428-1-T',\n",
    "                  'NSP3-1469-0-T', 'NSP4-438-1-C', 'NSP4-492-1-T', 'NSP5-15-0-A']) # Done\n",
    "\n",
    "iota           = np.unique(['N-234-2-A', 'S-5-0-T', 'S-95-1-T', 'S-253-1-G', 'S-484-0-A', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                 'S-701-1-T', 'NSP13-88-2-C', 'ORF3a-42-1-T', 'ORF3a-67-2-T', 'NSP2-85-1-T', 'NSP4-438-1-C', 'NSP6-106-0--', 'NSP6-106-1--', \n",
    "                 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',  'NSP6-108-1--', 'NSP6-108-2--', 'N-199-1-T',\n",
    "                 'N-232-2-A', 'ORF8-11-1-T', 'NSP15-214-2-G', 'NC-28270--'])\n",
    "\n",
    "D614G          = ['NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G']    # Done\n",
    "\n",
    "B1_1_318       = np.unique(['NSP15-320-0-A', 'S-575-2-C', 'S-1238-2-A', 'ORF7b-44-2--', 'NC-27887--', 'NC-27888--', 'NC-27889--', 'NC-27890--', \n",
    "                  'NC-27891--', 'NC-27892--', 'ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'NSP4-173-1-T', 'S-796-0-C', 'ORF8-2-0--', \n",
    "                  'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--', 'NC-28270-G', 'N-208-2--', 'N-209-0--', 'N-208-1--',\n",
    "                  'NSP3-378-1-T', 'NSP3-1693-2-T', 'NSP5-21-1-T', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', \n",
    "                  'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--', 'S-95-1-T', 'S-143-2--', 'S-144-0--', \n",
    "                  'S-144-1--', 'S-484-0-A', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-681-1-A', 'M-82-1-C',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C']) # Add N-234-2-T, Many others\n",
    "\n",
    "# need to add orf1b-314, 3037\n",
    "omicron        = np.unique(['NC-240-T', 'NSP3-38-1-G', 'NSP3-106-2-T', 'NSP3-889-2-G', 'NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--', \n",
    "                  'NSP3-1892-0-A', 'NSP4-492-1-T', 'NSP5-132-1-A', 'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', \n",
    "                  'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-189-0-G', 'NSP10-57-2-C', \n",
    "                  'NSP12-323-1-T', 'NSP12-600-2-T', 'S-67-1-T', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'S-95-1-T', 'S-142-1--', 'S-142-2--', 'S-143-0--', 'S-143-1--', 'S-143-2--', 'S-144-0--', \n",
    "                  'S-144-1--', 'S-144-2--', 'S-145-0--', 'S-211-1--', 'S-211-2--', 'S-212-0--', 'S-339-1-A', 'S-371-0-C', \n",
    "                  'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-446-0-A', 'S-547-1-A', 'S-614-1-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-856-2-A', 'S-954-2-T', 'S-969-2-A', 'S-981-0-T', \n",
    "                  'S-1146-2-T', 'ORF3a-64-2-T', 'E-9-1-T', 'M-3-1-G', 'M-19-0-G', 'M-63-0-A', 'ORF6-20-0-C', \n",
    "                  'NC-28270-T', 'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'S-375-1-T', 'S-477-1-A', 'S-478-1-A', \n",
    "                  'S-484-1-C', 'S-493-1-G', 'S-496-0-A', 'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NSP14-42-0-G',\n",
    "                  'ORF7b-18-0-T', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', \n",
    "                  'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A'])  \n",
    "\n",
    "# 3037\n",
    "ba2            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NSP4-492-1-T', 'NSP5-132-1-A', \n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'E-9-1-T', 'ORF3a-223-1-T',\n",
    "                  'M-19-0-G', 'M-63-0-A', 'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C', 'S-493-1-G', \n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'N-413-0-C', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-112-1-T', 'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C', 'S-1146-2-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T', \n",
    "                  'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', \n",
    "                  'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', \n",
    "                  'S-214t-2-A', 'S-214u-2-A', 'NSP1-135-2-G', 'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', \n",
    "                  'NSP4-327-1-T'])\n",
    "\n",
    "# deletion after 29734\n",
    "ba4            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', \n",
    "                  'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'NSP1-135-2-G',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "                  'N-32-1--', 'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C',\n",
    "                  'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-492-1-T', 'NSP4-327-1-T',\n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP5-132-1-A', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-145-2-G', 'ORF3a-223-1-T', 'M-19-0-G', 'M-63-0-A', 'E-9-1-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T',\n",
    "                  'S-1146-2-T', 'ORF7b-11-2-T', 'S-486-0-G', 'NSP8-23-2-A', 'N-151-0-T',\n",
    "                  'NSP1-141-0--', 'NSP1-141-1--', 'NSP1-141-2--', 'NSP1-142-0--', 'NSP1-142-1--',\n",
    "                  'NSP1-142-2--', 'NSP1-143-0--', 'NSP1-143-1--', 'NSP1-143-2--', 'N-413-0-C',\n",
    "                  'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C'])\n",
    "\n",
    "# deletion after 29734\n",
    "ba5            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A','S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', \n",
    "                  'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'NSP1-135-2-G',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "                  'N-32-1--', 'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C',\n",
    "                  'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-492-1-T', 'NSP4-327-1-T',\n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP5-132-1-A', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-223-1-T', 'M-19-0-G', 'M-63-0-A', 'E-9-1-T', 'M-3-0-A', 'S-486-0-G',\n",
    "                  'NSP8-23-2-A', 'N-413-0-C', 'S-1146-2-T', 'NC-28270-T', 'ORF3a-64-2-T', 'ORF7b-18-0-T'])\n",
    "\n",
    "ba2121         = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', \n",
    "                  'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'S-452-1-A',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NSP4-492-1-T', 'NSP5-132-1-A', \n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-189-0-G', 'E-9-1-T', 'ORF3a-223-1-T',\n",
    "                  'M-19-0-G', 'M-63-0-A', 'N-413-0-C', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-112-1-T', 'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C', 'S-1146-2-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T',\n",
    "                  'NSP1-135-2-G', 'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-327-1-T', \n",
    "                  'S-704-1-T', 'S-452-1-A'])\n",
    " \n",
    "# synonymous: 9593T\n",
    "eta            = np.unique(['NSP2-231-2-T', 'NSP2-334-2-G', 'NSP2-618-2-A', 'NSP4-13-2-C', 'NSP16-22-2-G',\n",
    "                  'S-888-0-C', 'ORF6-2-0--', 'ORF6-2-1--', 'ORF6-2-2--', 'N-2-1--', 'N-2-2--',\n",
    "                  'N-3-0--', 'N-12-1-G', 'N-142-2-G', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                  'M-82-1-C', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', \n",
    "                  'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',  'NSP6-108-1--', 'NSP6-108-2--',\n",
    "                  'S-52-1-G', 'S-67-1-T', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-484-0-A', \n",
    "                  'S-677-2-C', 'N-205-1-T', 'E-21-0-T', 'NSP3-1189-1-T',\n",
    "                  'NSP14-44-2-T', 'S-1062-2-C', 'N-142-2-G', 'NC-29542-T'])\n",
    "\n",
    "# confused about orf1a-T1567I\n",
    "kappa           = np.unique(['NSP12-323-1-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                   'S-154-0-A', 'S-1071-2-T', 'S-681-1-G', 'S-484-0-C', 'S-452-1-G', 'NSP13-206-0-T',\n",
    "                   'NSP13-429-2-T', 'NSP15-259-1-G', 'NSP15-261-0-G', 'N-203-1-T', 'N-377-0-T',\n",
    "                   'M-82-1-G', 'ORF3a-26-1-T', 'NSP3-749-1-T', 'NSP6-77-0-G', 'ORF7a-82-1-C',\n",
    "                   'NC-209-T', 'NSP3-246-2-T', 'M-53-2-T', 'NC-28270--'])\n",
    "\n",
    "# maybe S-144-0-A, S-144-1-C, S-144-2-C, S-145-0-A (actually there is an insertion here)\n",
    "# not positive about 'NSP3-237-0-G'\n",
    "mu              = np.unique(['S-346-2-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-95-1-T',\n",
    "                   'S-484-1-C', 'S-501-0-T', 'S-681-1-A', 'S-950-0-A', 'NSP3-237-0-G', 'NSP3-720-1-T',\n",
    "                   'NSP4-492-1-T', 'NSP6-160-1-G', 'NSP12-323-1-T', 'NSP13-419-0-T', 'N-205-1-T',\n",
    "                   'ORF3a-57-2-T', 'ORF3a-256-0--', 'ORF3a-256-1--', 'ORF3a-256-2--', 'ORF3a-257-0--',\n",
    "                   'ORF3a-257-2-C', 'ORF8-11-1-T', 'ORF8-38-0-T', 'ORF8-67-1-T',\n",
    "                   'NSP3-1106-2-T', 'NSP10-11-2-T', 'NSP14-280-0-T', 'NSP15-176-2-T', 'NC-26491-T',\n",
    "                   'NC-28271-T'])\n",
    "\n",
    "all_variants   = [epsilon, alpha, beta, gamma, lambda_new, delta, twentyE_EU1, D614G, omicron, ba2, ba5, ba4, ba2121]\n",
    "np.save(os.path.join(SARS_DIR, 'variants.npy'), all_variants)\n",
    "\n",
    "var_dic = {\n",
    "    'alpha' : alpha, \n",
    "    'epsilon' : epsilon, \n",
    "    'beta' : beta, \n",
    "    'gamma' : gamma, \n",
    "    'lambda' : lambda_new, \n",
    "    'delta' : delta, \n",
    "    '20E\\(EU1\\)' : twentyE_EU1, \n",
    "    'B.1' : D614G,\n",
    "    'omicron' : omicron,\n",
    "    'BA.2' : ba2,\n",
    "    'eta' : eta,\n",
    "    'iota' : iota,\n",
    "    'B.1.1.318' : B1_1_318,\n",
    "    'kappa' : kappa,\n",
    "    'mu' : mu,\n",
    "    'BA.4' : ba4,\n",
    "    'BA.5' : ba5,\n",
    "    'BA.2.12.1' : ba2121\n",
    "}\n",
    "\n",
    "variants_full = [var_dic[i] for i in var_dic]\n",
    "np.save(os.path.join(SARS_DIR, 'variants-all.npy'), variants_full)\n",
    "\n",
    "np.save(os.path.join(DATA_DIR, 'alpha-mutations.npy'),   alpha)\n",
    "np.save(os.path.join(DATA_DIR, 'gamma-mutations.npy'),   gamma)\n",
    "np.save(os.path.join(DATA_DIR, 'delta-mutations.npy'),   delta)\n",
    "np.save(os.path.join(DATA_DIR, 'omicron-mutations.npy'), omicron)\n",
    "\n",
    "# Saving the order in which the variants appeared for later early detection analysis\n",
    "group_order   = ['B.1', 'Alpha', 'Beta', 'Epsilon', 'Gamma', 'Eta', 'Iota', 'Kappa', 'Lambda', 'Mu', 'Delta', 'BA.1', 'BA.2', 'BA.2.12.1', 'BA.4', 'BA.5']\n",
    "variant_order = [[D614G], [alpha, beta, epsilon], [gamma, eta, iota, kappa, lambda_new], [mu], [delta], [omicron], [ba2], [ba2121, ba4, ba5]]\n",
    "\n",
    "new_groups = []\n",
    "old_muts   = []\n",
    "for i in range(len(variant_order)):\n",
    "    current_muts = []\n",
    "    for group in variant_order[i]:\n",
    "        new = [j for j in group if j not in old_muts]\n",
    "        new_groups.append(new)\n",
    "        for j in new:\n",
    "            current_muts.append(j)\n",
    "    for j in current_muts:\n",
    "        if j not in old_muts:\n",
    "            old_muts.append(j)\n",
    "            \n",
    "np.save(os.path.join(SARS_DIR, 'variants-exclusive.npy'), new_groups)\n",
    "\n",
    "#for i in range(len(new_groups)):\n",
    "#    print(i+1, new_groups[i])\n",
    "#    print('')\n",
    "\n",
    "# THE BELOW ARE INCOMPLETE LISTS OF MUTATIONS\n",
    "b12    = ['NSP5-89-0-T', 'NSP14-129-0-G', 'ORF3a-172-1-T', 'ORF8-24-1-T', 'N-67-0-T']    # this is B.1.2\n",
    "b11318 = ['NSP15-320-0-A', 'S-575-2-C', 'S-1238-2-A', 'ORF7b-44-2--', 'NC-27887--', 'NC-27888--', 'NC-27889--', 'NC-27890--',     # This is B.1.1.318, S:D796H\n",
    "          'NC-27891--', 'NC-27892--', 'ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'NSP4-173-1-T', 'S-796-0-C', 'ORF8-2-0--', \n",
    "          'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--', 'NC-28270-G', 'N-208-2--', 'N-209-0--', 'N-208-1--']\n",
    "b1429  = ['NSP8-3-2-T', 'NSP2-598-0-C', 'NSP4-131-2-T', 'NSP9-65-0-G', 'S-929-2-C', 'NC-27889-T', 'NSP2-530-2-T']\n",
    "\n",
    "\n",
    "# ADD ZETA, THETA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='preprocess'></a>\n",
    "# Data Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting the accessions, virus name, date, and location information from metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir /Users/brianlee/SARS-CoV-2-Data/metadata-2022-06-09 &&\n",
      "tar -xf /Users/brianlee/SARS-CoV-2-Data/metadata_tsv_2022_06_09.tar.xz -C /Users/brianlee/SARS-CoV-2-Data/metadata-2022-06-09\n"
     ]
    }
   ],
   "source": [
    "print(f'mkdir {METADATA_DIR} &&')\n",
    "print(f'tar -xf {METADATA_COMP} -C {METADATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/brianlee/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (5) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "100%|██████████| 11256004/11256004 [28:22<00:00, 6611.46it/s] \n"
     ]
    }
   ],
   "source": [
    "accessions = set()\n",
    "header     = ','.join(METADATA_COLS)\n",
    "\n",
    "use_cols = ['Accession ID', 'Virus name', 'Collection date', 'Location', \n",
    "            'Additional location information', 'Submission date', 'Host']\n",
    "df = pd.read_csv('%s' % METADATA_FILE, sep='\\t', usecols=use_cols)\n",
    "f = open('%s' % METADATA_INTERIM, 'w')\n",
    "f.write('%s\\n' % header)\n",
    "for df_iter, df_entry in tqdm(df.iterrows(), total=df.shape[0]):\n",
    "    #if df_iter % 10000 == 0:\n",
    "    #    print(df_inter / len(df))\n",
    "    entry = []\n",
    "    unique_id  = False\n",
    "    valid_date = False\n",
    "        \n",
    "    # check that host is human\n",
    "    if 'Host' in df_entry:\n",
    "        if df_entry.Host!='Human' and df_entry.Host!='human':\n",
    "            continue\n",
    "        \n",
    "    # get accession\n",
    "    acc_idx = METADATA_COLS.index('accession')\n",
    "    if 'Accession ID' not in df_entry:\n",
    "        continue\n",
    "    xformed = METADATA_XFORM[acc_idx](df_entry[METADATA_OCOLS[acc_idx]])\n",
    "    acc     = METADATA_XFORM[acc_idx](df_entry[METADATA_OCOLS[acc_idx]])\n",
    "\n",
    "    if xformed not in accessions:\n",
    "        #accessions.append(xformed)\n",
    "        accessions.add(xformed)\n",
    "        entry.append(xformed)\n",
    "        unique_id = True\n",
    "    #else:\n",
    "        #print('Excluding %s (duplicate accession number)' % xformed)\n",
    "        \n",
    "    # get virus name \n",
    "    if 'Virus name' in df_entry:\n",
    "        name_idx = METADATA_COLS.index('virus_name')\n",
    "        xformed  = METADATA_XFORM[name_idx](df_entry[METADATA_OCOLS[name_idx]])\n",
    "        xformed  = xformed.replace(',', ' ')\n",
    "        entry.append(xformed)\n",
    "    else:\n",
    "        entry.append('NA')\n",
    "            \n",
    "    # get date\n",
    "    if unique_id:\n",
    "        date_idx = METADATA_COLS.index('date')\n",
    "        xformed  = METADATA_XFORM[date_idx](df_entry[METADATA_OCOLS[date_idx]])\n",
    "            \n",
    "        try:\n",
    "            dt.date.fromisoformat(xformed)\n",
    "            valid_date = True\n",
    "        except:\n",
    "            valid_date = False\n",
    "            #print('Excluding %s (date %s is incomplete)' % (entry[0], xformed))\n",
    "        entry.append(xformed)\n",
    "            \n",
    "    # get location \n",
    "    if unique_id and valid_date:\n",
    "        loc_idx = METADATA_COLS.index('location')\n",
    "        xformed = METADATA_XFORM[loc_idx](df_entry[METADATA_OCOLS[loc_idx]])\n",
    "        xformed = xformed.replace(',', ' ')\n",
    "        entry.append(xformed)\n",
    "        \n",
    "        loc_add_idx = METADATA_COLS.index('location_additional')\n",
    "        xformed     = METADATA_XFORM[loc_add_idx](str(df_entry[METADATA_OCOLS[loc_add_idx]]))\n",
    "        xformed     = xformed.replace(',', ' ')\n",
    "        entry.append(xformed)\n",
    "        \n",
    "    # get submission date\n",
    "    if unique_id and valid_date:\n",
    "        sub_idx = METADATA_COLS.index('submission_date')\n",
    "        xformed = METADATA_XFORM[sub_idx](df_entry[METADATA_OCOLS[sub_idx]])\n",
    "        \n",
    "        try:\n",
    "            dt.date.fromisoformat(xformed)\n",
    "            valid_sub_date = True\n",
    "        except:\n",
    "            valid_sub_date = False\n",
    "        entry.append(xformed)\n",
    "        \n",
    "    # save data\n",
    "    if unique_id and valid_date and valid_sub_date:\n",
    "        f.write('%s\\n' % ','.join(entry))            \n",
    "\n",
    "# Add reference if not in the list of accessions\n",
    "if REF_TAG not in accessions:\n",
    "    entry = []\n",
    "    for i in range(len(METADATA_COLS)):\n",
    "        if METADATA_COLS[i]=='accession':\n",
    "            entry.append(REF_TAG)\n",
    "        elif METADATA_COLS[i]=='date' or METADATA_COLS[i]=='submission_date':\n",
    "            entry.append('2020-01-01')\n",
    "        else:\n",
    "            entry.append('reference')\n",
    "    f.write('%s\\n' % ','.join(entry))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging sequence data in the GISAID alignment with metadata on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/msa_0608.tar.xz blee098@cluster.hpcc.ucr.edu://rhome/blee098/bigdata/SARS-CoV-2-Data &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/merge-metadata-lmdb-parallel.py blee098@cluster.hpcc.ucr.edu: &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/merged-interim.csv blee098@cluster.hpcc.ucr.edu://rhome/blee098/bigdata/SARS-CoV-2-Data &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu: &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/job-merge-db.sh blee098@cluster.hpcc.ucr.edu: \n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/msa_0608.tar.xz blee098@cluster.hpcc.ucr.edu://rhome/blee098/bigdata/SARS-CoV-2-Data/msa_0608.tar.xz\n",
      "\n",
      "mkdir -p ./MPL/out &&\n",
      "sbatch job-merge-db.sh\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Write sequences to lightning-mapped database and extract location information from metadata\n",
    "msa_name         = MSA_NAME\n",
    "msa_file         = f'{msa_name}.fasta'\n",
    "msa_path_cluster = ppath.join(SSH_DATA, msa_name, msa_name + '.fasta')\n",
    "script_file      = 'merge-metadata-lmdb-parallel.py'\n",
    "msa_file2 = f'{msa_name}.tar.xz'\n",
    "msa_path2 = os.path.join(SARS_DIR, msa_file2)\n",
    "num_dirs  = int(11000000 / 500000) + 1\n",
    "\n",
    "job_file         = 'job-merge-db.sh'\n",
    "job_str          = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --job-name=merge\n",
    "#SBATCH -e ./MPL/out/merge-error-%a\n",
    "#SBATCH -o ./MPL/out/merge-out-%a\n",
    "#SBATCH --array=0-{num_dirs}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "tar -xf {SSH_DATA}/{msa_file2} -C {SSH_DATA}\n",
    "\n",
    "python {script_file} --meta_file {os.path.join(SSH_DATA, METADATA_INTERIM)} --msa_file {msa_path_cluster} --lmdbDir {DATABASE_DIR} --refFile {REF_FILE} --dbNumber $SLURM_ARRAY_TASK_ID --metaNew {METADATA_NEW}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SARS_DIR, job_file), mode='w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "meta_path      = os.path.join(SARS_DIR, METADATA_INTERIM)\n",
    "msa_path_local = os.path.join(SARS_DIR, msa_file2)\n",
    "print(f'scp {msa_path_local} {SSH_HOME}{SSH_DATA} &&')\n",
    "print(f'scp {meta_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('scp %s/%s %s '   % (SARS_DIR, job_file, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# commands to execute on the cluster\n",
    "print('sbatch %s' % job_file)\n",
    "print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find list of known gaps\n",
    "ref_df    = pd.read_csv(os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv'))\n",
    "ref_index = list(ref_df['ref_index'])\n",
    "ref_sites = [dp.get_label_new(str(i) + '-A')[:-2] for i in ref_index]\n",
    "variants  = [epsilon, alpha, beta, gamma, lambda_new, delta, twentyE_EU1, D614G, omicron, ba2, ba5, ba4, ba2121, kappa, mu, b12, b11318, b1429, B1_1_318]\n",
    "gap_sites = []\n",
    "for var in variants:\n",
    "    for site in var:\n",
    "        if site[-1]=='-':\n",
    "            gap_sites.append(ref_index[ref_sites.index(site[:-2])])\n",
    "for i in list('abcdefghijklmnopqrstuvwxyz'):\n",
    "    gap_sites.append(f'22203{i}')\n",
    "#print([dp.get_label_new(i + '-A')[:-2] for i in gap_sites])\n",
    "out_file = os.path.join(SARS_DIR, 'gap-list.npy')\n",
    "np.save(out_file, gap_sites)\n",
    "\n",
    "# transfer list of known gaps to the cluster\n",
    "print(f'scp {out_file} {SSH_HOME}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking sampling by region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/check-regional.py blee098@cluster.hpcc.ucr.edu: &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-check-regional.sh blee098@cluster.hpcc.ucr.edu:\n",
      "\n",
      "sbatch job-check-regional.sh\n",
      "\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/regional-2022-06-09.csv /Users/brianlee/SARS-CoV-2-Data\n"
     ]
    }
   ],
   "source": [
    "# finding the number of sequences in each region\n",
    "out_file      = f'regional-{DATA_DATE}'\n",
    "script_file   = 'check-regional.py'\n",
    "job_dir       = SSH_HOME + SSH_DATA\n",
    "job_file      = 'job-check-regional.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --time=0-3\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem-per-cpu=100G\n",
    "#SBATCH --job-name=regional-check\n",
    "#SBATCH -e ./MPL/out/region-error\n",
    "#SBATCH -o ./MPL/out/region-out\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python {script_file} --alignment {os.path.join(SSH_DATA, METADATA_INTERIM)} -o {os.path.join(SSH_DATA, out_file)}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), mode='w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer data processing scripts and job file to the cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run the job on the cluster\n",
    "print('sbatch %s' % job_file)\n",
    "print('')\n",
    "\n",
    "# transfer output data back to local directory\n",
    "out_path = os.path.join(job_dir, out_file + '.csv')\n",
    "print('scp %s %s' % (out_path, SARS_DIR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting regions and time series for analysis\n",
    "<a id='select'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regions that are selected manually because either\n",
    "# 1. They are broken up in an abnormal manner\n",
    "# 2. For some of the pandemic the case numbers in the region were too low for the model to be reliable \n",
    "#(or there was essentially no community transmission)\n",
    "\n",
    "north_cali = ['santa clara county, alameda county', 'san francisco county', 'sacramento county', \n",
    "              'contra costa county', 'humboldt county', 'sonoma county', 'stanislaus county', \n",
    "              'san joaquin county', 'solano county', 'merced county', 'santa cruz',\n",
    "              'madera county', 'southern san joaquin valley', 'placer county',  \n",
    "              'davis', 'nevada county', 'santa cruz county', 'san mateo county', 'butte county', \n",
    "              'mendocino county', 'shasta county', 'tuolumne county', 'napa county',\n",
    "              'marin county', 'el dorado county', 'lake county', 'san luis obispo', \n",
    "              'sutter county', 'contra costa', 'tehama county']\n",
    "\n",
    "south_cali = ['los angeles', 'los angeles county', 'orange', 'orange county', 'san diego', \n",
    "              'san diego county', 'ventura', 'ventura county', 'san luis obispo', \n",
    "              'san luis obispo county', 'imperial', 'imperial county', 'san bernardino county', \n",
    "              'riverside county', 'kern county', 'imperial county']\n",
    "\n",
    "regs = [['europe', 'united kingdom', 'northern ireland', None, None, None],\n",
    "        ['europe', 'united kingdom', ['england', 'wales', 'scotland'], None, None, None],\n",
    "        ['north america', 'usa', 'california', south_cali, None, None],\n",
    "        ['north america', 'usa', 'california', north_cali, None, None]]\n",
    "\n",
    "state_elim   = ['california']\n",
    "country_elim = []\n",
    "for entry in regs:\n",
    "    if entry[1]!='usa':\n",
    "        country_elim.append(entry[1])\n",
    "\n",
    "# Large countries or states, that must be broken up into smaller regions\n",
    "use_subregions  = [          'usa',        'mexico',        'canada', 'turkey', 'india',        'brazil', 'russia',     'argentina',         'chile']\n",
    "cont_subregions = ['north america', 'north america', 'north america',   'asia',  'asia', 'south america', 'europe', 'south america', 'south america']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join(SARS_DIR, f'regional-{DATA_DATE}.csv'))\n",
    "continent = [str(i) for i in list(df['continent'])]\n",
    "country   = [str(i) for i in list(df['country'])]\n",
    "combined  = [continent[i] + '-' + country[i] for i in range(len(continent))]\n",
    "df['combined'] = combined\n",
    "data           = df['combined'].value_counts()\n",
    "\n",
    "regions = np.array(list(data.axes[0]))\n",
    "counts  = np.array(list(data))\n",
    "mask    = counts > MIN_REGIONAL_SEQS\n",
    "regions = regions[mask]\n",
    "counts  = counts[mask]\n",
    "\n",
    "regions    = [i.split('-') for i in regions]\n",
    "continents = [i[0] for i in regions]\n",
    "countries  = [i[1] for i in regions]\n",
    "\n",
    "for i in range(len(regions)):\n",
    "    if countries[i] not in use_subregions and countries[i] not in country_elim:\n",
    "        regs.append([continents[i], countries[i], None, None, None, None])\n",
    "\n",
    "for i in range(len(use_subregions)):\n",
    "    continent = cont_subregions[i]\n",
    "    country   = use_subregions[i]\n",
    "    df_temp   = df[(df['continent']==continent) & (df['country']==country)]\n",
    "    data      = df_temp['region'].value_counts()\n",
    "    states    = np.array(list(data.axes[0]))\n",
    "    counts    = np.array(list(data))\n",
    "    \n",
    "    mask      = counts > 1000\n",
    "    states    = states[mask]\n",
    "    counts    = counts[mask]\n",
    "    \n",
    "    for j in range(len(states)):\n",
    "        if states[j] not in state_elim:\n",
    "            regs.append([continent, country, states[j], None, None, None]) \n",
    "\n",
    "regs.sort(key = lambda x: x[:2])      \n",
    "selected = regs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find number of infected individuals in each region over time using the IHME data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['africa', 'mauritius']\n",
      "['africa', 'reunion']\n",
      "['africa', 'seychelles']\n",
      "['asia', 'brunei']\n",
      "['asia', 'hong kong']\n",
      "['asia', 'india', 'jammu and kashmir']\n",
      "['asia', 'india', 'puducherry']\n",
      "['asia', 'iran']\n",
      "['asia', 'south korea']\n",
      "['asia', 'vietnam']\n",
      "['europe', 'gibraltar']\n",
      "['europe', 'kosovo']\n",
      "['europe', 'liechtenstein']\n",
      "['europe', 'russia', 'moscow']\n",
      "['europe', 'russia', 'saint petersburg']\n",
      "['north america', 'canada', 'new brunswick']\n",
      "['north america', 'canada', 'newfoundland and labrador']\n",
      "['north america', 'martinique']\n",
      "['north america', 'mexico', 'state of mexico']\n",
      "['north america', 'mexico', 'yucatan']\n",
      "['north america', 'mexico', 'nuevo leon']\n",
      "['north america', 'mexico', 'veracruz']\n",
      "['north america', 'mexico', 'queretaro']\n",
      "['north america', 'sint maarten']\n",
      "['north america', 'u.s. virgin islands']\n",
      "['oceania', 'northern mariana islands']\n",
      "['south america', 'argentina', 'buenos aires']\n",
      "['south america', 'argentina', 'santa fe']\n",
      "['south america', 'argentina', 'ciudad autonoma de buenos aires']\n",
      "['south america', 'aruba']\n",
      "['south america', 'bonaire']\n",
      "['south america', 'brazil', 'sao paulo']\n",
      "['south america', 'brazil', 'parana']\n",
      "['south america', 'brazil', 'goias']\n",
      "['south america', 'brazil', 'ceara']\n",
      "['south america', 'brazil', 'para']\n",
      "['south america', 'brazil', 'espirito santo']\n",
      "['south america', 'brazil', 'paraiba']\n",
      "['south america', 'brazil', 'federal district']\n",
      "['south america', 'chile', 'region metropolitana de santiago']\n",
      "['south america', 'chile', 'valparaiso']\n",
      "['south america', 'chile', 'biobio']\n",
      "['south america', 'chile', \"o'higgins\"]\n",
      "['south america', 'chile', 'antofagasta']\n",
      "['south america', 'chile', 'los lagos']\n",
      "['south america', 'chile', 'tarapaca']\n",
      "['south america', 'curacao']\n",
      "['south america', 'french guiana']\n"
     ]
    }
   ],
   "source": [
    "est_dir = os.path.join(SARS_DIR, IHME_DATA)\n",
    "file1   = os.path.join(est_dir, 'data_download_file_reference_2020.csv')\n",
    "data1   = pd.read_csv(file1)\n",
    "data1   = data1[data1['version_name']=='reference']\n",
    "for year in ['2021', '2022']:\n",
    "    df2 = pd.read_csv(os.path.join(est_dir, f'data_download_file_reference_{year}.csv'))\n",
    "    df2 = df2[df2['version_name']=='reference']\n",
    "    data1 = pd.concat([data1, df2])\n",
    "popsize = list(data1.inf_mean)\n",
    "dates_all = list(data1.date)\n",
    "locs      = [i.lower() for i in list(data1.location_name)]\n",
    "\n",
    "regions = selected\n",
    "new_regions = [list(np.array(i)[np.array(i)!=None]) for i in regions]\n",
    "\n",
    "eliminated = []\n",
    "full_regs  = []    # the full region name including continent, country, etc.\n",
    "new_regs   = []    # the name of the smallest scale of region used\n",
    "for i in range(len(new_regions)):\n",
    "    reg = new_regions[i]\n",
    "    if '2' in reg[-1] and '2' in reg[-2]:\n",
    "        last_loc = reg[-3]\n",
    "    elif '2' in reg[-1]:\n",
    "        last_loc = reg[-2]\n",
    "    else: \n",
    "        last_loc = reg[-1]\n",
    "    if last_loc not in locs and not isinstance(last_loc, list):\n",
    "        if last_loc=='czech republic':\n",
    "            full_regs.append(regions[i])\n",
    "            new_regs.append('czechia')\n",
    "        else:\n",
    "            print(reg)\n",
    "            eliminated.append(reg)\n",
    "    else:\n",
    "        new_regs.append(last_loc)\n",
    "        full_regs.append(regions[i])\n",
    "new_regs.append('czechia')\n",
    "\n",
    "dates_excluded = []\n",
    "for loc in new_regs:\n",
    "    if isinstance(loc, list):\n",
    "        dates_excluded.append([])\n",
    "        continue\n",
    "    pop   = np.array(popsize)[np.array(locs)==loc]\n",
    "    dates = np.array(dates_all)[np.array(locs)==loc]\n",
    "    dates = dates[~np.isnan(pop)]\n",
    "    pop   = pop[~np.isnan(pop)]\n",
    "    \n",
    "    window = 5\n",
    "    pop_smooth     = [np.mean(pop[i:i+window]) for i in range(len(pop)-window)]\n",
    "    dates_smooth   = dates[window:]\n",
    "    dates_smallpop = dates_smooth[np.array(pop_smooth)<100]\n",
    "    dates_temp     = np.array([(dt.date.fromisoformat(i) - dt.date.fromisoformat('2020-01-01')).days for i in dates_smallpop])\n",
    "    \n",
    "    new_dates  = []\n",
    "    dates_diff = np.diff(dates_temp) - 1\n",
    "    idxs       = [-1] + list(np.nonzero(dates_diff)[0]) + [-1]\n",
    "    if len(dates_temp)==0:\n",
    "        dates_excluded.append(None)\n",
    "        continue\n",
    "    intervals  = [dates_smallpop[idxs[i]+1] + '---' + dates_smallpop[idxs[i+1]] for i in range(len(idxs)-1)]\n",
    "    dates_excluded.append(intervals)\n",
    "    \n",
    "mask = [True if i is not None else False for i in dates_excluded]\n",
    "dates_excluded = np.array(dates_excluded)[mask]\n",
    "full_regs      = np.array(full_regs)[mask]\n",
    "new_regs       = np.array(new_regs)[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_new_regions(reg, dates_excluded):\n",
    "    #print(reg)\n",
    "    if any([isinstance(i, list) for i in reg]):\n",
    "        new_regs = [reg]\n",
    "        return new_regs\n",
    "    new_regs = []\n",
    "    for i in range(len(dates_excluded)):\n",
    "        new_reg = [j for j in reg]\n",
    "        if i==0:\n",
    "            continue\n",
    "            #new_reg[-1] = dates_excluded[i].split('---')[0]\n",
    "        else:\n",
    "            new_reg[-1] = dates_excluded[i].split('---')[0]\n",
    "            new_reg[-2] = dates_excluded[i-1].split('---')[1]\n",
    "        new_regs.append(new_reg)\n",
    "    last_reg = [i for i in reg]\n",
    "    last_reg[-2] = dates_excluded[-1].split('---')[1]\n",
    "    new_regs.append(last_reg)\n",
    "    return new_regs\n",
    "\n",
    "selected = []\n",
    "for i in range(len(full_regs)):\n",
    "    new_regs = make_new_regions(full_regs[i], dates_excluded[i])\n",
    "    for reg in new_regs:\n",
    "        selected.append(reg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__need to also take into account days that are present in the sequence data but not the IHME data, and regions not in the IHME data.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_dir = os.path.join(SARS_DIR, f'regions-{DATA_DATE}')\n",
    "if not os.path.exists(regions_dir):\n",
    "    os.mkdir(regions_dir)\n",
    "\n",
    "for group in selected:\n",
    "    new = np.array(group)[np.array(group)!=None]\n",
    "    contains_list = False\n",
    "    for i in new:\n",
    "        if not isinstance(i, str):\n",
    "            contains_list = True\n",
    "    if contains_list:\n",
    "        label = ''\n",
    "        for i in new:\n",
    "            if isinstance(i, str):\n",
    "                label += i\n",
    "            else:\n",
    "                if len(i) > 4:\n",
    "                    #print(i)\n",
    "                    if new[2]=='california':\n",
    "                        if 'riverside county' in new[3]:\n",
    "                            label += 'south'\n",
    "                        else:\n",
    "                            label += 'north'\n",
    "                else:\n",
    "                    label += '_'.join(i)\n",
    "            if i!=new[-1]:\n",
    "                label += '-'\n",
    "    else:\n",
    "        label = '-'.join(new)\n",
    "    f = open(os.path.join(regions_dir, label + '.npy'), mode='wb')\n",
    "    np.save(f, [group])\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/find-region-times-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/SARS-CoV-2-Data/regions-2022-06-09 blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-regions-parallel.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/regions-times-2022-06-09\n",
      "sbatch job-find-regions-parallel.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/regions-times-2022-06-09 /Users/brianlee/SARS-CoV-2-Data\n"
     ]
    }
   ],
   "source": [
    "# Transfer scripts, run processing, and extract files\n",
    "# Processing clips the time series in different regions according to set rules regarding the distribution of the number of sampled genomes over time\n",
    "out_folder    = ppath.join(SSH_DATA, 'regions-times')\n",
    "script_file   = 'find-region-times-parallel.py'\n",
    "region_data   = f'regions-{DATA_DATE}'\n",
    "\n",
    "job_file      = 'job-find-regions-parallel.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH -e ./MPL/out/region-error-%a\n",
    "#SBATCH -o ./MPL/out/region-out-%a\n",
    "#SBATCH --array=0-{len(selected) - 1}\n",
    "\n",
    "regions=({region_data}/*)\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "region=${{regions[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file {ppath.join(SSH_DATA, METADATA_INTERIM)} -o {out_folder} --regions \\\"$region\\\" \n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('scp -r %s %s. &&' % (regions_dir, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add london and south africa time-series for early detection\n",
    "reg_dir = os.path.join(DATA_DIR, 'additional-regions')\n",
    "if not os.path.exists(reg_dir):\n",
    "    os.mkdir(reg_dir)\n",
    "london_reg   = [['europe', 'united kingdom', 'england', 'lond', '2020-10-15', '2021-03-01']]\n",
    "safrica_reg  = [['africa', 'south africa',        None,   None, '2020-04-20', '2022-05-28']]\n",
    "london_name  = 'early-detection-london'\n",
    "safrica_name = 'early-detection-south-africa'\n",
    "np.save(os.path.join(reg_dir, london_name  + '.npy'), london_reg)\n",
    "np.save(os.path.join(ref_dir, safrica_name + '.npy'), safrica_reg)\n",
    "file_paths = os.path.join(reg_dir, '*')\n",
    "print(f'scp {file_paths} {SSH_HOME}{out_folder}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /rhome/blee098/bigdata/SARS-CoV-2-Data/regions-times-2022-06-09 | wc -l\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/extract-seqs.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-extract.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-extract.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/msa-reg /Users/brianlee/SARS-CoV-2-Data &&\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:EPI_ISL_402125.fasta /Users/brianlee/SARS-CoV-2-Data\n"
     ]
    }
   ],
   "source": [
    "# Processing with multiple lightning mapped databases storing the sequences\n",
    "out_folder    = ppath.join(SSH_DATA, 'msa-reg')\n",
    "script_file   = 'extract-seqs.py'\n",
    "region_dir    = ppath.join(SSH_DATA, f'regions-times')\n",
    "lmdb_dir      = DATABASE_DIR\n",
    "\n",
    "# run on cluster\n",
    "print(f'ls {region_dir} | wc -l &&')\n",
    "print('')\n",
    "num_regions = 500\n",
    "\n",
    "job_file      = 'job-extract.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=25G\n",
    "#SBATCH --time=6:00:00\n",
    "#SBATCH -e ./MPL/out/genome-error-%a\n",
    "#SBATCH -o ./MPL/out/genome-out-%a\n",
    "#SBATCH --array=0-{num_regions - 1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "regions=({region_dir}/*)\n",
    "region=${{regions[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file {METADATA_NEW} -o {out_folder} --regions \\\"$region\\\"  --lmdb {lmdb_dir}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer job file to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/msa-reg | wc -l\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/process-threaded.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-process.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/gap-list.npy blee098@cluster.hpcc.ucr.edu:\n",
      "sbatch job-process.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-data /Users/brianlee/SARS-CoV-2-Data &&\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:EPI_ISL_402125.fasta /Users/brianlee/SARS-CoV-2-Data\n"
     ]
    }
   ],
   "source": [
    "# Processing with multiple lightning mapped databases storing the sequences\n",
    "out_folder    = ppath.join(SSH_DATA, 'genome-data')\n",
    "input_dir     = ppath.join(SSH_DATA, 'msa-reg')\n",
    "script_file   = 'process-threaded.py'\n",
    "threads       = 20\n",
    "\n",
    "job_file      = 'job-process.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --cpus-per-task={threads}\n",
    "#SBATCH --mem=150G\n",
    "#SBATCH --time=5:00:00\n",
    "#SBATCH -e ./MPL/out/genome-error-%a\n",
    "#SBATCH -o ./MPL/out/genome-out-%a\n",
    "#SBATCH --array=0-{num_regions - 1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "files=({input_dir}/*)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file \\\"$file\\\" -o {out_folder} --refFile {REF_FILE + '.csv'} --threads {threads}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# moving south africa and london time series that are used for early detection to their own folders\n",
    "file_paths = ppath.join(SSH_DATA, 'genome-data', '*early-detection*')\n",
    "target_dir = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "print(f'mv {file_paths} {target_dir}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding time series with good sampling in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-data ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/trim-intervals-sd.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-trim-intervals.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-trim-intervals.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09 /Users/brianlee/SARS-CoV-2-Data\n"
     ]
    }
   ],
   "source": [
    "# Finding time series with good sampling\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-data')\n",
    "output_dir    = ppath.join(SSH_DATA)\n",
    "script_file   = 'trim-intervals-sd.py'\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "job_file      = 'job-trim-intervals.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=20:00:00\n",
    "#SBATCH -e ./MPL/out/sampling-error-%a\n",
    "#SBATCH -o ./MPL/out/sampling-out-%a\n",
    "#SBATCH --array=1-370\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {script_file} --input \\\"$file\\\" -o {output_dir} --trimDir genome-trimmed\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eliminating sites that never rise to a very high frequency or have a mutation only in a few sequences in a region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-trimmed ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/find-max-frequencies.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-max-freqs.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-max-freqs.sh\n"
     ]
    }
   ],
   "source": [
    "# Finding maximum frequencies in each region\n",
    "min_freq = 0.01\n",
    "\n",
    "input_dir   = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "max_dir     = ppath.join(SSH_DATA, 'max-freqs')\n",
    "out_dir     = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "max_script    = 'find-max-frequencies.py'\n",
    "filter_script = 'filter-sites-parallel.py'\n",
    "\n",
    "# run on cluster to find number of files\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 518   # the number of data files \n",
    "\n",
    "job_file      = 'job-filter.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/freqs-error-%a\n",
    "#SBATCH -o ./MPL/out/freqs-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {max_script} --input_file \\\"$file\\\" -o {max_dir} --refFile {REF_FILE + '.csv'}\n",
    "python {filter_script} --input \\\"$file\\\" -o {out_dir} --max_dir {max_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# move early detection regions to their own directories and remove their files from the max-freqs directory\n",
    "target_dir    = ppath.join(SSH_DATA, 'genome-filtered-variants')\n",
    "delete_paths  = ppath.join(max_dir,  '*early-detection*')\n",
    "london_dir    = ppath.join(SSH_DATA, 'genome-filtered-london')\n",
    "london_paths  = ppath.join(out_dir,  '*early-detection-london*')\n",
    "safrica_dir   = ppath.join(SSH_DATA, 'genome-filtered-south-africa')\n",
    "safrica_paths = ppath.join(out_dir,  '*early-detection-south-africa*')\n",
    "\n",
    "# Copy Great Britain files to their own directory for doing early detection\n",
    "uk_dir   = ppath.join(SSH_DATA, 'genome-filtered-uk')\n",
    "uk_paths = ppath.join(out_dir, '*england_wales_scotland*')\n",
    "\n",
    "# run on cluster\n",
    "print(f'cp {london_paths} {safrica_paths} {uk_paths} {target_dir}/ &&')\n",
    "print(f'rm {delete_paths} &&')\n",
    "print(f'mv {london_paths}  {london_dir} &&')\n",
    "print(f'mv {safrica_paths} {safrica_dir} &&')\n",
    "print(f'cp {uk_paths} {uk_dir}/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/combine-max-freqs.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-combine-max-freqs.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-combine-max-freqs.sh\n",
      "\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/nucleotide-counts.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# Combining maximum frequencies in different regions\n",
    "input_dir     = ppath.join(SSH_DATA, 'max-freqs')\n",
    "out_dir       = ppath.join(SSH_DATA)\n",
    "script_file   = 'combine-max-freqs.py'\n",
    "\n",
    "job_file      = 'job-combine-max-freqs.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=20:00:00\n",
    "#SBATCH -e ./MPL/out/freqs-combine-error\n",
    "#SBATCH -o ./MPL/out/freqs-combine-out\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python {script_file} --input_dir {input_dir} -o {out_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "print('')\n",
    "\n",
    "print(f'scp {SSH_HOME}{out_dir}/nucleotide-counts.npz {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='linked'></a>\n",
    "# Finding linked sites and classifying mutations as synonymous or nonsynonymous"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finding linked sites on cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "cd ./Archive-alt-vector\n",
      "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-find-linked-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-alt-vector blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-counts.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-linked.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-find-counts.sh\n",
      "sbatch job-find-linked.sh\n",
      "\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/linked-sites.npy /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/linked-sites-alleles.npy /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/linked-sites-anywhere.npy /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/nucleotide-counts.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# Finding single and double site counts in each region\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "count_script  = 'epi-covar.py'\n",
    "link_script   = 'epi-find-linked-parallel.py'\n",
    "out_dir       = ppath.join(SSH_DATA)\n",
    "temp_dir      = ppath.join(SSH_DATA, 'counts')\n",
    "\n",
    "# run on cluster to find number of files\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('cd ./Archive-alt-vector')\n",
    "print('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "num_files = 520   # the number of data files \n",
    "\n",
    "job_file1 = 'job-find-counts.sh'\n",
    "job_str1  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/counts-error-%a\n",
    "#SBATCH -o ./MPL/out/counts-out-%a\n",
    "#SBATCH --array=1-{num_files - 1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "cd ./Archive-alt-vector\n",
    "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "pwd\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{count_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --find_counts\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "# combining single and double site counts from the different regions to find linked sites\n",
    "\n",
    "job_file2 = 'job-find-linked.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=150G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH -e ./MPL/out/linked-error-%a\n",
    "#SBATCH -o ./MPL/out/linked-out-%a\n",
    "#SBATCH --array=0\n",
    "\n",
    "tols=(0.8)\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_dir}/linked-sites\n",
    "python {link_script} --data {temp_dir} -o $outfile --timed 1 --multisite -q 5 --link_tol ${{tols[$SLURM_ARRAY_TASK_ID]}} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file1)\n",
    "print('sbatch %s' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp %s%s/linked-sites.npy %s &&' % (SSH_HOME, out_dir, DATA_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NSP1-141-2\n",
      "['ORF9b-3-1']\n"
     ]
    }
   ],
   "source": [
    "print(dp.orf_to_nsp('orf1a-141'))\n",
    "print(dp.get_noncanonical_orfs(28290))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ORF9b-10-0']\n",
      "N-13-1\n"
     ]
    }
   ],
   "source": [
    "site = 28310\n",
    "noncanon = dp.get_noncanonical_orfs(site)\n",
    "print(noncanon)\n",
    "print(dp.get_label(site))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'NSP8-23-2'"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dp.get_label(12159)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 752,
   "metadata": {},
   "outputs": [],
   "source": [
    "traj_dir  = os.path.join(DATA_DIR, 'trajectories')\n",
    "sites_all = []\n",
    "times_all = []\n",
    "files_all = []\n",
    "for file in os.listdir(traj_dir):\n",
    "    df = pd.read_csv(os.path.join(traj_dir, file))\n",
    "    sites_all.append(df.columns.values)\n",
    "    times_all.append(list(df['time']))\n",
    "    files_all.append(os.path.join(traj_dir, file))\n",
    "nucs_all  = [i[1:] for i in sites_all]\n",
    "sites_all = [[dp.get_label_new(nucs_all[i][j]) for j in range(len(nucs_all[i]))] for i in range(len(nucs_all))]\n",
    "short_all = [[sites_all[i][j][:-2] for j in range(len(sites_all[i]))] for i in range(len(sites_all))]\n",
    "regions   = [os.path.split(i)[-1] for i in files_all]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 757,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZgAAAJOCAYAAAA6bWJ4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAD2L0lEQVR4nOzdeXxkeV3v/9enluzd6b1nprtnb2YFBhgYlgGGRQQERi+KLIoLgnLlooJ6Qa+KeO9VXK64IAo/kE1BFMQRRgdkEWSbhWGWnrVn757pvZPubJVavr8/TiWppLN1dbqTSl7PxyOTqlOnzvnWmU6+yTuf+nwjpYQkSZIkSZIkSccrt9gDkCRJkiRJkiS1JgNmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWZIkSZIkSZLUFANmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJZOUERcGRHfioj+iDgUEd+MiKdOs9+ZETEw5SNFxNsb9nltRDwUEYMR8bmIWDfNcbZHxEhEfGKOcf1eRNwWEZWIeNeCvFhJklrQcpqrI+KciKhFxPvns78kSa1iuczXkXlrRNxeP/+uiPjHiHj8PC+F1HIMmKUTEBGrgc8DfwGsA7YAvwuUpu6bUno4pdQz9gE8HqgBn6kf6xLgb4CfBDYDQ8BfTXPa9wE3zGN4O4FfB75wnC9LkqRlYxnO1a8HDgM/HhHtx/E8SZKWrGU2X/8Z8EvAW+uv5XHA54AfmufzpZZTWOwBSC3ucQAppU/W7w8DX5znc18PfD2l9GD9/uuAf00pfR0gIn4LuDMiVqWUjta3vRroA74FnD/bwVNKH60/53XzfTGSJC1Dy2aujoioj+l/Ae8CXg780zxfiyRJS9mymK8jYjvwi8AzUkrXNzz0d/N8LVJLsoJZOjH3ANWI+GhEvCQi1s7nSQ2/IH60YfMlwC1jd1JK9wGj1Cfa+l903w28bYHGLknSSrCc5uorga3Ap4BPAz91ks4jSdKptlzm6xcAu6aEy9KyZ8AsnYCU0hGyX/YS8EFgf0RcExGb53jqlWRv1WmsOuoB+qfs1w+sqt/+PeBDKaVdJzxwSZJWiGU2V/8U8G8ppcPA3wMvjohNJ+lckiSdMstovl4PPHYSjistaQbM0glKKd2ZUvrplNJW4FLgDOC9EbGjYcGBZ0952k8Bn0kpDTRsGwBWT9lvNXA0Ii4DXgj86XRjmONcc+4fEb/RcP+v5/fKJUlqDctkru4Efoz6W2xTSt8GHgZeO9exJElqBcthvgYOAqfP5/VKy4k9mKUFlFK6KyI+Avx8SumS6fZp+AXxR6Y8tAN4YsN+5wLtZG8VegNwNvBw9g4geoB8RFycUnryTOeaZZxT9/8G8H+P5xiSJLWiVp2rI+K1ZL8c/1VE/EV98xqyX6zfezzHliRpqWvh+XoP8L6IuDyldOPxHEtqZVYwSycgIi6MiLdHxNb6/W3Aa4DvzPK0HyFb/f2rU7b/HfDyepVSN1lPqM/WFyH4AHAecFn946/JVrD9wVnGVoyIDrKv80JEdERE/vhfpSRJrWsZzdU/BXwYeHzDOZ4FPDEiHj/La5EkaclbLvN1Sule4K+AT0bEVRHRVt//1RHxjtmvgtS6DJilE3MUuAL4bkQMkk1+twNvn+U5PwV8PKWUGjemlHYAv0A2Ge4j6w/13+uPDaWU9ox9kL3lZySltH+W83yQbOXd1wC/Wb/9k8f/EiVJamktP1dHxBayRYPe23iOlNJNwL/jYn+SpNbX8vN1g7cCfwm8D+gD7iMLw/91ludILS2mfB1KkiRJkiRJkjQvVjBLkiRJkiRJkpoyZ8AcER+OiH0RcfsMj0dE/HlE7IyIWyPiyQs/TEmSNBfnbEmSlj7na0nScjOfCuaPAC+e5fGXANvrH28C3n/iw5IkSU34CM7ZkiQtdR/B+VqStIzMGTCnlL4OHJpll6uBj6XMd4A1EXH6Qg1QkiTNj3O2JElLn/O1JGm5KSzAMbYAjzTc31Xf9tjUHSPiTWR/gaW7u/spF1544QKcXpKkY910000HUkobF3scS8y85mzna0nSqeScfYxF+x376OhRdg3sOqFjtJpcynH2SJbfJyBg/L+JNLFf5AiglhKJVN9jvoIgeLj9Mcq5Cr2VHjaU15KoHedRchwq9HO4eOS4ngeQ6uM+s3Q67amNvcUDDBSGxx+vpRr5yBMRVGtVEok11VVsKK/lSH6QQ8U+qtQmXROAYipSo0Y1qgQx6fFCrsD6jvWsaltF7lCVNFpjNEYZyo9wOH+EWi6xtn0tQ5Wh+iCAHFRqFSq1ConE5rZNrMuvgQTV0QppuEo1V6OUH6Ut2ih2tFFob8v+l9WgOjhKbagy7TWoFmvkcwVSuUY0XPooBCRI1Wzs0fA/NyWOETGxPQo5cu15Ih+kWso+hiukNPk4s/+/mf4cc+0bxRxUahNjaXhOLTfKaNcx3zJmG0V2jFQkanlqhWFIJ3HZtqgR1fbszPnR+umDaS7FAho7+vF99TY+PyJRqxaISESuSkox5XiJWsqx5+hmKuTpjiHaY5S1fYMcLXSRO+MMNvS0w3AfHH6gPpw8FNpg44XctruffC64+PTVTb/Kk2m2+XohAuZ5Syl9APgAwOWXX55uvPHGU3l6SdIKEhEPLfYYWpXztSTpVHLObp5z9sL6/r7v857r38Njg4+xqm0VV265kjc+4Y28/t9ez0NHHuLs3nN4oP8B3vj4n+OXnvxL8z7uyM7DHPj/bmfDGx9Px3lr6P/SQxz9ysNs+T9XErn5h12P/cH1tJ29mvWvPv4/JIweHOL+L3yPnjuykG30aZ2c8yNPIeqp5Nce+RpffeSr5CPP6rbVPG/rVWz6yCjVwyUIaD+nl/bXnsnNh2+hUqtQTVUuWHcBxb/aQ61vlAee2s+t5z5ELpejPd/Ohs4NfODWD/DQkYd49qpn8BvX/ySPFQ9Q6Gpj49FeWF3gl8//I+4evIfnbXwep1c38tZv/gi3b36QT27/Em996i/x1a9eyytvuZJimoiu7ul4iELKc25pK0O5YbpqnZNfaD5Y9eytdD1lE7XBMjfvvpkvfucato2exhMHL2AgP8ThwhG+uek2nvWE53LR7jNZ29dF6f4+8ms6WP/aC2nbumriuj02SPXwCFHMEYUcbdtWEYUctaEyR77yCJX9Q5Qe6CeN1oi2PNGRp23rKta8/FwKazvm9f+mNlql/OgAtcEK1SMl2s5cTduWnmn3rRwc5tA/3UsaqZDrLFDY0Enl4DC9Lz132ucM7zhAef8wqVSlNlBm1VVbKazvPGa/crmPhx76G4aGH6JU2kNHxxYuvuiPyefbx/c5emiEowdHuO/mfYwMlOkq19iYavRu7qJybx+5c3qpXLqBarnGtovWkSsE931vP0cODFNsz9PRXeScyzaQyyd27vx9BgbvJpdro61tE6cdfg1D1xxh45seT/u5a+a8ZqX7+yme1kWuqzivawzwsY99jIGdO3n+Nf/Khl/4Bar9/Qx///tUDhwgv3o1Izt2sPV9f8mqF7xg2uenVOX+B/6MwcF7aWvbSGfnmWzd8lry+a7xfT7/no9RvPAvGGir8LmdP8R/7b6CRI4tPQ/wqpF/5zOrX8Hv/djLufzAF+ALb4e33w2rTgPgyEiZJ7zriwDc+Ac/NO/XdSrNNl8vRMC8G9jWcH9rfZskSVpanLMlSVr6nK8XyWWbLuOTL/vkMdv/+ep/5roHr+Od33gnAJeuv/S4jpvvzUK66pFRAGpDZXKdheMKlwGKm7uo7B2adZ9USwx861E6L1pHYX0n5T2DjO4eoO+a++gppazqNR/0DnWNh8sAV227iqu2XTU+zoOfvIvRwyXW/+TFpHKVQ/9wN/kv7OXJZ5xPee8gq56zlcKqTnb3PQzAOTf08pTtr6LrCRPFjS8/9+X8wfV/wOC3skrabz/3ft76wrczcl8fBz54G++98ZfpftN21p57GkPf38ch7ubSvWfzRx1vZ/1zLmLdw0M8VtzDBx73Ofqq/ZzXfwYHzxzh6ot/hMvPeBIfve3P+N4t3+ayjsfzPy59C+0d7bSdtZr8qrZsABvhS499g3/d/A0+e/Vn2T+0n/v338qWVefyf7e9gUJuIhKr9I2Qa8sfE1i2nd4Np3cfc51zXUXWvOzc7JpXatSGKxPnPU65tjztZ/fOa9/C+k42/fwT5n3szks2cGycfKxicQ3nn/8/Z91n1boOVq3r4Izta+Z9/vOfsmna7Y973G9Nul8brVL62o30X/cQm948+/FTpcaBj+2g85INrPuxx82+b0ocOnSI9evXMzg4yKp160gjI+x/73sByPX0UNi8mdJ997Hx7W+j5/nPn/FYEXnOO/dts55vU3uZm/7jd/jBtw5wZscov/nUAm+79l7uGDiP/82b2V7byf/73Me58pv38vKutWzrXDf+3CPD5VmPvdQtRMB8DfCWiPgUcAXQn1I6njp8SZJ0ajhnS5K09DlfLzHFXJGXnfsy9gzu4c++92c8cdMTj+v54wFzXwmA2mD5uCovxxQ2dTFyXz+plmYMp0sP9NP/+fsZ+PajbHzj49n73u8BEO15Nr/tKeS6ixz+9N1UDo5M+/xUSxz56sOMPtBP2zm9dFy8joig9EA/g9/dAzfvA2Do+/vofck5DORh1XO2kr/jEIf/6V6Km7rI9RQZ/O4ecj1Ffqj2PHYfuYMH2ndzyfbLAOg4bw09V25h4L92w81HKeW7OPLlh6GQY81Lz6Hvmvt47P9+lyIw8twO1nRs5IzCWbzuotdxyYZLxsf6m8/4Tb665au89atvpTrazq88/lfobM9C3qHyEL/29V/j67u+znO3Ppdtq7axbdU2nrz5ydNf2zXzqzaeThRyTYfLyuTa8vRcuYX+ax+gcmCYwobJsXgqVzn8mXspPXyUXEeeNFKl8wkb5jzufffdxyc+8Qn+23/7bwwMDLBl+3ZOe9fvUNiwge5nPhNyOXIdHaRKhSiceES6tftebiifw7/9yURV8ztX30DuSTX++b5h/qnvCdwLfOv8S7iOC3j3Hb/C5k0votDzYg4cLZ3w+RfTnFcvIj4JXAVsiIhdwO8ARYCU0l8D1wIvBXYCQ8DPnKzBSpKkmTlnS5K09Dlft66fe/zP8bqLXkdnYT41oRNybXmis0Dp/j7Sc7dSG6qQ6zr+MKv9rNUMfGM3+/7yZtZcfT7tZ030aa2NVqkcHGH4lv1EMUe1r8SeP7hh/PF1r7qA4qYs9Cqs76T04BFSSuNVzKmWOPyZexm+/QCpVKX9cWvZ8PqLxx/vuHh9FjADm37xMg5fcx/9/3o/V/3gKnriCHf97KXs+YPrGbxpL8O3HRgP09cB69jOJzZ8gTef9s7x8ax52blUj5QYvGEPgzdkx20/r5eeZ55BYX0HB/52B+TgeT/4Mp6fe/mM1+R5Zz6PZ295Np+59zM8cvQRPvSDHwLgP3f9J1/f9XWuPu9qfuUpv3Lc11qnXucTNtB/7QMM3XaAVK4y+J3H2PjmJ1Lc2MXQrQcY+v5+Chs6KT86SK67QMd5a+Y8Zn9/PwDf/va3swrm3l7W/vAPH7PfQoTLAGcUbuMVFxzmPw/9NOdetpGdN+2jb+gpvHztPp75m/+H8t/8B5ft+FNuuuQ8Pv/gi/nlz3fxyu3v5/o93+Hru69ckDEsljmvYErpNXM8noBfXLARSZKkpjhnS5K09Dlft7bjDZfHpOEKpXv7OPr1XdQGy+NVzcej45L19DxnCwPffJQDH7qd09/5NHKdWaxz8ON3ULq3D4Cup2ym64kbOfKVh+l68iZ6nnb6pOMU1ndkPXmPlsmvbiPVEiN3HmTopr3keorkeoqsvmorUZhY5K3jvDX0PPMMOi5ZT9u2Vax/9QXs+aOs5/dAShR62yms72TgG1k3l3Wvu5DSzj4Gv7uHwqZO3vD6X6a3fXIbiJ5nbaE2UqXjvF6KZ/SM9wbuuGAdm9/2FKItN682Ir9xxW/w2i+8luv3XM+OAzu49oFruf3A7azvWM+7n/VucnESF6vTgims6aDtnNUc+dKDjK1/OfrgEYobuyjt7CPXXWDz255CefcAue7ipH+fMxkZySr1H3ssexNId/ex7U4W1NFH2bZ1Ez/xK88AYLRU5d7/2k46+B3i+g/ypz/3M/z7b93IGz/2MTpfXOEfj7yMP7/55ynkJi9M+X+vvZPfeOlFJ3esC+yULvInSZIkSZK00vQ86wwGvvkoI3ceojZUoThNX9+5RARrXnouHRes48AHb6P00JFskbcDw5Tu66vvBKuu2kpxYxcdj1s77XGK9cXgRncfpXP1egavf4y+z90HwGlvv3w8tJ507kKONa84b/x+YX0np/3Pp8L378qOVatR2NRF5cAw+d42Oi/ZQNfjN7LmZedBPqYNitvPWs3Gn52+n/VYtfV8bF21lc9e/Vle9s8v49VfePX49lduf6XhcotZ9eytHHzgjvH75b1DpJQY2dlH+3lriFzQtm3VLEeYcO2113L99ddP2nZSA+aU4OgeuOCM8U2nnbuaHV/v5NAdt7P+3n8jV2jnJed/jT2PDPGzf/81Dj1pM18+66nUCH7/yndz3YPP42u7ns3/9437eeOzz2XjquP/Q9Ri8StNkiRJkiTpJFrz8vPoefYWRncdpdpfItd9/D2Yx7SfuQrywfCt+9n//ls4+JEdUIMNb7iUM37nGRQ3zh7OFs/ogYDRXQMADN9xCIC2M1dNGy7PJDVUYT88Mkr7uVmF8qoXnDkeKEdxflXIJ2pD5wb+8Dl/yMXrLx7f9vwzZ16wTUtTx0XrWPuqx9H7Q+dQPKOb8r4hKvuGqB0dpWP79H8wmclYuNzV1cULX/hCANatWzfbU05M6QiUh6Bn8/imdfU/JPVX66HzDR8kHvkO6087CsAvf/8f+YUn/C2/+4w/oJCr8Ph191AAagmu27Hn5I31JLCCWZIkSZIk6STresLG8RYSzSzyNyaKedq29DD0vWzBve4rTqPaV6L97NVEMT/n83NteYqbuxh96AgAlb1DtD9uLetfe+FxjaOvMvG2/geGSpz3zDPovGjdeKuLU+05W5/Dc7Y+h0/f/Wn+8ua/5IrTr1iUcah5EUH3k7OAtvzoIEM37+PAviEA2s9f09QxR0ZGuPLKK3n84x9Pb2/v3E9o1kD29ciq08Y3ra5/LRypbgYC9twGQFtPlYefcRpnfnsPP7391exdU+Cz3/oMz97yHX5y9R1cy5O46aHD/MTTzzp5411gVjBLkiRJkiSdZG3bVtFxwVpy3UU6LzqxSsq2s7OgLL+2nbU/sp0NP3PpvMLlMR0Xrae0s48DH7uDan8pG1fH8dUgHipXx2/fNThC5GLRwuVGr7rgVXz91V+nPd867QV0rFXP20bPs84gt6qNjgvXUVjbcVzPH2uHUatlDZ1PargMWXsMmFTB3N5doK0jz5HLfh021Xsq95zG7c95P3+2+pUAtN01whPOv5rv3f5KRvq2ccVlf8eazhw7Hu0/ueNdYAbMkiRJkiRJp8D611/C6b9xBcXTTqwXbPvZqwGabj/R9ZQsBBu542B2//Ebj/sYh8oTFcz/uq+PP3twLwdHK7M8Q5q/4qYu1rz8PDb/4mVs+OlLjvv5q1bN0av5nuvg98+E6z84/4OODsI3/xyq5YlttRp8+MXwtd+vn3iigjkiWLWhkyN9NVhf72Heu4VLn/9aPvcnbya/fj0jd95JZbTGUwZ72H/Lj1HsOErf0Qe5Z+8Aw6NVWoUBsyRJkiRJ0ikQ+SDyJ96TuO2sLGBeddW2pp5f3JAt0tf15E30PGcr+dVtx32MsYB5Q7HArQPD/P4Dj/H5/X1NjUdaaNVqFs7+7M/+7LEPjhyBf/lFKPXDrhvmf9Cv/QF86bfg9s9m9w/dD/97Ezz8bXjom9m2nk2TnrJ6fQcP3XaQbx96JdVUgKHsjzpthTwdF1xA6a67eOTOQ2ys5Xhs8Fxq1QIXrMkW3fylT918fC96ERkwS5IkSZIktZB8d5Etv38l3U89be6dZ1BY28G6V13Ampee09TzD9dbZDx//USlaGNVs7SYRkdHeeITn8iZZ5557IMPfQtG6i0oRo7MfJBvvw8eaQig992Rfe5/OPt8z3VQK09+TseaSXcveubpAHzvll52jT4BznjS+GPtF15IaedOhvpGAOi9dDMjh8/i1Y/7HC97fC//cedejoxMOf4SZcAsSZIkSZLUYiJOvBL6ROwplQngf2/fylefegHd+dx46CwtttHRUdraZqjMv+DF8Cs74JznwPChmQ9y3W/Ah14IKUGlBLu/l23/yv+Gf3kL/Ps7svvPfjus3gLdG2HK1+U5T9zIC3/mYgAOPu1P4BV/Mf5Yx4UXkEZHGXx0PwCv+dELufu+l9DeNsLlG/6DWoKvfOwadr/tbZT37m3uQpwiBsySJEmSJEk6Lg8Olzi9vcjqQp6LejpZU8hzuGIFs5aGWQNmyFpZdK6DoYaA+bZ/gv33ZLfriwMCsOdWuO0fszB6bPHImz9eP85p8ILfhrfeDG+5cdpTXXDFafSsbefg4XaqhYn+64VNWTuN4UODFIo51q3uYPOq53L44adyRvE62vLBf/3HDRy59t/Y94d/dNzX4FQyYJYkSZIkSdJxeWh4lLM6JwK8dcWCFcxaEqrVKtVqdfaAGaBr/XhPZGo1+Mwb4ANXZfcrIxP7PXI9fO09cNrj4X/cBK/8EFz4suyxgT3Z50I7dK6Z8VTrt/Zwz/V7+etf/Br7HzkKQK4nay8zMjBKe0+RI1/6Emu/8EmG9lxMnn2c3TXKA90biSdvof8/v0j1yCztPBaZAbMkSZIkSZKOy4MjJc7ubB+/v7aY57A9mLUEjI6OAswjYF4HI31ZuDwWNJcHs8+NAfO1v5r1XX7R/4E12+DxPwo//gnY8hT44ffPa0xrNneN3354R3au/OosYO6/8TYKA4c4/Hd/T6EywuCjl5EocHr3/TzQezq7f+4BDv/4MKX77pvXuRZDYbEHIEmSJEmSpNYxWK2yf7TC2R2NAXOBXSPDizgqKTPvgLlzHaRaFjIffTTblqtHpeUp/5Yv+CE497kT9yPgjV+Z95h6N3SO3x46ko0vtyoLmMvFbnL7dlE+tJtC9Sxqoz2M8Aw2r76D/+y4gKOj3eTPHqTa3z/v851qVjBLkiRJkiRp3vaWskrlMzqK49vWFgtWMGtJOK4KZsh6L++6Ibvdvjr7PBYwF+uVxz/0xyc0ptUbJwLmg7sGAMj39GSnKnZTKA9S3rWLfKUEwMNH38pTills++8PvoBcX1Bbwi0yrGCWJEmSJEnSvPXVF/NbU8iPb1tbyNNXqVJNiXzEYg1Nmn/A3LM5+/xvvzaxrT2rKqZSD5hf8Rdw/gugc+0JjamxgnnfQ0epVmvki9kfaMqFboqVQUiJQjU778H+4JUHLuAHjnyH6x58AU/cfieX9lnBLEmSJEmSpGWgv76Y35riRN3iOV3tJOA7fQOLNCopMxYwF4vF2Xc8+9lw6Y9O3taWVRVTrvdg7lhzwuEywKr1HeO3y6Uq+x7MFvqrRY5ysZu20ex+vt77+cDhEdLgMG+5+Wts6inzmX0vpdLfd8LjOFkMmCVJkiRJkjRvfZUsYO5tqGB++cY1rC8W+OijBxdrWFphBgcH2bFjB0NDQ5O279mzB4D169fPfoB8AX70Q/CcX5/YVs1aVFCuH7PYcezzmpAv5PjBN17Kj77jcgB2330IgFL7WogcG658Emtf/5MU6uc/1F+iNjBAV2eRH7uszM7+8/jPTf9GuXx4Qcaz0AyYJUmSJEmSNG9jAfPa4kTA3JnP8eTVXdw/VFqsYWmF2bt3L//4j//I3r17J21/4IEHWLduHb29vfM70DnPnrg9Oph9rlcSU+g8dv8mnf+UTWw+ezUbtvWw664sKB7uyELw01/+fDb/+q+Tr2bnPXKkRHVwkHx3N1ecdzYAj5Dn4Yc/vGDjWUgGzJIkSZIkSZq3/vpifo0VzACntxd5tDS6GEPSCjQWIPf3T/QmLpVKPPjgg5x99tnzP9DZz4Zf2QFXvHkiYB6vYF64gHnM1gvW8tj9/VRGq4zUA+Y1W9YQhQLFyL62opoYOXKUXFc32894AgBH+jYxPPzwgo9nIRgwS5IkSZIkad76KlU6cznacpNjpS3tbRwqVxmq1hZpZFpJVq9eDUwOmG+55RZGR0d50pOeNP8DRUDvVmjrzgLmlCZ6MC9Qi4xGWy5YS62SeOy+fkZWnQapRvfadgAKbdnXVFsKRo8MkOvpYeOqdvKpxuGjaxkZ2b3g41kIBsySJEmSJEmat/5KlTXF/DHbT+/IFlXbUyqf6iFpBSoWi3R3d3PkyJHxbbfddhubN29m27Ztx3/Atm5IVaiUoDKcbVvAFhljzti+hlwu2HX3YfIvfiU9vW3k81lEW+goEqnKj9/7deKB+8h1d5PPBRsocXiwl5HSYws+noVgwCxJkiRJkqR56ytXj2mPAXBGexYw2yZDp0pvb+94BfPAwACPPPIIF110UXMHa+vJPo8OQrkeMJ+EFhltHQU2nb2aB289wIE9o2w4e834Y/nOTvKVEoXIEaURct3dAGzKV/n68JM5ePQotdrS+wOOAbMkSZIkSZLmra9SYc20AXMbAI9awaxTZPXq1eMB8+7dWfuI8847r7mDtWVhLqMDJzVgBnjC87dy6NFB+vcPs2Frz/j2XGcn+WqJaj5rmTEWMJ/dli2s+fn7X0SptOekjOlEGDBLkiRJkiRpXkq1GrceHWZ797G9ac9oL/K/t2/hyau7FmFkWolWrVrFwMAAMNGLec2aNc0drK3+73Z0ECojQEC+7cQHOY3tl29m7WnZ+dad0T2+PVWrkwLmtjPPBOBXzyjRWz7Kzr5zGRi4+6SM6UQYMEuSJEmSJGlevnl4gMFqjRetX33MYx35HD+3dSPndy38wmjSdNrb2ymVSqSUOHLkCLlcju7u7rmfOJ2O3uzzjn+Gr/8RkLIFAE+Sl/zC49l28Tq2XbRufFvp7rvJV0vcsWE71/3e37L2J38CgN51q/jBh77Lw0e3su/QLSdtTM0yYJYkSZIkSdK83HY0ax3wzLU9c+wpnXwdHR3UajUqlQr9/f2sXr2aXK7JuHPb07PPX//DhRvgLNae1s0r3noZHd3FSdvz1RLV9m4OFLuJesCdW72ai/c/RDXlufGB3adkfMfDgFmSJEmSJEnzcqBcpiefozt/bA9m6VRrb89aSZRKJY4cOcLq1cdW1s9bWxc84y3QcxpseBz0blugUc7fGe/5AzrP2EhbLs+R4cr49nzvGi49eD+5qHHjI8Ho6KFTPrbZFBZ7AJIkSZIkSWoNB8tV1heNk7Q0jAXMIyMj9Pf3s23bCYbCP/B78MLfhchBqi3ACI9P79VX07P/dtpu3Uf/8MRimfne1XRXSly0eogdB87lG//1VM4+682cd96vnvIxTscKZkmSJEmSJM3LwdEK69sMmLU0jAXM+/bto6+vj02bNp3YAXM5yBcmPi+CYnueYgoGB47y4PdvAiBfr8x+UleNh45spVwr8Miuj5EWIQSfjgGzJEmSJEmS5uVguWIFs5aMsYD5xhtvBODSSy9dzOEsiGJ7gVxliGd880/5zO//Dn/y918mtzpbgPDSVKWaCvTzA1Srg4yMLI1+zAbMkiRJkiRJmhcDZi0lHR0dADz88MOsWbOGtWvXLvKITlyxI0+BTu7u3g7Add+6lVx3F1EscuHQPgD6c68GYGDgrkUbZyMDZkmSJEmSJM0ppcQBW2RoCRmrYK5UKvT09CzyaBZGsT0PCW7e9gMk4PSRvfznxz9E8cILWHPXLazrbuPeg11AGDBLkiRJkiSpdRyt1iinxAYrmLVEjAXMAN3d3Ys4koVTbM8DsKqtg4F8D084ejs3feFzHD33bIavv54LO6vc8dgg289/J+vWPWuRR5sxYJYkSZIkSdKcvnV4AIBzu9rn2FM6NRoD5q6urkUcycIZC5h7i3n6iqvHt3dfeCEAW771Je7Ze5TTzvgZenufvChjnMqAWZIkSZIkSXP6xGMH2dJe5AXrVs+9s3QK5PN5CoWson65VTCvLuR5qPPM8e0dT30qq170Irb37aJcTXz7/oOLNcRjGDBLkiRJkiRpTvtHy1zY3UkhF4s9FGnc2EJ/yy1g7snnubn3Mm5e/QQAarUaG/77m7lizw5Ob4d3fOZWvnb3vsUc6jgDZkmSJEmSJM2pkhJFkyQtMevXrweWX4uM7nwOIri353wAqtUy7eefT+fqVfz6rf9IZyF44MDgYg51nN8WJEmSJEmSNKdKgnxYvaylZcuWLYs9hAVV7MgC5mLK7lfJ7lcrFaJQ4LR3vYsL7/wu//ysLn7i6Wct1jAncdlPSZIkSZIkzalSSxQNmLXEXHXVVbS3t3PJJZcs9lAWxFgFc66aJcy1yOqDa5UKAB0XXwRAemw3xfzSqB1eGqOQJEmSJEnSklZOiYIBs5aYtrY2nvvc544v9tfqiu3Z68jVsvvVesD88W/ez7d2HqB4+umQyzG6a9diDfEYBsySJEmSJEmaU9WAWTrpxiqY82MVzPUWGTc9cIC/++7DRLFI4bTNlHftXrQxTmXALEmSJEmSpDlZwSydfIViDgKiHjCPVTDnUpVSpQpA29ZtlHe3WMAcES+OiLsjYmdEvGOax8+MiK9GxM0RcWtEvHThhypJkmbjfC1JUmtwzlarqqREIWfALJ1MkQsKbXny9RYZtahXNKcaj/aNANB21lmM3n8/KaXFGuYkcwbMEZEH3ge8BLgYeE1EXDxlt/8FfDql9CTg1cBfLfRAJUnSzJyvJUlqDc7ZamWVlCiYL0snXbE9z5NO7+WcDd3jFcx5qjzWPwxA+4UXUO3ro7Jnz2IOc9x8KpifBuxMKd2fUhoFPgVcPWWfBKyu3+4FHl24IUqSpHlwvpYkqTU4Z6tlVRK2yJBOgWJ7VsH8/p94MtV6BXNve47DQ2WGR6t0XJT9XXLkzrsWc5jj5hMwbwEeabi/q76t0buAn4iIXcC1wP+Y7kAR8aaIuDEibty/f38Tw5UkSTNwvpYkqTU4Z6tlVWr2YJZOhWJ7nnKpSkchT60e327tbQPgsf5hOi54HEQwcucdiznMcQu1yN9rgI+klLYCLwU+HhHHHDul9IGU0uUppcs3bty4QKeWJEnz5HwtSVJrcM7WklRJiaIBs3TStdUD5vZiDiKokmNbPWC+f/8gue5uNv3ar9H9zGcu8kgz8wmYdwPbGu5vrW9r9Abg0wAppW8DHcCGhRigJEmaF+drSZJag3O2WlItJWpA3oBZOunGKpjbC1l7jFrkOGtNOwD37DsKwPqf/Rm6nvSkRRtjo/kEzDcA2yPinIhoI1tg4Jop+zwMvAAgIi4im/x8f44kSaeO87UkSa3BOVstqZwSgBXM0ilQbM9THqnQUcyi22rkaMslTu/t4N69AwDc+tVHOPTY4GIOc9ycAXNKqQK8BbgOuJNsJdsdEfHuiHhFfbe3A2+MiFuATwI/nVL9O48kSTrpnK8lSWoNztlqVZX6P8FCzoBZOtnau4uMDFVoy2fRbY08tUqFi9b18MgjRzj46ADf+PS93HP9nkUeaaYwn51SSteSLSzQuO23G27fATxrYYcmSZKOh/O1JEmtwTlbrahSqwfM5svSSdfd28bw0VHGvtyKbUWqlSpPvGuYwkiNz/3JzRTb8lz2gjMXdZxjFmqRP0mSJEmSJC1TlXoNfcEWGdJJ172mHRIMHxnlxv/1Qtav7qJWKVMYqQEwMljmJT//eDp6ios80owBsyRJkiRJkmY13iLDgFk66bp7swX9BvpKrOsskqpBtVKZtM+2i9ctxtCmNa8WGZIkSZIkSVq5Ki7yJ50y3WuygHmob5Rb7t3FQF+Zo4eGiEKQKokznrl5kUc4mRXMkiRJkiRJmtVYwJw3YJZOurGAebC/xOCREpBn+OgIqZL4ekeZtsvWLu4ApzBgliRJkiRJ0qzGK5hzBszSydbZUyRywWBfifbOApBjdLgEwEAusf9oaXEHOIUBsyRJkiRJkmZVHq9gXuSBSCtA5ILu3jYG+0uUS1WIPKXhUQCG83BgYHSRRziZAbMkSZIkSZJmVanZg1k6lbp62xnsK1EaLBPkqZbLAJSLQalSBSClxJ+8+uVc/y//tJhDNWCWJEmSJEnS7CpZvkzBgFk6JXrWtDPYP8rIUAUiB9SAsYA5uz06PAQp8Y2//8jiDRQDZkmSJEmSJM1hrAezAbN0anT3tmUVzENlIA9kVcu1Yo7ResA81N+3aONrVFjsAUiSJEmSJGlpOVqpcvORIZ6zbhXQsMifAbN0SnStaac0VGGwbxTIQ8oC5va2HKOjZf72bW/mzEufsLiDrDNgliRJkiRJ0iS/sOMhvnzoCLc/61I2tBXGA+a8AbN0SnT3tgPQt3eIQnsHo4MlANryOXJ77+PQ7kc4tPuRxRziOFtkSJIkSZIkaZKbjw4CMFjNqibHK5hzBszSqdCzpn38dlfvGkjDpJRoK+TpfOS2Y/ZP9a/RxWDALEmSJEmSpElKtSysGqjWFxarjVUwL9qQpBWla03b+O3ejeuABGmEjlyNVXvuOGb/0eHhUzi6yQyYJUmSJEmSNMloPVA+WskqmKv14kh7MEunxliLDID1WzcCsHbTXTzplo+Tr5RYd8bWSfsP9h06peNrZMAsSZIkSZKkScppcsA8dr9gwCydEu1dE0vnrd28HoA9936V7qN7ANh+xTMn7V8eGTl1g5vCgFmSJEmSJEnTOlpvkVE1YJZOqWj4Wlt7xsZjHt9w5tkAbH9aFjRXK+VTMq7pFObeRZIkSZIkSSvB3lKZ/zh4ZPy+FczS4lt/xoZjtj3u6c9i7WnvpTQ0xL3Xf4tqefECZiuYJUmSJEmSBMB7HniMt9/9yPj9YwLmnAGzdKp0rs4W+uvq7T3msVwuz+ZzzydfyOqHq5UKAAOHDvLv738v5dHSKRunFcySJEmSJEkCYE1hclR0bIuMUz4kacV67e9cQblUJZfP87Jffgennbedv/iTv+L63Fm8vb5PvlgEJlpk/OcnPsxd3/xPzrnsci54xpWnZJxWMEuSJEmSJAmAsQ4Yf3bhmfQW8hMVzDVbZEinWkd3kVXrOgC44BlX0rtpM6Vn/Bi7uraN7zNewVxvkTE6PARMfC2fCgbMkiRJkiRJAmCoWmNtIc+Pn76OVYUcR6tZwFzN8mWKBszSomor5CiVq+P3JyqYsxYZlXrQXBoaOmVjMmCWJEmSJEkSAMPVGp35LC5alZ+oYB6pZa0yivZglhZVWyHHaL11DUC+UA+Y68Hy2OehI/2nbEwGzJIkSZIkSQJgqFajqx4w9+TzDFSyIGuoWiOArpxRkrSY2vI5ytVErd62ZmKRvyxYHhk4CsCwAbMkSZIkSZJOteFqjc56iNyRD0bri/sN1bLK5rBFhrSo2ovZ1+dYFfPUFhlHDx4ADJglSZIkSZK0CIaqExXM7bkcI9WJCmarl6XF11b/+ixVpgTM5TIjAwPji/wNHT1yysbkdwZJkiRJkiQBMFybqGBuzwUj9bfhD1VrdOeNkaTF1l6oVzCPBczjLTIq9O/bM77fUH/fKRuT3xkkSZIkSZIEZC0yxiqYO3I5SrWGCmYDZmnRtRfyAJTqC3Dm8hM9mMcC5jMuuJjDjz1KrVY9JWPyO4MkSZIkSZKALEjuzE9UMJfqFcyD1aoBs7QEtE2pYI4I8sUi1XKZvr1ZwHzhs55DeWSYg7seOSVj8juDJEmSJEmSgKxFhhXM0tI1HjDX+6ND1ibjpi/8C/d/7wY6Vq3m7Cc+GYDH7rnrlIzJ7wySJEmSJEkCJi/ml/VgngiY7cEsLb5zNnTz0888m97O4vi2yOWoVSvsvmsHm885jzWbT6fY3sHB3aemgrlwSs4iSZIkSZKkJS2llC3y11DBPFJLpJQYqtXoyucXeYSSLjp9Ne96xSWTtlVGR8dvP+6KZxER9KzfwMChg6dkTP7pSZIkSZIkSYymRDUx3gqjPRckoJwSgw2VzZKWlmq5DEBHdw8XPfsqAFatW8/RQwdOyfn9ziBJkiRJkiSG6z1dO3MTFcwApVqyB7PUAl78i2+j2N4BwCormCVJkiRJknQqDY0FzGMVzPXPI7WaPZilFtDe3T1+u2fdegYPH6JWq5708/qdQZIkSZIkSQzXF/RrbJEB0Feukhq2S1qaOroaA+YN1KpVhvr7T/p5/c4gSZIkSZKkhhYZWbA81iLjULmSbTdglpa09p6e8dur1q8nly8w1N930s9bOOlnkCRJkiRJ0pI3XEtAQ4uMetB8sB4w9xgwS0taR9dEwHzOky7nlz/xWeIULM5pwCxJkiRJkqTxHsxdUxb52zuaBcxrCsZI0lJWaG8fv53L5U/Zef3TkyRJkiRJkiZaZEypYN5bKgPQWzx1gZWk4xcRi3JeA2ZJkiRJkiSNL/I3FjCPVTC/96G9AKwpGDBLS1FbZ9einn9eAXNEvDgi7o6InRHxjhn2eVVE3BEROyLi7xd2mJIkaS7O15IktQbnbC1VE4v8Ta5gHrPGCmZpSXrDn3+QN/7lhxft/HM2z4mIPPA+4AeAXcANEXFNSumOhn22A+8EnpVSOhwRm07WgCVJ0rGcryVJag3O2VrKhuoVzF1TKpjH9NqDWVqSulb3Lur551PB/DRgZ0rp/pTSKPAp4Oop+7wReF9K6TBASmnfwg5TkiTNwflakqTW4JytJWtqBfOUAmY6p26QJOYXMG8BHmm4v6u+rdHjgMdFxDcj4jsR8eLpDhQRb4qIGyPixv379zc3YkmSNB3na0mSWoNztpasoWqNYKI1xpkd7bxo/erxxxdrATFJS9tCLfJXALYDVwGvAT4YEWum7pRS+kBK6fKU0uUbN25coFNLkqR5cr6WJKk1OGdrUQzXanTmc+NBcjEXfPTx5yzyqCQtdfMJmHcD2xrub61va7QLuCalVE4pPQDcQzYZSpKkU8P5WpKk1uCcrSVruFobb48xxqplSXOZT8B8A7A9Is6JiDbg1cA1U/b5HNlfVomIDWRv57l/4YYpSZLm4HwtSVJrcM7WkjVcq40v8Nfox09bxxu3bliEEUlqBXMu/5lSqkTEW4DrgDzw4ZTSjoh4N3BjSuma+mMviog7gCrwaymlgydz4JIkaYLztSRJrcE5W0vZ0DQVzAB/dtGZizAaSa1izoAZIKV0LXDtlG2/3XA7AW+rf0iSpEXgfC1JUmtwztZSNVxNdOZtiSHp+CzUIn+SJEmSJElqYcO16SuYJWk286pgliRJkiRJ0vK2qa1AhwGzpONkwCxJkiRJkiT++pKzF3sIklqQf5aSJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDXFgFmSJEmSJEmS1BQDZkmSJEmSJElSUwyYJUmSJEmSJElNMWCWJEmSJEmSJDVlXgFzRLw4Iu6OiJ0R8Y5Z9ntlRKSIuHzhhihJkubD+VqSpNbgnC1JWk7mDJgjIg+8D3gJcDHwmoi4eJr9VgG/BHx3oQcpSZJm53wtSVJrcM6WJC0386lgfhqwM6V0f0ppFPgUcPU0+/0e8B5gZAHHJ0mS5sf5WpKk1uCcLUlaVuYTMG8BHmm4v6u+bVxEPBnYllL6wmwHiog3RcSNEXHj/v37j3uwkiRpRs7XkiS1BudsSdKycsKL/EVEDvh/wNvn2jel9IGU0uUppcs3btx4oqeWJEnz5HwtSVJrcM6WJLWa+QTMu4FtDfe31reNWQVcCnwtIh4Eng5c4yIEkiSdUs7XkiS1BudsSdKyMp+A+QZge0ScExFtwKuBa8YeTCn1p5Q2pJTOTimdDXwHeEVK6caTMmJJkjQd52tJklqDc7YkaVmZM2BOKVWAtwDXAXcCn04p7YiId0fEK072ACVJ0tycryVJag3O2ZKk5aYwn51SStcC107Z9tsz7HvViQ9LkiQdL+drSZJag3O2JGk5OeFF/iRJkiRJkiRJK5MBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJaooBsyRJkiRJkiSpKQbMkiRJkiRJkqSmGDBLkiRJkiRJkppiwCxJkiRJkiRJasq8AuaIeHFE3B0ROyPiHdM8/raIuCMibo2IL0fEWQs/VEmSNBvna0mSWoNztiRpOZkzYI6IPPA+4CXAxcBrIuLiKbvdDFyeUnoC8E/AHy70QCVJ0sycryVJag3O2ZKk5WY+FcxPA3amlO5PKY0CnwKubtwhpfTVlNJQ/e53gK0LO0xJkjQH52tJklqDc7YkaVmZT8C8BXik4f6u+raZvAH4t+keiIg3RcSNEXHj/v375z9KSZI0F+drSZJag3O2JGlZWdBF/iLiJ4DLgT+a7vGU0gdSSpenlC7fuHHjQp5akiTNk/O1JEmtwTlbktQKCvPYZzewreH+1vq2SSLihcBvAs9NKZUWZniSJGmenK8lSWoNztmSpGVlPhXMNwDbI+KciGgDXg1c07hDRDwJ+BvgFSmlfQs/TEmSNAfna0mSWoNztiRpWZkzYE4pVYC3ANcBdwKfTintiIh3R8Qr6rv9EdAD/GNEfD8irpnhcJIk6SRwvpYkqTU4Z0uSlpv5tMggpXQtcO2Ubb/dcPuFCzwuSZJ0nJyvJUlqDc7ZkqTlZEEX+ZMkSZIkSZIkrRwGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWrKvALmiHhxRNwdETsj4h3TPN4eEf9Qf/y7EXH2go9UkiTNyvlakqTW4JwtSVpO5gyYIyIPvA94CXAx8JqIuHjKbm8ADqeUzgf+FHjPQg9UkiTNzPlakqTW4JwtSVpuCvPY52nAzpTS/QAR8SngauCOhn2uBt5Vv/1PwF9GRKSU0gKOdVrvuf493HXorpN9miXlJfdewekD6xZ7GEvC3561kQe720/4OGcPlviZh/YvwIhOvq5iF9tWnbnYw1hwj274ECPtDyzY8SrlGql20r8FzVs1oJTPUaoNUqMy8461GjQMO5dyFCtFSDUmPTAf89z9048+j10jm47jwDHD6aowx7f903MHeVl1eNZ9Np11Ls/76Tcdx3hU53y9xDhfT7YQc7bz9eJb6PkaltacPe/5GibN2Utzvobp5mzn6yVhSc/ZkiQdr/kEzFuARxru7wKumGmflFIlIvqB9cCBxp0i4k3A2E8hAxFxdzODPgEbpo6pFX2EjyzEYZbFtVggG4ADf7vYo1h8/puYsMKuxX/O9MBJuQ6fmM9OP/PzC3GqsxbiIC1kOc3XsAy+Dp2vF5zz9QT/XWRW2HWYcb6Gk3AtTuF8Dc7ZrTxnr7Cvw1l5LTJehwleiwlei0yrX4cZ5+v5BMwLJqX0AeADp/KcjSLixpTS5Yt1/qXEazHBa5HxOkzwWmS8DivXYs/X4L+/MV6HCV6LCV6LjNdhgtdi5VrsOdt/exO8FhmvwwSvxQSvRWY5X4f5LPK3G9jWcH9rfdu0+0REAegFDi7EACVJ0rw4X0uS1BqcsyVJy8p8AuYbgO0RcU5EtAGvBq6Zss81wE/Vb/8o8BV7Q0mSdEo5X0uS1BqcsyVJy8qcLTLq/Z7eAlwH5IEPp5R2RMS7gRtTStcAHwI+HhE7gUNkE+RStKhv911ivBYTvBYZr8MEr0XG69BCltl8Df77G+N1mOC1mOC1yHgdJngtWsgym7P9tzfBa5HxOkzwWkzwWmSW7XUI/wgqSZIkSZIkSWrGfFpkSJIkSZIkSZJ0DANmSZIkSZIkSVJTllXAHBEdEXF9RNwSETsi4nfr28+JiO9GxM6I+If6QgpERHv9/s7642cv6gtYYBGRj4ibI+Lz9fsr9To8GBG3RcT3I+LG+rZ1EfGliLi3/nltfXtExJ/Xr8WtEfHkxR39woqINRHxTxFxV0TcGRHPWGnXIiIuqP9bGPs4EhG/vNKuw5iI+JX698vbI+KT9e+jK/J7hU4d5+vJnK8zztcTnK8zztkTnK+1GJyvj+WcnXHOzjhfZ5yvJ6zk+XpZBcxACXh+SumJwGXAiyPi6cB7gD9NKZ0PHAbeUN//DcDh+vY/re+3nPwScGfD/ZV6HQCel1K6LKV0ef3+O4Avp5S2A1+u3wd4CbC9/vEm4P2nfKQn158B/55SuhB4Itm/jxV1LVJKd9f/LVwGPAUYAv6ZFXYdACJiC/BW4PKU0qVki8y8mpX9vUKnhvP1ZM7XE5yvMyt+vgbn7DHO11pEztfHcs6e4JztfA04X49Z8fN1SmlZfgBdwPeAK4ADQKG+/RnAdfXb1wHPqN8u1PeLxR77Ar3+rWRfwM8HPg/ESrwO9df0ILBhyra7gdPrt08H7q7f/hvgNdPt1+ofQC/wwNT/tyvxWjS8phcB31yp1wHYAjwCrKt/7X8e+MGV+r3Cj8X5cL52vm64Fs7Xyfl6luuyYuds52s/lsLHSp+v66/JOXviWqz4Odv5esbr4ny9Qufr5VbBPPaWle8D+4AvAfcBfSmlSn2XXWT/02Hifz71x/uB9ad0wCfPe4FfB2r1++tZmdcBIAFfjIibIuJN9W2bU0qP1W/vATbXb49fi7rG69TqzgH2A39bf1vX/xcR3azMazHm1cAn67dX3HVIKe0G/hh4GHiM7Gv/Jlbu9wqdQs7X496L8/UY5+uM8/X0Vuyc7XytxeR8Pcl7cc4e45ztfD0T5+sVOl8vu4A5pVRNWVn+VuBpwIWLO6JTLyJeBuxLKd202GNZIq5MKT2Z7G0YvxgRz2l8MGV/LkqLMrJTqwA8GXh/SulJwCATb1EBVtS1oN736BXAP059bKVch3oPrKvJfjg6A+gGXryog9KK4XztfD0N5+uM8/UUK33Odr7WYnK+zjhnH8M52/n6GM7XK3u+XnYB85iUUh/wVbLy8zURUag/tBXYXb+9G9gGUH+8Fzh4akd6UjwLeEVEPAh8iuwtPH/GyrsOwPhfkUgp7SPrA/Q0YG9EnA5Q/7yvvvv4tahrvE6tbhewK6X03fr9fyKbEFfitYDsh6HvpZT21u+vxOvwQuCBlNL+lFIZ+CzZ948V+b1Ci8P52vl6jPP1OOfrY630Odv5Wotuhc/X4Jw9iXM24Hw9HefrFTxfL6uAOSI2RsSa+u1O4AfImqx/FfjR+m4/BfxL/fY19fvUH/9K/a8qLS2l9M6U0taU0tlkb0/4Skrpdayw6wAQEd0RsWrsNlk/oNuZ/JqnXovX11c1fTrQ3/CWjpaWUtoDPBIRF9Q3vQC4gxV4Lepew8Rbd2BlXoeHgadHRFdEBBP/Jlbc9wqdWs7XGefrCc7XE5yvp7XS52znay0K5+sJztkTnLMzztfTcr5eyfN1WgKNoBfqA3gCcDNwK9k3uN+ubz8XuB7YSVaq317f3lG/v7P++LmL/RpOwjW5Cvj8Sr0O9dd8S/1jB/Cb9e3ryRZouBf4D2BdfXsA7yPrLXYb2eqfi/46FvB6XAbcWP8a+RywdiVeC7K3qhwEehu2rbjrUH99vwvcVf+e+XGgfSV+r/Dj1H44X097TZyvna8br4fz9cS1cM5Oztd+LM6H8/WM18U52zl77Fo4X09cC+frtLLn66i/KEmSJEmSJEmSjsuyapEhSZIkSZIkSTp1DJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWZIkSZIkSZLUFANmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWZIkSZIkSZLUFANmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWZIkSZIkSZLUFANmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWZIkSZIkSZLUFANmSZIkSZIkSVJTDJglSZIkSZIkSU0xYJYkSZIkSZIkNcWAWWpSRFwZEd+KiP6IOBQR34yIp06z35kRMTDlI0XE2xv2eW1EPBQRgxHxuYhYN81xtkfESER8Yo5x/V5E3BYRlYh41zxfy/aI+FRE7I+IIxFxb0T8RURsnc/zJUla6lp53o6IZzeMZbA+nsbxndnEJZEk6ZSJiAcjYl9EdDds+7mI+Fr99tUR8f3676MHIuIrEXFO/bF3RUS5Puf11efzZ9Qfi4j4zYh4uP7cT0XE6jnGsjEi/r7+M8HhiPi7hsc+EhGjU+bZ/CzHWhcR/1yfnx+KiNfO41r4+7eWHQNmqQn1CevzwF8A64AtwO8Cpan7ppQeTin1jH0AjwdqwGfqx7oE+BvgJ4HNwBDwV9Oc9n3ADfMY3k7g14EvzPO1nA98F3gUeFJKaTXwLOA+4Mr5HEOSpKWs1eftlNI3GsZzSX3zmoZxPjyP80iStNjywC9N3Vj/nfRjwNuBXuAcsnm02rDbP9TnwY3AfwGfjYgAXk82Jz8LOAPoJJvvZ/NZYA9wJrAJ+OMpj/9h488CKaXqMUeY8D5glOxngtcB76//rDAtf//WclVY7AFILepxACmlT9bvDwNfnOdzXw98PaX0YP3+64B/TSl9HSAifgu4MyJWpZSO1re9GugDvgWcP9vBU0ofrT/ndfMcz7uAb6aU3tZwjH3Ae+f5fEmSlrrlNG9LktSq/gj49Yj4q5RSX8P2y4AHUkpfrt8/Sv0Pu1OllMoR8VHg14D1wMuBD6WUHgGIiPcAX4mIN6eUhqY+PyJeBGwDrmoIjm9u5sXUq7FfCVyaUhoA/isiriELvN8xw9Pehb9/axmygllqzj1ANSI+GhEviYi183lSw19YP9qw+RLglrE7KaX7yP4C+rj6c1YD7wbexsnxQmaYvCVJWiaW07wtSVKruhH4GvCrU7Z/D7gwIv40Ip4XET0zHSAi2oGfBh5JKR0Y29y4C9AObJ/hEE8H7gY+GhEHI+KGiHjulH3+e2TttG6KiFfO8noeB1RSSvc0bLuFiXcbTcffv7UsGTBLTUgpHSF7+0oCPgjsj4hrImLzHE+9kuytM//UsK0H6J+yXz+wqn7798j+IrvrhAc+vQ1kbw8CICLeUu9rNRARHzxJ55Qk6ZRZZvO2JEmt7LeB/xERG8c2pJTuB64ia2H1aeBAvRdyY9D8qojoAx4BngL8SH37vwM/FxFnR0Qv8D/r27tmOP9W4EXAV4HTgD8B/iUiNtQf/3OycHoT8FvARyLiWTMcqwc4MmVb488E0/H3by1LBsxSk1JKd6aUfjqltBW4lKzf03sjYkfDYgDPnvK0nwI+U3/7zJgBYOoiBKuBoxFxGdlfOP90ujHMca757n8QOL3hdf1lSmkN2Vt0inMdU5KkVrCM5m1JklpWSul2snUR3jFl+3dSSq9KKW0Eng08B/jNhl0+nVJak1LalFJ6fkrppvr2DwOfJKuM3kEWHAPsismL5O6obx8GHkwpfSilVE4pfYostH5WfRzfSykdTClVUkrXAn8H/DeAiPi3huO9jll+Jqjv7+/fWjHswSwtgJTSXRHxEeDnU0rTvh0mIjqBH2PiL61jdgBPbNjvXLK39NwDvAE4G3g4e5cuPUA+Ii5OKT15pnPNMs7p9v8y2YT5t8dzLEmSWlWLz9uSJLW63yFri/En0z2YUrohIj5L9gfhWaWUavXj/Q6M91jeDeyu92We2m7jVrK+zZMOM9spqLfgSCm9pPGBeg/mQkRsTyndW9/8RLKfFfz9WyuKFcxSEyLiwoh4e0Rsrd/fBrwG+M4sT/sR4DATf1Ed83fAy+t/Xe0m69v42fpCQR8AziNb9OAy4K/JVpn/wVnGVoyIDrKv70JEdEREfpZxvQt4dkT8v4jYUj/GBuCiWZ4jSVLLWGbztiRJLS2ltBP4B+CtABFxZUS8MSI21e9fCLyC2edp6vuui4jzInMx8P+Ad9eD5+n8M7A2In4qIvIR8aNkbTO+WT/ej0ZET0Tk6mH1TwDXzPA6BoHPAu+OiO56K42rgY/PMuR34e/fWoYMmKXmHAWuAL4bEYNkE9/twNtnec5PAR9PKU3662hKaQfwC2S/sO4j69f03+uPDaWU9ox9kL0FZySltH+W83yQ7G0/ryF7S9Ew2Sq206ovSHAF2aR6S0QcJZtcHyXrOSVJUqtbNvO2JEnLxLuB7vrtPrJA+baIGCDrq/zPwB/O4zgbgGuBQeDfgA+nlD4w084ppUP1c/0qWb/kdwBXNywY+EtkFdB9wB8Bb0wpfW2W8/93oJPsZ4JPAm+u/6ww0/n9/VvLUkz5mVmSJEmSJEmSpHmxglmSJEmSJEmS1JQ5A+aI+HBE7IuI22d4PCLizyNiZ0TcGhFPXvhhSpKkuThnS5K09DlfS5KWm/lUMH8EePEsj78E2F7/eBPw/hMfliRJasJHcM6WJGmp+wjO15KkZWTOgDml9HXg0Cy7XA18LGW+A6yJiNMXaoCSJGl+nLMlSVr6nK8lSctNYQGOsQV4pOH+rvq2x6buGBFvIvsLLN3d3U+58MILF+D0kiQd66abbjqQUtq42ONYYuY1ZztfS5JOJefsY/g7tiRpyZltvl6IgHneUkofAD4AcPnll6cbb7zxVJ5ekrSCRMRDiz2GVuV8LUk6lZyzm+ecLUk6VWabr+fTg3kuu4FtDfe31rdJkqSlxTlbkqSlz/laktRSFiJgvgZ4fX2l26cD/SmlY966I0mSFp1ztiRJS5/ztSSppczZIiMiPglcBWyIiF3A7wBFgJTSXwPXAi8FdgJDwM+crMFKkqSZOWdLkrT0OV9LkpabOQPmlNJr5ng8Ab+4YCOSJElNcc6WJGnpc76WJC03C9EiQ5IkSZIkSZK0AhkwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkSWqKAbMkSZIkSZIkqSkGzJIkSZIkSZKkphgwS5IkSZIkSZKaYsAsSZIkSZIkAalcozowutjDkFqKAbMkSZIkSZIEHPj4HTz2v7+72MOQWooBsyRJkiRJkgSU7jkMQCpXF3kkUuswYJYkSZIkSZIaVI+WF3sIUsswYJYkSZIkSZIa2IdZmj8DZkmSJEmSJKlBzQpmad7mFTBHxIsj4u6I2BkR75jm8TMj4qsRcXNE3BoRL134oUqSpNk4X0uS1Bqcs6UlLLJPVjBL8zdnwBwReeB9wEuAi4HXRMTFU3b7X8CnU0pPAl4N/NVCD1SSJM3M+VqSpNbgnC0tcfksKqsNWMEszdd8KpifBuxMKd2fUhoFPgVcPWWfBKyu3+4FHl24IUqSpHlwvpYkqTU4Z0tLVKolqNQAK5il4zGfgHkL8EjD/V31bY3eBfxEROwCrgX+x3QHiog3RcSNEXHj/v37mxiuJEmagfO1JEmtwTlbWqLSSGX8dm2oMsuekhot1CJ/rwE+klLaCrwU+HhEHHPslNIHUkqXp5Qu37hx4wKdWpIkzZPztSRJrcE5W1oEteGJUDmNVhdxJFJrmU/AvBvY1nB/a31bozcAnwZIKX0b6AA2LMQAJUnSvDhfS5LUGpyzpSWqNjIRKhswS/M3n4D5BmB7RJwTEW1kCwxcM2Wfh4EXAETERWSTn+/PkSTp1HG+liSpNThnS0vUeAVzIUijtcUdjNRC5gyYU0oV4C3AdcCdZCvZ7oiId0fEK+q7vR14Y0TcAnwS+OmUUjpZg5YkSZM5X0uS1Bqcs6WlK5WyquX8qjZqVjBL81aYz04ppWvJFhZo3PbbDbfvAJ61sEOTJEnHw/lakqTW4JwtLU1joXJ+VRvVo6OLPBqpdSzUIn+SJEmSJElSy0ojWYuM3Ko2W2RIx8GAWZIkSZIkSSteraFFhov8SfNnwCxJkiRJkqQV4cDH7uDIVx6e9rFUqkJArrtIKtdINVufS/NhwCxJkiRJkqQVYeSOgxz54kPTPpZKVaI9T64tn90v2yZDmg8DZkmSJEmSJC1bwzsOzli13KhWqpJrzxNtWVxmmwxpfgyYJUmSJEmS1NLK+4dmfOzgx++YsWq5URrNKphjrILZgFmaFwNmSZIkSZIktazhuw6x909uYujW/Sd0nFqpSrQXxiuYa6O2yJDmw4BZkiRJkiRJLauyL6teHn346Kz7pTT7on2p3iIjZwWzdFwMmCVJkiRJktSyxnsml2cPhNMcFcmpVCHabJEhHS8DZkmSJEmSJLWsyI8FzLMHyLXh8uyPjy/yZ8AsHQ8DZkmSJEmSJLWsVM1aX8wZMA9WZj9Oqb7IX3sWMNdKBszSfBgwS5IkSZIkqWWNVRqnco3ygWEqfaVp95tvBXOuo17BPGLALM2HAbMkSZIkSZJa1ljlcnn/EHv/+Eb2/MH1pOqx1cy1oZkrmFOlBtVEtOfJdRSy/Udmr3iWlDFgliRJkiRJUssaq2CuHhyZ2FZvm9Fo1oC5foxoyxOFHBTCgFmaJwNmSZIkSZIktaxpey9PFzDP0iKjVg+Yc/X+y7mOgi0ypHkyYJYkSZIkSVLLGguHG6VaQ8Ac9f1mWeQv1Rf0i4aAeayCeXT3AKWHjizQaKXlx4BZkiRJkiRJLWuuCuYoZPFXbWiigjmlyRXOtdJEiwyA6MhTq1cw7/uLm9n//lsWdMzScmLALEmSJEmSpJaV5qpgztcD5uGGCuYpLTTGKpgnt8iwB7M0HwbMkiRJkiRJalnTBcxUJ6qao5D1yGhc5C9VJ1c9Ny7yB5DryLvInzRPBsySJEmSJElqWdO1yJhUwVy/2dgio3RvH6OPHB2/X5tSwRwdhfEWGZJmZ8AsSZIkSZKkljVtBXNDwJzq7TAaA+aDn7iTfe/7/jHHiFlaZKTKNL2eJRkwS5IkSZIkqXXVyrXx1hZjUmOP5VoWDNcGZ255kUrHtshIo7VJxxmrcpY0mQGzJEmSJEmSWlYarZLrLkzeOE0F82xqpSoERDGLysaC5lSeCJWTAbM0LQNmSZIkSZIktaw0UiXf0zZ5W2MP5trcAXMqVYlinshlCwJGWxaZNfZ3dtE/aXoGzJIkSZIkSWpJtVKVVK6RX9cx+YF61XKqpfFF/qYzVqGcRmvj/ZcBolivYB61glmaiwGzJEmSJEmSWlLt6CgAhSkB83gF8xztMWpDWVVyrVQhNylgPraCue/aB+bVbkNaaQyYJUmSJEmS1JKqMwTMExXMtalPmbzbWMA8UiU6GgLmeg/mWkMFc/mRo5Tu7zvRIUvLjgGzJEmSJEmSWtJYwDy1RcbUCuZjFgGsqw2Vs/1LVXIdE/uMVzCPTG6LURson/igpWXGgFmSJEmSJEktabxFxvrOyQ809mAGct3F6Z8/XsFcmdSDOTdWwTw8eWG/yqGREx+0tMwYMEuSJEmSJKklVQfKkAvyq9smbT+2grn+eA6iowC5AKA2XK9gHplSwdyWqz+eBcz5Ne2AAbM0HQNmSZIkSZIktaTq4RHyPUWiHhiPGX34CPv+5paGHspZ0Nxx4Xq2vOsZnP6bV2Sb64v41UamLvI3uYJ51VXbaDt7NZVDwyfx1UitafoGNJIkSZIkSdISlmqJkZ19tJ/be8xjA9/YDUB59wAAHReso3pohN4XnQVAFOo9lquJVEuk0SmL/BUnVzBHISis7aB0f//Je0FSizJgliRJkiRJUssp7x6gNlCm86L1M+6T6i0yCms7OP2dV4xvj3yMP57KVUhMaZGRhc1pPGDOkV/dRnVglFRLx1RMSyuZLTIkSZIkSZLUciqHs37IxTO6AVjzI+ez9ke3T9onVbIWGGOB8rix+9UatZGsjcb0Fczl+v45cqvboJqoDZUX9HVIrc4KZkmSJEmSJLWcVO+fPNbuoueK06keHZ28T6neg3lKxXFEQD6yCuaRrEo5195QwZwLKAS1oYkWGWMLCVaPlsn3TF5UUFrJrGCWJEmSJElSyxmvTi5OxFtTK5Vr9fD4mArm+rZUmahgzjVUMAPk2vINPZhz5FdloXLtSGnyOUarjNx7+EReitTSDJglSZIkSZLUcqZWMAPHVCqPBcRMEzCTz0E1jVc5R8fkN/pHMTcRMOdz5Fe3AxxTJd332Xs58KHbqRwcbvq1SK3MgFmSJEmSJEktZz4VzGmsv3Lu2Ags8kGq1qgNZj2Vc51TA+aJCmYKMV7BXD0yOWAefWwQaAizpRXGgFmSJEmSJEmtpx4wk5+lgnlk5grmyOdIlUT5wDAEFNZ1TH68LTdR3ZzPEcUc0Z6nNjD9In/j55JWGANmSZIkSZIktZxUqUE+sgX5xszQImO6HswUAqo1KgeGya/tmNxqAyis7xy/HYXs+bmOArWxhQPHHqsfujZowKyVyYBZkiRJkiRJLSeVa8eEwhExKe1KY1XFuRkW+asmKgeGKWzoPObxtm2rJvatnyfa86TS1CA5O3ZtqEzl8EgTr0RqbfMKmCPixRFxd0TsjIh3zLDPqyLijojYERF/v7DDlCRJc3G+liSpNThnSwsjVWqT+i+Pa+i3XN47BEC+t/2Y3bIWGVkFc3G6gPnMiYCZesCc68gzfPtBHnvP9Y0jAeDIlx9mz3tuYHT3QBOvRmpdhbl2iIg88D7gB4BdwA0RcU1K6Y6GfbYD7wSelVI6HBGbTtaAJUnSsZyvJUlqDc7Z0sKZroIZIHJRj3yBBLnuIvnu4rEHyAfVI6OkUpX8+o5jHm7buoqOi9eT7ymOL/AX7XkAqodL4/vV6gsJjvVmrhwcpm1Lzwm8Mqm1zBkwA08DdqaU7geIiE8BVwN3NOzzRuB9KaXDACmlfQs9UEmSNCvna0mSWoNztrRAZqxgntJvubDp2OpkyNpejO7Kqo2na5ERhRwbXn/xpG25esAMkGqJVKlRG5p+0T9ppZhPi4wtwCMN93fVtzV6HPC4iPhmRHwnIl483YEi4k0RcWNE3Lh///7mRixJkqbjfC1JUmtwzpYWSKqkGSuYGxU3dU37/MgHVGrA5AX9ZhPtE7WafZ/byaO//S3SaG3KwBLSSrJQi/wVgO3AVcBrgA9GxJqpO6WUPpBSujyldPnGjRsX6NSSJGmenK8lSWoNztnSPKRyddqAeWoFc/GMGdpV5OvPzUFhzbE9mqfTWME8eP2e6cc1NXCWlrn5BMy7gW0N97fWtzXaBVyTUiqnlB4A7iGbDCVJ0qnhfC1JUmtwzpYWSKrUxhffaxST82U6zl8z7fOjHkTn13RMH1RP95yGgHmSwsRJx3oySyvFfL56bgC2R8Q5EdEGvBq4Zso+nyP7yyoRsYHs7Tz3L9wwJUnSHJyvJUlqDc7Z0gJJlTRtD+ZUm9yiYqb2F2Ohcn5127zPOVPAXFg7sUhgKlXmfTxpOZhzkb+UUiUi3gJcB+SBD6eUdkTEu4EbU0rX1B97UUTcAVSBX0spHTyZA5ckSROcryVJag3O2dICKteIVdMEzOWsRUXPlVvofurmmZ9fr2DOdcwZj42LqeXRdYV1HVT2DwNWMGvlmddXUErpWuDaKdt+u+F2At5W/5AkSYvA+VqSpNbgnC0tjFSpTV/BXA94Oy5cR3Fz94zPj3oP5uiYoe3FdOesTd9fOb+usYLZgFkry0It8idJkiRJkiSdMqlcm7V3ctu2VbM+P5qoYKaapt1caAiYa7bI0ApjwCxJkiRJkqSWkyo1ojB9ywqA3EwL8o0ZD5jnX8Hcdk7vtNsbezDbIkMrzXH8iUaSJEmSJElaGrIWGceGwxt//gmzVjaPqy8GGMdRwdxx3ho2/OylHPjw7ZO22yJDK5kBsyRJkiRJklpOVsF8bJDcPkOV8THPry8GeDwVzAC57uIx2/Kr28Zv10ZskaGVxRYZkiRJkiRJaimplqCaZm2RMecxKvWAuf346i+nO2eus5hVQgekYQNmrSwGzJIkSZIkSWotY4vt5aePtmrVGv37h2Y9xFgFcxSPLx6brmo68sGWdz2DnmdvoTpkwKyVxYBZkiRJkiRJLSXt+FcgC3anc8uXd/GJ3/oOB3cPzHyMSpMB8yz757qKUKlRG7UPs1YOA2ZJkiRJkiS1lPTPv5jdmCFgPnJgGID7v79/5mOMVzAfXw/m2RYQzHVl7TZqtsnQCmLALEmSJEmSpBaThcIxQ4uM9nrQ++i9fTMfob4w31goPG8NAXO05Sls6pw4Zle2AGBtsHx8x5Ra2HF+BUmSJEmSJEmLK5EFuTO1yBitVxAPH5056F3737bTeckGiqd1H9e5GyuYz3jXMyY9Nl7BbB9mrSAGzJIkSZIkSWotqd7WIjd9wFwayQLeWrU24yFyHQW6nrjxuE8dDeeMKefPjVUwD1nBrJXDFhmSJEmSJElqKaleMxmFmSqYs0X2qpWZA+aTwQpmrUQGzJIkSZIkSWopYwEzuemjrbEWGdH2KN/97ksplw+fknFZwayVyIBZkiRJkiRJLWZskb8ZWmTUA+aeM7/AwODd7Nt33SkZVRRyRFvOCmatKAbMkiRJkiRJainzXeSvOtoJQKVyZEHP3/tD57D+py/J7jx2K/z9j0N5OBtTR4FUqi7o+aSlzEX+JEmSJEmS1GLqi/xFmvbRsYC5Ujo5AfOqZ2+duPOV34N7vwg3fAge+Dq5tl+lVrKCWSuHFcySJEmSJElqHdUyKdUX+YtjK4VTSoyOZNsrpS4AygscME+y8YLs8xd/E+69jigfoDZiBbNWDgNmSZIkSZIktY7KyMQifxxbKXzrV3aRaom2jjypmu3XVAXzTR+FWz89936d6ybdzZUP2iJDK4oBsyRJkiRJklpHeYTxRf44Nsi9+7t7iICNZ60eb6HRVMD8vY/BLZ+ae79qedLdqB6mNmKLDK0c9mCWJEmSJElS66iMTCzyR/mYh3/snZcDWSXz7d+rZU8pNxEwFzqgOjr3flP2yTFoBbNWFCuYJUmSJEmS1DoqExXMTBMwRwQRQb4QEFnAXK4cPf7zFNrq55rD1IA5DdiDWSuKAbMkSZIkSZJaR3l4vAfzdC0yxuQKOaIeMDfVIqPQMc+AeUqLjNpR0miFlNLxn1NqQQbMkiRJkiRJah2VEqR619c0cwuLfEPAXC4fPP7zFNqzc83lmBYZA1CDVK4d/zmlFmTALEmSJEmSpNZRGSaNL/I382J6+UJuvEVGSlUqlcHjO0+ho6mAORjKzmkfZq0QBsySJEmSJElqHZXSxCJ/aeYAOJcPIjcR8pZKjx3fefJt8wyYJ7fIyEUWMNdGZg6/peXEgFmSJEmSJEkto3ywRH/lTWP3ZtwvX8xBTPRBHjj60PGdaL49mGtTejBbwawVxoBZkiRJkiRJLePwNzrHb0eaJWBu6MEMsOO2/0m5fHj+Jyq0H9P+YlrVUejeOHHeOARA5eDw/M8ltTADZkmSJEmSJC05Rw8dYN+D90/zSMPiebWZW1jk8wExUUWc8ofZccevzX8AYxXMKc2+X7UMPZuz29ueTjEeINqgtLN//ueSWpgBsyRJkiRJkpacv3vnr/Dx//lWatWprSYmAt+ozVxhnC/miEhAgWL5cqrl4NChb85/AIU2SDWozdFLuToKxU54Vz8859eIqNG+JcfIfX3zP5fUwgyYJUmSJEmStKSkWo3BvqydxaP33Dn10Ymb1ZlbZMRwhULUCPJceMHvs//W9aQ0SpqrInlMoSP7PFcf5mo5WxAQoJg9p9BTpTYw89ik5cSAWZIkSZIkSUvK/ocfHL/9D+96Bwd3PzJ+v7GvMtWZW2RU/vMRTm9PQI4N286CVASgNktbjUnGA+Y5+jBXRyFfrD8n6w8dVEjV2ixPkpYPA2ZJkiRJkiQtKfd857+IyHHp814EwIGGwHlyBfMs4W+CRA3Ik8vn6V6TLcRXq81RkTym0J59nrOCefSYCmYoQy3Nv1paamEGzJIkSZIkSVoyUkrs+Np/cM6TnsKzX/tTAAz29TXuMHF7luriSJByVSAPwKo12UJ8865gzs83YG5okVGveg4qWQ5uEbNWAANmSZIkSZIkLRnVcpmBw4c443EX0dmzisjlGOo/PP54rdIQZ81awZxIkbXIAMjX21dUq8dbwTxHIN3YIqPYWT93fVw1E2YtfwbMkiRJkiRJWjLKo1mgW+zoIHI5unrXTKpgTvMMmGOsRUbK9s/lsuri+QfM9XYXs/R5Hh/D1ArmlC3wl6q2yNDyZ8AsSZKkk2r3yChXfvdOrt3ft9hDkSRJLaBSygLdQltWQdzVu2ZSBXOq5Cd2Lg/NfKCUICZaZOTzWfhbLg/ObyDzrmAuH1PBHPUKZgNmrQQGzJIkSTqp8hHsHCpxYLSy2EORJEktoFwPmIvtWcDbvWbtpArmWmPAXBqY+UC1RMolUi0AyOWy41VGjzdgPo5F/saqnsdaZFRtkaHlz4BZkiRJJ1VnLvulbtgehJIkaR4q9RYZhbGAuXcNQ/19QLYAYKpmAXOx8CDMFhbXIJdP1Kr1Hsz5rLq4suAVzA0BcwQUOoiUPccKZq0EBsySJEk6qTrz2Y+cw1bwSJKkeRivYG5rp3JohAsfvYwYqGXhcrkGBL0dn2Tz6X8JozNXMKdagnyNWv1NVOMBc2W+AXO9Gnm2gHnkCIz0T7TIgHrYXD+pAbNWAANmSZIknVRtuRyFgOHa5F+w+ssVLvvmDq7vm+WtrZIkqSUcfmw31/31n1Mpl0/4WOM9mNvbGfreXgqVAmd2XkRpcJDqoaxdRS53FNp6Zg2YqSUin6hVgpQS+XpgXC0Pz28gcwXMgwfhD7Zlt8cqmAFyBYKxRf78A7uWv3kFzBHx4oi4OyJ2RsQ7ZtnvlRGRIuLyhRuiJEmaD+drLWWduRxD1eqkbTccGWLPaJn/9+DeRRqVJC0O52wtR//+V+/l9q9+kUfvvvOEj1UenahgHmsxUaPKYN9hhu86BEBH+93Q1j1ri4w0FjDXcpSGKuSLXdnxK7MsDNhoLDSeqQfz9z567L5QD5izCmZbZGglmDNgjog88D7gJcDFwGsi4uJp9lsF/BLw3YUepCRJmp3ztZa6znyO4Sm/YI3UK3ra87EYQ5KkReGcreXq6MEDAOx/6P4TPlallAW6hfaGgDnVGOo7xPD391PsPkS+OFCvYJ6l3UU1C5hJOQb7ShTqLTJqlXlWMBc76wOaIWAePjRxO9ew8GCuAPUKZltkaCWYTwXz04CdKaX7U0qjwKeAq6fZ7/eA9wBzLK0pSZJOAudrLWmdudykRf6+f2SIn9vxIAAdObu2SVpRnLO17JRLIwwcOgjAnvvuPfHjjVUwt7dD/Q/StVRl5O4+ynsG6dp4J7sHu+oVzHP0YI4aKQXVSo1CsRuAQ3sf5osf+AvSXAsQj7XImKmlRunoxO2hxrC5QKTKxBikZW4+P81vAR5puL+rvm1cRDwZ2JZS+sJsB4qIN0XEjRFx4/79+497sJIkaUbO11rSsgrmiV/i3n73w+O3i2EFs6QVxTlby87w0aOklM3z+x964ISPN96DuaFFRpCj7/ZHSCS+dNeNfOr20zhSaYfS7D2YydWglqdaSRQKWUXy/Td9k9u+fB1HDszxdTNWwTwWMB+4Fx78r4nHG899ZPfE7VyesQrmVLEHs5a/Ey4XiYgc8P+At8+1b0rpAymly1NKl2/cuPFETy1JkubJ+VqLrSs/uYJ5W8dEn8KBKb2ZJWklc85WK9j/8IMMHekfv18eyQLY7jVrOXJgPymdWNVuuTRRwTwW0BbybQztPcxIZZAHHssK+/uG83P2YM4qmHPUKjXyxU5Sgihkxzy859HZB5LLQ74dxlpqvP9Z8JEfgrGfaRqrp9tXTdzOF8cX+cMKZq0A8wmYdwPbGu5vrW8bswq4FPhaRDwIPB24xkUIJEk6pZyvtaRli/xNHzD3lQ2YpaXq4O5HKI/YoWGBOWer5X3s197C377tzeP3R4ezAHb9trMojwxTGpw+9N1995188C1voDQ0S99koDJaoquwmgN/dhtDN2aLAbe3d9Ge72Q0jdDTkb37qW8osvC3NsPPErVEihqkHNVqjUKxSK0SFDuLABx+dNfcL7bYMVHBXM2Cbw7cA4/dAvvvgjOfCa/6OPzg7088J1eAVK9gtgezVoD5BMw3ANsj4pyIaANeDVwz9mBKqT+ltCGldHZK6WzgO8ArUko3npQRS5Kk6Thfa0mb2iKj0vC7Vl/FgFlaimq1Kh9525v5lz/5P4s9lOXGOVvLwsjRI+O3xwLmDVvPBODIgX3TPue7n/0UR/bv5ZE7bqdcmvmPV+VSic3dZ1M9NLHPtgseT1fHGnrPPJ03vqibXCT6B6epJG5USxCpXsGcyBeKpEqOQlsWUB96dPf0z2tU7ILyUHZ7zVnZ513Xw988Bw4/CO09cPEroGP1xHNyeSKNLfI3c4uM8t5Bhm6Z/lpJrWTOgDmlVAHeAlwH3Al8OqW0IyLeHRGvONkDlCRJc3O+1lI3dZG/Uq3G5rYCrz5tHf0GzNKSVC1n4chDt968yCNZXpyztRyNjmQB7PptYwHz9L2Ne9ZvAOBb//Bx3vezr2bXnbdPu19ltERHW8+kbes2bWHTaWfTuaGX3cNFetoT9963j0otpm2TsfeB+zhaOgxRHa9gzheL1KpBqrevOPjIQ3O/uGInlOtBd2/9zQe7b5p4fMo4gWyRP0aB2SuY9773exz65N0n3FJEWmyF+eyUUroWuHbKtt+eYd+rTnxYkiTpeDlfaynrzMekFhmlWqIjl6O3mLdFhrREVcuVxR7CsuWcrVbWGIamlIiIiQrmbVmFb98MvY27VvcCWQ9ngG988mO85t1/eMx+5VKJjkL35POOVqkNlqkUEx+69zw2tQ1z5PAA382dybOmBMx779/JJ975ywBcXqqRUqFewVwgVYOxBfj2PnAfqVbj9q8/ypmXrKN3Y9exgy50TlQwV+pB894dE4+3TxcwF6GWBczM1iKj/lAqVYmOeUV00pJ0wov8SZIkSXPpyucnVTCP1Gp05HOsKWTbSzVXWJeWmmqlPMN2g2dpJUuN70iq91IerS/y17vpNNaesZVvffrvGDh08JjnVsoT31c2n3v+jEF0pVSivdBJFCdiq1SqUhuuUClmqezwaWeTz+e448jZHPjC5DYT3/rHvxu/PXw0Ta5gruQgl/1xe3R4iMfue5Cvf+oern3/bdO/4GIn3PV5+MKvQqXeg3nXDROPt6069jm5ApHmrmAeUz06Ouc+0lJmwCxJkqSTrjMXk3owl2qJ9lzQk88DMDhLf0JJi2OmgPnf/vJP+Nj/fOspHo2kpaLxe8PRgweAiR7MbZ2dvPjNv0y5NMJj9959zHPL9SA6cjm2XHAxldHpg9XR4SHaCp3kuots+sXLyK/voNpfggTlfD0cLnTwnB98NkfKNT77pb8Zf+6e++7l/u/dwAVPf3a23+DkHsy1SpArTIS+D912JwCplhgYGKBcnvK9r9gJQP83P8YX72zLWnI0KrQf+wJy+YaAeZafceqHqhkwq8UZMEuSJOmk68znGK4lavW31Y5Ua3TkcnTlsx9HhwyYpSVnpkrlo4cO0tHVPe1jkpa/xu8NYxXI5ZERInIU2trZeObZABzYdWx/43KpRM+69fyPj3yaQns7ldHStOfo37eXzrZV5DoLtG1bRWFdB5W+bN9SLjt/KeW55PLL6Mqv5mj5EIMHDwHw/es+T1tnF1f++E8B9fbMKUe1UssW+avmyOUTa047HYDDj2bVz1297fzxH/8xn/zkJycPph4wf+KBJ3Hb3i725c+d/HhlmteQK0A9YKY2cwVztGd/aK8enf4PelKrMGCWJEnSSdeZy37sHGuTMVbB3G3ALC1Z1alVfHUDhw6ML9QlaeWpNQTM+x64D8gW+Wvr7CQiKHZ00LtpMwcfefiY55ZHRujoWUWxvYNCWxu1apVadfJaDLValf59e+godJHrzPoSRzFPGsn2GxkLbgkY6ubpm14OwFc/8kFGR4a557vf4nFPfxar1q6nEG2UBiDVctSqWQ/mWjWIQo0oFMnlCwwcPgJA56oiAPfff//kQRc7Ga3mGallj490bZ3yooaOvUiNLTIqMwfMubaxgNkKZrU2A2ZJkiSddGOVysP1PoSlWo12K5ilJW26gDnVagwcOsSqdesXYUSSloLGCuZ7vvst9j/8IKPDwxQ7O8e3r996Jgd3TRMwl0YotmctJQpt2efKaImh/j7e+7of5uHbb+HogQNUKxUK0T4RMLdNxFdD1ZHx24MHYW3bZgDuvv4bfOsf/57yyDDbn/ZMqEFXYRWlQUiTejAHuXxi/8FDVIDBviP11zXDosOFTo6UJ9pgDLVtnPsi5YuQ6uOcbZ2JfNYjozZgwKzWZsAsSZKkk64zP7mCeaRewTwWMA9WZ/ilTtKiaQyRvn/dFwAYOtJPrVqhx4BZWrEavzcc2v0IH/u1t3DPt/+Lto6JgLl382kcObDvmOeWR0YotncAUGhrA6AyOspjO++hWqnwX5/6GIfrbTfyo0GuK6saHqv0BRgoD4/fHu6vUcgVecbGVwBwx39+GYBN55wHtUR3oZehgzlSLahVErl8nlTNkyvUIIKULzB89CgApZEZ2lSk6uSAuVKYeOyCl8Lz/9exz8nliVSuP33mCuZUzn4uqg7YIkOtzYBZkiRJJ11XbnKlcqlWozOXO2a7pKWjcSGvL3/4/ZSGhhg4dBDAFhnSCjb2veFpP/xjXPq8FwFZZXJbQwVz95p1jA4PUx4ZmfTccmmEYsfkgLlcKjHYdxiAx+69m8/8n9/Kdh6F6BprkTERXw00tKQY7c9+fjiz5yJW925k+OgROlf30r1mLdQSZ/VcQulInqF9VUbLZe7vu59UyxH5NB4wl4YG6uOYCM6PfOlLPPCqH6f0wAOMDA5xw6GJthjDjS2Xf/ivoHua74e5AtSyHWcNmEvZH9hTxZ+D1NoMmCVJknTSTbTImNyDeayyeWi2t49KOiX+4YaHed9Xd47fn9oiY+cN3+bowQMArFprBbO0Uo31YD7t3O085yd+Znx7YwVz95q1AAz2HaZWq3Lj5/+ZkYGBeouMLGAujrfIGKV/355J59jUcSYRMV65PFbJDHC0NMDqXBYyVwdq5COreF67Nlu0b9PZ5xIRpGpiW/eFtK+q0v9AiUf7H+Pqf7maWjVHrpBIkYNoI9WyiuhyqUqpUOQ/LrqcO975m4zceivDN3+f6+8eZNfQGgB6CiWGSw0/s7Svnv4i5QpEqkAuYIaAOdXSeAUzBsxqcQbMkiRJOumm9loeqfdgdpE/aen48p37+NdbHh2/P/Y2+PMuv4J8scjeB3ay/6EHIIK1Z2yd6TCSlrmx7w25QoG2ejUyQLEhYO6pB8yP7bybb/z9R/nPj3+IL/7Nn1MulRoqmCd6MPfv2wvAhc96Lle9/o08fePLsgPVexTnV7eNH3twZIjNuX4gwRAUYzeRr3HWGU9gw5lnc8UP/1i2Yy2RixybnzBIqb8Cg1krjFotRy6ftciI6CSlEXo3dlIu/f/snWeAXFd99n/nlum7s72pd8mWXOUi2xgXXMCmN1NCL4EASUh5CQkBkgABAoEEQu8dDDZg3HHvlmTZ6m3Vttfp5bbzfjizM7vqtiWtBOf3ZWduOffcO9Itz33O8/fZ3DGHHW0z+fk1KnLDT6exqMV4xSyHQtmH9uVqglGL7piCYUHgIUyBPMQ9jnRr7R7O5azRnApYR15Eo9FoNBqNRqN5fkSNqU5lVeRPEDPVg5kWmDWa6cfxA0JWzYM0MQz+4te9mczwEOnBAcb7+2iZOZtwLDZd3dRoNNPMhMBsWhamZWOYJoHvT4nIiFUE5lv/97+q0/asfwrgoBnMmaFBZq84i+s+9A8A9Dz4oFqpckoyk5MykAsFukSRhBVglgSmGEFYPgtnnct5f31DdTkZKNE21lbJtCjkIKYEZmFJECCMMARZYvVlxobBCpTom48n1L6mUxC4gM2lbd3syTdSLLrw9tsgP3zog2RY4HtKID+Ug9mp3ftogVlzqqMdzBqNRqPRaDSa487kiAwpJaVAEjGMA5zNGo1m+nC8gJA5WWCuuRSTbR2kBgcY2L6VzkVLpquLGo3mJCCovHwyLeVZnBCMJ0dkJBqbDljPKRZxisWDCsypoQGSbe0H2dpBHMz5PDGK1FkS2zMwGEeYPoGjxOGc57M2k4eKwGzH1bnML+bVX18gBBiBq0RgYPfaL+G7AWblRXg2pProp9N4DQuwhM95MwtETZdiyYVIPTQvOPRBqjqYjUM7mMuTChzriAzNKY4WmDUajUaj0Wg0x53oJCHZlRIJKoPZEAi0wKzRnAyUvYDwpEJaExnMlm2TbO9grHcfpXyOJh2PodH8WVN9+WROCMzKXTzZwRytq0cYUyWnieUtW+UpT0RkFLJpipk0ybaOAzdWEYknC8xBEBCnSF1Iic9CuBiGV3UEv3vjbl6yZjt5Twm4dswBwC2mqQvV4UqBm7cwczmMwl4AZODhuxKn0sdcZV+CdAa3ZTlWPAkNc7BEgB8chdu4IjBjikO6k4OyjsjQ/OmgBWaNRqPRaDQazXEnVnnILAYB5cqDWcQwEEIV+tMCs0Yz/RzgYHYnXIo2DZOEn1iy4UR3TaPRnERMjsgAMCuC8WQHszCMaqG/lllzeO3HPsXFr38zAOF4HKg5mMd6lMh7MAfzhPAqolMTXmMUqi+vhQBhesiKg/mpjCoAmK0IzAgfM2ziFFI0RZooFw02/nhRrTERQZpxAk/iWhMCs+qjn07jOg52OApWGFNIgqO5ZzEsCHwM26j2a3+G/ldFhnCYnGaN5lRBC8wajUaj0Wg0muPO5CiMUmX4abgyLW4a1WxmjUYzfTje/hnMtYiM+ta26vRoffKE902j0Zw8BJPODQCyYr6d7GAGuOLt7wXgqvd8kNnLz+T8l7+GN/z7f7H88quAmvN5tGcfcHCBecLBLIRyKxNTtRtiskBk4oWYIRDCRbrqXsI21LLpisAsCbCjFnv9FF70HMr5qVKYFV0I+AgMXFOJ5SMVcdxPp/HKZaxwGMwQhgjwj0pgNiHwMBsj+KnyAbMnC8qhmXWHzGnWaE4VtMCs0Wg0Go1GoznuRCdlMJcqD4vhysNizNAOZo3mZEAV+TOr3yeK/Fm2TaKpuTo9pgVmjebPGt+rjW4AkLIi7O4nMC86/yI+8L1f0rV4aXVa1+KlkzKYlcA8UnUwHxiRkb1nL35GCbQd/3Q+hVeqbOeYzBOunK+EKRCGW3UK25X7i4xbE5jNSAi3lGJd+FV4xcp5LmQggXS8DqRa1pkQmBub+fG1LydbLOE6ZexQuOpg9v2DO5KnYNoQeFhNEbyx0gGzg4IS6RtevgCzPoTUGcyaUxwtMGs0Go1Go9FojjumEIQNQSEIyFUcRfGKUzJmGuSP5mFNo9EcV8quT3iSg7nmUrSpa2mtTtcOZo3mz5v9IzJkxcI8OSJjgnAsdsh2JiIyxvt6sCNRonX1By4kYfzmnWr5ZJi8VwQgFuQIm0ooDgxDOZj3E5jTvo+q+uAjk3U0pUeJlAoEJYNwsszs0+chgO+95DIy8QRA1cEM8J2X38AvVqzEK5ewI2Eww5giwPeO4p7FsCBwMZsiBDl3St4yQFCsnF9jFuIwOc0azamCFpg1Go1Go9FoNCeEqGFQ9ANyFbdyfeXBMGwY1VxmjUYzfSgHc+0R0atmMFtEKuILaAezRvPnglMqctNnP8lo774p0/39IjI4jMB8OCYEZlDxGNUYjP0o70whK/cJhYLKV47LLJHKfYRnWBiyQFDIg+dUIzIyfoCT6FX70qZifs5d/yjmeIhIo0M4psTkWDHHWEMLUkpcyybi1CIt4uNjuKWycluf/WYMIZFSEgRHEJkNC2SA1aBc2vu7mIOCOr8aMRtMA/RILs0pjhaYNRqNRqPRaDQnBEsIvts7wmOpHAB1laGtlgBfaoFZo5luyvsX+fNcTMtCCDFF+JksCmk0mj9ddj+9lu61T/LAj787ZXpQGXW0v4N5/4iMI2GYFkKoc87B4jEmkGWfG7/7M3K5HIVCAdM0CeESqmzfMyxEehuFcgke/3otIsP3GV70KyyjgbH4pfiGyUVr7gUBHecNY4RUv2PFPCONbYBPEIrQlhqt9VEGuKWiyote+hLMF31MHYMjuZiNyj1OoxKx/fH9BeaKSB+1EJZ2MGtOfbTArNFoNBqNRqM5IYy46mHqU939ACQqTklLCFwtMGs0J5TUYAHX8XFKRZ65+3bG+nq5fO/vCeNUlwk8F8OyD9OKRqP5UyY7MqI+7OcsDvbPYK4U6n22DmYhRPWF1UEL/E0ivyfF6tWrKRQKxGIxBBCqOJhdTNY0JLnkqg4eeuaP2Kh7iowf4EaHSMZWkg61MdA6A4Bi0iHaVEZYSiSOFfOMNLWB9HBtm6hT5nV33QJA2Q7hFovV3OgJUX3CxX1IDHVsjKg6dhOC8gSTHczCNJCevg/SnNpY090BjUaj0Wg0Gs2fJxMRGZYQOFpg1mhOGL4f8JOPP8ac5c20zdrOgz/9PgBzAPq2AWcD4Lkepl0TmF/6tx8hnxo/4f3VaDQnntRAP/f98FsAlPO5KfMOmcEcPXTe8qGIJOpwy6UDBWYBSGh4xQLGf7uD5iBBNBqlv7+fWDQCWQiZavuuV+bxVhWB8cCZHyRAibqpkosbHSVid5AKQb5rLjMG95KNq/mFmIVrWpyz4VFuu+xV+EYB17KIOGXe+btf8MurrsexQ7i5MlZYRV1MvHSbKHR4SAzVtwmXdFDaX2CuZTCroVw6IkNzaqMdzBqNRqPRaDSaE8KNZy3gfbNqhcImIjJsQ+DqDGaN5oRRzCiX8p4NowfMs7wihUyazPAQvutWBSSAxRdewtnXvvSE9VOj0Uwf9//4O9XPY7091c9SSga6dwAHZjDbFRH22bDiiqsBiCUbpmyjYkLG7ojjmZIQFpZlkc/niVeiOCYiMlyvBFQiK5qXUqjcUwzvGUaaDuFwJ5mQYF/XPLU/MbWtfNjj/lXXMnNgL7bv8dS8KI5pEXXK2J6HkJJyKITnlLFDat+syku34IgOZtU3YQcgakX9JggKHhgCETaVg1lHZGhOcbTArNFoNBqNRqM5IVzSWMdLWhuq3+NmLSJDZzBrNCeOXKpWwKq0nzNRZEa46bOf5FsfeAeFTKo6BF6j0fx5MbS7G4AlF11KMZuhmM3Qu2UTv/nPT7Dt0QcBMCojkVa99o0AhCcVAz1aLnjV67j+bz7C4gsurk2cJLYK00AiMRC4rqsiMqJK7A2Z6vzk+h5yQmAOJMVKZMcDrSFcLCKhLvIm7J65kJ++/N0kgpfilw0KhsueGfMBmLNvB73NYGVTRMplBBChEpHhupMczEcbkaGOjZABImQiD3AwuxixSsa9KSCQ1UKGGs2piI7I0Gg0Go1Go9GcMJrt2u2nUcl01BnMGs2JpZCq5SwX0qmpM9PDDOzZDkD3mido7Jp5Anum0WhOBlKDA2SGh7j8be8l2dbO1kceYHTfXn7xyY9MWW6i+OfZ11zP2ddc/5y2ZRgmS1ZdMmXaFDevIQhkgIFg9erVjI2NsXB2JwC2FaIMuBiUTdWXkhtQqMRNjIVsHpSXsdLuoGDCzExAT+ccBoZSODkb0ZlkvNxCIAzm9O5kxsC9hN0yG865Gl8IIkIJzDIIqu7so89grtzvBB5G1DrQwVzyMCKVZSaKq/oSjKl51xrNqYJ2MGs0Go1Go9FoThjNtnnANEsIPC0wazQnjHy65mAe3juEMFuZefYl5MwY/tBeIrF4df5zGfKu0WhObZ747a8wTJMF555H0wz1kunJ3//6xHVgUh7xGzd0sycKQgpGKkUHYyElZYXCyjFdwmQ0pIrwjbseRT/glUPK0TxEB2tvyZI3obEQ0Jzx6W2JUTfjdPIiQBoGfihBx1AvYVedG6PlInecMR/KJUqVIoQ9fQNATWAOjpTBXHFXEyghOSj5U2bLso8IV1zOFXFc6hxmzSmMFpg1Go1Go9FoNCeMeutgAjNaYNZoTiD5SREZqYFRhIgx+8I38kDzCwgK2SmxGYnGpunookajmUa2PfoQyy65jGRbB8nWdgzTonvtk4SiUWadtuK4b39yVMQ+1+M786MIas5ev1wAwAjXAfD5067l150rARh2PRwp6SgGdDg5xmhifFCSNyThUsCMUY++ZgthNJOuCMpOYiaxSpsAgZNXH4r5qsDcO9Cvtlkt8nf0DmYRNQ90MDsqOgMmC8z6Xkhz6qIFZo1Go9FoNBrNCWNiOO1kLENHZGg0J5LJDmbPyYERw3UDdsXmHLBsLNl4Irum0WimGbdcolzI09g5A1A5yw0dKpJi7pnn8rqPf+a492Gy0OoJeLoxhD0pD74zqURfTFXs76GOBdV5/a4ScsOepNUqMUYzm4wHyBqSUMmnNeOTjxiMZJrIBeqeRERaSNU1VNsQFYE5YghKYeWMDoQSg2sRGUdwME+OyIhYB2QwS8fHqDiYsSYiMrSDWXPqogVmjUaj0Wg0Gs20YguBNu1oNCeOyQ7mwCsgRJR8xiEQJna9cixPFLSKN2iBWaP5c2Iil33y//2LXvtG6ppbOe3SKwC49v1/y7nXveL4dWI/gXk0YuJUHMxXXXUVy9qUwCwrAvNktlaE34graSXPZrGcre1zCYQgWpY05lRUxfbhJAViGIHEkgFSGPzx4usACJdKAFieSylSEZiN/QXmoyvyN5HB7Pbn8Sade6XjIypRH6KSwSw9fTOkOXXRRf40Go1Go9FoNCeUe89bQtio+RwsIXB15XSN5oSRTzs0tMcYH8gBLkKEKWZU4b95r343A7d8j2RbB/s2PkO8oWFa+6rRaE4s+dQ4MFVgXrLqBSxZ9YLq99NfeOXz3k5Q9hG2gThIUbvJWcRe5XZhKGKBA42NjYhSt5poRvHJHbA+gOFL6lGxF/ctPQeAqCNpzFUKAIY7KBAQ9nykIYiUi5Rt9WIt4hZVG06ZcuVlm18ZgWU8hyJ/GKqN4a8/TedHzlf7WPZrERmWzmDWnPpoB7NGo9FoNBqN5oSyLBFlfqxWOMwSAl9HZGg0J4x8qkxTVxyoFJ0SJuWscv01L1rGu7/yXSJxVTwrUlc/Tb3UaDTTQVVgPo756zKQ9H38EVI37zj4Ar5kQmrtKikBeShiUrJsgnAEyhnVjhlhOKLE2avjUf55Y4kFhokBLM76hEVpSrNKYFbnvUFZT4EYUd9FGpKwU8INqXuTkKNeuBlOiXIlg9mrCMy1In9HEphrRf4mHNl+qoz0AqTrEzg+RkVgppLBrIdzaU5ltMCs0Wg0Go1Go5lWLKEzmDWaE4Xn+JQLHk2dcahKOCblnBKYw5VCnFKqeYahHxk1mj8n8qkUcHzjcSbyiPNPDBx8vi8pVbTXtmIagKFoiN+deQkv2pth90SesRGhL6rOUS+PWryyx+XOaCv7Ljyd5emA14onaQzGq+1GHEnYg1gpYEgkKBIjIstIQ2BIiWMrMdmuFP+zPI+Spab5FfnMrBb5O9oMZp/ki+ciwiZmMsTg/6yl9xOPqoiM8ESRv4mIDO1g1py66LsFjUaj0Wg0Gs20oh3MGs2JY6LAX31LFGTFwYyBWxWY1SPiWVdfD0DX4mUnvI8ajWb6yKfGEcIgWn/8Ri8EhZr7V+53/ZeBJHvfPryKqTdZzBHxJfsSIcYSSQAe9uJgx5FS0BdVC7YJJc5KL6g6gcMyw0Xlx6ttR1w1va4YkDEi5IkTk6WqMlYOqbxl23Ww/ADLc3FsCyOQ1fuUZ53B7LuY9WHiF3Ti5128oaLqX0AtIsOciMjQ90KaUxedwazRaDQajUajmVYsgXYwazQniHxKDf1ONISZiMiIJCJ4BQ8iNYF5zhln8Xe/uGW6uqnRaKaJ7MgQsWQSY0IgPQ4ExZo464+WsFpqxfr8TJni+hEm5G2zlKezHLBm0jJDvoBwHfiS/oqDuUmo85n0gqoTOKBEJHCq68VLanqsLCnETAxZR5PMIQ2JRFQdzFbgEwoktu/hmDZ2UGkvCKoO5iNGZNgx9dfNq/1I2LBfET+jUuSPioMZncGsOYXRDmaNRqPRaDQazbRiGQJPHuhi0mg0x54JB3OsIVSNwWhoTxCUfAwJLYnw4VbXaDR/wkgp2bPhabqWHN+RC37B5fNLw6xrMPEzztSZ+4mwVqlAVzlgJFLzRw4HJoTrVJZzxCBeLiEqIzKkGyBddW7zKRHxa+3HHNV21AkoiRBZ6qkjowRmy8KpFPkTgInA8lxcy8LyAgwvQqlUOvoif9FKxEgxBYARtw9Y5MAif/o+SHPqogVmjUaj0Wg0Gs20YlcK5+jnKo3m+JNPKYE5ngwzkcHc2F6n/gqDhtiBIohGo/nTQwYBa2/7PaV8rjptdN8ecqMjzD3z3OO67cF8mV/MCfGuC2IEhalZxoFbc/G6AqxSni5HTbOAudEQQzKkHMyBpC8mqC8VcDwXjAkHs7qhCGSJ8KSsZLPSdKwsKYQMciJOHVmkCJCmhROqvWALmzaW5+JZNr7dRPPI+YwOpo4+IiPaoP4WVQa0kTiIwLx/BrO+EdKcwmiBWaPRaDQajUYzrVgVgVnHZGg0x5982sG0DcIxi4mIjFhSDT1/RypMeqg4jb3TaDQnip1rnuDe73+DR371k+q0dXfeimlZLDj3/OO23f6yw19mRqvfJ8dlAEjXr362JURch86K83hRPEJn2GZERJSD2Zf0RwzqSwVc10WYBtIPwKs5mMP+gcX4YmVJKWzgC4uEyJJILyQSbuANN3+rukxdKIoVBPimxWD7cgIBo4OZSREZRyjyF2lQfysOZvNwDma7IjCX/QOW0WhOFbTArNFoNBqNRqOZVqyqg1kLzBrN8SafKhNPhhBCcN37TwcgWhepzt+9fmS6uqbRaE4goz17ATBM5cgd7+9lw713cvoLX0S8ofG4bfdLuwdZJ2uicm++PGW+dKbmEJtSUi/VfcKFDQnaQjarI3PoiXQS+AGDEUFdqYjjOAjbUBEZFYHZ8wqEDuI0jpVr26gjTaTcim3EaB0fAmHSlinQ1dRK2CkB8M3XXMc3r65n70iW/+4bxzdMdnV3861vfevQ8V5WCEIJKKWAgzuYjfKQ+lunsp/9rHPAMhrNqYIWmDUajUaj0Wg004p2MGs0J458qky8IYyUkkLFwRytrwnMG+7vpahFDo3mT57U4AAAicYmAB6/6ZeYts2q177xuG43Yk6VofaV9hOY3QNdvJdnfd49GPDR+Z3YQuAJk9c0v5F8IPENQchzcBwHLAM8OaXIX8g/UGCOlmv3G3UiXfmk+mVaSbJhm332KLFCtrpc0crz+7V38sAdf2D3jAVs3bKZ3t5ehoeHD7OzDdWIDDMROmC2zKl1jZgFlsDPlA9YRqM5VdACs0aj0Wg0Go1mWrEMJTB7WmDWaI47+bQSmH+9tpf3fP9xAGKTBOb0cJH7f7Z1urqn0WhOEOnBfqCWJZwZHqJ1zryq4Hy8GCmoF1gvHFIRE72ZWga0n3cp78kcsE7UNPmrPkmdZbIwpnKSd5tJCoESkkPSVxEZllHJYA6QSITlkLAPvLfY0/BQ9XO9OU5jZxwZqHsRjC6KYZtdfeNEnFpk0JUP/4GlGx/lRQ/dwoJ92zCKeQC6u7sPvbPRxqrALKya/Ba370SQx46qqBAhBGZ9+MCChxrNKYQWmDUajUaj0Wg008pEkT8tMGs0x5982iGeDLO5P4NZKfJX9AXfrSthv2425754DjvXDjPamztCSxqN5lQmO6ricHxXiZqlfI5Iou64b7d/vMgZ4z4fWZ9hQdYn7dfiKsZ/tY3cfT0HrGOYBgTqHuF9s9u4JLWODlkkXxGYI0KoiAxL1ARmw0UISV1I3WPYXu0eY3fTJIFZpGidncD3K3FdViO252NKk7HWmvu5ITPGztmLSdU1ABC4DsJ12Lt376F3NtpQzWCeTOTi85gReT2mV3M/m/Uh/LQWmDWnLlpg1mg0Go1Go9FMK2bFNOQGWmDWaI4nvhfglX0icYsZDVEMqYai96YdRk3John1nHXlbFa9cgF1zZEjtKbRaE5lChkVDeG5yklcyuWIxI+/wDwofdqLPq1eiF88UmD5nloshJ86eESEYRnIyj1CGDgrs4lRwhSkEpijBvs5mCXSVGKt4XpcPNjPW++pOaOj/lD1c505xk9C/8vm1jHVh1A9S/qVs3gsvKu6XLyQIxev549XvAaA0WQziR3PMDQ0eOidjTZUHcwAmGofjLY56nu+lnlv1ocIdESG5hTmqARmIcS1QoitQogdQoiPHGT+h4UQm4QQzwgh/iiEmHPsu6rRaDQajeZw6Ou15lTFrhb5m+aOaDR/4rhlJSjbYQshwKyIM3srw7IXt9cRSdicc80cQhFr2vr554C+ZmumE89xcIqFymclapZyWSKJ+HHf9pAM6CrVLviLUjUHs5k8MKcYpjqY8Yo0u+O4wmCw4mCOGRMOZiUw4wYEhtov1xG8ZHyYzvFatnNImFy6bguLsv0kjDRPeQ/x4GmrkS3vxA9FCXlq2RIqRkQEAdFSnnw0wVXb1jGWbCEwTBw7RKqvF98/MDcamBKRofZDLWc0tUG0CQqTBWYVkXHIooEazUnOEQVmIYQJfBV4MXAa8AYhxGn7LfYUsFJKeQZwI/C5Y91RjUaj0Wg0h0ZfrzWnMrrIn0ZzYqgKzBGTkhtUHcx7x8skwhZdyQjO3r2Mfue7eKOj09nVP2n0NVsz3RQyqepn33HxPRe3XDruDua875NF0lE8+PVeHmIkk2lZtXlOgSZXua97mHAwGzWB2Q3wHZ/AVAJzLu8SiU4dkWEbNi3j23jf9gcxkIQrI6kG6vdgBhbhisAcddSMaKmAISXzhw1EYBD2XMJOid6OOYhcmlQqdUCfnZKHDCehVJsnKq5qYvUQb4H8pIiMuhDSDZDOIcRqjeYk52gczOcDO6SU3VJKB/g58PLJC0gp75VSFipfHwNmHttuajQajUajOQL6eq05ZbGqDmYtMGs0xxO3NOFgNil7fjWDuSfjMK8ljhCC/OOPM/T5zxPk89PZ1T919DVbM60U0unqZ891KOVU5vrxzmAeKqtM47ZSMGX6hGs3yKu4DkfALR0WWzwVa2FYZs3B7OZpdlMA9KLOaXHTUBEZtoH0JW7BrUZkuI4gGo1O2Z4pLdKNWwl85Zi+ovky2jPz8A2PqFtXdTC3owoexgtZAGaNxYgWO7B9n0ipwGhDC2apwMjIyJT2pZR8628eYMfGMnglcEsANM17iJC9DaspCrEWyNde5Blxu3IMPDSaU5GjEZhnAPsmfe+pTDsU7wRuO9gMIcR7hBCrhRCrh4eHD7aIRqPRaDSa54a+XmtOWbSDWaM5MdQiMkzKXoBRicjYNVaiM6kcfuXNWzDiceyZWs88juhrtuZ5kRsfY/PD9z/n9Sc7mD3HoZRXAnM4kXi+XTss/WUlILc4+13vKxlZfs4hb0LRhNX1Bk8FSrg17UkCs5OnueJg7hXqHJawTBzHAdMAN8DNe1UHcxBYxBNTBWYjCLBmrkUGShJb0bKYl2x5D+2FOVhBqOpgnmfP5NUPP8Gb79oBgDBiiCCEgSRaKjCWbEF4Ltlsdkr7TlGJxL37KtbokupvWGygbfZPEJZxgIPZiKtYogmRXaM51TimRf6EEG8GVgKfP9h8KeU3pZQrpZQrW1tbj+WmNRqNRqPRHCX6eq052bAM9QDmaYFZozmuuBX34JqeFF+7b2c1IqM/69LVoASY0ubNhJctRRi6HvzJgL5ma/bH9zy+8b63cuv/fJ7M8NCRVzgIhUqkQzgWn+JgjsaPr8A86CjxtNFRwvDG+opzueyzbWCQl59hc3eHTckUdBQDpKHOUaZtTonIaHZU//cZqp24adYczF6AmynhmEUAfN8iVjdVYJbFAFfCwGk/BaCr+G2Wvug/6Yx2ITEwA4lpWTRu6aB9xKIxU3EoizhCGgSRGFbgk2lsRXguucrxmyA7ViIQ4MhKpnVFYCY3BIl29bm+C7L91XUmHMy+Fpg1pyhHc9fQC8ya9H1mZdoUhBAvAv4ZeJmUUpe+1Gg0Go3mxKKv15pTlokif94hshc1Gs2xYcLB/Lk/bgNqRf4CYdKZjCClpLR1K5Gly6atj38m6Gu25jkztGsnVF7IDnRvf05tjPX1AFDX0orvupRyyoF7vCMyBioO5mRFYM7ZFedyweW3vaPsi5tEfIkdwIt7XKTwCYVCiMlF/pwcnc4wUSHZbqppdSGbUqlULfJXSuUpWyrmJwhM4onYlH5YgY0TCHZXtNwgyBNO9tMxLwnCQACxuga8sacJl0oIvyIgi0hVYAbIxesphsJ8w7H495191fY/uneAz7+yEdesHM+JHObcICTa1OfkTChnquKzWY3ImCowSy8g+0AP0p8aK6LRnGwcjcD8JLBICDFPCBECbgB+N3kBIcTZwDdQF77n9gpNo9FoNBrN80FfrzWnLGZlBKmOyNBoji8TAnOlbhV2ZXi5Lww6G6J4w8PIQoHQ/HnT1cU/F/Q1W/OcGdrdXf082L3jWa/ft20zT/7+NzTNmEW0rh7PKVM+AREZq9N5essOUQmhiljsVxzI6VyZB4s+s/MB543laDYNujMeUnhKYDZEzcHsFrClz7lRgSdASEkyFqVYLCIsgfQCvFx5ioM5Go0QjlnVvtgYOBJGPTGlj8m2KFIYSODyN74DAKs8AjKPRCJEBAMLaaq2LLfMt173Ie61E3x17xCPp3IMll1+V87j2IJdzR2q4VIafA/yIxCvCMz1lVSctBL7JxzM47/ahtNXc0TnHu4lfesu8o8PHINfQaM5fhxRYJZSesAHgDuAzcAvpZQbhRD/JoR4WWWxzwMJ4FdCiHVCiN8dojmNRqPRaDTHAX291pzK2NUif9PcEY3mT5wJgdkV6j9bY8QEwK84mN0eJXSEdP7ycUVfszXPh+E93YTjcdrmLnhOAvODP/0BiYZGbvi3z2HZNp7j4hSVGBuOxo6w9rNnqOzyv3sGuX7tdr7dM0JrUBN1DVMJzMPDeW54Msenny7S5EQwYjYZH6Twsa0QCDHJwaxqX16QUIJsNIB4NEoQBPhIpBsgyx6uqZbzA4twOEy0LlTdbn1dhGIgSPtTJTE7bIIQBIZNaqQVYdQjvHEISiAkQggEJtJSAvMN6zfi2bV2X/7UDs58ZGNVQP/j7IqIXEwp9zKSsVwlNiNZGcRQEZhF2Ky2k/3j3urnoKjO20FBR2doTm6sIy8CUspbgVv3m/avkz6/6Bj3S6PRaDQazbNEX681pyq6yJ9Gc2I40MGsvl+7opPTu+pxn1FChy7wd/zR12zNc2Vozy5a58yjqXMm2x57CCmV8HnU6+/uZvllLyKaqCONwb5sjgVFJcbakcgx7+87N+zmyUy++r01AIm63ocqERf1v9nFpZNXquxOYFQEZkNAgNpXR7l7L0zGYShFLIBYTAnjQUwgyz4xTLyJIn9+RWBO2KQGVbsx22S4IGixpt532BWR1zdDPHFHL8JsRXojSOkhlXwNvsCvi4OwCHXv4cXi16xZeSWh1lZ6SkoEdiq1JR7rqmd772wWlVIwuBGA++4yueLCAg3Jynk2rep9Tv4NzeZJv0Nlsr5F0pzs6MoNGo1Go9FoNJppZaLIn6+fnjQnGT09Paxfv366u3HMqDqYJyb4PsIw+L83n0csZOFUHMz2jBnT00GNRnNECukUdU0ttC9YSCmfIz00eNTryiDAKRUJx5WLdlPZxymX2Z3OIISBFQof8/5uL5SmfF9RSROXAprrDr69lrecVlnGxzJthDGhOAOuEsPPaajHksrBHI2qIn5OUi1nYeCbDqAiMkzTJFJxPANEwyY+gvJ+tx1mJRPaM6MgDAyzHRmMUzILIHzKZgGkAYZBKDmLwB9m+danuOHhW2m2p/o3L1x7H1IINiQWqQzm/nUAjHjzKKQdVezPsGHbHfD0z6d2xKt1TDrqvC2L3kGPlUZzsqAFZo1Go9FoNBrNtKIdzJqTlW9/+9v8+te/nu5uHDPcko9pG8gJR1zgY1pKdJFSUt68Gau1Fcc4qoGuGo1mGnBLJexIhPb5iwAY2Lnt6Nctl0BKQpUojGQ0guV79GWy2JHIUTmhB8su79u4m5R7cMHz5sFxftg7Uv1eZ5lT5p+fr/iXLcG8rrbq9NWNarlxkcNuVwK4FD6mYdeKNUiJdIp40iYeibOsBAkpqg7mtXtqLwSl7aIG7Rskk0lmLG4EYNayRuKVeKCsP3V/rZASpT1LtWeGVwAWruljBpCODiOkWjcUmwuygAQcN8/nF3bRHqqdO+f07ARgfWihymDuW0fWmIUrY+TTZTAMaJwL2++Em94LUtL+t+cAU+Mw/Jz67GcdpKsL/WlOXrTArNFoNBqNRqOZViYEZk8LzJqTiCAIDvr5VMYt+9Uh4ADC9zArWaL5Bx4ge9fd/ObKt7HqM3+kL1Wcrm5qNJrD4BSLhKIxmmeoDN/04NEXfysXlPt3QmA2bRvLcxnO5QkdZTzGLwbGuGkoxf/uPbD25FDZ5cNb9/HxHb3kPZ9yENBbcnhdRyOfXTyTzrDNirwPSLAF9S0N1XW7E0qeKuBUp0nhYRpW1cEsA8njT3fwjcFf4vkGH+12+Wg+VHUwr938NAHqfC1sH8uK8YlPfIJoNMqZV87idf98Hte8ZwWhiq7sI3ACMDy174ZVsVfH61QbRhzDngtA2AuRD6WYyKxYdtH5ALj15/DrMx+gKyhzx8ol1b53Dewlkc+wOTQPHv8mbP0DY6GzAciNVbbTvLB28Irj2O1x7BkJgklu5SCnjkdp2zi9H3uY4sbRw/4+Gs10oQVmjUaj0Wg0Gs20MlHkzwu0wKw5eRgcrA07L1REmVMdt+xjhtQjYMQvMWt8C4apBOf8E0+wrmMpXyu1c87sRp5FpKtGozlBBIGP55SxwxGsUAg7HKGYzRz1+k4lazlccfy6po3pe8hyCTsSPao2whWxd006P2V63vf5f9t6KPgBxUBy52iGf9vRRwBc2ljHW2e08NRFpxOacOGGDMxYzfE7EKncCwgf11Wu3d7GJB9ZeDpjvl85AJL125XrOZd2WTzkcl44XHUwAzxgb1aLhlxMc2rRwtZZdYSjFrZQfbCFJGRAbFRFchghJfyKRJxtLU8yEu3FMFsBSJhRXMPBr+RHn/WSMwCbaCV6qHe8l/aQxQ2OxVt+9VWswCeZGaO/cQHIgB35dp4qXwZAdrwSG9I0v9a5n90AT/0EI2ZNEZgnHMyysp3StrGD/zAazTSjBWaNRqPRaDQazbTSaJuYAjbktGNSc/KQSqWqn/+kBGbbACl5c89PCfslitkMbn8/qZ/9nD+suJq2ujBffdM5dCaPTmzSaDQnDrekBNBQxbEbqat7VgLz/g5m17SwfA9RLh91gb9hR4mfW/JTs5Xfs2EPt42keceMFjrDNr8cGON7laiMVQ2J6nKB5yMQiIiFEbXZHwmUy2UQsK2jC4BtlRxn6UtEReDNDRWQjo+RCBGZ1Pcd5gDfC9+LjPgYxsH3aaLAadSsxyw1EJiq/exdWwEoWjb3LPox47EBhNms1onaNNc1EggomAUe6H4Uw55F4O5CBIK+dB+BJ1l883baR/sBaMiM0RtpIPX2x/jt3sXs3HST6vtY5dhZoVqn9j0Ov30/hlkmKEx2MNfiMgBEaGrkiEZzsqAFZo1Go9FoNBrNtNJoW1zX2sBP+8e0i1lz0lAul6uf8/n8YZY8dXDLHsI2iPlFokFt//r/+V8ICgW2xTtYtaCZiK0FDI3mZMQpVQTiits4mqinlMse/frFqQKzY1oYUmIV81XR+khMCMwpzyfj+dXpe0tlWkMW/7ZwBi9ra+DesSwB8LXT5jAjUhNSpRdgIDAbwhiTHMz1bu36Xy6XEabE9tW28l7F9RxIhKgIzP1qX8yErYr4RSKYhoXpxYjHk7TNiB/gYJ7AZkJgjmKXG/Ft1ZYsqWNZNAQhI4SNjTBbVP9MixUdy7F8m+vfdj1rdj6DYc9HBhlmjrbSl+kjny7ju7nqdubt3U7WDvGdNZsAEIY6xoVMJQZk0dUH9M0o9RIUXaSUeGMl5WY2akNKtMCsOVnRArNGo9FoNBqNZtq5qCFB2vMZPUTRII3mRPOnKTD7YAmSXnrK9MJTT5EOxRgixKIWnz17v4XjjByiFY1GM124JeV8nXAbR+vrn1tERkVMdmzlIA7lM1XR+kgMOzVH7d5i7TyZ8nyubUliGYJXtDVWpy+NT3UR+66PiUGoJYoRqzmYJ8SpnChSLpfxpVMVmLPlyr3BpJfQpRG1LzsHPg/ABz7wAVZ2voxZwSr+5u8/iJ8eJOgfI3PrrbgDU3OqbVR7YRHCKjfgW6otYar9KZk2HfEOGq1mDLOJC70ky8P1JON1WEGI/vwA/YMjGJZyWF+76dVk76+nkHGQfu38uqR7A4lCnvu8CVFY/fWcijA/5yL4yL5ax1oWY+R3EhQ9cg/0MPC5JyGQWC2130ZYOr9Ic3KiBWaNRqPRaDQazbTTbCsXkxaYNScLf6oCc2AaJF0lgDzU8gJe9Lq/QBaLjH/onwCYVbebHTv+E8/709hnjeZPCaeooqSqERmJZxmRMeFgrmQW5+INAMQz49jho4vIGHE82kLqmr2npJy4UkpSrk+DpQTUs+qizImEsAQsiIWnrC9dFZERboljTCo6Om4r4fRJayeFQhHXyuEbSrLq89S9ge8HlF217XIlZiLjrSOf38H6uwbpXjNC+5wGDMOg3Lcbv2eQ3g//Hb1/9/dT+mChRPKwEcEqNeKHlHNZWKpNxzRpibZQFyQBqLMTGK5LfawOA4Pd47spjDkIEa7sv4Pd30wh7eC7W4klm1j2gsuxjHqSmVFGK05vGRSVM9mZVDg2Ul/73HEGRnEvBJB7tL/W30kCs3T/NIrOav700AKzRqPRaDQajWbaqQrMjhaYNScHpVIJwzAQQvxJCcwDmXGuGrkXgGfqljGvIjBtS3QCMCO2BdOME43Omq5uajSaQ+CWlMBshysRGXV1lLLPIiKjMCFQK4E5U99QnXe0Rf6GXY+V9XEANueUIJv3A1wpaahcy4UQ/M3cdv6iq4WQsZ/sVInCsJqiYCpRuScqeOtuJVb7IuCuu+4k07gRx1IO561S3RuUMg5SqvbcVGX5cIah4TtYe8ceAFpmJpBS4osyodZZhBYswBsentIFC6fy18QqNyHtSg0IS71YLJsWC3adjzGssqNdK4R0HEJhtX8P7XmYqFNP1ZgsVXvZsRyBu5tF51/CSz7wdzTOvJZoMUfGCiGMEOADLt7+IvGLPwdv/jW0LsUqb1T7laq95LRaauK/Fpg1JytaYNZoNBqNRqPRTDvNIe1g1pxclMtlotEora2t7Nq1a7q7c0xwSj4DA1sA8CN1/OQ9F+OnxgF4Op2nI57Gyd1LIr4YIfSjokZzsjDau4/utU/ilKY6mKN19ZQKeQLfP9zqVWoZzGr9VKKhOu9oMpidIGDE8VgQC9Nkm/zX7gGeSOUYrzh0Gyflt7+hs5nPLJ55QBvCUwKz2RhGCCUwzyxKGlwoVZzFg4ODanumujfwK6kQxUp2ccgo4mWVAOuHMhTzPdX2O+YnCTIZZCjAijQQO28lwX4ivFHJoO/N9GL2OFQSM6oRGWXDoPmZZbX9tiyk42CF1HlxODOK7YdxLLWirAjMg7tHAEnzTBWd0ThjMfFCjkIsgTDnATB3eQzP3e/3uuC9sPBF0LqEkLH5gGOmHcyaUwF916DRaDQajUajmXYmHMwjWmDWnCSUy2XC4TArVqxg3759pFKp6e7S88Yt+4zGFtPw5n/mHf/+GS6c34yfVsPrnxnMMaduO8XibhKJpdPcU41GM5nvf/h93PTZT5KvnIcmMpgjiXqQklI+d5i1a5SLBexwBMNQQnAmFMGxVAG+UOTwERnrMgVm3/8MrpScl4zz4zPmU2cafG7XAKnKtbvROnIBOlHRR60GtT2zUcVMrDN3c2P40SnLTjiYvYrAPFEcryu6E9Mr4JtFpOFRLKqM5XOumc2sZU24vb0EYbCjjaRtDy+bQcpJ+c2eEpLb0wLzV7dTf6OSxqRVQgQerjFVbHdMuyIwq/2Ll2xCfhjXdAG76mDe+phyUUcSyvlc35KkPmiiGIkjwkuRSNxIGt85hEjcuhRTZDHjU++FzMSkIonO0b1M0GhONFpg1mg0Go1Go9FMO422iUBHZGhOHiYE5rlz5wIwNDT0/Br0K4WxPAcmCR0niiCQ+G6AK6Bp5ixaZs9V3cqkGUy2MVIIMz+5G8OI0tn5qhPeP41Gc3DG+3urn7vXPglMcjBXhMxS7uhiMkq5LOF4vPq9GEA+pto4UkTGg+O1bVzaWMc59XHeNbOVh1I5rlq9DYDGysviw2EEgkBIhK3kqKbXLgbgN7NjbJo51fHsWDYLRl2uyKlle7MqkqMzvouI5Vezk8tl5XhumVWn1uvrRUbASjTz4303ITwfOSlXP4Fq57JBte3IZhNBCN/OE/IK+KF5U/thmEjXIZ5UQu8nft5MZyqMZzhgKIHZNUo42Z8DVI9xfXOEhNOJFAKvIcm6RWn+NfRRcjI7RfCu0rwQ4q00dv2RuhfWjoUIm0SWNgHawaw5edECs0aj0Wg0Go1m2jGFoNE2dUSG5qShVCoRiURIJlWRp3Q6/dwbc0vwP+fAz94I/9EKD33xGPXy6Nn3kX8BwAHqI3Z1epDJ8PTM0wG4dMlMLn3BkyST55zw/mk0moPz+M2/qn7eufoxAEIVMTgUU0LmRPG/I5EdGaKupbX6vRgErDnjIoab2ulctOSw61qVOIt/W9hFxFRS0umJqaJ0g314B7OUEkMKgkk6dHh+A99pf4pbFy/gzoVnsmHWQiakV8e0iGFQX1RT9hYrDub4bkJGgB/KUFe3AscdBALCUdWw07cPTBg2XAqVGoPl1BgAQeAyO+zw7+VVXDmgYjCEC5YVx7dzNGR2ICwlMJ/9ug7VnjDxU2la220sWzDWtAwRWASWR6IxQUCZQPgglXAdjk04mKPESkoQNjqb2TBfjRgpWXn8gwnFhgGLriYy/EPqXthVmxwyaXnb6dgzEtrBrDlp0QKzRqPRaDQajeakoMW2GXLc6e6GRgPUHMyJRALDMMhkMs+9sa23QnovbP1D5fvtx6aTR4l0HMZv/yMArpDUR2sCs5/O8NSMBSRDac6YswLTPLpCXxqN5vhTSKfY9MA9nP3ilzLztOXV6RMRGeFKsb6eTespF45cjDQ9NEiytb36vRQEPLX8Qr7/ug/SsXT5YdaEgq8E0XfOrAnUi+NTYzUarcM7mEulEgYCaYvqtHHX4yennV/9/tD85WQiFeHcskkIg2heiaqDjoMhfBoi44SFgR/KUl9/Bn6QYv6L/4VwTJ3bcgmVydwbZKsCc++Acln7vhLj53ttOM6ImumBacbJt6ynse3+al/a2htVPwwTb3iY7lXnUz+4kfGGxWBEkHZAPFlPwRzElLV9j1QczMnWKImSEsdzMxvxTfXZMUtTCv2lXI9i5fiy8EVQSiNGn6nOF5XsZ2Eb2sGsOWnRArNGo9FoNBqN5qTgjLooj6fyeMGJjw/QaPZnQmA2DIO6urrn52DecTdEm+DKj6vvbcsOv/zzZP19PTzymx3V70GxiG8qlUUGHnV+bai4n8mwKdbFosZukskzj2u/NBrNs2PLw/cjg4AzrryWzkUqGz3R3FLNUA7FlMB8/4+/y42f+tgB63+3Z5jv9gwDEPg+mZFhkm1KYPalpBxI4hU3shMcXrgsBAFhQ2CKmjg8Lxqufl4YC1frKRyK1HgKgagKpgAPD2fIhKYK1X0NLapPlk2dYUBOjW4a8X2idgnLAluaeHaOaERFSYTqhgnH1PZ3NvwAgLu3DZGz1Uuz3sHtar99JcQbfgSzoLKrjRIYRgQ/nIHXbKr2IxJTBV6dUO3YhMtpPCuGFCGwfEKRKEL6WEHtWITjNQdzW9onmff5v7bavY1jFvEm5TAvfWgDr1m3g6IfcF1hCY8kz0R850XV+aKS/SxCJoEWmDUnKVpg1mg0Go1Go9GcFFzTkmTc83kifaALa322wE/7Rin5+sFKc2IolUqEw0owqK+vf34O5lIa6jrhBR+GhjngHt1w9ue0qbzLAz/fxlN37iWovKyZLDD//eofIa+/EndQZUqvcyIMUc+ixl7i8YXHrV8ajebZ07NlIw3tnbTMmkNrJTd99mkrqvNDFQczwMCObQes/9HtvXx0ey++lGRHR5BBQH3FwTxxPU1WCvM5R8iGL/gBMWOqhGQbAlPABck4D12wDMsQh1hbkR0cRyCwojaBH7Bn4yiPrVfnojc9dgenZUYBJTCH8zNwTYvGmE0+raIxxmVAxC6STV9LOIgR2C5tbS+edDymRnSkRi8kl3kFAEODuwHw/QIAhh/GLqr7DeEJDL8mjlumEp57+74MQNBRO2cbgUdgWAgZoixKZCgggol4L5PxhMOb7387/bl+InGbiCt54YYiplfL8XesEp7rs6dY5uxHNgKwJlOgu1hmTd7l3ad9csp+VAVm29ARGZqTFi0wazQajUaj0WhOClYm1YPyjkLpgHl3jGT48NZ9HOHZVaM5JjiOQ7lcJlEpoJVMJkmlUs+jwTyEKoW17Bh4x09g/uMPNlc/p4eUkBIUingVgdn0yyAl3kA/rh/w17OvB+CMLokQh89P1Wg0JxbPcapu2IXnXcjKl76KF77lXdX5E8X+DkbeqwmRa9J5UkOqEN5EREah4liunxCYjzB6qOAHxMyahOQO5un5yINsmjuXG886updTmaEUAGY0xDP39nDL/z7Nxr4MIdcjUS7yhqFuLrNhqG0m0luEFIL5LfFqXdSMIYnZBfKZiystholGZxF23wmAFfGmFM8r+WHGfZWjPDSg9t/zKg5mL4JZzFWX9Yu1G4xLlnfz6o/OIFtaDcJHTjo3CukSGBaGDBFYHk+lnsGQ6lgGwmJvR4Gd6W6+uu6rALzo7adxzaWzMb2BahuOWcR3A345MEZ/uRYNlqrUoRgNNcJ1X6hON74wB8Z3Y+iIDM1JjBaYNRqNRqPRaDQnBQ2V7MaUd6A7p7/s0BqyCBn69lVz/JlwK08U+Js5cyapVIrR0dHn1uAUgTmiiv4dBzIjRXY/M8L8s1RG6kiPEk+CYqHqYDYr8RhBPs9AWvVjZfNGTu8MHZc+aTSa547nOFghlStshyO88M3vIFafrM6f7GDen6352nnmI9t6+MXOvQCsFzZ9JYdbhlXsT0NFYC4fISKjGEwVmEvbUwDIZ0axj+Lt75o1a1j98BNqX2Ih8mmHnmaLR+eGqC+WEIBlWbx64WzSQrB1ptrvZTOS+JXmA6AuPI4w1Hks3bkbKSWFlFpWygKuWztPl7wIeUuJ8KNDKirErbwsNPwwsjiOCCXwTJPt/bW+trUk8ViNEGCYLshabj34BKaNKUNIy8e1JCJQ9y2/vmIQ31AC9xMDal+XXNDBpSs7Md3aBsqWisjw9tP0RyYVOh46462TNpmHvY8hQibS1Q5mzcmJvkPXaDQajUaj0ZwURE2DsCFIHeThqa/s0hW2D7KWRnPsmchbnhCYlyxZAsDmzZsPuc5hmSwwW9HjFpHRs3UcgJUvmYswYHjXGAByUkRG3lb/j/x8nv5x5XC+dPaDRKNdx6VPGo3mueO5DqZ96Jc/lj31uug5KkpCSsmvB9X54LOLZ7K9UObOPuXg3erBdWu380/bVCG8OZUcZfcoIjKikwRmYSnVVx4hukpKSSAlGzduJFwRakcGCxiG4MaLlDvbNdULZtM0ubhBTds4W+33vLoIoYQSlJtdyaw565FBmKF5v8c1w/RtSzG4U7mAS6V9FIs1IbfkhynYKt85yKrzXXlwBDyQTzxKkBtANMwmnUxiGjVxV0ZhdOAhrFIjppBIavEZInCRwsKSYaTl41kSs3IMrCDEQJPqa3lS1n2LbWF6AwRCvRBwzCKe67N6Z/eUY3VT5TcDuHmo9lmYFgxvqURkBDDWDUEA+RG4+xPge2g0040WmDUajUaj0Wg0Jw1JyyTtHfigpARm7bDUnBgmBOb6+noAGhsbmTVrFmvWrCE4gsvvoDg5CCnRBDt63CIyereOE60P0TwjRjg/wuDtDwFTM5i/c9lb1LR8nt4BJUA3RseIRGYclz5pNJrnju+4B4jIh+PbH3oX9/3w23xu8y6+0zvCq9sbeUtXM19eNpuwowTPm1KlKbEMFzeqc9NRRWRMGkUkKmKz3N+Gux+f3NlH131PE43HCUsl1A7sy9G7PUUppETqc7uVKGyaJklbOaqH69XfGZt/Qbz4FAGSlrJkyGgEIIiMIt0msmMlfFe5lFeveS07995Y3XbJi1CwwgQIGFUjO8qj40Q2CIKNj6mFGmaSamgAIfnhptfx/x74OH7IJ11YS3RsKYYRUAhCfP/KEI8tEeRCtfO3Zzjkoh4C8IpPYAY2Y/VK5C+4hdqxEgLTHcC3u5CEKRvj5NMZtqUyhMtFFlWyo28fUaNnViSifKtnpHbsm+bTP7QekduLdDzkl8+GP34SbvtHeOi/ofu+w/4GGs2JQAvMGo1Go9FoNJqThgbLOmhERl/J0Q5mzQljIiJjQmAGOP/88xkfH2fPnj1H35BTgPxoxcFcGcpuR49LRIaUkp4t48xc0ojM54mWRskVlHgTFIp4lnLxxZvr1LRcnt4hJaQ3RlJEItrBrNGcbHiug2da/O+ewSNGWDR2dpEfH2PNH27mnscf4yUtSb6ybDZCCF7V3sgbGpUIuysQXJCMV9drOIoif9/uGeaRVG5KRMZE1rH0Dt+vr+9T0RR7i2Xa6prVfknYMV7AtQSv2eJwXp/KRe7p6SFaEbHLIYMmwyB2y4dImv2UJLSUJQOyTbURyuDll+OUPAK3lkU9Pvzz6ueQWQQhKFkhQmWbbMnFzWUw8rVIj3JjO5tOP409u8/i/p5LGCk1k7dSuMEIkewchBng9FzILP8LfOVlcW49p3YNcGWBLbMylG0D39uNFdh4ljouJb+EF9RemM+yUsyIz0Aaccqih4FdWxhPNrN0+2ryfV+Zcsw+tWgG+0oO+TolyH+2IcbV7jaym38AUhBQD49/A/Lq2BK4aDTTjRaYNRqNRqPRaDQnDQ22SXq/iIys55P1A7oi2sGsOTGk02kSiQSWVRsWvWjRIoQQ7N69++gb+sPfwefnQ2FkUkRGBCY5244V4wMFChmHmUsaCbJZIqURShEl5qgMZiXAtHRW3H/5PFuHVhO1ikStErHonGPeJ41G8/zwXYecMPlUdz93Vdyth+LtX/w6f/mNHwFg5DJc1VKPEDUhdaYIsMMRtl92Fr89ZxE3nb2Qe89bgl1ZZrKDeWehxPs27mZzrsg/bevhX7b3AkyJyJBORVg+QkTGBLt8QbwSEeEDT3UqYfsDf7GC2S2LATjzzDMxhCBcWacddT8QMgqUfckLh1xEfiYApVIXfm4RTtGfIjBDTdSNWMpNXDJDRIMc24dyeIUM5lDtuGxvsSnE47iT2tg8sg+AcHYWkTrVhsCgrtyEbKxtqeCmkAYUogbIElYw9UV4wVPnei/wGCkM0VU3g8CMUTILbO/bjWeHiBWHKOfXExIC25W8/w8pmrsLLI5F+Oy1rXR+7EKeDqlr0d0t6jfyZJcaCbPrAbWh/AgazXSjBWaNRqPRaDQazUlD0jIPcDA/mlKFyhbGwgdbRaM55lxwwQW87GUvmzItEonQ2dnJrl27jq4RKeGZX9S+T4nIOHYOZrfss/6+Hnq2qLiLGUsa8bNZIqUxnHCSno99kty99+FZYVwkM9obwLLwc3k2jNq0xYaZHX0T8fjCY9YnjUZzbPBcl65EnPaQxS8rkTaHQhiGKgBomCTyGVbWx6fMd4oFQrFY1YW8qiHBskSUkDEhMNeE4rc8s4ubhlJc/uRWvtdbEy+nOJgdda0+UkRGpNL+rkBUM5g3z7B56LQoFxQMliaixBJRFotruOaaawAIV0TvuK/iKNoWtuFJaHDhvJ7ZALjEcF2fh3c/ipTRKdsseSFc32Jhw27efWGZghUm6gRs3rEHr5TDqgjMT65cyfq6PNFCgbmT8pC3VUaxhLOzWPqCrew562sARN16okakupwVqHN5OBpDBmWsYOqL8ImYjJHiCL70mZHoRBoximGfbekBAJpSIxhBlg/euIeWrE9zLuD2b27g3PoojxWLGDELx1L3P3tnqZeGI85/4Nhn1zaUGzzsb6DRnAi0wKzRaDQajUajOWlQAvPUDObv9Y4wI2xzRVP9IdbSaI4tHR0dLF68+KDTR0dHj9zA6E747FyQk16WTDiY7WNb5O/x33fzwM+38eAvt1PXHCHZGsXPZIiWVD8HbrmH7B134JsRHCGZ1RzHiMd5LGuwK9PJFc7jzO54+zHrj0ajOXb4joMdCnFNS7L6snWCXw6M8S/be6ZME4ZBUFdPsphjwX4vZZ1CgVA0dsA2QpVIigkHc3/ZYWexzIXJOI2V+IwJTGrO3wkH84TQDOqF186nhqrfU65HqdLu7pZOhoQSZ7vbbNrSPu93VX8icYty3q86riNMFZjnvPNfaYs7U/riiRjlssPe0R4Kk2I69mRm8mj/eZT8MM2RcS5r/j5mwibiSi69M4NvF6oCc/fCBQA0jKdw3dq+7ZExDCeO5dYTqwuRiioxeJafwC3V7kVmVBzZ7S0zD3Awd6UXsXHDVnVM8ypjemlyJoERx7Eteiv72DaiivmVrBzX7K0dy7NEiDHXZ2ehQH9Ord8byapjTgRn0btqByNXO+YazXShBWaNRqPRaDQazUlD40EiMroLZS5oSGAb4hBraTQnhlAohOseRdblg1+EUmq/lSdFZHglNaTZP7Cg5bMlP64KdyGVexkgyGYJOUqIcEMqc9kzI5QFzGqKkWmv59+cTtpiQ1xeeAKzoeF590Oj0Rx7PNfFtG1mR0Jk/YDcpBE+n+3u59s9I1h/+wle//H/rE534/Us27KW737oXex5Zl11ulMsEI5OdfoCNQdzJVP5qYxy3X5sQRebLlnOoklCdWbS9ieE5aBQO4/d95Mt3P6NDYz2KTF8Q18/y/p2Ey0XGU0kwTPIUoRSjnfelWH+THV+iiRsSnm3muscmhCYvQLEmsGO0LxiO3HzFupn/wN7z/s0kiaMwOTshnPxosPVPmwcXUrELFPyItT5GYrFHchomKjr0t+6htTM+yAtGFu8HFMqSWz51c0Mz6q5j/dmZmE5SQCMkEPaVufTGbIOWa6rLpe0lYDc0NoGuFh+TWB+2aYPsPmHShSfEIjPTM7EiHyIGemXMNbYgumVaUrnEBIifoxZW/PV9TsD1bcNqUE8qY5xMJiuzt/iT4pa+jN3MN8ylOKft/UceUHNcUULzBqNRqPRaDSak4akZZH1A7xJWZDjnneAi0qjmQ5CoRCO4xAcodgWuQFoXQav/g50nlVZeSIiI6YE5s8vgO9cpaI0ngdy0v+VjnnKWedns9iuEiqMpWeqaVYER0CDPcC3rjibcRnhvSt+QGzcw6yrO7BhjUYzrUgp8VwHKxSis1Lktr+sXnC5gWTQUZ/7mzqYedry6nrlkBKE00OD3PnN/+UXn/gI+zY+Q7lYPLiDeb8M5tXpArYQnJ6IIoTgJ2fM56pmdW4ZnzTCKJgQmIu1l279O5QAWkgpYfXu732HF25fx/l9KlqoyQkYNbI0DD2OALoWNQAQiYcIfIlbVm2GKwJzzM1AvSpAGuko0mh/Hdccpdi4jVTvTkq5EaJBgrOu/FK1D9vGFxCxSpT8MOwLEFaZcCRKxHfYPO+3SKOEWYBIfSsCWMQcLn7tZ1gTV4Ly3Po9dKfnINwIpbo9+GaOnHDwzCKxcj1t2blIw+Hrq/6aWNc+kAFdbSrDPuKorOS2zDwAhG+SdbL05lSGdf4RwQdvzTDcsphNi86iPrMDAwg7BjF36iit+sqhfnpc5UG3RdtoXl/gvv5fcHffjxlNF+EVX4PGecrBnJ0qMhezDlJK+nek8I9QiPFU510bd/Od3hFKR5kHrjk+aIFZo9FoNBqNRnPSUH2Irjw4e4Ek4wU02tbhVtNoTgihkHK4ed4RnMeFMSWKrHgNRJRogV0Rduxafid9a2HjTc+rT/l0bdh4yywlFAeZLFZFYBaLTld9NiP4oRx7tr6GLfl5LGvaxtzkPqxSDGHp/18azclG4HsgJZYdoqNybRyoCMy7i2Umoo8fTeXYkK25We1RFZdw1jXXkxkepGfzBp6+6zaVwXzQiIxaBvPWfIkf9Y3wgsYEkUre8uxomP83rwNgygijCQezn3bo/+wTlHelKeVV/7LjpSmjPVbGVP+bXElJuBiBGnkRTyoxPJJQ56BSTq1jVXYukk1QNs9VjVhRsnGTVFSd80a6N5IeuA/h1V5An3P2T+nJdtJg5Sl5YeZ1tVb6M0jUKzPi1LFpxxL6u7owOufiiQATi4/8+hlEvgWA0xp2kHHq6QsX2LPq4/SP/IqiZ1GOjpAvh0gaHuWIyqUONcL1MQ8zos7rYVftx2U7X1ft0xU/vIoNIxuYXTeb/q0q29mJz8SzLGbs+yUAieKBv0uicvgeGdpKS7SFVV2riKYD8vEchVCW9OAAnPVGmHkewZ7H2PWpNyF7VgNKXP7uPzzEfT/Zyk1ffIrVt+4+oP0/RbYXjl19A82zRwvMGo1Go9FoNJqThjlRJeDtKaqHzwm3VKOtHcya6WdCYHYc5/ALFsfUsG6AcMUd7FfWsfYbon7Xx8Gd9FDcuwa+dMZRZ2qmR2p5zs1dKobDz2awvcpQ6/lL1TQrQuvMtbieQ3+2kwV55TQMeTrbXKM5GfEqL1pN26YzrM49Ey9fJ4S0Fttia77Ei1Zvq673xBWvIHfa2Vzxtvdw4atvAMAtl1RERuxAIdOuOJjHPZ9rV28j6wf8zZz2KcssikdYnojyb4tmVKdJJ8CxA7qNQfzxMqlbd1UdyNnRErlcLTN6SUM92y9ZTsiHEmofXv3PZ1TnR+JKgJ4QqMmra79d7GB452sqHY3wxLmN7J6pRFyvYCIDiVuqid621Uy9F9BiZ6hzy1xy9dUACNNnbEYb6S1ns2ngfH588cv4SbQSuyVNfv7kPko5gZABi+J7AdgyrooJDrgCv9BIKdoPIsC386yd+QcAwoHgRc0eu/o/BcDpm/fRkWnB9mvn+Y7sfNYNr+PM1jOxw8ph+8KBfVw8/knsoJLtPDa3urxhVLKts2WihmBHbozr5l1HkjiJrMGZV17LgpUXMtKj+sn572FN/jXcmvooex7dNOU4bnqoDxlIFp8/9ff8U2N+VL2o2JTTAvN0ogVmjUaj0Wg0Gs1Jw5zKQ8LuohLjUhW3VJN2MGtOAo5aYC6MQ6xJfW6cO3XeZAfzG34O6b3wxDdgz6Ow5vvwrSsgtQe67ztif8oFl2JG9cW0DayQehETZHNYYRPDEniReqLnnINrRemcuZaMPBsXi8XrBmn/nwbqW8468o5rNJoTju+q/9uTHcz9pQkHs5p3UWPigPXWz1hM8Ka/RBgGF7/uzSy+8BLG+noOWeQvXCnyd8tQimIQ8L/LZnN+Q+KAZe4+bwmXNNbidKTjM+ZluCe0AfsFbbj7stgVzTY7VqKnp5aJa0tJ3AcDQVlUYj6ovRyrCsw5F9fx8Sq5zlEfnNgAw733EZi1fGPfFcjAoJzbyOD2HzCQb2PT6GL2dj9Jl8hhmi6GaRCvb8bqE7TnyhA26fcbuK10Hve5C7nHnQuA4xo0uCkQgnqnQIuZJRlKs2F0Gf+1+q/4n33NXN5/OYFZYm7LUgICFmWW0FJswdw9EwTEYzUxfVF/J6Gglls9Z2w55zzzUpaHzsYpqZfnF4TqKXs9WI0SYYSYk1aRGhc8/gmKo18HYHxghLgok09cxfIZ1xFNKeG5Zc48WmbNpphJU8ikYdZ57Iy8Xm1sfI86PtkxLqn7NnXmEDOXNtLYET/gd6eYAqdw4PRTkPawukfckPvT2J9TFX2nrtFoNBqNRqM5aegK29hCsHvCwexqB7Pm5OGoBGbfhXIaohWB+fJ/hroOOO3l6vtEVIYwYcmLYf5l8OAXoJSe2k6mT/0tjEG4HswDH91Ge5VL+bq/OoPZpzXVupDNYNXVEYmrwllzfvwjCn97J7ObtrPL/2sA5vbvxEwXSP7Nq57dQdBo/gzo3bKJgZ3bOPe6V0xbH7yKwGyGbGKmQaNl0luuxEM4HuH9Ct/6UvJ/e4coBgHNk17KNs2YybbHHgIgVp88YDsx00AAj6XzNNkmr2xrPKr+BW6AK9RL4BxFwkBIgCuVg3nL/Y9Ul83uHWFo6zhQczCn02lmzZoFTp7I2i8DL6CUd8kMF5lI0o35kp5zvkDumRw9219P14IE0WgO6Vg0zTiTOcUWDMPgb1ZfRKrcgLfxq8wwL8E0XQI3jJ8LaP6KhZkpkUylGJY14dwRPhk7g+lYzM/vIG3XkyzniObKnN6yhXVDyyl4UZbFa+fWiy44gx13PIXhRVmWWobss2HZHqxwzUW9tH8mMqzuWcpmgWXDFwIgHzPJp1xAkBsJMdpksazeYe5ZZyJ2NmHKLDvOd5GpElL6ZEbSjLXXgzB529Yyf1cIYQB2cz3NlVij0Z69RJctZ3REvSTwR/fBvidI/v4vaYnvpKUrQtv7X3vgj+eV4bNzYMEV8BfPL6bpZMCt5IffN5ad5p78eaMdzBqNRqPRaDSakwZTCGZFQuwpOvSUHLYVlNDcoDNiNScBRxSYpYSRylD1CQdzKAYXfRCMyksSrzKE9+w3qb8X/GVNXG5fAQuvUp/v/jis/SF8bh48+F8HbMopeTx60w4AWkJ7Mczao503NIzZ3KQE5pzLI1uGmXn29zBNjy1jC6iLWKy45BxaPvRBEi984XM7GBrNnzBP330b9/3oO5QL+Wnrw0REhmWr885piSjrKlnLY65Hk23xgkkO5pTr86nufoApAvOMpSqHfdEFF3HWtdcfsJ06y+RnZ87nQ7Pb+OqyOVj7CdeHIii5eChhtbtPOWdtATvPT/Kz2BAjIyqnuD6I0tfbT+bnGwGqDuZMRuURs+12Ihu+BUDxqVvJ9AwimXAwSwKzzOjIbPrGfLp3ngdAzI7QYDayOLmSuc3zybvqxd0vvFez1WvFNF18zybzlV9hVE6vZhCQzwuSwRiXja/H77yJu2beRYGApeVdlCMNtBfG6Ng3yOy6XgpeHDBoLKtzeZdM8t9/6GWoSZ3jQ34Ib+g01v/+vVjRWi6/W9iEN/ZVvNIT7Gh5qjp9fMs9ZIbWIoM8qUGfQtGkySySnBGG/FawHqH19UOYIR9kifx4nuagt7r+Pr8BAKM+SvMsVVRwtGcfO9bU4pQe2mHzxOfehfBKuEGYDvcRbHs/2U9KFc0EsPOeo/qtT3ZKFYF5e6FcNShoTjxaYNZoNBqNRqPRVJFSIqWc1j60hy2GHZcXPL6Fv9+qqqdrB7PmZOCIAvPN74OvXaQ+x5oOvszS6+GiD8E1n1Hf500SeG/4Cbz5xtr3331Q/d376AHN3PvjLQx0K4Em/tPLof+Z6rzyli1EFi2uOphv/OE9JGevZtfY+Tyw0+CCec3M+cLnaX3/+xHi6MQkjebPiezoMEhJ39bNx3U7rlNmy8P3EwT+AfMmR2QArEzG2ZQrUvCDisBs8ubOZv59ocpF7i/Xzkv2JJF47hln8+Gf/Y6XffijROIHRmoAXNZUz0cXdHF589FnsntFF7ciMG/btxMA0/T56TyTB5Z2YEoDQwpMM0zKyPOgtV7tFwGhUIh0uqL8+h5hkQcCStseJ7vxcQKpnKhRHwJXUCypaI5sthkpwXabOM+8QPUjlKHRTBFziwx4Haz3u8h5UUIDAc4t95G/LIC6MJbnMWg20JHKMCM3jFWv8op/UvcYUa+Ps8d7uSH9KN1951LPxIsFn0anATcmeEl5JeGSpBAItia3UufWkW5YiF1ciZNZSP3sVxJpWalWE+CVVrM38TC+swUZZCmPrcEr3E85/Q0CL0VnZj7zQgHD478h8HbjlNUxPPP0fUCRVCrLnOxPuDL/I0yg141QCAX0DaRINDUTikYZ7dnLurv2ViNGxp0kDw7NZvDc/+bh7NswC/2QrkWVqH8oT8PjX1OfZ6w86t9bvUDdfuBom5OAoh+wOKbip57K6JiM6UILzBqNRqPRaDSaKvnH+un9p4cISt6RFz4I0peUdqYY/MpTlLuf20NIk23RW3YpBsGUaRrNdDMhMP/2t789uMi89Lra5+ghBOZYE1z97xCuCD2hSZmoDaqoFNd9ceo6hj3lq+v47Fg9RH1rlKuavooQQEoVfPLGxvCGhwkvWUK0zqZvX5Zl9bsA+NaW1zGad3jNuTOPan81mj9XsiPDAPRs3nDctuF7Lr//wqf5w/98np2rHz9g/uSIDIBz62N4EtZnC4y5Hs22hRCCxXElrG2vjPg5tz7Gy9saprQljGMv/RiBoKtrLsmxFcQalDCdbqwJ2yHqiBFmzFRibRi1H0YY6hOJmsCcH8YQAWGRpxgkGBnLEm4cBSDqSYJollxOnU9dN0q5HIe8ivEoBJKyneadj9zOr//wMf7l7C8AsCM9n87HxsG2yb7EJ/BcLMclY8U4fXQ37fnRaj9Hu+5i0zkNdDUOs/7MM+ies5DxZ9S5+H0vbCPpJAm8FAGSFfUmZjmhojUw8SyVvdzz0Acx7CiRZDsJJ0o00Q6yxOWrLXKRpxiK3oMUAok6PoG3lqv3vpLk4Bn0P94KCNxCCd8VBOcXsWNZcrkie1KbKG4Zoz7r8/CsFXzzzR/j4X2jCCFonjGbwV3DDO3JsuKySvFFoY5x0eok41cK++0vMA8qJzntK8DJcdRsuQW+shK+f6ALfropBQGnJdT/g+6CdjBPF1pg1mg0Go1Go/kTQHrBMXEep+9Uw1ydfc8txy77QA8j31qP25Mjffuu59RGk22xr6QerN/U2cR/LZlFnaUdzJrpZ0JgzmazbNy48cAFll4PZ79ZicutS46+4XffA6//CUy4ic97J7z3AWhaAG2nQW5gyuLZERWzccFLZrI4dLeamOklCCSFTcpxGVm6hNkXtiPKAc1t23C9MKOFMNec3s61yzue3Y5rNH9GyCAgO6oEyJ4tm47LNlynzI3/8TF2rVsDQPfaJ9X0combPvtJRvbtwa9GZKiCcbOj6vwz6HiMViIyAJoqI3y25dV54YOz26uF+44X0pfIokfz7GZCTiMzW88AoFxXexnsW4IoEcrzZwEQlkr89Pp3Y2/ezEB/P77vQ36YdXIJq6NFvGv+gDzncxhhdT9jhUaQAvLZZurqVORGLtuMkCaCAgXzm2TlL7m0V43guPjj/Xz5/i+xfngZdX1FMvEw7dkSXhjskkPEK/PynQ+yvHc7F2ypvcQ2AklDJk99RfR+9cP3My/dy+bt6wkFIZav2cvTW77PPudBQn6Y4egwrnDxwg9w+cBdBF4EPBsZRInUvxhPnquOQSSG0zYDo7GR/MKzKSxeRXROQOA8hZGPMPLYSgJXYEVWgYT8gHrhGIqmCAUJjMAg7MXwTHVtcENRnnTVsWmbv5CxgflYIYNlF3epHakIzCUayPktalp636QfTsLAejDD0H4aOM8iAqZntfqb6T38ctNAMQhIWiYzwja7dETGtKEFZo1Go9FoNJpTHC9Vou/fH2PkuxuQwXMXmaWU4KkHLmfvsxeYZSDJP6mEsNCcepy9WfzcYYqhHYLJbuUPz+3gzV3Nz7oNjeZQFJ4eZuyXWxn4r9Xknxg48gqTmBCYAYyDCThCwMu/Cv9vF9R3HX3DM86FZfu5wjrPhA+thZkrITs4ZVZmtAhAvVXL3mRoE9/9v8/w+f+9GYDwsmUMhsFr2U5y7iM8s+98vEByxsyGo++XRvNnSD6dIvA9IvEEAzu24TrHXrB65q7b6Nm8gRf/1YdZevEL2fLQ/dzx9S+zc/XjdK99kj9+92t4le1aFQdzsvKiNeP5jLn+JIFZ/d1WUAJzS+jYjvhx9mUZ//V2pFsTZIOCCxKirVESjWG2PKIcvp5t0JoZp76YR/olpBlhmZ+ifvNq8oHDbmOISHaUedt3kEqn+eYnPsrd92/l6+mruCMUZ19mJlu3XkQ45XDa2s3s6/wZjz7yevzAYtZM9VJv8+YXkiqESIb/kbPCv+eCwlMElZdzjh1h8XgPm3eeTijjItttTt+Wo7U9R0M5w1ef+SJtpTSd2TH+7qaAsCMRUuL3hrjs/gco9Kp7n1RDA9eNrsYdU9eImUPDLNzyKPXpNG1GlKJV5KmWp8jV1fHLs8KMt6xm2BgkVbbJRucgzTiFWYspzDutdiCFSWA6mHMTyEBiJ36BWxgFEUOYzSAkmx6YSXGPINwygmHWEy8qgfmlTw2z8umHqMv2sqOS5982ZwnC7KSUeQjfTQE+YsLBnM2SDVrVdicLzPd/VsVjxFshkoTyUdzrBQEUx2Go8rLFLT7Lf0HHn6IviZoG82Nhns4WcCaNgNOcOLTArNFoNBqNRnOKk3uwF1n2KW9P4aef+4OwP16uPkDmnxxg7OdbyNy77whr1fCGCvhjJRpfvYjktXMBKK4fedaid+Mkt3JX2MbPOQTF5xbZcSx5621v5d8f/ffp7obmeSD9gLFfbqWwdgg/75JfM3jklSYxWWAuFAq4rks+f5yLgCU6oDACkzJaM8MVgfl3r6wtt+b7vGvksywf3oQ1axZ7s3ewvedpFl7yFdxCE7ucvwDgtK6jz1jVaP4cyY6qeIzFF15C4Hvs2/DMEdZ49mx+6D7a5y/ktEuv4LK3vIu2+QvZcO9drL/3LkC5qD1XOZjNSgZzfeXaOOp4pL0DBeatFQdz8zGMlJKBZPRHm8g/OUDmvtr9QJBXfTMSNte8ezkLzuygjMtlY4J54+O8cevjeLJEYz7A/fUfkMBeerk7tJ7l730Pi7u6CJVKjOaLPL09R9vwLiQGn1z7t2zoPxN3tUH3cD2Pb13JoIyx2WvjrPvfQZeh7gXKfSUSYje3di5l846ZGFLyxwtewG8ueCcAV46uwwwkVtwnQGBEJPiCrr2jU/Zv5XbJNWskH79VOcgfXjaC4fsMt7aSO62LRZWM6QfmzgXg/H4Y61N50L3xXiZubzyrQCkyQi65nVLjGja0juMn1Lk20ttNYvNqGkZmE7OLFMZaqJuVpzCUIXC6MaxZeMWHQQoiOYPeJ0LYsTEgxPzhxbTPf5Crz/4Qlz96O3WZHfSHE+zeMkaiVY1E8Uob+P7fvQ8ZOCRtdXyyY8N4MoKMNUNq0n3cht+ov9FGCCWOzsG84Ub47FzYfqf67hbAd4+83glCSkkpCIgaBvOiYbYXyvzL9pPPZf3ngBaYNRqNRqPRaE5hpJQUnhmBSkGfoPDchdjyLjU0NPHCmfipMoV1w2Tu2I10/SnOpUMxEasRmltPaKbKl039dieF1c9OxGuquK/qLQMhBCPf3kDfJx/FGz3xrhkpJZ994rNcdeNVrB1ayy+3/RIvmH6xW/PsCEoe6dt3U1g7BL6k8bWLSVzYibMv86zyxm27loWcz+f54he/yJe+9KXj0ONJ1LWDDCA/XJ2UGUxjiRJRY2rOeao7yhl93Tw+ewG/fOhGZvofIhIq8OPtL6PXC2MagrNnNRzf/mo0pzjDe1S805lXv4T61jbu//F38b1nL6iV8rmDFu8b6+thsHsHyy65DIB4QyOv+9dPY5gme9evAyDw/GoG80SRv5hhYAn4zK5+AJor18qoaTArEmJHJXv2WDmYpZQM/vca/IzqR/6J/uoLYz9byYeOh+iYn+Qlf3kGUUK0leG1AxZvW5xECjALAdkJwdtT/Zu7dCldn/oPQoUsQTRBU7bAef3dXN7yCAC3OsvYFajM5T1BI3c7S3jcm8O/hE0e9E9nxqInWD66kXTR5hMbP4BYq+5PstEQv0p2AvDSfd0AzGisJ2+GuTNcK6YaVFQwX8C8Qck5O2ovwfvn7KBjdic7Fy2sTqvvf4bU0nYKkQjN624nPa6E/kAEbPK7YKTAvP56YrnZhEpN5MIFZoRS2OND1G97mqe6dnDPuUMsuvo7NLZ3k/GaaWrKEbggrFas6CpO37OF2e3qHD+WjxDbOw7AqoH30jzSgFdQ22zIb8Azbf7nZxvYtf2PBP44yBy+6yKDFM0RF5Ds2HwnwhCQnAk7/wjlStbyxMia13wXQnEI3OrvckiGJsXENM5Tf0uZw69zAnGkRAIRw+CdM5Vre+Jli+bEogVmjUaj0Wg0mlOUwPFVQb6sQ+zsNjUt/9xdJc7uDCJikrxmLi3vXkHd5So3sfdjjzD87fVHzHh2erKIsInVHEXYJnZFZC6sHz7sevtTV8mXTVoW0g1wB5TDJjcpziBwDnxoPx7cvONmfrz5xwzka9teO7j2hGxbc+wY/812svftY/zX2wGw22OEFzZAwLMqRimEQFSGYj/66KMUi0Vc1z0m+eeHJFEp1JSt/Rsc78uQNPurkc1c9wVyrWeT2qXyO7/YdjFff+btfOXpdxFIwdrxhWweyHLRgmYaYiE0Gs2h6du6mWhdPa1z5nHlO97HWO8+Vt9y87NqIwh8vvqOG/jvN7ycH//T33Dz5/+D3NgoUkpW33ITCMGSVS+oLm9aFo2dM6rfMyND9G/bgjAMovXKCSuEqLqYAVY1xKufXzGpqF/CPDYyjxCCxEVdtFwxRGfXJwiyRZwe9SK55mCuidlDcZ/RebcwM7mT7z+hXLOeYREYgqjZSrRvJy+86EKampqgpZnAKxPYIS68eIz4Pw3zF6c9xqX2egCWpoe4MNZNxqljhjHAJb1P88nb/p3C3j7EfSHKt0Pfb1v54LpfAzDrihFmtAxQsCMU61toqRSys9s9vJhgpLml2s9b/vta3vz3JgOtHczqacQsNlbn+aag/Qz1O1hOgebNv+J7q7ZhejZ3LbkQgLd1b0AEknO2BzTu7CXXH/BoNgu52dSnTqchUgIkHaE8F6eKtAYv4lzjfMLJfpLJQSQmydsbGbnmXMJ1r8cwm7hwyV4WtVYij2SIaG682idrzWKKoyqHu81fhxGU6T5jH6PDdxM4W6vLee4uHmq4hLtfcDaZ7qfxwwXEOW+B1F7cJ7+tFiqOwaKroXUxhJUT+4gu5sIYxNvgr56EF/ydmlZKHX6dE0jRVy8YoqZgSTzCS1qSpLwTc4+omYoux63RaDQajUZziuKP1RwasTNbKawZVLmIz7adjMPoTzfj7M4QPaMFYQgiCxqwW6Jk7++BQOLsyVDeNk55TwZnd4aGly/Abo9PacftyxOakVCuGaD13SuUg3ntEAOffxIjbtP6njMQVu3h1+nPk3+8n/oXzcZMhMg+3Iv/4B64IEaDZeIO1h58is8ME1nQQObuPTh7s4RXtRJ/yWxidmxKPzL37sXZmyW+sgMMiCxqRFgGUkrK3WnCc+sRR/EAfs/ee/jXR/6VZU3L+NiFH+OuPXfxk80/4d5993J+5/nP+jhrpgdvvERx/QgIoKIDW20xhCEQtkF5R4rw/CRuXw57Zh1G6PAFJT/+8Y/z1a9+leHh2ouTUqlENBo9PjtQV3GcZfuBs5CBZGBviUX2Nrj8n6HzLIZbL2LdP/4GsaCJL1z+ZsacRhrCKTaNLuUfH/gkRS+ObUref9nCw25Ko9FA79ZNzFh6GkII5p9zHrOXn8HG+/+I9H2sUIiVL33VEdsY76sN0R/s3sFg9w7mnnE24Xic9X+8gwUrLyTRNLW+QF1zC6M9ewHIp8ZZe9vvWH75VcTqk9Vl4qbJmOvzt3PaWRqvnXPeMqOFXw6MMSsSqr4EOxYkVnXBp/8anCyW2Utx/RxKW8bI3rMPW3RjffvNuGe/j5GWF/GPqzz+ybgRgFj2esr5ZvJeDoQkXncapcx97Lrju5yX+gPmxl/xiro6Hp2/hIG2HFaPIPmLHv5q7Gaunb2Vhf0D/GLxuTjrXT71yLeq/fnI2p+SKBeq31cObWV12xKeal7CDca9fK0rRc98WLROzTeNYbyoxdxLQ3xVvoJ3Wt/nxwN3c0H76Sxq3cWMHRlyQZStDbPIv//tID9N0BSw44wdbEhvwF/QQuAG/LK4CuYZmDmP63c/zrVrJG+/OwB+We3LE2c0kWs6nSASsN6IcubiIulNcWaWLbrau/F9i6E7L4AuwcblLyC1U5Du2MPfBl8naRSZbQiMiI9VbkakhqFyi2XIRnY/PAcDj9ZSwF/8+su0jWbYByDWEq6XlDICv/QEnr+cp05/DZFSMyvGt/J/doSu+kY+1/0DvnvatSwtjKvCsaAczABODmJNh/5HUBiFeIsSpcd2qmmlo38xe7wpVVz1kUpdhJaQxePp4xxdpTko2sGs0Wg0Go1GM824vosbPAdhuJK33Pi6xdhd6kHhuURkpH63A2d3htjKdhpftag63UyG6fh/59HxDysRUYvRn20le88+yt1pBv97LYNfXos/yTHtjRaxWmsPvEbYov7K2VitUbzREs7eLIWnhqZsO3vvXvKP9TP20y2Uto+TuX03LWXlRnlFe2PVXZq8fj5+qszIdzdUCxCWHx3mv+74zJT2grJP5u69lDaPMfqjTYz+YBOFp4aQrk/6lm5GvrWe9B27j3hMRooj/PW9fw3ADUtvYEXrCj688sOs6lrFTzb/hJu233R8XauaY0bh6WGQ0PbBswnNriM8P4kRMhGWQWhektwTAwx8cQ3D31xP+tZdR9XmhLi8YsUKgOObw1zfiSdDPHl/ge9/5GHW3LEHpwwdoS2w6Cr62i7hrz/9A/IvGOMTC9/DDm8O57at459X/ZjZdXsZLzcAcMN5s1m1QBfM1GgOh1sqkRrop31+7Vq44NwLGO/r4eFf/pj7f/zdo2pnaJcS4s665nr+8hs/IhyL8/hNv2DdHX8A4KV/+5ED1pl39nkAXP83HyEcjxOKxrjkhrdM7V9FTGsP21Omz4qEWHfR6dxy7uKj3NNnga/uNWJdo+Qe7CV7zz7Cxhrawx9COCnsxz/DT3/1XTaYLVhPfAqAzq5tnDNcwmcDZ757C3XnbmTJtVnmXLeG0XU3ExVFFkeHWNqwHmuPSevnbKxBgbB9zt7wGHVNLdQnxrmwXzmR981Q0QeTxeUn39/Oe6/4e750yetInPEmQsLjOn8HGxrV+dhOeDS5eymHI8yZ08cT0WV0dyjx/ZPP3Esytg/KgoRbYvecTmauuhSA0dIog8Eg1y26jjtefRsr5GcBg7qwyRNdi5G+WxGXIbRoLvL6KLFFZSxDjQ65ed/lrBlbwWe2v4QdrzBpX7oG05I8vOciftO6lIJnkW40WGqP8Sr/5ywytgAQdQPi88/CkIKnQ3Galvwar/QUoq4Vw1H3dnMe76A+O0nclSUic4qsXr4KCKgffwqAR8+9nC2JfXzt6a/xyaYEWenxrw//K0OlMYgqMfk7Xgf/uOjD9Oay/LB3hL7SIYoy54chVrl2RCovO04igbnmYFbyZrNtMe56+Poe7YSjHcwajUaj0Wg000ggA15288uYVTeLb179zWe1rp9WDwPh+UmMqHrY9J9DREZ5T4bYue00vebAB1MrqYZlRk9vprB6kPiqTiILGxn90Sbc/jyp3+2k4fr5al8KHlbTVBen1Ryl/W/PBQn9n3pMVaKXksT5nUg/oLwjpfrQnabcncZqj7F0QQO33dfP4gbJ+O27CM2tJ3FRF6FZdRQ3jRKakeAnu37Gix89G7YWcF/iYhuqGGD/fz4BvqT5baeDlIz+YBOlHSkKzwxT3q62lXu8n2/U/YIep4+FjQt52+lvIxmuOcSklHzuic8hEHzxsi9y5ewrq/PeteJdbBvfxr8+8q9ErAgvnvfiZ328NScOKSXF9SPYs+oIdSVoe/9ZU+Ynr57D+G93QiAhkBSfGabhpQsQ5uEdgPPmzWPXrl2cccYZrF+/nvGdOxG//R2Jyy/HbGzATNb+PQnjeXp64m08lH0HG9c2AmUe/203ITsga49xz0O7+PcNq9llzuUR52PgwP9ru40Lr8ixZPE3uPjCFj5280b+uGWIlXMbj7gpjeZ443se+fEx+rZvoaG9k44Fi4680nFg2+MPE4knmLH0NLY8/ACLzl9FKBpjfKAPYEpcxbxzzuPeH9RctIHvY5iHHunQv30rt37lCwBc9pZ3YVoW8YZGxvp6yI2P0Tp7LqZ1oBRz9jXXceZV12JaNrNOW46UknjD1P+3pUCJae0HyVk+Fs5lZ69yUAs/g33n+1UxOL+St5x9HDgHgzRN9n8BkCssJFbXx2uaH+LR3nOYvTXDnjktdHVuRV6SIrrVRAhILnxQ9XuozJxQlsHWEM1jLmc9nWXrlhUgdzLyNz7lfpO670PikgtIxO/hjOERnmhfyta3X8Art9xM7Hdp6s8u84X8a7i591KoF/xFOsxLX3MlcuRLvDK1nf+uRIfYMRWT8Nuei5gzv8iQbKLXtjClpN0tY8yF/icbAPjgwt+Slu8lYScYKY4wVhyj2U7QlXmG7/3FOfz9bzbxgXPDtNz8j0TEKMPr62hcWKBxgcqN5lzYs+0xUizisw/8JwMNCf5v2fUYDxp453s0Pewy1DmTi52nOEv2sf7MMzjj0XWcHt/BwJwOeqzzWRn/HYsvWMX9+SESewbpe2wPsAfLvAojdA7SSkGhmwcv7aE8eD7njMxl5miK/8pfQOrSuZy5cw/R7A5esesn3DzvTWyb3YnhGZjS4/1elG+ld/Ha9iSfociC/CCfyTeR63o5P9zuAj28Y0YLn14888B/FPkR6DxTfQ7X4wNmeWoGs5TymDrnnw0T/ycmHMzNIYsAGHf9Y5ZHrjk6jupoCyGuBb4MmMC3pZT/ud/8MPBD4FxgFHi9lHL3se2qRqPRaDSaw6Gv16ceI8URvvH0N+jJ9dCT68EPfEzj8MPzJ+OlyyDArAshTIGIWs86IiMouARZF7s9dtjl6q+ag90WI3HxDDCg+W2nU+5OkXugl+LTwxh1yrljNUcOWHciMqPpTcsY/eEmMnfsIbailfEbtxEUPJretJTixlHsdtV+eUeK1kf6GP/FNoyYRctbT0cYgvCcesJz6nF9l59uuYl5kQYuyZ7NhpENnNV6Fpk794AnsWbFEfNjREIRQnPrKT49XN2HyOJGhr66jvijLrm5KR7e/ke2PbaO0LIGPrDir/B8j399+GPE+03+Y84/sWp0BWW3EqthGZzVdha3v/p2Lvn5JTzY8yDzk/Ppy/WxM72TF8x4AUuallT3W0qJO1DAbo9Vj8Gz/W2wjCNGNmgOTXnbOG5vjoaXLzjo/NDMOtr/6iwAihtHGP3RZsrdKSKLDi/G3nDDDTiOQ6Gg3HR9P/s53u23M/T5z6t2583DGx3FrKtjxpf+m2jF6fycMC163LOYPfsZFr+8gexQCG9skNc++RFYCyGjhQs7n6RFzuSiW37Ha2+IY5/1MwBiMbAqYnnoGOWyao4Nf47X7HV33sofv/N/1e/Jtnbe+T/fPuHClOuU+f0X1egX07bxXZfM8BCrXvMGxvsnBOau6vKNHV28+yvfZfcza7nrm18hMzxEQ0fnIdvf9vjDACy9+IVVIXn+uecz1tcDQH1b+0HXE4aBWRHJYsmGgy5TE5ht/GwWs67uqPY5yOcJHAersfGgYqDb38/wl75E+re/A6Bhfp7O86e6VM2BOwiVLOyEixnJsuveeez+yk9I57fyorvey407/g45x2TeVp/yDotSLA1LVH2IkY0vZXnjjZy+I8uoiNCbPJtU3dO07zZpza4nbTcQt1fhz30QsJB7vsibciXGc0numruSRRGf5LIh5pnj/Ma/lJu9F2IZgiYHmv1KNv5Zb+T0Oz7KJ+wBgs562q6Zxeqma/jFrvO4dN9aQLC65Uw6rBxWexIWX8tn4mcSG1rN1+0v0/S7t9E8Zw57xrbhBA5Nj38bxv6T6OIX89Wr/w1+9ApkkOFjbX/Bf1yjMo1/5V1KP228xHqSi+fexML19zIYaWTpviH+Z9//qAOnfnYWD6r4kLHGRtafeQb3r7qM7Jo6nD0hCq2CzlgbxUe/zc7ghczsaiaWHSSaHcXLP4QQEaQzjiNs+nv/mnSpjXmBTWfCYI6xj3M3PUmpvZVwTrKq9CBP7z2PXbOvBq7m3MxDvGPrF7jijT/j7297K+8dvBv/N0+Sm/kVXjtwO8V5V3BLMUR34eDF/tKFUe4xSnT2P8bXVn8Jo6ON701yMA/mB3nl717Jma1n8volr+cFM17wrO5lny/FqsCs/k23VIpKjrren53A7Odd3J4sTn8eZ2+W+qvmEOqMH3nFY8QRj7YQwgS+ClyF+q/xpBDid1LKSaUkeScwLqVcKIS4Afgs8Prj0WGNRqPRaDQHoq/XJzfffOab/H7n78m7eZoiTXzonA/hBi7/eP8/4gQOCTtBzs3Rne5mUePRu7n8dBkjEarmCZsx61lHZLhDSiCz2g4vMFvJMHWX1pwt0aVNRJc24c8Nkf1VN1alonw6VmDCw1z2y/xo04/Yld5FzIrRmejkujdcifv9vfR98lFARV8Mzs7RnewlkAHGwHbObzlH7Z8VsPYF+9ix8VFaoqpAT97N883138QLPIIlUZY+3cHGbz7NTpEmkjUZWe7wbj6M8SuD8zrO4yxzHtdyPt9fcTvDIsc/NPwDdzY8yivGLucVY5dX9+fh8XU4d+8k5kf4tPhLwjIE+2CcberYNoZpuG4+ZlMEWfJ5q3gN6ceH2XbvA8xyOphrhXg8dAv/t+TLvKP5zTQMRgj3SryhArGz20heNw8zoUR46UvKO8aRgDAFZiKE3aEeAMp7MuSfHMDtz+P25kCAiFgYMYvYma3Uv2jOUYvVXqqEN1IkyLkERY/YWW0Y0YPf/udXD5B7uA8MgZkMY4RNZCBpeu3iKZnZAEHZI//kIPgBXqqMEbOpv3L2QfuVTq+lr+9XlMr9OM4IM2a8kZkz3jhlmZGeLGP9eXo2j+M5Ps0lj5ZkiLAp8LrTmFfOphwPEa0L0T6vnnLRY/PDfUgJdtikqTNO16IGCoVdbN/+aYKgjGFGSNafRfSeVZiNYeLndRzxeEUWN9HyzuWE5zcccdlwOEw4HMaoiEHpffuYd9EqwosWIz2P3EMPEpo5E290lH1/+T4W3nkHRvzgD3nF4l66u79MIB3C4Q4aG86jtfXq6vxS3iVo7yF2/je4feMSbtt9JVvG1IuM18RupaMwyub4K/mvt76GSP2NYART2jcqQpIeLnzy8Od6zd54310AXPH295IdHeHJ3/2a1EDfFLfwiaBv62YAZp22gvq2djbedzd7Nz7Nqte8gVTFwby/gFzf2kZTl7oGjg/0HVZg7t26ic7FS7nuQ/9QnXbJDX9BKBLlkV/9BBUI/9yYyJt1/vav2fbIw9hdXZgNDYhIBGHbhGbPJjRnNs6evYhohMzvbyE0dy7e4CBeKkXsvJWUN20mcsYZxC++CAKJ1dTI4Gc/hzc4iAiHaX7nO0hkbsTLb2D3Xc2YdkCkK0rHWcO0+b8FwPPCxK69gYdi9Xx1ZA7nr/gcv1v3QQQ+a1hOR1CkI7+TFWMjPF23gq5dgtNHsxBvxX7L/Zxd2I38/iu5Tb6cFZk7CM9uozX8Rnj6boYNE3vAxc6rfb2u6wlWbdlByCtSMG3+u/xq5iQCfvaBK7npI4/VDs5Zb+LRtd9k1fBueOEYnPFaTnvRp+n5+G2sHlAjrUZswczkAni9Eoj/cWmabzwwi8Gmetqf+BzNKZNtuQEwoCnWBnNfAmt/CNtug1gz4h23IR+Fr60b4/SLruPSVddjGYIv/O//MVdsJX7mEG+y/sjQM3WM74jjnynIp6N0No0T7BIUgwjnrdrMQDCD9cbprD333Gr3u1F9nJ/IIoIIXv1cjG2Ccr2FjETxzSZs3+JvHr2Nzy9/HVHDo5AcZ5W9h2JOIKI2TnQGuFt51V0/Zu/SC7jlostZW3cR4+7nWCBC/LRvkK9ecAPfSisH8hsGbuWis07nb5Jnctdo5sCXD57Dl+KCG3Mb4M53q2nRCHsye5lTWeSuPXeRdbI81PsQD/U+REu0he9f+33m1M/hRLB/RMaEqDziuCyJH2h6+FMlKPsM/c9T1fg8AGEbNL9h6QnrgzhSdpwQYhXwCSnlNZXv/wQgpfzMpGXuqCzzqBDCAgaAVnmYxleuXClXr179vHfgs098li1jW553O6cSL95+AZ25w4Sw/xnxvTmt7I6Hn3c7c/Nl3r7n2VW4ny5idoxZdbOnuxvHnL6W71AKH13u4dHguQEyOHkepHwBZdOgHOQJOIz4EwTVAkQAhjSwPRtkwJQZR8NRLv7LvsvpKbU9i4YPflMs8eEI15ROY5Tr/eJhl2mbM5/L3/aeZ9GfgyOEWCOlXPm8GzpFONmv1x958CPcu/fe593OqcTl4yt5Z98rAPXf0RQmAoEvfWTlP6ghDMKmuo4VveKzfuQMBTY7o73800Llkvn0jg+wsDgLxzh6F7MpTULS5n1LPs1QaOxZ9gBKfonOUgvf7v4EAQGvWvJhzMrNvS99yn6ZtlgbRa9I1sliYPDpfR+iq9zKDztu4eGGdRS9qecFU5jMKXWxO9RDICSGMAhkTTS7oOMCrp57NS9uvIrUlzYgg4Cn4lt4tGE9t9U/yJKmJSxpWsL64fVEzShusUy8ro51w+swMPACj5vP+BntbjPCMhi+fwcy59NXP8JIPMPSusW0nzmPWLIOYRt4QwXSd+7BGyxwMMYXOjT5Dbj7clieesAJREBkVhJ3oIB0fDBAWMrRI/0A/Kn/7YRtgBBIVw3pDc2pJ7KwASlBFj280SKlrePV5Y6MRDpThUZMccjihtJR2zWbIghT4A0Xp/RrgqH5v2C86+5av1H/vg/Vo8AsY3hRQsV2pOFSTvRgeFMf+I7V1VoaLoYfws51EYSyuPEhhB8ikOKYbWPqBtWfQKhjJzAPPA5HsWFheCAkbq4ZK5bCsFwCN8Tkoyosh2c2ncNXet5Mk59hVXwNZ47s4OI/7uATF7+bL3/x/XQkI/DzN8GWWyCUUCte/Nd8PPUSfvDoHr7z1pVcuezgzsnpRl+zT55rdvaee+j7+3/g7rltuEeIinm2+IbB0pEMC8fz5G2Te+e2YQSS4/Q/9JBIBFLANTsHsaRkc3MdO5sSmEFAIAQhP+CqXUMHrFcyDe6e337EPvuGwfzxHKeNZKdMH4mGeGxmM7PSBc4cOjDD9sfzL+PGuRcfst3bY/+PZiurXjx6hztmk+cd+Bv23pIkKB9pRIPEx6SMPakNiUVAmP/f3r3G2FHXYRz/Pt2lRZC0tFxsaYWitQ0h0AKhRRqiiIrEgDa8gBglEdOYlABeYjQkmlYTY6JcNKRqRF4QA40I2jRKrRRfSITacimXdqVAA6290AKtwcKye36+mP/ZObvndFuW3XP2nP/zSSadW9uZpzPz+3X2zJxe+pnA2zGxuP6LgfkA7zCRChOYqF66ojJQIypoYFlVF/1M7HuXE05/mw8tOkC3Kjzzl1l0/bev2M9jYN5Vu5Bg81knsCHm8qMnv8mSM/u5demV3Pn19QAs++WlAKz45wr6Nt3Nin2vs3/xTVyxZy2Hevuo/rdQXb0smbOE5R9fXr/Le7fy7Ye+ytoo/t1WXraSxTMuhr//GLY/Cpf9AGZdSKUSVCLoblBPf/bnzVzy0q2cc84CenvW8fqOFzhduwHY/8VV/OLBx5jV/x+u71rFISZxS/+NHNofnH/wRZjYxce29vDSh2dz4lsHeekjszkwZQoAE/p7UWUyla53CfWjECBC/ZyyZzcL//EY/577BXrOTv+XPJofKLbmjRZH1NfVTeU9v1pKPDqjh0MTttB7zAx2n7YCVXqByhF/ZzMc99Ympu2/e+z/oqjpJ1OvederP2Rq/2ROWXZu3Zdzj8Rw9fpobjBfDVweEV9L018GFkbEDTXrPJvW2ZGmX0zr7BvyZy0FqncN5gI9I9ulETsJ2HfEtfLgLErOouAcSs6i0O45nB4RJ7d6I5qlw+o1tP/xN1qcQ8lZlJxFwTmU2j0L1+z2rtnvR7sfu2PBmdRzJvWcSWPOpd5oZnLYet3UF5JExK+B9/btNaNI0sacfjI+HGdRchYF51ByFgXnkK9W12vw8VflHErOouQsCs6h5CzyNR5q9vvhY7eeM6nnTOo5k8acS71mZXI0nzvfCcyqmZ6Z5jVcJz2+M5niiwjMzMysOVyvzczM2oNrtpmZdZSjucH8L2COpNmSJgLXAKuHrLMauC6NXw2sH+7dUGZmZjbqXK/NzMzag2u2mZl1lCO+IiMi+iTdAKwFuoDfRsRzklYAGyNiNXAXcI+kbcDrFAVyPGrbR4fGgLMoOYuCcyg5i4JzaCMdVq/Bx1+Vcyg5i5KzKDiHkrNoIx1Ys98PH7v1nEk9Z1LPmTTmXOo1JZMjfsmfmZmZmZmZmZmZmVkjR/OKDDMzMzMzMzMzMzOzOr7BbGZmZmZmZmZmZmYj0lE3mCUdK2mDpKclPSdpeZo/W9LjkrZJWpW+SAFJk9L0trT8jJbuwCiT1CXpSUlr0nSuOWyX9IykpyRtTPOmSlon6YX064lpviT9PGWxWdJ5rd360SVpiqT7JW2VtEXSRbllIWluOhaqw0FJN+eWQ5Wkb6Tr5bOS7k3X0SyvFdY8rteDuV4XXK9LrtcF1+yS67W1I0mzJD0i6fl0/N6U5md3Dle5Bzo890P13BvVc4802HjqlTrqBjPwDnBpRJwLzAcul7QI+AlwW0R8FHgDuD6tfz3wRpp/W1qvk9wEbKmZzjUHgE9GxPyIuCBNfxd4OCLmAA+naYDPAXPSsBRY2fQtHVt3AA9FxDzgXIrjI6ssIqInHQvzgfOB/wEPklkOAJJOA24ELoiIsym+ZOYa8r5WWHO4Xg/mel1yvS5kX6/BNbvK9draWB/wrYg4C1gELJN0Fpmdw0O4Bzo890ONuTcazD1SjXHVK0VERw7AccATwEJgH9Cd5l8ErE3ja4GL0nh3Wk+t3vZR2v+Z6SC6FFgDKMcc0j5tB04aMq8HmJ7GpwM9afxXwLWN1mv3AZgMvDz03zbHLGr26TPAo7nmAJwGvApMTef+GuCzuV4rPLRmcL12va7JwvU6XK+HySXbmu167aFTBuBPwKdzO4eHySPrHmhIFu6HGufi3mjwvrtHGj6flvZKnfYJ5upjFU8Be4F1wIvAmxHRl1bZQdGkQdmskZYfAKY1dYPHzu3Ad4BKmp5GnjkABPBXSZskLU3zTo2IXWl8N3BqGh/IIqnNqd3NBl4D7k6PHv1G0vHkmUXVNcC9aTy7HCJiJ/BT4BVgF8W5v4l8rxXWRK7XA27H9brK9brget1YtjXb9do6QXqNwQLgcTI7h4dyD9TQ7bgfasS90WDukYbX0l6p424wR0R/FB8NnwlcCMxr7RY1n6TPA3sjYlOrt2WcWBwR51E8CrBM0iW1C6P4sU20ZMuaqxs4D1gZEQuAtygfkwCyyoL0Dq8rgd8PXZZLDuk9TFdRFOoZwPHA5S3dKMuG67XrdQOu1wXX6yFyr9mu19buJH0Q+ANwc0QcrF2Wwzk8lHugwdwPDcu90WDukQ5jPPRKHXeDuSoi3gQeoXiUYoqk7rRoJrAzje8EZgGk5ZOB/c3d0jFxMXClpO3AfRSPmdxBfjkAA5/6ICL2UryL5kJgj6TpAOnXvWn1gSyS2pza3Q5gR0Q8nqbvp7g455gFFEX6iYjYk6ZzzOEy4OWIeC0i3gUeoLh+ZHmtsNZwvXa9rnK9HuB6XS/3mu16bW1L0jEUN5d/FxEPpNm5ncMNZd4D1XI/dBjujeq4Rzq8lvdKHXWDWdLJkqak8Q9QvN9pC8VF++q02nUU734CWJ2mScvXpzv7bS0ivhcRMyPiDIqPyK+PiC+RWQ4Ako6XdEJ1nOKdNM8yeJ+HZvGV9M2ai4ADNY8VtLWI2A28KmlumvUp4HkyzCK5lvLxEcgzh1eARZKOkyTKYyK7a4U1l+t1wfW65Hpdcr1uKPea7XptbSkdr3cBWyLi1ppFuZ3DA9wD1XM/1Jh7o3rukYbV+l4pxsGLqEdrAM4BngQ2U5x430/zzwQ2ANsoPi4+Kc0/Nk1vS8vPbPU+jEEmnwDW5JpD2uen0/AccEuaP43iSwReAP4GTE3zBdxJ8R6sZyi+rbvl+zGKecwHNqZz5I/AiTlmQfFo6X5gcs287HJI+7cc2JqumfcAk3K8Vnho7uB63TAT12vX69o8XK/LLFyzw/XaQ3sOwGKKx7I3A0+l4Yocz+GaTNwDDZ9P1v3QkCzcGzXOxT1SfSbjoldS+gvMzMzMzMzMzMzMzN6TjnpFhpmZmZmZmZmZmZk1j28wm5mZmZmZmZmZmdmI+AazmZmZmZmZmZmZmY2IbzCbmZmZmZmZmZmZ2Yj4BrOZmZmZmZmZmZmZjYhvMJuZmZmZmZmZmZnZiPgGs5mZmZmZmZmZmZmNyP8BA8VR2on+gaYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1800x720 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "known_site = 'NSP9-65-0-G'    # the site that we know is a defining site of the variant.\n",
    "site       = 'S-704-1'      # the site that we are interested in looking at the trajectories for, for different mutations.\n",
    "\n",
    "fig, axes  = plt.subplots(2, 3, figsize=[25,10])\n",
    "\n",
    "locs_known = []\n",
    "ax1 = axes[-1, -1]\n",
    "ax1.set_title(known_site)\n",
    "ax1.set_ylim(0,1)\n",
    "for i in range(len(sites_all)):\n",
    "    if known_site in sites_all[i]:\n",
    "        site_nuc = nucs_all[i][sites_all[i].index(known_site)]\n",
    "        col = list(pd.read_csv(files_all[i], usecols=[site_nuc])[site_nuc])\n",
    "        ax1.plot(times_all[i], col)\n",
    "        \n",
    "sites_present = [site in short_all[i] for i in range(len(short_all))]\n",
    "locs_present  = np.array([i for i, x in enumerate(sites_present) if x])\n",
    "site_idxs     = [short_all[i].index(site) for i in locs_present]\n",
    "times_present = np.array(times_all)[locs_present]\n",
    "for j in range(2):\n",
    "    for i in range(3):\n",
    "        if j * 3 + i < 5:\n",
    "            idx = j * 3 + i\n",
    "            ax  = axes[j, i]\n",
    "            ax.set_ylim(0,1)\n",
    "            ax.set_title(site + '-' + NUC[idx])\n",
    "            new_site = site + '-' + NUC[idx]\n",
    "            for k in range(len(locs_present)):\n",
    "                site_nuc = nucs_all[locs_present[k]][sites_all[locs_present[k]].index(new_site)]\n",
    "                times = times_all[locs_present[k]]\n",
    "                col   = list(pd.read_csv(files_all[locs_present[k]], usecols=[site_nuc])[site_nuc])\n",
    "                ax.plot(times, col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 697,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['europe-croatia---2021-7-4-2022-2-19.csv'\n",
      " 'europe-estonia---2021-7-14-2021-8-26.csv'\n",
      " 'europe-bulgaria---2021-7-18-2021-12-9.csv'\n",
      " 'europe-estonia---2021-9-15-2021-10-15.csv']\n"
     ]
    }
   ],
   "source": [
    "print(np.array(regions)[locs_present])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 611,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P\n",
      "S\n"
     ]
    }
   ],
   "source": [
    "print(dp.codon2aa('ACT'))\n",
    "print(dp.codon2aa('TCT'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Classifying mutations as synonymous or nonsynonymous on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/find-nonsynonymous.py blee098@cluster.hpcc.ucr.edu:\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-nonsynonymous.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-find-nonsynonymous.sh\n",
      "\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/synonymous.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/synonymous-prot.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/protein-changes /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "script_file   = 'find-nonsynonymous.py'\n",
    "out_dir       = ppath.join(SSH_DATA)\n",
    "\n",
    "job_file      = 'job-find-nonsynonymous.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --time=5:00:00\n",
    "#SBATCH -e ./MPL/out/error-synonymous\n",
    "#SBATCH -o ./MPL/out/out-synonymous\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python {script_file} --data {out_dir} --multisite --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "print('')\n",
    "\n",
    "print('scp %s%s/%s %s &&' % (SSH_HOME, out_dir, 'synonymous-prot.npz', DATA_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NSP5-47-0-A NSP5-47-2-T\n",
      "NSP2-88-0-- NSP2-88-1--\n",
      "S-144-0-- S-144-1--\n",
      "S-68-1-- S-68-2--\n",
      "NSP6-197-1-C NSP6-197-2-A\n",
      "NSP7-33-0-T NSP7-33-1-G\n",
      "S-136-0-- S-136-1--\n",
      "NSP12-671-0-- NSP12-671-1--\n",
      "S-1101-0-- S-1101-2--\n",
      "NSP12-185-0-C NSP12-185-1-A\n",
      "NSP12-185-0-- NSP12-185-1--\n",
      "NSP12-261-1-G NSP12-261-2-C\n",
      "NSP15-206-1-- NSP15-206-2--\n",
      "N-2-1-- N-2-2--\n",
      "S-156-1-- S-156-2--\n",
      "NSP13-443-0-G NSP13-443-1-A\n",
      "NSP15-232-1-C NSP15-232-2-G\n",
      "S-218-0-- S-218-2--\n",
      "NSP15-215-0-A NSP15-215-1-T\n",
      "NSP15-222-0-T NSP15-222-1-C\n",
      "N-208-1-- N-208-2--\n",
      "ORF8-3-0-- ORF8-3-1--\n",
      "NSP16-293-0-T NSP16-293-1-C\n",
      "S-240-1-- S-240-2--\n",
      "S-91-0-G S-91-1-G\n",
      "S-104-1-T S-104-2-T\n",
      "S-385-1-A S-385-2-C\n",
      "NSP2-493-0-C NSP2-493-2-C\n",
      "S-943-0-C S-943-1-C\n",
      "ORF3a-163-0-A ORF3a-163-2-T\n",
      "ORF3a-219-0-A ORF3a-219-1-A\n",
      "ORF6-61-0-C ORF6-61-1-T\n",
      "ORF8-1-0-C ORF8-1-1-A\n",
      "ORF8-14-0-C ORF8-14-2-G\n",
      "ORF8-80-0-G ORF8-80-1-A\n",
      "ORF8-108-0-A ORF8-108-1-A\n",
      "N-1-0-G N-1-2-C\n",
      "N-202-0-T N-202-1-C\n",
      "N-203-1-A N-203-2-A\n",
      "N-361-0-T N-361-1-C\n",
      "N-420-0-C N-420-1-T\n",
      "NSP1-39-0-A NSP1-39-1-A\n",
      "NSP1-77-0-G NSP1-77-2-T\n",
      "NSP3-1439-0-A NSP3-1439-1-G\n",
      "NSP4-111-0-A NSP4-111-1-A\n",
      "NSP4-442-0-G NSP4-442-2-A\n",
      "S-142-1-A S-142-2-T\n",
      "S-142-1-A S-142-2-T\n",
      "S-142-1-A S-142-2-T\n"
     ]
    }
   ],
   "source": [
    "# Changing the amino acid for nucleotide mutations that are linked to other nucleotide mutations in the same codon\n",
    "dp.fix_aa_changes(os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "                  os.path.join(DATA_DIR, 'linked-sites.npy'),\n",
    "                  os.path.join(DATA_DIR, 'synonymous-prot-test'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='trajectories'></a>\n",
    "# Finding frequency trajectories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-traj-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-traj.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "\n",
      "sbatch job-find-traj.sh &&\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/trajectories /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# Finding frequency trajectories\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script   = 'epi-traj-parallel.py'\n",
    "temp_dir1     = ppath.join(SSH_DATA, 'trajectories')\n",
    "temp_dir1     = ppath.join(SSH_DATA, 'trajectories-nosmooth')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 518   # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-traj.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/traj-error-%a\n",
    "#SBATCH -o ./MPL/out/traj-out-%a\n",
    "#SBATCH --array=1-{num_files}\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {freq_script} --data \\\"$file\\\" -o {temp_dir1} -q 5 --timed 1 --window 15\n",
    "python {freq_script} --data \\\"$file\\\" -o {temp_dir2} -q 5 --timed 1 --window 0\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print(f'scp -r {SSH_HOME}{temp_dir1} {DATA_DIR}')\n",
    "print(f'scp -r {SSH_HOME}{temp_dir2} {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-traj-variant.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-traj-delta.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "\n",
      "sbatch job-find-traj-delta.sh &&\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-south-africa-sd /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# Find delta trajectories\n",
    "# Finding frequency trajectories\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script   = 'epi-traj-variant.py'\n",
    "out_file1     = ppath.join(SSH_DATA, 'delta-traj')\n",
    "out_file2     = ppath.join(SSH_DATA, 'ba2-traj')\n",
    "out_file3     = ppath.join(SSH_DATA, 'lambda-traj')\n",
    "\n",
    "job_file1     = 'job-find-traj-variant.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/traj-delta-error\n",
    "#SBATCH -o ./MPL/out/traj-delta-out\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python {freq_script} --data {input_dir} -o {out_file1} -q 5 --timed 1 --window 15\n",
    "python {freq_script} --data {input_dir} -o {out_file2} -q 5 --timed 1 --window 15 --variant BA.2\n",
    "python {freq_script} --data {input_dir} -o {out_file3} -q 5 --timed 1 --window 15 --variant Lambda\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print(f'scp -r {SSH_HOME}{out_file1} {DATA_DIR} &&')\n",
    "print(f'scp -r {SSH_HOME}{out_file2} {DATA_DIR} &&')\n",
    "print(f'scp -r {SSH_HOME}{out_file3} {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='inference'></a>\n",
    "# Inferring selection coefficients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inferring selection using all sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered ! -name \"*sites*\" -print0)\n",
      "echo $(#files[*])\n",
      "cd Archive-test\n",
      "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-new/job-find-freqs.sh blee098@cluster.hpcc.ucr.edu:\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-new/job-infer.sh blee098@cluster.hpcc.ucr.edu:\n",
      "\n",
      "sbatch job-find-freqs.sh &&\n",
      "sbatch job-infer.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-06-09 &&\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-* /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09 /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script    = 'epi-covar.py'\n",
    "inf_script     = 'epi-inf-parallel.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('cd Archive-vector')\n",
    "print('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "num_files     = 518   # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-freqs.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "cd ./Archive-vector\n",
    "pwd\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=10:00:00\n",
    "#SBATCH -e ./MPL/out/inf-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-out-%a\n",
    "#SBATCH --array=0-10\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "regs=(1 5 10 20 30 40 50 60 70 80)\n",
    "outfile={out_file}/infer-g-${{regs[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {inf_script} --data {temp_dir} --timed 1 -o $outfile -q 5 --g1 ${{regs[$SLURM_ARRAY_TASK_ID]}} --refFile {REF_FILE + '.csv'} --saveCovar\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s' % (SSH_HOME, out_file, DATA_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Eliminating sites that are infrequently observed (frequency never about 5% in any region and total observations > 40)\n",
    "window    = 5\n",
    "count_min = 40\n",
    "cutoff    = 0.05\n",
    "\n",
    "# determining which sites pass the 2 cutoff criteria\n",
    "traj_dir  = os.path.join(DATA_DIR, 'trajectories-nosmooth')\n",
    "traj_muts = []\n",
    "for file in os.listdir(traj_dir):\n",
    "    traj_file = os.path.join(traj_dir, file)\n",
    "    df        = pd.read_csv(traj_file)\n",
    "    df        = df.drop(columns=['time'])\n",
    "    muts_temp = df.columns.values\n",
    "    traj_muts.append(muts_temp)\n",
    "traj_muts = list(np.unique([traj_muts[i][j] for i in range(len(traj_muts)) for j in range(len(traj_muts[i]))]))\n",
    "freq_mask = [False for i in range(len(traj_muts))]\n",
    "for file in os.listdir(traj_dir):\n",
    "    traj_file = os.path.join(traj_dir, file)\n",
    "    df        = pd.read_csv(traj_file)\n",
    "    df        = df.drop(columns=['time'])\n",
    "    muts_temp = df.columns.values\n",
    "    for site in muts_temp:\n",
    "        mut_traj = list(df[site])\n",
    "        mut_mask = np.any([np.all(mut_traj[i:i+window]) for i in range(len(mut_traj)-window+1)])\n",
    "        if mut_mask:\n",
    "            freq_mask[traj_muts.index(site)] = True\n",
    "            \n",
    "regs = [1, 5, 10, 20, 30, 40, 50, 60, 70, 80]\n",
    "for reg in regs:\n",
    "    inf_data  = np.load(os.path.join(DATA_DIR, f'infer-{DATA_DATE}', f'infer-g-{reg}.npz'), allow_pickle=True)\n",
    "    alleles   = inf_data['allele_number']\n",
    "    selection = inf_data['selection']\n",
    "    errors    = inf_data['error_bars']\n",
    "    s_ind     = inf_data['selection_independent']\n",
    "    locations = inf_data['locations']\n",
    "    times     = inf_data['times']\n",
    "            \n",
    "    mask_new  = [i for i in range(len(alleles)) if freq_mask[traj_muts.index(alleles[i])]]\n",
    "    selection = selection[mask_new]\n",
    "    s_ind     = s_ind[mask_new]\n",
    "    errors    = errors[mask_new]\n",
    "    alleles   = alleles[mask_new]\n",
    "        \n",
    "    mask      = selection!=0\n",
    "\n",
    "    selection = selection[mask]\n",
    "    alleles   = alleles[mask]\n",
    "    errors    = errors[mask]\n",
    "    s_ind     = s_ind[mask]\n",
    "\n",
    "    count_data = np.load(os.path.join(DATA_DIR, 'nucleotide-counts.npz'), allow_pickle=True)\n",
    "    counts     = count_data['mutant_counts']\n",
    "    alleles2   = count_data['allele_number']\n",
    "\n",
    "    mask1      = np.isin(alleles2, alleles)\n",
    "    alleles2   = np.array(alleles2)[mask1]\n",
    "    counts     = counts[mask1]\n",
    "\n",
    "    mask2      = np.isin(alleles, alleles2)\n",
    "    alleles    = alleles[mask2]\n",
    "    selection  = selection[mask2]\n",
    "    s_ind      = s_ind[mask2]\n",
    "    errors     = errors[mask2]\n",
    "\n",
    "    mask3      = np.nonzero(counts>count_min)[0]\n",
    "    alleles    = alleles[mask3]\n",
    "    selection  = selection[mask3]\n",
    "    s_ind      = s_ind[mask3]\n",
    "    errors     = errors[mask3]\n",
    "    \n",
    "    np.savez_compressed(os.path.join(INF_DIR, f'infer-{DATA_DATE}-observed-g-{reg}.npz'),\n",
    "                        selection=selection, selection_independent=s_ind, error_bars=errors,\n",
    "                        allele_number=alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "site NSP14-93-2-T already found in a different group\n",
      "site ORF8-68-2-C already found in a different group\n",
      "all fields calculated, putting data into dataframe...\n",
      "nucleotide number 10398\n",
      "protein 10398\n",
      "amino acid number in protein 10398\n",
      "amino acid number in reading frame 10398\n",
      "synonymous 10398\n",
      "amino acid mutation 10398\n",
      "nucleotide 10398\n",
      "reference nucleotide 10398\n",
      "selection coefficient 10398\n",
      "error 10398\n",
      "independent model coefficient 10398\n",
      "linked sites 10398\n",
      "type of linked group 10398\n",
      "total coefficient for linked group 10398\n",
      "error for linked group 10398\n"
     ]
    }
   ],
   "source": [
    "ref_file = os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv')\n",
    "dp.website_file2(os.path.join(INF_DIR, f'infer-{DATA_DATE}-observed.npz'),\n",
    "                 os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "                 os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR, f'selection-observed.csv'),\n",
    "                 ref_file=ref_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inferring selection using only nonsynonymous mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/Python/MPL/epi-inf-fast.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/6-29-20-epidemiological/infer-2022-06-09/infer-g-40.npz blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/2022-06-09/nucleotide-counts.npz blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/2022-06-09/synonymous-prot.npz blee098@cluster.hpcc.ucr.edu: &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-nonsyn.sh blee098@cluster.hpcc.ucr.edu: \n",
      "\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-nonsyn.npz /Users/brianlee/Python/MPL/6-29-20-epidemiological &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-nonsyn-unmasked.npz /Users/brianlee/Python/MPL/6-29-20-epidemiological &&\n",
      "scp blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/nucleotide-counts.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# Inferring the selection coefficients only for nonsynonymous sites \n",
    "inf_script    = 'epi-inf-fast.py'\n",
    "count_file    = 'nucleotide-counts.npz'\n",
    "input_file    = ppath.join(f'infer-{DATA_DATE}', 'infer-g-40.npz')\n",
    "inf_file      = input_file\n",
    "syn_file      = 'synonymous-prot.npz'\n",
    "out_file      = ppath.join(SSH_DATA, f'infer-{DATA_DATE}-nonsyn')\n",
    "\n",
    "job_file2 = 'job-infer-nonsyn.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/inf-nonsyn-error\n",
    "#SBATCH -o ./MPL/out/inf-nonsyn-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "\n",
    "python {inf_script} --data {SSH_DATA}/{input_file} -k 0.1 -R 2 --pop_size 10000 --timed 1 --g1 40  -o {out_file} --syn_data {syn_file} --count_data {SSH_DATA}/{DATA_DATE}/{count_file} --infFile {SSH_DATA}/{inf_file} --eliminateNS --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "input_path = os.path.join(INF_DIR, input_file)\n",
    "syn_path   = os.path.join(DATA_DIR, syn_file)\n",
    "job_path   = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {input_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print(f'scp {syn_path} {SSH_HOME} &&')\n",
    "print(f'scp {job_path} {SSH_HOME} ')\n",
    "print('')\n",
    "print(f'scp {SSH_HOME}{out_file}.npz {INF_DIR} &&')\n",
    "print(f'scp {SSH_HOME}{out_file}-unmasked.npz {INF_DIR} &&')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39500 39500\n"
     ]
    }
   ],
   "source": [
    "# Masking out the rest of the synonymous mutations\n",
    "inf_data  = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-nonsyn-unmasked.npz'), allow_pickle=True)\n",
    "alleles   = inf_data['allele_number']\n",
    "selection = inf_data['selection']\n",
    "errors    = inf_data['error_bars']\n",
    "s_ind     = inf_data['selection_independent']\n",
    "syn_data  = np.load(os.path.join(DATA_DIR, 'synonymous-prot.npz'))\n",
    "syn_muts  = syn_data['nuc_index']\n",
    "types     = syn_data['types']\n",
    "types     = types[np.isin(syn_muts, alleles)]\n",
    "syn_muts  = syn_muts[np.isin(syn_muts, alleles)]\n",
    "new_muts  = []\n",
    "new_types = []\n",
    "for i in range(len(alleles)):\n",
    "    if alleles[i] not in syn_muts:\n",
    "        if alleles[i][-1]=='-':\n",
    "            new_types.append('S')\n",
    "        else:\n",
    "            new_types.append('NS')\n",
    "    else:\n",
    "        new_types.append(types[list(syn_muts).index(alleles[i])])\n",
    "    new_muts.append(alleles[i])\n",
    "types = new_types\n",
    "\n",
    "assert all(np.array(new_muts)==alleles)\n",
    "idxs_keep = np.array(types)=='NS'\n",
    "alleles   = alleles[idxs_keep]\n",
    "selection = selection[idxs_keep]\n",
    "s_ind     = s_ind[idxs_keep]\n",
    "errors    = errors[idxs_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-3-1b00c599459e>:26: RuntimeWarning: invalid value encountered in greater\n",
      "  mut_mask = np.array(mut_traj) > freq_cutoff\n"
     ]
    }
   ],
   "source": [
    "# eliminating sites that don't rise above 5% frequency for at least \"window\" consecutive time points in at least 2 regions\n",
    "window      = 5\n",
    "freq_cutoff = 0.05\n",
    "\n",
    "traj_dir  = os.path.join(DATA_DIR, 'trajectories-nosmooth')\n",
    "traj_muts = []\n",
    "for file in os.listdir(traj_dir):\n",
    "    traj_file = os.path.join(traj_dir, file)\n",
    "    df        = pd.read_csv(traj_file)\n",
    "    df        = df.drop(columns=['time'])\n",
    "    muts_temp = df.columns.values\n",
    "    traj_muts.append(muts_temp)\n",
    "traj_muts   = list(np.unique([traj_muts[i][j] for i in range(len(traj_muts)) for j in range(len(traj_muts[i]))]))\n",
    "freq_mask   = [False for i in range(len(traj_muts))]\n",
    "mask_counts = np.zeros(len(traj_muts))\n",
    "for file in os.listdir(traj_dir):\n",
    "    traj_file = os.path.join(traj_dir, file)\n",
    "    df        = pd.read_csv(traj_file)\n",
    "    df        = df.drop(columns=['time'])\n",
    "    muts_temp = df.columns.values\n",
    "    for site in muts_temp:\n",
    "        mut_traj = list(df[site])\n",
    "        mut_mask = np.array(mut_traj) > freq_cutoff\n",
    "        mut_mask = np.any([np.all(mut_mask[i:i+window]) for i in range(len(mut_traj)-window+1)])\n",
    "        if mut_mask:\n",
    "            mask_counts[traj_muts.index(site)] += 1\n",
    "for i in range(len(mask_counts)):\n",
    "    if mask_counts[i] >= 2:\n",
    "        freq_mask[i] = True\n",
    "\n",
    "mask_new  = [i for i in range(len(alleles)) if freq_mask[traj_muts.index(alleles[i])]]\n",
    "selection = selection[mask_new]\n",
    "s_ind     = s_ind[mask_new]\n",
    "errors    = errors[mask_new]\n",
    "alleles   = alleles[mask_new]\n",
    "\n",
    "np.savez_compressed(os.path.join(INF_DIR, f'inf-data-nonsynonymous-{DATA_DATE}.npz'), allele_number=alleles, \n",
    "                    selection=selection, error_bars=errors, selection_independent=s_ind)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all fields calculated, putting data into dataframe...\n",
      "index 1259\n",
      "nucleotide number 1259\n",
      "protein 1259\n",
      "amino acid number in protein 1259\n",
      "amino acid number in reading frame 1259\n",
      "synonymous 1259\n",
      "amino acid mutation 1259\n",
      "nucleotide 1259\n",
      "reference nucleotide 1259\n",
      "selection coefficient 1259\n",
      "error 1259\n",
      "independent model coefficient 1259\n",
      "linked sites 1259\n",
      "type of linked group 1259\n",
      "total coefficient for linked group 1259\n",
      "error for linked group 1259\n"
     ]
    }
   ],
   "source": [
    "ref_file = os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv')\n",
    "dp.website_file2(os.path.join(INF_DIR, f'inf-data-nonsynonymous-{DATA_DATE}.npz'),\n",
    "                 os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "                 os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR, f'selection-nonsyn-observed.csv'),\n",
    "                 ref_file=ref_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking sensitivity of inference to changing the criteria used to eliminate infrequent sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-trimmed ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/filter-sites-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-filter-sensitivity.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-filter-sensitivity.sh\n"
     ]
    }
   ],
   "source": [
    "# eliminate sites from a time-series if there are less than min_seqs total mutations at this site in all sequences in the time-series\n",
    "min_seqs  = 5\n",
    "max_dir   = ppath.join(SSH_DATA, 'max-freqs')\n",
    "out_dir1  = ppath.join(SSH_DATA, 'genome-filtered-5pct')\n",
    "out_dir2  = ppath.join(SSH_DATA, 'genome-filtered-0.5pct')\n",
    "input_dir = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "script    = 'filter-sites-parallel.py'\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "num_files = 518\n",
    "\n",
    "job_file  = 'job-filter-sensitivity.sh'\n",
    "job_str   = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=3:00:00\n",
    "#SBATCH -e ./MPL/out/filter-sens-error-%a\n",
    "#SBATCH -o ./MPL/out/filter-sens-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {script} --input \\\"$file\\\" -o {out_dir1} --max_dir {max_dir} --min_count {min_seqs} --refFile {REF_FILE + '.csv'} --min_freq 0.05 \n",
    "python {script} --input \\\"$file\\\" -o {out_dir2} --max_dir {max_dir} --min_count {min_seqs} --refFile {REF_FILE + '.csv'} --min_freq 0.005 \n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-233-e82491299f2a>, line 16)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-233-e82491299f2a>\"\u001b[0;36m, line \u001b[0;32m16\u001b[0m\n\u001b[0;31m    num_files1 =\u001b[0m\n\u001b[0m                 ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir1     = ppath.join(SSH_DATA, 'genome-filtered-5pct')\n",
    "input_dir2     = ppath.join(SSH_DATA, 'genome-filtered-0.5pct')\n",
    "freq_script    = 'epi-covar.py'\n",
    "inf_script     = 'epi-inf-parallel.py'\n",
    "temp_dir1      = ppath.join(SSH_DATA, 'freqs-5pct')\n",
    "temp_dir2      = ppath.join(SSH_DATA, 'freqs-0.5pct')\n",
    "out_file1      = ppath.join(SSH_DATA, f'infer-{DATA_DATE}-5pct')\n",
    "out_file2      = ppath.join(SSH_DATA, f'infer-{DATA_DATE}-0.5pct')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir1} ! -name \"*sites*\" -print0)\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "num_files1 = 518\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir2} ! -name \"*sites*\" -print0)\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "num_files2 = 518\n",
    "print('cd Archive-vector')\n",
    "print('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "job_file1     = 'job-find-freqs-sensitivity.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-sens-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-sens-out-%a\n",
    "#SBATCH --array=1-{num_files1-1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-vector\n",
    "\n",
    "readarray -d '' files1 < <(find {input_dir1} ! -name \"*sites*\" -print0)\n",
    "file1=${{files1[$SLURM_ARRAY_TASK_ID]}}\n",
    "python ../{freq_script} --data \\\"$file1\\\" -o {temp_dir1} -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1\n",
    "\n",
    "readarray -d '' files2 < <(find {input_dir2} ! -name \"*sites*\" -print0)\n",
    "file2=${{files2[$SLURM_ARRAY_TASK_ID]}}\n",
    "python ../{freq_script} --data \\\"$file2\\\" -o {temp_dir2} -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "\n",
    "job_file3 = 'job-infer-sensitivity.sh'\n",
    "job_str3  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-sens-error\n",
    "#SBATCH -o ./MPL/out/inf-sens-out\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python {inf_script} --data {temp_dir1} --timed 1 -o {out_file1} -q 5 --g1 40 --refFile {REF_FILE + '.csv'}\n",
    "python {inf_script} --data {temp_dir2} --timed 1 -o {out_file2} -q 5 --g1 40 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file3), 'w')\n",
    "f.write('%s\\n' % job_str3)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path3 = os.path.join(SCRIPT_DIR, job_file3)\n",
    "print(f'scp {script_path3} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file3)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp %s%s %s &&' % (SSH_HOME, out_file1, DATA_DIR))\n",
    "print('scp %s%s %s'    % (SSH_HOME, out_file2, DATA_DIR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Checking sensitivity of inferred coefficients to different window sizes for calculating delta x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-dt-10 &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-dt-20\n",
      "cd Archive-test\n",
      "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-inf-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-test blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-freqs-dt.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-dt.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-find-freqs-dt.sh &&\n",
      "sbatch job-infer-dt.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-06-09 &&\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-* /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-dt-10.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09 &&\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-dt-20.npz /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script    = 'epi-covar.py'\n",
    "inf_script     = 'epi-inf-parallel.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer-{DATA_DATE}')\n",
    "\n",
    "# run on cluster\n",
    "print('cd Archive-vector')\n",
    "print('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "num_files     = 519   # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-freqs-dt.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "cd ./Archive-vector\n",
    "pwd\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "tempout1={temp_dir}-dt-10\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout1 -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --delta_t 10\n",
    "\n",
    "tempout2={temp_dir}-dt-20\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout2 -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --delta_t 20\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer-dt.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=10:00:00\n",
    "#SBATCH -e ./MPL/out/inf-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-out-%a\n",
    "#SBATCH --array=0-1\n",
    "\n",
    "delta_t=(10 20)\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}-dt-${{delta_t[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {inf_script} --data {temp_dir}-dt-${{delta_t[$SLURM_ARRAY_TASK_ID]}} --timed 1 -o $outfile -q 5 --g1 40 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s &&' % (SSH_HOME, out_file + '-dt-10.npz', DATA_DIR))\n",
    "print('scp -r %s%s %s'    % (SSH_HOME, out_file + '-dt-20.npz', DATA_DIR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='errors'></a>\n",
    "# Error analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='bootstrap'></a>\n",
    "## Bootstraping sequence data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "cd Archive-test\n",
      "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar-bootstrap.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-test blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-freqs-bs.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-bs.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-find-freqs-bs.sh &&\n",
      "sbatch job-infer-bs.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-bs\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-bs*.npz /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-bs\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data//rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-bs /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script    = 'epi-covar-bootstrap.py'\n",
    "inf_script     = 'epi-inf-parallel.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs-bootstrap')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer-bs')\n",
    "nruns  = 100\n",
    "sample = 20\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('cd Archive-vector')\n",
    "print('g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "num_files     = 518    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-freqs-bs.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=3-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-bs-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-bs-out-%a\n",
    "#SBATCH --array=1-{num_files - 1}\n",
    "#SBATCH -p highmem\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-vector\n",
    "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "pwd\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --nruns {nruns} --bsSample {sample}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer-bs.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=75G\n",
    "#SBATCH --time=3-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-out-%a\n",
    "#SBATCH -p highmem\n",
    "#SBATCH --array=0-{nruns-1}\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "dirs=({temp_dir}/*)\n",
    "dir=${{dirs[$SLURM_ARRAY_TASK_ID]}}\n",
    "outfile={out_file}/$SLURM_ARRAY_TASK_ID\n",
    "python {inf_script} --data $dir --timed 1 -o $outfile -q 5 --g1 40 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s/%s %s' % (SSH_HOME, SSH_DATA, out_file, INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nmask1 = np.isin(muts1, muts2)\\nmuts1 = muts1[mask1]\\ns1    = s1[mask1]\\nmask2 = np.isin(muts2, muts1)\\nmuts2 = muts2[mask2]\\ns2    = s2[mask2]\\n\\ndata = {'directory' : bs_dir,\\n        'out_file'  : out_file}\\ndp.find_bootstrap_stats(**data)\\n\\nfig, ax = plt.subplots(1,1, figsize=[10,10])\\nax.set_xlim(-0.04, 0.08)\\nax.set_ylim(-0.04, 0.08)\\nax.scatter(s1, s2, marker='.')\\nax.plot(np.linspace(np.amin(np.append(s1, s2)), np.amax(np.append(s1, s2)), 100), np.linspace(np.amin(np.append(s1, s2)), np.amax(np.append(s1, s2)), 100))\\nax.set_xlabel('Subsampled data')\\nax.set_ylabel('Full data')\\n\""
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = os.path.join(DATA_DIR, f'infer-bs')\n",
    "for file in os.listdir(data_dir):\n",
    "    if 'covar' in file or '.csv' in file:\n",
    "        os.remove(os.path.join(data_dir, file))\n",
    "\n",
    "data2 = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}.npz'))\n",
    "muts2 = data2['allele_number']\n",
    "s2    = data2['selection']\n",
    "muts  = []\n",
    "for file in os.listdir(data_dir):\n",
    "    muts.append(np.load(os.path.join(data_dir, file), allow_pickle=True)['allele_number'])\n",
    "muts_all = [muts[i][j] for i in range(len(muts)) for j in range(len(muts[i]))]\n",
    "muts_new = [i for i in muts2 if i in muts_all]\n",
    "\n",
    "out_file = os.path.join(SARS_DIR, DATA_DATE, 'bootstrap-stats')\n",
    "data = {'directory'    : data_dir,\n",
    "        'out_file'     : out_file,\n",
    "        'mutant_sites' : muts_new}\n",
    "dp.find_bootstrap_stats(**data)\n",
    "df = pd.read_csv(out_file + '.csv')\n",
    "errors = list(df['standard_deviation'])\n",
    "\n",
    "#var_out   = os.path.join(SARS_DIR, DATA_DATE, 'variant-bootstrap-stats')\n",
    "#var_names = ['Alpha', 'Gamma', 'Delta', 'B.1',   '20E-EU1', 'Beta',   'Lambda', 'Epsilon', 'BA.4', 'BA.5', 'BA.2',  'BA.1', 'BA.2.12.1', 'Mu']\n",
    "#linked    = [  alpha,   gamma,   delta, D614G, twentyE_EU1,   beta, lambda_new,   epsilon,    ba4,    ba5,    ba2, omicron,      ba2121,   mu]\n",
    "#var_data  = {\n",
    "#    'directory'     : data_dir,\n",
    "#    'out_file'      : var_out,\n",
    "#    'linked_sites'  : linked,\n",
    "#    'variant_names' : var_names}\n",
    "#dp.bootstrap_variant_stats(**var_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all fields calculated, putting data into dataframe...\n",
      "index 1259\n",
      "nucleotide number 1259\n",
      "protein 1259\n",
      "amino acid number in protein 1259\n",
      "amino acid number in reading frame 1259\n",
      "synonymous 1259\n",
      "amino acid mutation 1259\n",
      "nucleotide 1259\n",
      "reference nucleotide 1259\n",
      "selection coefficient 1259\n",
      "standard deviation 1259\n",
      "error 1259\n",
      "independent model coefficient 1259\n",
      "linked sites 1259\n",
      "type of linked group 1259\n",
      "total coefficient for linked group 1259\n",
      "error for linked group 1259\n"
     ]
    }
   ],
   "source": [
    "# regular selection for old data\n",
    "ref_file = os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv')\n",
    "dp.website_file2(os.path.join(INF_DIR, f'inf-data-nonsynonymous-{DATA_DATE}.npz'),\n",
    "                 os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "                 os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR, f'selection-nonsyn-bootstrap.csv'),\n",
    "                 ref_file=ref_file,\n",
    "                 bootstrap_file=os.path.join(DATA_DIR, 'bootstrap-stats.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=section_subsample_regions></a>\n",
    "## Subsampling regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-inf-parallel.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-subsample.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-subsample &&\n",
      "sbatch job-infer-subsample.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-subsample /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "inf_script = 'epi-inf-parallel.py'\n",
    "temp_dir   = ppath.join(SSH_DATA, 'freqs')\n",
    "out_dir    = ppath.join(SSH_DATA, f'infer-subsample')\n",
    "\n",
    "job_file2 = 'job-infer-subsample.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=75G\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH -e ./MPL/out/inf-subsample-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-subsample-out-%a\n",
    "#SBATCH --array=0-100\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_dir}/$SLURM_ARRAY_TASK_ID\n",
    "python {inf_script} --data {temp_dir} --timed 1 -o $outfile -q 5 --g1 40 --regPCT 80 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print(f'scp -r {SSH_HOME}{out_dir} {INF_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nmask1 = np.isin(muts1, muts2)\\nmuts1 = muts1[mask1]\\ns1    = s1[mask1]\\nmask2 = np.isin(muts2, muts1)\\nmuts2 = muts2[mask2]\\ns2    = s2[mask2]\\n\\ndata = {'directory' : bs_dir,\\n        'out_file'  : out_file}\\ndp.find_bootstrap_stats(**data)\\n\\nfig, ax = plt.subplots(1,1, figsize=[10,10])\\nax.set_xlim(-0.04, 0.08)\\nax.set_ylim(-0.04, 0.08)\\nax.scatter(s1, s2, marker='.')\\nax.plot(np.linspace(np.amin(np.append(s1, s2)), np.amax(np.append(s1, s2)), 100), np.linspace(np.amin(np.append(s1, s2)), np.amax(np.append(s1, s2)), 100))\\nax.set_xlabel('Subsampled data')\\nax.set_ylabel('Full data')\\n\""
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = os.path.join(DATA_DIR, f'infer-subsample')\n",
    "for file in os.listdir(data_dir):\n",
    "    if 'covar' in file or '.csv' in file:\n",
    "        os.remove(os.path.join(data_dir, file))\n",
    "\n",
    "data2 = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}.npz'))\n",
    "muts2 = data2['allele_number']\n",
    "s2    = data2['selection']\n",
    "muts  = []\n",
    "for file in os.listdir(data_dir):\n",
    "    muts.append(np.load(os.path.join(data_dir, file), allow_pickle=True)['allele_number'])\n",
    "muts_all = [muts[i][j] for i in range(len(muts)) for j in range(len(muts[i]))]\n",
    "muts_new = [i for i in muts2 if i in muts_all]\n",
    "\n",
    "out_file = os.path.join(DATA_DIR, 'subsample-stats')\n",
    "data = {'directory'    : data_dir,\n",
    "        'out_file'     : out_file,\n",
    "        'mutant_sites' : muts_new}\n",
    "dp.find_bootstrap_stats(**data)\n",
    "df = pd.read_csv(out_file + '.csv')\n",
    "errors = list(df['standard_deviation'])\n",
    "\n",
    "var_out   = os.path.join(SARS_DIR, DATA_DATE, 'variant-subsample-stats')\n",
    "var_names = ['Alpha', 'Gamma', 'Delta', 'B.1',   '20E-EU1', 'Beta',   'Lambda', 'Epsilon', 'BA.4', 'BA.5', 'BA.2',  'BA.1', 'BA.2.12.1']\n",
    "linked    = [  alpha,   gamma,   delta, D614G, twentyE_EU1,   beta, lambda_new,   epsilon,    ba4,    ba5,    ba2, omicron,      ba2121]\n",
    "var_data  = {\n",
    "    'directory'     : data_dir,\n",
    "    'out_file'      : var_out,\n",
    "    'linked_sites'  : linked,\n",
    "    'variant_names' : var_names}\n",
    "dp.bootstrap_variant_stats(**var_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all fields calculated, putting data into dataframe...\n",
      "index 1259\n",
      "nucleotide number 1259\n",
      "protein 1259\n",
      "amino acid number in protein 1259\n",
      "amino acid number in reading frame 1259\n",
      "synonymous 1259\n",
      "amino acid mutation 1259\n",
      "nucleotide 1259\n",
      "reference nucleotide 1259\n",
      "selection coefficient 1259\n",
      "standard deviation 1259\n",
      "error 1259\n",
      "independent model coefficient 1259\n",
      "linked sites 1259\n",
      "type of linked group 1259\n",
      "total coefficient for linked group 1259\n",
      "error for linked group 1259\n"
     ]
    }
   ],
   "source": [
    "# regular selection for old data\n",
    "ref_file = os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv')\n",
    "dp.website_file2(os.path.join(INF_DIR, f'inf-data-nonsynonymous-{DATA_DATE}.npz'),\n",
    "                 os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "                 os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR, f'selection-nonsyn-subsample.csv'),\n",
    "                 ref_file=ref_file,\n",
    "                 bootstrap_file=os.path.join(DATA_DIR, 'subsample-stats.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name\tn dupes\ttotal s\n",
      "NSP6-L105-\t3\t-0.097\n",
      "NSP6-S106-\t3\t0.150\n",
      "S-P25-\t3\t0.040\n",
      "S-G142-\t2\t0.135\n",
      "S-G142D\t2\t0.064\n",
      "S-V143-\t3\t0.087\n",
      "S-Y144-\t3\t-0.052\n",
      "S-F157-\t3\t0.036\n",
      "S-N440K\t2\t0.078\n",
      "S-N679K\t2\t0.049\n",
      "ORF6-D61L\t3\t-0.052\n",
      "ORF8-F120-\t3\t0.044\n",
      "N-D3L\t3\t0.084\n",
      "N-E31-\t3\t0.073\n",
      "N-R32-\t3\t0.072\n",
      "N-R203K\t2\t0.047\n"
     ]
    }
   ],
   "source": [
    "# Find mutations that are in the same codon and strongly linked to one another and sum their selection coefficients\n",
    "data  = np.load(os.path.join(SARS_DIR, DATA_DATE, 'linked-sites.npy'), allow_pickle=True)\n",
    "nucs  = [np.array([data[i][j][-1] for j in range(len(data[i]))]) for i in range(len(data))]\n",
    "idxs  = []\n",
    "for i in range(len(data)):\n",
    "    if len(np.array(nucs[i])[nucs[i]=='-']) != len(nucs[i]):\n",
    "        idxs.append(i)\n",
    "data  = data[idxs]\n",
    "split = [[data[i][j].split('-') for j in range(len(data[i]))] for i in range(len(data))]\n",
    "cods  = [[split[i][j][1] for j in range(len(split[i]))] for i in range(len(split))]\n",
    "prots = [[split[i][j][0] for j in range(len(split[i]))] for i in range(len(split))]\n",
    "    \n",
    "#linked_same_codon = []\n",
    "linked_groups     = []\n",
    "for i in range(len(data)):\n",
    "    group   = data[i]\n",
    "    g_cods  = cods[i]\n",
    "    g_prots = prots[i]\n",
    "    prot_unique = np.unique(g_prots)\n",
    "    for j in range(len(prot_unique)):\n",
    "        temp_cods   = np.array(g_cods)[np.array(g_prots)==prot_unique[j]]\n",
    "        temp_labs   = np.array(group)[np.array(g_prots)==prot_unique[j]]\n",
    "        cods_unique = np.unique(temp_cods)\n",
    "        for k in range(len(cods_unique)):\n",
    "            labs_same_cod = temp_labs[temp_cods==cods_unique[k]]\n",
    "            if len(labs_same_cod)>1:\n",
    "                linked_groups.append(labs_same_cod)\n",
    "\n",
    "linked_groups     = [[linked_groups[i][j] for j in range(len(linked_groups[i])) if linked_groups[i][j][-1]!='-'] for i in range(len(linked_groups))]\n",
    "linked_groups     = [i for i in linked_groups if len(i)!=0]\n",
    "linked_all        = [linked_groups[i][j] for i in range(len(linked_groups)) for j in range(len(linked_groups[i]))]\n",
    "\n",
    "# collapse identical mutations and sum contributions\n",
    "df_sel = pd.read_csv(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-subsample.csv'), memory_map=True)\n",
    "\n",
    "f = open(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-subsample-collapsed.csv'), 'w')\n",
    "\n",
    "headers = list(df_sel.columns)\n",
    "f.write('%s\\n' % ','.join(headers))\n",
    "\n",
    "print('name\\tn dupes\\ttotal s')\n",
    "\n",
    "dup_names = []\n",
    "for iter, entry in df_sel.iterrows():\n",
    "    if entry['protein']=='NC':\n",
    "        continue\n",
    "    loc      = str(entry['amino acid number in protein'])\n",
    "    anc      = entry['amino acid mutation'][0]\n",
    "    mut      = entry['amino acid mutation'][-1]\n",
    "    var_name = str(entry['protein']) + '-' + anc + loc + mut\n",
    "    label    = dp.get_label_new(str(entry['nucleotide number']) + '-' + entry['nucleotide'])\n",
    "    if var_name in dup_names:\n",
    "        continue\n",
    "    \n",
    "    df_same = df_sel[(df_sel['protein']==entry['protein']) \n",
    "                     & (df_sel['amino acid number in protein']==entry['amino acid number in protein'])\n",
    "                     & (df_sel['amino acid mutation']==entry['amino acid mutation'])]\n",
    "    \n",
    "    s = entry['selection coefficient']\n",
    "    if len(df_same)>1:\n",
    "        s = np.sum(df_same['selection coefficient'])\n",
    "        if var_name[-1]!='-' and label not in linked_all:\n",
    "            s = np.max(df_same['selection coefficient'])\n",
    "        dup_names.append(var_name)\n",
    "        if np.fabs(s)>0.03:\n",
    "            print('%s\\t%d\\t%.3f' % (var_name, len(df_same), s))\n",
    "        \n",
    "    for i in range(len(headers)):\n",
    "        if headers[i]=='selection coefficient':\n",
    "            f.write('%s' % str(s))\n",
    "        elif headers[i]=='linked sites':\n",
    "            f.write('\"%s\"' % str(entry[headers[i]]))\n",
    "#         elif headers[i]=='nucleotide number':\n",
    "#             f.write('%s' % entry[headers[i]][:-2])\n",
    "        else:\n",
    "            f.write('%s' % str(entry[headers[i]]))\n",
    "        if i!=len(headers)-1:\n",
    "            f.write(',')\n",
    "    f.write('\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "# Change nucleotide indices to be indexed starting with 1 instead of 0 for publishing\n",
    "df = pd.read_csv(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-subsample-collapsed.csv'), index_col=False)\n",
    "nucs = list(df['nucleotide number'])\n",
    "new_nucs = []\n",
    "for i in nucs:\n",
    "    num, letters = dp.separate_label_idx(str(i))\n",
    "    if letters=='':\n",
    "        new_nucs.append(str(int(num) + 1))\n",
    "    else:\n",
    "        new_nucs.append(str(int(num) + 1) + letters)\n",
    "nucs = new_nucs\n",
    "df['nucleotide number'] = nucs\n",
    "df.to_csv(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-subsample-corrected.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NSP5-8-1-A\n",
      "NSP2-78-2-C\n",
      "NSP13-574-0-C\n",
      "NSP13-574-1-T\n",
      "NSP6-20-2-A\n",
      "S-532-0-C\n",
      "NSP3-210-0-C\n",
      "NSP8-134-2-G\n",
      "NSP4-320-0-G\n",
      "NSP14-376-1-C\n",
      "NSP3-1181-2-G\n",
      "NSP14-313-1-T\n",
      "NSP2-367-2-A\n",
      "NSP3-1451-1-C\n",
      "S-234-2-G\n",
      "S-235-0-C\n",
      "S-235-1-A\n",
      "S-236-1-T\n",
      "S-236-2-A\n",
      "NSP2-487-0-G\n",
      "M-118-2-A\n",
      "ORF8-3-1-A\n",
      "N-209-0-G\n",
      "NSP3-332-0-A\n",
      "NSP3-332-1-A\n",
      "NSP3-570-1-C\n",
      "NSP3-614-2-C\n",
      "NSP3-616-0-T\n",
      "NSP3-616-1-C\n",
      "NSP3-623-0-G\n",
      "NSP3-623-1-C\n",
      "NSP3-624-2-A\n",
      "NSP3-627-1-G\n",
      "NSP3-628-1-A\n",
      "NSP3-628-2-A\n",
      "NSP3-629-0-T\n",
      "NSP3-629-1-T\n"
     ]
    }
   ],
   "source": [
    "# Variant coefficients \n",
    "args = {\n",
    "    'infer_file'    : os.path.join(INF_DIR, f'infer-{DATA_DATE}-observed-g-40.npz'),\n",
    "    'link_file'     : os.path.join(DATA_DIR, 'linked-sites.npy'),\n",
    "    'syn_prot_file' : os.path.join(DATA_DIR, 'synonymous-prot.npz'),\n",
    "    'variant_file'  : os.path.join(SARS_DIR, 'variants-all.npy'),\n",
    "    'bootstrap_dir' : os.path.join(DATA_DIR, f'infer-subsample'),\n",
    "    'out_file'      : os.path.join(DATA_DIR, 'linked-coefficients-new.csv'),\n",
    "    'ref_file'      : os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv')\n",
    "}\n",
    "\n",
    "dp.linked_analysis_alt(**args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fixing nucleotide index to start at 1\n",
    "df = pd.read_csv(os.path.join(DATA_DIR, 'linked-coefficients-new.csv'), index_col=False)\n",
    "nucs = list(df['nucleotides'])\n",
    "nucs = [i.split('/') for i in nucs]\n",
    "nucs_new = []\n",
    "for i in nucs:\n",
    "    nucs_temp = []\n",
    "    for j in i:\n",
    "        num, letters = dp.separate_label_idx(str(j))\n",
    "        if j=='not_present':\n",
    "            nucs_temp.append(j)\n",
    "        elif letters=='':\n",
    "            nucs_temp.append(str(int(num) + 1))\n",
    "        else:\n",
    "            nucs_temp.append(str(int(num) + 1) + letters)\n",
    "    nucs_new.append(nucs_temp)\n",
    "nucs = ['/'.join(i) for i in nucs_new]\n",
    "df['nucleotides'] = nucs\n",
    "df = df.rename(columns={'standard deviation' : 'standard_deviation'})\n",
    "df.to_csv(os.path.join(DATA_DIR, 'linked-coefficients-new-paper.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a csv file containing the frequency trajectories for the different variants in different regions\n",
    "delta_file  = os.path.join(DATA_DIR, 'delta-traj.csv')\n",
    "ba2_file    = os.path.join(DATA_DIR, 'ba2-traj.csv')\n",
    "lambda_file = os.path.join(DATA_DIR, 'lambda-traj.csv')\n",
    "dp.linked_traj_csv(os.path.join(DATA_DIR, 'linked-coefficients-new.csv'), \n",
    "                   os.path.join(DATA_DIR, 'trajectories'), \n",
    "                   os.path.join(DATA_DIR, 'linked-trajectories-new'),\n",
    "                   delta_file=delta_file,\n",
    "                   ba2_file=ba2_file,\n",
    "                   lambda_file=lambda_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=early_detection></a>\n",
    "# Early Detection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=null_distribution></a>\n",
    "## Finding Null Distribution and Early Detection of Linked Groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-tv\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-tv\n",
      "chdir Archive-tv-int\n",
      "g++ src/main-test.cpp src/inf-test.cpp src/io-test.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
      "\n",
      "ls /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-regional | wc -l\n",
      "scp /Users/brianlee/Python/MPL/epi-covar.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel-tv.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-tv-int blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-freqs-tv.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-move-data.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-tv.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-find-freqs-tv.sh &&\n",
      "sbatch job-move-data.sh &&\n",
      "sbatch job-infer-tv.sh &&\n",
      "\n",
      "scp /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-tv /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "input_dir     = ppath.join(SSH_DATA, 'genome-filtered')\n",
    "freq_script   = 'epi-covar.py'\n",
    "move_script   = 'make-dirs-tv-inf.py'\n",
    "inf_script    = 'epi-inf-parallel-tv.py'\n",
    "temp_dir      = ppath.join(SSH_DATA, 'freqs-tv')\n",
    "out_file      = ppath.join(SSH_DATA, f'infer-tv')\n",
    "freqs_dir     = ppath.join(SSH_DATA, 'freqs-regional')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print(f'chdir Archive-tv-int')\n",
    "print('g++ src/main-test.cpp src/inf-test.cpp src/io-test.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "num_files     = 517    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-freqs-tv.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-tv-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-tv-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-tv-int\n",
    "pwd\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "outfile={temp_dir}\n",
    "\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $outfile -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --tv_inference --delta_t 15\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file3 = 'job-move-data.sh'\n",
    "job_str3  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH -e ./MPL/out/move-error\n",
    "#SBATCH -o ./MPL/out/move-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python ./{move_script} --input {temp_dir} -o {freqs_dir} \n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file3), 'w')\n",
    "f.write('%s\\n' % job_str3)\n",
    "f.close()\n",
    "\n",
    "print(f'ls {freqs_dir} | wc -l') \n",
    "\n",
    "job_file2 = 'job-infer-tv.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-tv-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-tv-out-%a\n",
    "#SBATCH --array=0-130\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "dirs=({freqs_dir}/*)\n",
    "dir=${{dirs[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "tempout1=$(basename \"$dir\")\n",
    "tempout2={out_file}/\n",
    "outfile=\"$tempout2$tempout1\"\n",
    "\n",
    "python {inf_script} --data \\\"$dir\\\" --timed 1 -o \\\"$outfile\\\" -q 5 --g1 10 --delta_t 15 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "job_files = [job_file1, job_file2, job_file3]\n",
    "for file in job_files:\n",
    "    job_path = os.path.join(SCRIPT_DIR, file)\n",
    "    print(f'scp {job_path} {SSH_HOME} &&')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "for file in job_files:\n",
    "    print(f'sbatch {job_file}')\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp %s %s' % (out_file, INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inference-europe-belgium---2-29-5-18.npz -0.05975845691275847 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "infer-tv-uk-2021-10-10.npz 0.05949373801986084 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-europe-spain---2-26-4-30.npz 0.06104866981184767 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-north-america-usa-washington---2-24-7-7.npz 0.09881704555828634 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "max null distribution = 0.09881704555828634\n",
      "inference-europe-belgium---2-29-5-18.npz -0.05975845691275847 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "infer-tv-uk-2021-10-10.npz 0.05949373801986084 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-north-america-canada-quebec---3-10-4-14.npz 0.0538696059213304 -0.008148500370654205 N-204-0\n",
      "inference-europe-spain---2-26-4-30.npz 0.06104866981184767 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-oceania-australia---6-9-9-26.npz 0.058996962233346634 ['S-613-2', 'NSP3-1607-2']\n",
      "inference-north-america-usa-michigan---10-11-11-13.npz 0.07047688041334618 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-minnesota---3-9-4-18.npz -0.06286156097614015 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-washington---2-24-7-7.npz 0.11796985612157637 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-washington---2-24-7-7.npz 0.09881704555828634 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "max null distribution = 0.11796985612157637\n",
      "inference-europe-belgium---2-29-5-18.npz -0.05975845691275847 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "infer-tv-uk-2021-10-10.npz 0.13183354155350271 ['NSP16-199-2', 'NSP1-60-2', 'NSP3-1189-2', 'M-93-2', 'N-220-1', 'ORF10-30-0']\n",
      "infer-tv-uk-2021-10-10.npz 0.05949373801986084 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-north-america-canada-quebec---3-10-4-14.npz 0.0538696059213304 -0.008148500370654205 N-204-0\n",
      "inference-europe-spain---2-26-4-30.npz 0.06104866981184767 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "inference-oceania-australia---6-9-9-26.npz 0.058996962233346634 ['S-613-2', 'NSP3-1607-2']\n",
      "inference-north-america-usa-michigan---10-11-11-13.npz 0.07047688041334618 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-minnesota---3-9-4-18.npz -0.06286156097614015 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-washington---2-24-7-7.npz 0.11796985612157637 ['NSP5-89-0', 'NSP14-129-0', 'ORF3a-172-1', 'ORF8-24-1', 'N-67-0']\n",
      "inference-north-america-usa-washington---2-24-7-7.npz 0.09881704555828634 ['N-203-1', 'N-203-2', 'N-204-0']\n",
      "max null distribution = 0.13183354155350271\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIQAAAI/CAYAAAAGDwK6AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAiGUlEQVR4nO3df5Dt913X8dfbXluwjIcfLQhJrxu8oXoZHNEl1cEfHX6UxOsSplRN8Ac/YtY6U2eUQVmmzACCcmFQmA5BZkc6ZfyRGDqoWW8wQrWGwYymrQgJIfQSLuRGNC1ljlNQQujbP/Z0WK73prt7zu537/08HjM7yX7POd/zvvncu9l93u+P6u4AAAAAMI7fNfUAAAAAABwvQQgAAABgMIIQAAAAwGAEIQAAAIDBCEIAAAAAgxGEAAAAAAZzauoBkuRVr3pVr62tTT0GAAAAwA3jve997we7+9VXe+xEBKG1tbW85z3vmXoMAAAAgBtGVf3itR5zyhgAAADAYAQhAAAAgMEIQgAAAACDmTQIVdVGVW3P5/MpxwAAAAAYyqRBqLt3untzNptNOQYAAADAUJwyBgAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBuMsYAAAAwGDcZQwAAABgME4ZAwAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwbjtPAAAAMBg3HYeAAAAYDBOGQMAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMGcmnoAAJjK2taFle7v0vlzK90fAAAclUmPEKqqjarans/nU44BAAAAMJRJg1B373T35mw2m3IMAAAAgKG4hhAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYCYNQlW1UVXb8/l8yjEAAAAAhjJpEOrune7enM1mU44BAAAAMBSnjAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBCEIAAAAAgxGEAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBTBqEqmqjqrbn8/mUYwAAAAAMZdIg1N073b05m82mHAMAAABgKE4ZAwAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgVh6Equr1VfXjVfX9VfX6Ve8fAAAAgOXsKwhV1dur6vmqeuKK7bdX1dNVdbGqthabO8mHk3xcksurHRcAAACAZe33CKF3JLl974aqelmS+5LckeRskrur6mySH+/uO5J8fZJvWd2oAAAAAKzCvoJQdz+a5ENXbL4tycXufqa7X0jyQJI7u/sji8d/NckrVjYpAAAAACtxaonX3pTk2T2fX07yuqp6Y5IvSfKJSb73Wi+uqs0km0ly+vTpJcYAAAAA4CCWCUJX1d0/nOSH9/G87STbSbK+vt6rngMAAACAq1vmLmPPJXnNns9vXmwDAAAA4ARbJgg9nuTWqrqlql6e5K4kD61mLAAAAACOyn5vO39/kseSvLaqLlfVPd39YpK3JHkkyVNJHuzuJw/y5lW1UVXb8/n8oHMDAAAAcEj7uoZQd999je0PJ3n4sG/e3TtJdtbX1+897D4AAAAAOJhlThkDAAAA4DokCAEAAAAMRhACAAAAGMykQchFpQEAAACO36RBqLt3untzNptNOQYAAADAUJwyBgAAADAYQQgAAABgMIIQAAAAwGBcVBoAAABgMC4qDQAAADAYp4wBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDDuMgYAAAAwGHcZAwAAABiMU8YAAAAABnNq6gEA4EaxtnVh5fu8dP7cyvcJAACOEAIAAAAYjCAEAAAAMBhBCAAAAGAwbjsPAAAAMBi3nQcAAAAYjFPGAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAmDUJVtVFV2/P5fMoxAAAAAIYyaRDq7p3u3pzNZlOOAQAAADAUp4wBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADGbSIFRVG1W1PZ/PpxwDAAAAYCiTBqHu3unuzdlsNuUYAAAAAENxyhgAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYzKRBqKo2qmp7Pp9POQYAAADAUCYNQt29092bs9lsyjEAAAAAhuKUMQAAAIDBCEIAAAAAgxGEAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBCEIAAAAAgxGEAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBCEIAAAAAgxGEAAAAAAYjCAEAAAAM5kiCUFW9sqreU1V/7ij2DwAAAMDh7SsIVdXbq+r5qnriiu23V9XTVXWxqrb2PPT1SR5c5aAAAAAArMZ+jxB6R5Lb926oqpcluS/JHUnOJrm7qs5W1Rcn+Zkkz69wTgAAAABW5NR+ntTdj1bV2hWbb0tysbufSZKqeiDJnUk+IckrsxuJ/k9VPdzdH1ndyAAAAAAsY19B6BpuSvLsns8vJ3ldd78lSarqq5J88FoxqKo2k2wmyenTp5cYAwAAAICDOLK7jHX3O7r7377E49vdvd7d669+9auPagwAAAAArrBMEHouyWv2fH7zYhsAAAAAJ9gyQejxJLdW1S1V9fIkdyV5aDVjAQAAAHBU9nvb+fuTPJbktVV1uaru6e4Xk7wlySNJnkryYHc/eZA3r6qNqtqez+cHnRsAAACAQ9rvXcbuvsb2h5M8fNg37+6dJDvr6+v3HnYfAAAAABzMkV1UGgAAAICTaZnbzgMAR2xt68JK93fp/LmV7g8AgOuTI4QAAAAABjNpEHJRaQAAAIDjN2kQ6u6d7t6czWZTjgEAAAAwFKeMAQAAAAxGEAIAAAAYjCAEAAAAMBgXlQYAAAAYjItKAwAAAAzGKWMAAAAAgxGEAAAAAAZzauoBAGC/1rYuTD0CAADcEBwhBAAAADAYdxkDAAAAGIy7jAEAAAAMxiljAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGLedBwAAABiM284DAAAADMYpYwAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwkwahqtqoqu35fD7lGAAAAABDmTQIdfdOd2/OZrMpxwAAAAAYilPGAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBCEIAAAAAgzk19QAAwPFZ27qw0v1dOn9upfsDAOB4THqEUFVtVNX2fD6fcgwAAACAoUwahLp7p7s3Z7PZlGMAAAAADMU1hAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGc2rqAQC4Ma1tXZh6BAAA4BocIQQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwUwahKpqo6q25/P5lGMAAAAADGXSINTdO929OZvNphwDAAAAYChOGQMAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAzm1NQDAADXr7WtCyvd36Xz51a6PwAArs4RQgAAAACDEYQAAAAABiMIAQAAAAxm5UGoqv5QVX1/Vb2zqv7GqvcPAAAAwHL2FYSq6u1V9XxVPXHF9tur6umqulhVW0nS3U9195uT/IUkn7/6kQEAAABYxn6PEHpHktv3bqiqlyW5L8kdSc4mubuqzi4e+9IkF5I8vLJJAQAAAFiJfQWh7n40yYeu2Hxbkovd/Ux3v5DkgSR3Lp7/UHffkeQvrXJYAAAAAJZ3aonX3pTk2T2fX07yuqp6fZI3JnlFXuIIoaraTLKZJKdPn15iDAAAAAAOYpkgdFXd/e4k797H87aTbCfJ+vp6r3oOAAAAAK5umbuMPZfkNXs+v3mxDQAAAIATbJkjhB5PcmtV3ZLdEHRXkq9YyVQAHLu1rQtTjwAAAByT/d52/v4kjyV5bVVdrqp7uvvFJG9J8kiSp5I82N1PHuTNq2qjqrbn8/lB5wYAAADgkPZ1hFB3332N7Q9niVvLd/dOkp319fV7D7sPAAAAAA5mmWsIAQAAAHAdEoQAAAAABiMIAQAAAAxm0iDkotIAAAAAx2/SINTdO929OZvNphwDAAAAYChOGQMAAAAYjCAEAAAAMBhBCAAAAGAwp6Z886raSLJx5syZKccAuC6tbV2YegQAAOA65aLSAAAAAINxyhgAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABjNpEKqqjarans/nU44BAAAAMJRJbzvf3TtJdtbX1++dcg4A4GRY27qw8n1eOn9u5fsEALjeOWUMAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAG47bzAAAAAIOZNAh19053b85msynHAAAAABiKU8YAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYE5N+eZVtZFk48yZM1OOAdwA1rYurHR/l86fW+n+AAAATpJJjxDq7p3u3pzNZlOOAQAAADAUp4wBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADGbSIFRVG1W1PZ/PpxwDAAAAYCiTBqHu3unuzdlsNuUYAAAAAENxyhgAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAzm1NQDAACMbG3rwsr3een8uZXvEwC4sThCCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABuMuYwDADW3Vd/FyBy8A4EbgCCEAAACAwQhCAAAAAIOZ9JSxqtpIsnHmzJkpxwA4Fqs+bQUAAOCwJj1CqLt3untzNptNOQYAAADAUFxUGuAqHM0DAADcyFxDCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDcVFpAIADcNF5AOBG4AghAAAAgMEIQgAAAACDccoYMAmnXAAAAEzHEUIAAAAAg3GEEADADWbVR2FeOn9upfsDAKbnCCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwZw6ip1W1ZclOZfk9yb5ge7+90fxPgAAHL21rQsr3d+l8+dWuj8A4OD2fYRQVb29qp6vqieu2H57VT1dVReraitJuvtfd/e9Sd6c5C+udmQAAAAAlnGQU8bekeT2vRuq6mVJ7ktyR5KzSe6uqrN7nvKNi8cBAAAAOCH2HYS6+9EkH7pi821JLnb3M939QpIHktxZu74jyY909/tWNy4AAAAAy1r2otI3JXl2z+eXF9v+ZpIvSvKmqnrz1V5YVZtV9Z6qes8HPvCBJccAAAAAYL+O5KLS3f22JG/7GM/ZTrKdJOvr630UcwAAAADw/1v2CKHnkrxmz+c3L7YBAAAAcEIte4TQ40lurapbshuC7kryFUtPBQDADctt7AFgege57fz9SR5L8tqqulxV93T3i0nekuSRJE8lebC7nzzAPjeqans+nx90bgAAAAAOad9HCHX33dfY/nCShw/z5t29k2RnfX393sO8HgAAAICDW/YaQgAAAABcZwQhAAAAgMEIQgAAAACDmTQIuag0AAAAwPGbNAh19053b85msynHAAAAABiKU8YAAAAABrPv284D41rbujD1CAAAAKyQI4QAAAAABuOi0gAAAACDmfSUse7eSbKzvr5+75RzAABw/TqKU5svnT+38n0CwEnilDEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADMZdxgAAAAAGM2kQ6u6d7t6czWZTjgEAAAAwFKeMAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYNx2HgAAAGAwbjsPAAAAMBinjAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGAEIQAAAIDBTBqEqmqjqrbn8/mUYwAAAAAMZdIg1N073b05m82mHAMAAABgKE4ZAwAAABiMIAQAAAAwmFNTDwAAACfN2taFle7v0vlzK90fACzLEUIAAAAAgxGEAAAAAAYjCAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMJMGoaraqKrt+Xw+5RgAAAAAQ5k0CHX3TndvzmazKccAAAAAGIpTxgAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwZyaegAAAGB6a1sXVrq/S+fPrXR/q54vWf2MANcTRwgBAAAADEYQAgAAABiMU8bgEE76IdUAAADwUhwhBAAAADAYQQgAAABgMIIQAAAAwGAmDUJVtVFV2/P5fMoxAAAAAIYyaRDq7p3u3pzNZlOOAQAAADAUp4wBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwmFNTDwAka1sXph4BALiO+N4BgGU5QggAAABgMIIQAAAAwGCcMgYAAKyc09oATjZHCAEAAAAMRhACAAAAGIxTxrjhOVwZAICrWfX3iZfOn1vp/gCOkiOEAAAAAAYjCAEAAAAMZuVBqKo+s6p+oKreuep9AwAAALC8fQWhqnp7VT1fVU9csf32qnq6qi5W1VaSdPcz3X3PUQwLAAAAwPL2e4TQO5LcvndDVb0syX1J7khyNsndVXV2pdMBAAAAsHL7CkLd/WiSD12x+bYkFxdHBL2Q5IEkd654PgAAAABWbJnbzt+U5Nk9n19O8rqq+pQkfz/J51bVN3T3t1/txVW1mWQzSU6fPr3EGAAAAMCo1rYurHR/l86fW+n+TqplgtBVdfevJHnzPp63nWQ7SdbX13vVcwAAAABwdcvcZey5JK/Z8/nNi20AAAAAnGDLBKHHk9xaVbdU1cuT3JXkodWMBQAAAMBR2e9t5+9P8liS11bV5aq6p7tfTPKWJI8keSrJg9395EHevKo2qmp7Pp8fdG4AAAAADmlf1xDq7ruvsf3hJA8f9s27eyfJzvr6+r2H3QcAAAAAB7PMKWMAAAAAXIcEIQAAAIDBCEIAAAAAg9nXNYSOSlVtJNk4c+bMlGMAAMCRWtu6MPUIAPA7THqEUHfvdPfmbDabcgwAAACAoThlDAAAAGAwghAAAADAYAQhAAAAgMFMGoSqaqOqtufz+ZRjAAAAAAzFRaUBAAAABuOUMQAAAIDBCEIAAAAAgxGEAAAAAAYjCAEAAAAM5tSUb15VG0k2zpw5M+UYAAAAJ87a1oWV7u/S+XMr3R8c1qp/b3M47jIGAAAAMBinjAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMIIQAAAAwGDcdh4AAAD2adW3TL90/txK9wf75bbzAAAAAINxyhgAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYzKkp37yqNpJsnDlzZsoxWNLa1oWV7u/S+XMr3R8AADCmVf+schRO+ox+PrtxTXqEUHfvdPfmbDabcgwAAACAoThlDAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwp6Z886raSLJx5syZKcdYqbWtC1OP8JIunT839Qgf00n/bwgAAFfj+1jgejLpEULdvdPdm7PZbMoxAAAAAIbilDEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAGIwgBAAAADEYQAgAAABiMIAQAAAAwmEmDUFVtVNX2fD6fcgwAAACAoUwahLp7p7s3Z7PZlGMAAAAADMUpYwAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYjCAEAAAAMBhBCAAAAGAwghAAAADAYAQhAAAAgMEIQgAAAACDEYQAAAAABiMIAQAAAAxGEAIAAAAYzKlV77CqXpnk+5K8kOTd3f3PV/0eAAAAABzevo4Qqqq3V9XzVfXEFdtvr6qnq+piVW0tNr8xyTu7+94kX7rieQEAAABY0n5PGXtHktv3bqiqlyW5L8kdSc4mubuqzia5Ocmzi6f91mrGBAAAAGBV9hWEuvvRJB+6YvNtSS529zPd/UKSB5LcmeRydqPQvvcPAAAAwPFZ5hpCN+W3jwRKdkPQ65K8Lcn3VtW5JDvXenFVbSbZTJLTp08vMQYAAAAfy9rWhalH4Drk982Na+UXle7uX0vy1ft43naS7SRZX1/vVc8BAAAAwNUtc0rXc0les+fzmxfbAAAAADjBlglCjye5tapuqaqXJ7kryUOrGQsAAACAo7Lf287fn+SxJK+tqstVdU93v5jkLUkeSfJUkge7+8mDvHlVbVTV9nw+P+jcAAAAABzSvq4h1N13X2P7w0kePuybd/dOkp319fV7D7sPAAAAAA7GbeEBAAAABiMIAQAAAAxGEAIAAAAYzKRByEWlAQAAAI7fpEGou3e6e3M2m005BgAAAMBQnDIGAAAAMBhBCAAAAGAwghAAAADAYFxUGgAAAGAwLioNAAAAMBinjAEAAAAMRhACAAAAGIwgBAAAADAYQQgAAABgMO4yBgAAADAYdxkDAAAAGIxTxgAAAAAGIwgBAAAADEYQAgAAABhMdffUM6SqPpDkF6eeY0mvSvLBqYdgMtZ/bNZ/bNZ/bNZ/bNZ/bNZ/bNZ/bNfT+v/+7n711R44EUHoRlBV7+nu9annYBrWf2zWf2zWf2zWf2zWf2zWf2zWf2w3yvo7ZQwAAABgMIIQAAAAwGAEodXZnnoAJmX9x2b9x2b9x2b9x2b9x2b9x2b9x3ZDrL9rCAEAAAAMxhFCAAAAAIMRhA6gqj65qn60qt6/+OcnXeN5X7l4zvur6iv3bH95VW1X1c9V1c9W1Zcf3/Qsa9n13/P4Q1X1xNFPzCots/5V9Xuq6sLiz/2TVXX+eKfnsKrq9qp6uqouVtXWVR5/RVX9y8Xj/6Wq1vY89g2L7U9X1Zcc6+CsxGHXv6q+uKreW1U/vfjnFxz78CxtmT//i8dPV9WHq+rrjm1oVmbJr/9/uKoeW/w//6er6uOOdXiWtsTX/99dVT+4WPenquobjn14lrKPtf/TVfW+qnqxqt50xWMv+XPgSSQIHcxWknd1961J3rX4/Heoqk9O8k1JXpfktiTftOcHx7cmeb67PyvJ2ST/6VimZlWWXf9U1RuTfPh4xmXFll3/7+ruP5jkc5N8flXdcTxjc1hV9bIk9yW5I7tfs++uqrNXPO2eJL/a3WeSfHeS71i89mySu5J8dpLbk3zfYn9cJ5ZZ/yQfTLLR3Z+T5CuT/NPjmZpVWXL9P+ofJfmRo56V1Vvy6/+pJP8syZu7+7OTvD7Jbx7T6KzAkn/+/3ySVyy+/v+xJH/9yljMybXPtf+lJF+V5F9c8dqX/DnwpBKEDubOJD+4+PcfTPJlV3nOlyT50e7+UHf/apIfze4PA0nyNUm+PUm6+yPd/cGjHZcVW2r9q+oTknxtkm87+lE5Aode/+7+9e7+j0nS3S8keV+Sm49+ZJZ0W5KL3f3MYt0eyO7vg732/r54Z5IvrKpabH+gu3+ju38hycXF/rh+HHr9u/u/dff/WGx/MsnHV9UrjmVqVmWZP/+pqi9L8gvZXX+uP8us/xuS/FR3//ck6e5f6e7fOqa5WY1l1r+TvHIRBj8+yQtJ/vfxjM0KfMy17+5L3f1TST5yxWtfqgOcWILQwXxad//y4t//Z5JPu8pzbkry7J7PLye5qao+cfH5ty4OMfuhqrra6zm5Dr3+i3//1iT/MMmvH9mEHKVl1z9JsvhasJHdo4w42T7meu59Tne/mGSe5FP2+VpOtmXWf68vT/K+7v6NI5qTo3Ho9V/8BdDXJ/mWY5iTo7HMn//PStJV9cjie/6/ewzzslrLrP87k/xakl/O7pEk39XdHzrqgVmZZb5/uy6/9zs19QAnTVX9WJLfd5WH3rr3k+7uqjrILdpOZfeIgP/c3V9bVV+b5LuS/JVDD8vKHdX6V9UfSfIHuvtvO2z05DrCP/8f3f+pJPcneVt3P3O4KYHrRVV9dnZPI3jD1LNwrL45yXd394cXBwwxllNJ/mSSz8vuXwK+q6re293+ImgMtyX5rSSfkeSTkvx4Vf2Y7/s4qQShK3T3F13rsar6X1X16d39y1X16Umev8rTnsvuucIfdXOSdyf5lez+T+GHF9t/KLvnnnKCHOH6/4kk61V1Kbt/7j61qt7d3a8PJ8YRrv9HbSd5f3d/z/LTcgyeS/KaPZ/fvNh2tedcXgS/WXa/3u/ntZxsy6x/qurmJP8qyV/t7p8/+nFZsWXW/3VJ3lRV35nkE5N8pKr+b3d/75FPzaoss/6Xkzz60UtDVNXDSf5oHBl8PVlm/b8iyb/r7t9M8nxV/USS9SSC0PVhme/fPtbPASeSU8YO5qHsXhwyi3/+m6s855Ekb6iqT1pcROoNSR7p7k6yk9/+TfKFSX7maMdlxZZZ/3/c3Z/R3WvZ/VujnxODrjuHXv8kqapvy+43C3/r6EdlRR5PcmtV3VJVL8/uRaIfuuI5e39fvCnJf1h8vX8oyV2Lu5DckuTWJP/1mOZmNQ69/otTQy8k2erunziugVmpQ69/d/+p7l5b/D//e5L8AzHourPM1/9HknxO7d5h9FSSPxPf819vlln/X0ryBUlSVa9M8seT/OyxTM0q7Gftr+WaPwecaN3tY58f2T0v9F1J3p/kx5J88mL7epJ/sud5X5PdC4heTPLVe7b//iSPJvmpxX5OT/1r8nF867/n8bUkT0z96/FxfOuf3b8h6CRPJfnJxcdfm/rX5GNf6/5nk/xckp9P8tbFtr+X5EsX//5x2T3i82J2g89n7nntWxevezrJHVP/Wnwc3/on+cbsXkPiJ/d8fOrUvx4fx7P+V+zjm5N83dS/Fh/Hu/5J/nJ2Lyj+RJLvnPrX4uP41j/JJyy2P5ndEPh3pv61+Fj52n9edo8E/LXsHhX25J7XvuTPgSfxoxaDAwAAADAIp4wBAAAADEYQAgAAABiMIAQAAAAwGEEIAAAAYDCCEAAAAMBgBCEAAACAwQhCAAAAAIMRhAAAAAAG8/8A1wd7rzPNWZYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1440x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIcAAAI/CAYAAADtOLm5AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAiS0lEQVR4nO3dfYyl110f8O8Pbx1KKJdADC1+6Thak9alVRBTB6lqlVJe7C4TI4jAFi2BBk/TypX6H4Ogqopadek/baOkikZgTPpik6a8eLpGLm9pUBWK7RDAxjXZmEVek9YmQbctoKRpfv1jrmHY7npn5t7dZ2fO5yM92rnnPvfc3+wePXP3O+ecp7o7AAAAAIzps6YuAAAAAIDpCIcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGdmLqApLkta99ba+trU1dBgAAAMCx8eSTT/52d99wufMmDYeqaiPJxsmTJ/PEE09MWQoAAADAsVJVv7mf8yZdVtbdO929OZvNpiwDAAAAYFj2HAIAAAAYmHAIAAAAYGCThkNVtVFV2/P5fMoyAAAAAIZlzyEAAACAgVlWBgAAADAw4RAAAADAwIRDAAAAAAOzITUAAADAwGxIDQAAADAwy8oAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIG5WxkAAADAwNytDAAAAGBglpUBAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMzK3sAQAAAAbmVvYAAAAAAzsxdQEAsF9rW2dW2t+506dW2h8AABxF9hwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAY2KThUFVtVNX2fD6fsgwAAACAYU0aDnX3TndvzmazKcsAAAAAGJZlZQAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAA5s0HKqqjarans/nU5YBAAAAMKxJw6Hu3unuzdlsNmUZAAAAAMOyrAwAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIFNGg5V1UZVbc/n8ynLAAAAABjWpOFQd+909+ZsNpuyDAAAAIBhWVYGAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADGzl4VBVvamqfr6q3l1Vb1p1/wAAAACszr7Coap6oKperKqnLmi/s6qeraqzVbW1aO4k/zvJZyc5v9pyAQAAAFil/c4cejDJnXsbquq6JO9KcleS25PcW1W3J/n57r4ryXcl+UerKxUAAACAVdtXONTdH0jyiQua70hytruf6+5PJXk4yd3d/ZnF87+T5FUrqxQAAACAlTuxxGtvTPL8nsfnk7yxqr4xydcl+fwk77zUi6tqM8lmktxyyy1LlAEAAADAYS0TDl1Ud/9okh/dx3nbSbaTZH19vVddBwAAAACXt8zdyl5IcvOexzct2gAAAAA4IpYJhx5PcltV3VpV1ye5J8kjqykLAAAAgKthv7eyfyjJB5O8vqrOV9XbuvvTSe5P8liSZ5K8t7ufPsibV9VGVW3P5/OD1g0AAADACuxrz6HuvvcS7Y8mefSwb97dO0l21tfX7ztsHwAAAAAc3jLLygAAAAA44oRDAAAAAAObNByy5xAAAADAtCYNh7p7p7s3Z7PZlGUAAAAADMuyMgAAAICBCYcAAAAABiYcAgAAABiYDakBAAAABmZDagAAAICBWVYGAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAzM3coAAAAABuZuZQAAAAADs6wMAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYG5lDwAAADAwt7IHAAAAGJhlZQAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMbNJwqKo2qmp7Pp9PWQYAAADAsCYNh7p7p7s3Z7PZlGUAAAAADMuyMgAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgU0aDlXVRlVtz+fzKcsAAAAAGNak4VB373T35mw2m7IMAAAAgGFZVgYAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMAmDYeqaqOqtufz+ZRlAAAAAAxr0nCou3e6e3M2m01ZBgAAAMCwLCsDAAAAGJhwCAAAAGBgJ6YuAACmsrZ1ZqX9nTt9aqX9AQDA1WDmEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMLATUxcAwPG0tnVm6hIAAIB9MHMIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABjYFQmHqurVVfVEVX39legfAAAAgNXYVzhUVQ9U1YtV9dQF7XdW1bNVdbaqtvY89V1J3rvKQgEAAABYvf3OHHowyZ17G6rquiTvSnJXktuT3FtVt1fV1yT5tSQvrrBOAAAAAK6AE/s5qbs/UFVrFzTfkeRsdz+XJFX1cJK7k3xukldnNzD6/ap6tLs/s7qSAQAAAFiVfYVDl3Bjkuf3PD6f5I3dfX+SVNW3J/ntSwVDVbWZZDNJbrnlliXKAAAAAOCwrtjdyrr7we7+j6/w/HZ3r3f3+g033HClygAAAADgFSwTDr2Q5OY9j29atAEAAABwRCwTDj2e5LaqurWqrk9yT5JHVlMWAAAAAFfDfm9l/1CSDyZ5fVWdr6q3dfenk9yf5LEkzyR5b3c/fZA3r6qNqtqez+cHrRsAAACAFdjv3cruvUT7o0kePeybd/dOkp319fX7DtsHAAAAAId3xTakBgAAAODaJxwCAAAAGNik4ZA9hwAAAACmNWk41N073b05m82mLAMAAABgWJaVAQAAAAxMOAQAAAAwMOEQAAAAwMBsSA0AAAAwMBtSAwAAAAzMsjIAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDuVgYAAAAwMHcrAwAAABiYZWUAAAAAAxMOAQAAAAzsxNQFAMBxsbZ1ZuV9njt9auV9AgDAXmYOAQAAAAxMOAQAAAAwMLeyBwAAABiYW9kDAAAADMyyMgAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGNmk4VFUbVbU9n8+nLAMAAABgWJOGQ929092bs9lsyjIAAAAAhmVZGQAAAMDATkxdAADXhrWtM1OXAAAATMDMIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGNik4VBVbVTV9nw+n7IMAAAAgGFNGg519053b85msynLAAAAABiWZWUAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADOzE1AUAAJe2tnVmpf2dO31qpf0BAHD0mTkEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADmzQcqqqNqtqez+dTlgEAAAAwrEnDoe7e6e7N2Ww2ZRkAAAAAw7KsDAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGAnpi4AgMNZ2zozdQkAAMAxYOYQAAAAwMCEQwAAAAADW3k4VFV/tqreXVXvq6q/s+r+AQAAAFidfYVDVfVAVb1YVU9d0H5nVT1bVWeraitJuvuZ7n57km9O8pdWXzIAAAAAq7LfmUMPJrlzb0NVXZfkXUnuSnJ7knur6vbFc29OcibJoyurFAAAAICV21c41N0fSPKJC5rvSHK2u5/r7k8leTjJ3YvzH+nuu5J86yqLBQAAAGC1lrmV/Y1Jnt/z+HySN1bVm5J8Y5JX5RVmDlXVZpLNJLnllluWKAMAAACAw1omHLqo7n5/kvfv47ztJNtJsr6+3quuAwAAAIDLW+ZuZS8kuXnP45sWbQAAAAAcEcuEQ48nua2qbq2q65Pck+SR1ZQFAAAAwNWw31vZP5Tkg0leX1Xnq+pt3f3pJPcneSzJM0ne291PH+TNq2qjqrbn8/lB6wYAAABgBap7+u1+1tfX+4knnpi6DIAjZW3rzNQlQM6dPjV1CQAAXEJVPdnd65c7b5llZQAAAAAcccIhAAAAgIGt/Fb2B1FVG0k2Tp48OWUZAFeFZWAAAMC1aNKZQ929092bs9lsyjIAAAAAhmVZGQAAAMDAhEMAAAAAAxMOAQAAAAxs0nCoqjaqans+n09ZBgAAAMCwbEgNAAAAMDDLygAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgblbGQAAAMDA3K0MAAAAYGCWlQEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAzMrewBAAAABuZW9gAAAAADOzF1AQDA0bW2dWbqEi7r3OlTU5cAAHBNs+cQAAAAwMDMHAIAjrVVz24yEwkAOG7MHAIAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABjZpOFRVG1W1PZ/PpywDAAAAYFiThkPdvdPdm7PZbMoyAAAAAIblVvbAJNxaGgAA4NpgzyEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGA2pAaOBRtcAwAAHI6ZQwAAAAADEw4BAAAADGzScKiqNqpqez6fT1kGAAAAwLAm3XOou3eS7Kyvr983ZR0AAPtljzMA4LixrAwAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYCemLgDgWrS2dWbqEgAAAK6KSWcOVdVGVW3P5/MpywAAAAAY1qThUHfvdPfmbDabsgwAAACAYdlzCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBnZi6AACAka1tnVl5n+dOn1p5nwDA8WXmEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMDcrQwA4JhZ9R3Q3P0MAI43M4cAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgJ65Ep1X1DUlOJfm8JD/Y3f/pSrwPAAAAAMvZ98yhqnqgql6sqqcuaL+zqp6tqrNVtZUk3f3j3X1fkrcn+ZbVlgwAAADAqhxk5tCDSd6Z5D0vN1TVdUneleRrkpxP8nhVPdLdv7Y45XsXzwNH2NrWmalLAAAA4ArZ98yh7v5Akk9c0HxHkrPd/Vx3fyrJw0nurl3fn+Qnu/tDqysXAAAAgFVadkPqG5M8v+fx+UXb30vy1UneUlVvv9gLq2qzqp6oqideeumlJcsAAAAA4DCuyIbU3f2OJO+4zDnbSbaTZH19va9EHQAAAAC8smVnDr2Q5OY9j29atAEAAABwBCwbDj2e5LaqurWqrk9yT5JHli8LAAAAgKvhILeyfyjJB5O8vqrOV9XbuvvTSe5P8liSZ5K8t7ufPkCfG1W1PZ/PD1o3AAAAACuw7z2HuvveS7Q/muTRw7x5d+8k2VlfX7/vMK8HAAAAYDnLLisDAAAA4AgTDgEAAAAMbNJwyJ5DAAAAANOaNBzq7p3u3pzNZlOWAQAAADAsy8oAAAAABiYcAgAAABiYcAgAAABgYDakBgAAABiYDakBAAAABmZZGQAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwsBNTvnlVbSTZOHny5JRlAADwCta2zqy0v3OnT620PwBgOe5WBgAAADAwy8oAAAAABjbpsjIAAODgVr3UL7HcD2BkZg4BAAAADEw4BAAAADAw4RAAAADAwCYNh6pqo6q25/P5lGUAAAAADGvSDam7eyfJzvr6+n1T1gEHtepNIG0ACQAAwFQsKwMAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIFNGg5V1UZVbc/n8ynLAAAAABjWpOFQd+909+ZsNpuyDAAAAIBhnZi6AAAAuNasbZ1ZaX/nTp9aaX8AsEr2HAIAAAAYmJlDAABwha16JhIArJKZQwAAAAADEw4BAAAADEw4BAAAADAwew4BAHCk2c8HAJZj5hAAAADAwCYNh6pqo6q25/P5lGUAAAAADGvScKi7d7p7czabTVkGAAAAwLAsKwMAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIGdmLoAYPXWts5MXQIAAABHhJlDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAbEgNAMBV5cYJAHBtMXMIAAAAYGCThkNVtVFV2/P5fMoyAAAAAIY1aTjU3TvdvTmbzaYsAwAAAGBY9hyCa4C9FwAAAJiKPYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBCYcAAAAABnZi6gKOm7WtMyvt79zpUyvtb0Sr/jcBAACA48TMIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIGtPByqqtdV1Q9W1ftW3TcAAAAAq7WvcKiqHqiqF6vqqQva76yqZ6vqbFVtJUl3P9fdb7sSxQIAAACwWvudOfRgkjv3NlTVdUneleSuJLcnubeqbl9pdQAAAABcUfsKh7r7A0k+cUHzHUnOLmYKfSrJw0nuXnF9AAAAAFxBy+w5dGOS5/c8Pp/kxqr6wqp6d5Ivr6rvvtSLq2qzqp6oqideeumlJcoAAAAA4LBOrLrD7v54krfv47ztJNtJsr6+3quuAwAAAIDLW2bm0AtJbt7z+KZFGwAAAABHxDLh0ONJbquqW6vq+iT3JHlkNWUBAAAAcDXsa1lZVT2U5E1JXltV55P8w+7+waq6P8ljSa5L8kB3P32QN6+qjSQbJ0+ePFjVAADANW1t68xK+zt3+tRK+wPgD+0rHOruey/R/miSRw/75t29k2RnfX39vsP2AQAAAMDhLbOsDAAAAIAjTjgEAAAAMLCV38r+IOw5BAAAwFFiPy2Oo0lnDnX3TndvzmazKcsAAAAAGJZlZQAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwNyt7BpnJ3wAAIDDWfX/p+C4crcyAAAAgIFZVgYAAAAwMOEQAAAAwMCEQwAAAAADsyE1AADANego3JzmKNQIXJ4NqQEAAAAGZlkZAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwNzKHgAAuOat+pbpyXi3Tb8Sf4fA8eBW9gAAAAADs6wMAAAAYGDCIQAAAICBCYcAAAAABiYcAgAAABiYcAgAAABgYMIhAAAAgIEJhwAAAAAGdmLKN6+qjSQbJ0+enLKMoaxtnVl5n+dOn1p5nwAAcNRcic/aAFfDpDOHununuzdns9mUZQAAAAAMy7IyAAAAgIEJhwAAAAAGJhwCAAAAGJhwCAAAAGBgwiEAAACAgQmHAAAAAAYmHAIAAAAYmHAIAAAAYGDCIQAAAICBnZjyzatqI8nGyZMnpyyDJa1tnVlpf+dOn1ppfwAAXN6qP9PBYRiHMI1JZw519053b85msynLAAAAABiWZWUAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAzsx5ZtX1UaSjZMnT05ZBteYta0zU5cAAAAAw5h05lB373T35mw2m7IMAAAAgGFZVgYAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAxMOAQAAAAwMOEQAAAAwMCEQwAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAM7MSqO6yqVyf5V0k+leT93f1vV/0eAAAAAKzGvmYOVdUDVfViVT11QfudVfVsVZ2tqq1F8zcmeV9335fkzSuuFwAAAIAV2u+ysgeT3Lm3oaquS/KuJHcluT3JvVV1e5Kbkjy/OO3/rqZMAAAAAK6EfYVD3f2BJJ+4oPmOJGe7+7nu/lSSh5PcneR8dgOiffcPAAAAwDSW2XPoxvzhDKFkNxR6Y5J3JHlnVZ1KsnOpF1fVZpLNJLnllluWKAMAAODg1rbOTF0CrHwcnjt9aqX9HQX+Dpe38g2pu/t3k3zHPs7bTrKdJOvr673qOgAAAAC4vGWWfb2Q5OY9j29atAEAAABwRCwTDj2e5LaqurWqrk9yT5JHVlMWAAAAAFfDfm9l/1CSDyZ5fVWdr6q3dfenk9yf5LEkzyR5b3c/fZA3r6qNqtqez+cHrRsAAACAFdjXnkPdfe8l2h9N8uhh37y7d5LsrK+v33fYPgAAAAA4PLeaBwAAABiYcAgAAABgYMIhAAAAgIFNGg7ZkBoAAABgWpOGQ929092bs9lsyjIAAAAAhmVZGQAAAMDAhEMAAAAAAxMOAQAAAAzMhtQAAAAAA7MhNQAAAMDALCsDAAAAGJhwCAAAAGBgwiEAAACAgdmQGgAAAGBgNqQGAAAAGJhlZQAAAAADEw4BAAAADEw4BAAAADAw4RAAAADAwKq7p64hVfVSkt+cug6uitcm+e2pi+DIMF44COOFgzJmOAjjhYMwXjgI44WDOsiY+dPdfcPlTromwiHGUVVPdPf61HVwNBgvHITxwkEZMxyE8cJBGC8chPHCQV2JMWNZGQAAAMDAhEMAAAAAAxMOcbVtT10AR4rxwkEYLxyUMcNBGC8chPHCQRgvHNTKx4w9hwAAAAAGZuYQAAAAwMCEQ6xEVX1BVf1UVX1k8edrLnHeWxfnfKSq3rqn/f1V9WxVfXhxfNGi/VVV9SNVdbaq/mtVrV2lb4kraJnxUlWfU1Vnquq/VdXTVXV6z/nfXlUv7RlH33m1vidWr6ruXFwXzlbV1kWev+T1oaq+e9H+bFV93X775Og67Hipqq+pqier6lcXf37Vntdc9GcTR98S42Wtqn5/z5h4957XfMViHJ2tqndUVV3Fb4krbIkx8617xsuHq+ozVfWGxXOuMcfUPsbLX6mqD1XVp6vqLRc8d6n/L7nGHFOHHS9V9Yaq+uDi/0S/UlXfsue5B6vqN/ZcX95w2UK62+FY+kjyz5JsLb7eSvL9FznnC5I8t/jzNYuvX7N47v1J1i/ymr+b5N2Lr+9J8iNTf6+OacdLks9J8lcX51yf5OeT3LV4/O1J3jn19+dYyRi5LslHk7xu8e/8y0luv+Cci14fkty+OP9VSW5d9HPdfvp0HM1jyfHy5Um+ZPH1lyV5Yc9rLvqzyXG0jyXHy1qSpy7R7y8m+cokleQnX/7Z5Dj6xzJj5oJz/nySj+557BpzDI99jpe1JH8hyXuSvGVP+yv9f8k15hgeS46XL01y2+LrL0nysSSfv3j84N5z93OYOcSq3J3khxdf/3CSb7jIOV+X5Ke6+xPd/TtJfirJnQfo931J/pqU/Fg49Hjp7t/r7p9Lku7+VJIPJbnpypfMVXZHkrPd/dzi3/nh7I6bvS51fbg7ycPd/cnu/o0kZxf97adPjqZDj5fu/qXu/q1F+9NJ/nhVveqqVM1Ulrm+XFRV/akkn9fdv9C7n8rfk4v/bONoWtWYuXfxWo63y46X7j7X3b+S5DMXvPain39dY461Q4+X7v717v7I4uvfSvJikhsOW4hwiFX54u7+2OLr/57kiy9yzo1Jnt/z+Pyi7WU/tJjy9g/2/DD9g9d096eTzJN84UorZwqrGC+pqs9PspHkZ/Y0f9NiWuX7qurm1ZXMVXbZf/9c+vpwqdfup0+OpmXGy17flORD3f3JPW0X+9nE0bbseLm1qn6pqv5zVf3lPeefv0yfHF2rusZ8S5KHLmhzjTl+lvm88UqfYVxjjqeVfD6tqjuyO/Poo3ua/8ni/0X/fD+/+Dpx0DdlXFX100n+5EWe+p69D7q7q+qgt8H71u5+oar+RJL/kORvZjcR54i6wuMlVXUiux+w3tHdzy2ad5I81N2frKq/nd3f4H3VpfoAeFlV/bkk35/ka/c0+9nEhT6W5Jbu/nhVfUWSH1+MHXhFVfXGJL/X3U/taXaNAZa2mFn2r5O8tbtfnl303dn9Jfz12b3t/Xcl+b5X6sfMIfatu7+6u7/sIsdPJPkfi0H58uB88SJdvJBk70yOmxZt6e6X//xfSf5ddqfX/ZHXLMKAWZKPr/67Y9Wu5HhZ2E7yke7+F3ve8+N7fuP/A0m+YoXfElfX5f79/8g5F1wfLvXa/fTJ0bTMeElV3ZTkx5J8W3f/wW/cXuFnE0fbocfLYrnqx5Oku5/M7m9ov3Rx/t4lzq4vx8tS15iFe3LBrCHXmGNrmc8br/QZxjXmeFrq82lVfV6SM0m+p7t/4eX27v5Y7/pkkh/KPq4vwiFW5ZEkL++m/9YkP3GRcx5L8rVV9ZravTvV1yZ5rKpOVNVrk6Sq/liSr0/y8m9V9vb7liQ/u1hny9F26PGSJFX1j7P7oevv733By4HTwpuTPLPasrmKHk9yW1XdWlXXZ/dD9SMXnHOp68MjSe6p3TvH3Jrktuxu4rifPjmaDj1eFstTz2R3k/z/8vLJl/nZxNG2zHi5oaquS5Kqel12ry/PLZZK/8+q+srF0qBvy8V/tnE0LfMzKVX1WUm+OXv2G3KNOdaW+bxx0c+/rjHH2qHHy+L8H0vynu5+3wXPvfyL+Mru/lSXv74cZPdqh+NSR3bXVP9Mko8k+ekkX7BoX0/yA3vO+1vZ3Rz2bJLvWLS9OsmTSX4lu5uB/ssk1y2e++wk/35x/i8med3U36tj8vFyU5LObvDz4cXxnYvn/uliDP1ykp9L8mem/l4dS42Tv57k17P7m/nvWbR9X5I3L76+5PUhu8sXP5rk2ey5m8fF+nQcj+Ow4yXJ9yb53T3Xkw8n+aJX+tnkOPrHEuPlmxbj4cPZvSHCxp4+17P74fujSd6ZpKb+Ph3Tj5nFc29K8gsX9Ocac4yPfYyXv5jdvWV+N7szzJ7e89r/7/Pvot015pgehx0vSf5Gkv9zwWeYNyye+9kkv7oYM/8myedero5avBAAAACAAVlWBgAAADAw4RAAAADAwIRDAAAAAAMTDgEAAAAMTDgEAAAAMDDhEAAAAMDAhEMAAAAAAxMOAQAAAAzs/wHtdFu2udGHbwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIQAAAI/CAYAAAAGDwK6AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAhNUlEQVR4nO3db4ys110f8O8P3zoUKEMghhY77jq6Jq1LqyC2DlLVKqUQ7F4WI4jAFi2hDdmmVSr1HYugL4pa9dI3bSMiRSsIBqrapCl/vL1GLv/SoCqUOCGAXdfkxlwUm7Q2CZq2gJJaOX2xYzJc9tp7d2bvs3t/n4/0yDvPPHPmN94zz8z97jnnqTFGAAAAAOjjs6YuAAAAAIBrSyAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNnJm6gCR5xSteMTY2NqYuAwAAAOC68YEPfOB3xxg3HXTfiQiENjY28uijj05dBgAAAMB1o6p++0r3TTplrKq2qmp3Pp9PWQYAAABAK5MGQmOMvTHG9mw2m7IMAAAAgFYsKg0AAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZlxlDAAAAKAZVxkDAAAAaMaUMQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA04ypjAAAAAM24yhgAAABAM6aMAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGYmDYSqaquqdufz+ZRlAAAAALQyaSA0xtgbY2zPZrMpywAAAABoxZQxAAAAgGbOTF0AANenjZ0La2/z0vlza28TAAA6MkIIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACamTQQqqqtqtqdz+dTlgEAAADQyqSB0Bhjb4yxPZvNpiwDAAAAoBVTxgAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANDNpIFRVW1W1O5/PpywDAAAAoJVJA6Exxt4YY3s2m01ZBgAAAEArpowBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM1MGghV1VZV7c7n8ynLAAAAAGhl0kBojLE3xtiezWZTlgEAAADQiiljAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoJm1B0JV9bqq+qWqekdVvW7d7QMAAACwmkMFQlX1zqp6tqoeu2z/XVX1ZFVdrKqdxe6R5P8m+ewkT6+3XAAAAABWddgRQvcnuWt5R1XdkOTtSe5OckeS+6rqjiS/NMa4O8l3Jfln6ysVAAAAgHU4VCA0xnhvkk9ctvvOJBfHGE+NMT6V5MEk94wxPr24//eSvGxtlQIAAACwFmdWeOzNST66dPvpJK+tqm9K8nVJviDJD1zpwVW1nWQ7SW699dYVygAAAADgaqwSCB1ojPETSX7iEMftJtlNks3NzbHuOgAAAAA42CpXGXsmySuXbt+y2AcAAADACbZKIPT+JLdX1W1VdWOSe5M8tJ6yAAAAADguh73s/ANJ3pfk1VX1dFW9aYzxfJK3JnkkyRNJ3jXGePz4SgUAAABgHQ61htAY474r7H84ycNHffKq2kqydfbs2aM2AQAAAMBVWmXK2MrGGHtjjO3ZbDZlGQAAAACtTBoIAQAAAHDtCYQAAAAAmhEIAQAAADQzaSBUVVtVtTufz6csAwAAAKAVi0oDAAAANGPKGAAAAEAzAiEAAACAZgRCAAAAAM1YVBoAAACgGYtKAwAAADRjyhgAAABAMwIhAAAAgGYEQgAAAADNCIQAAAAAmhEIAQAAADTjsvMAAAAAzbjsPAAAAEAzpowBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzLjsPAAAA0IzLzgMAAAA0Y8oYAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANDMpIFQVW1V1e58Pp+yDAAAAIBWJg2Exhh7Y4zt2Ww2ZRkAAAAArZgyBgAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANDMmakLAOBk2Ni5MHUJAADANTLpCKGq2qqq3fl8PmUZAAAAAK1MGgiNMfbGGNuz2WzKMgAAAABasYYQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANDMpIFQVW1V1e58Pp+yDAAAAIBWJg2Exhh7Y4zt2Ww2ZRkAAAAArZgyBgAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgmTNTFwAAh7Wxc2Gt7V06f26t7QEAwGlhhBAAAABAMwIhAAAAgGYEQgAAAADNCIQAAAAAmjmWQKiqPreqHq2qrz+O9gEAAAA4ukMFQlX1zqp6tqoeu2z/XVX1ZFVdrKqdpbu+K8m71lkoAAAAAOtx2BFC9ye5a3lHVd2Q5O1J7k5yR5L7quqOqvraJP89ybNrrBMAAACANTlzmIPGGO+tqo3Ldt+Z5OIY46kkqaoHk9yT5POSfG72Q6I/rKqHxxifXl/JAAAAAKziUIHQFdyc5KNLt59O8toxxluTpKq+I8nvXikMqqrtJNtJcuutt65QBgAAAABX49iuMjbGuH+M8Z9e5P7dMcbmGGPzpptuOq4yAAAAALjMKoHQM0leuXT7lsU+AAAAAE6wVQKh9ye5vapuq6obk9yb5KH1lAUAAADAcTnsZecfSPK+JK+uqqer6k1jjOeTvDXJI0meSPKuMcbjV/PkVbVVVbvz+fxq6wYAAADgiA57lbH7rrD/4SQPH/XJxxh7SfY2NzfffNQ2AAAAALg6x7aoNAAAAAAnk0AIAAAAoBmBEAAAAEAzkwZCFpUGAAAAuPYmDYTGGHtjjO3ZbDZlGQAAAACtmDIGAAAA0IxACAAAAKAZgRAAAABAMxaVBgAAAGjGotIAAAAAzZgyBgAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzbjsPAAAAEAzLjsPAAAA0IwpYwAAAADNCIQAAAAAmhEIAQAAADQjEAIAAABoRiAEAAAA0IzLzgMAAAA047LzAAAAAM2YMgYAAADQjEAIAAAAoBmBEAAAAEAzZ6YuAICj2di5MHUJAADAKWWEEAAAAEAzAiEAAACAZkwZA6CtdU+7u3T+3FrbAwCA42KEEAAAAEAzkwZCVbVVVbvz+XzKMgAAAABamTQQGmPsjTG2Z7PZlGUAAAAAtGLKGAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzkwZCVbVVVbvz+XzKMgAAAABamTQQGmPsjTG2Z7PZlGUAAAAAtGLKGAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQzKSBUFVtVdXufD6fsgwAAACAViYNhMYYe2OM7dlsNmUZAAAAAK2YMgYAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaOTN1AQBwvdjYubD2Ni+dP7f2NgEAwAghAAAAgGYEQgAAAADNCIQAAAAAmhEIAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGbOTF0AQBcbOxemLgEAACCJEUIAAAAA7QiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgmbUHQlX1F6vqHVX17qr6h+tuHwAAAIDVHCoQqqp3VtWzVfXYZfvvqqonq+piVe0kyRjjiTHGW5J8S5K/tv6SAQAAAFjFYUcI3Z/kruUdVXVDkrcnuTvJHUnuq6o7Fvd9Q5ILSR5eW6UAAAAArMWhAqExxnuTfOKy3XcmuTjGeGqM8akkDya5Z3H8Q2OMu5N82zqLBQAAAGB1Z1Z47M1JPrp0++kkr62q1yX5piQvy4uMEKqq7STbSXLrrbeuUAYAAAAAV2OVQOhAY4z3JHnPIY7bTbKbJJubm2PddQAAAABwsFWuMvZMklcu3b5lsQ8AAACAE2yVQOj9SW6vqtuq6sYk9yZ5aD1lAQAAAHBcDnvZ+QeSvC/Jq6vq6ap60xjj+SRvTfJIkieSvGuM8fjVPHlVbVXV7nw+v9q6AQAAADiiQ60hNMa47wr7H84Kl5YfY+wl2dvc3HzzUdsAAAAA4OqsMmUMAAAAgFNo7VcZA7gebOxcmLoEAACAY2OEEAAAAEAzkwZCFpUGAAAAuPYmDYTGGHtjjO3ZbDZlGQAAAACtmDIGAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGZcZQwAAACgGVcZAwAAAGjGlDEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANOMqYwAAAADNuMoYAAAAQDOmjAEAAAA0c2bqAgCAK9vYubDW9i6dP7fW9gAAOJ2MEAIAAABoRiAEAAAA0IwpYwDQiCloAAAkRggBAAAAtDNpIFRVW1W1O5/PpywDAAAAoJVJp4yNMfaS7G1ubr55yjoAgKMxBQ0A4HQyZQwAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaGbSQKiqtqpqdz6fT1kGAAAAQCuTBkJjjL0xxvZsNpuyDAAAAIBWTBkDAAAAaObM1AUAPW3sXFhre5fOn1trewAAANczgRAAcGKsOyxOBMYAAAcxZQwAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJpxlTHgJR3HVX8AAACYjhFCAAAAAM1MGghV1VZV7c7n8ynLAAAAAGhl0kBojLE3xtiezWZTlgEAAADQiiljAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoJlJA6Gq2qqq3fl8PmUZAAAAAK2cmfLJxxh7SfY2NzffPGUdwOm3sXNh6hIAAABODVPGAAAAAJoRCAEAAAA0IxACAAAAaGbSNYQAAI7butcYu3T+3FrbAwCYghFCAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzZ6YuAADgNNnYubDW9i6dP7fW9gAADsMIIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJo5cxyNVtU3JjmX5POT/NAY4z8fx/MAAAAAcPUOPUKoqt5ZVc9W1WOX7b+rqp6sqotVtZMkY4yfGmO8OclbknzreksGAAAAYBVXM2Xs/iR3Le+oqhuSvD3J3UnuSHJfVd2xdMj3Lu4HAAAA4IQ4dCA0xnhvkk9ctvvOJBfHGE+NMT6V5MEk99S+70/yM2OMD66vXAAAAABWteqi0jcn+ejS7acX+/5xkq9J8oaqestBD6yq7ap6tKoefe6551YsAwAAAIDDOpZFpccYb0vytpc4ZjfJbpJsbm6O46gDAAAAgD9p1RFCzyR55dLtWxb7AAAAADihVg2E3p/k9qq6rapuTHJvkodWLwsAAACA43I1l51/IMn7kry6qp6uqjeNMZ5P8tYkjyR5Ism7xhiPX0WbW1W1O5/Pr7ZuAAAAAI7o0GsIjTHuu8L+h5M8fJQnH2PsJdnb3Nx881EeDwAAAMDVO5ZFpYFpbexcmLoEAAAATrBV1xACAAAA4JQRCAEAAAA0M2kgZFFpAAAAgGtv0kBojLE3xtiezWZTlgEAAADQiiljAAAAAM0IhAAAAACaEQgBAAAANCMQAgAAAGjGVcYAAAAAmnGVMQAAAIBmTBkDAAAAaObM1AUAAHS2sXNh7W1eOn9u7W0CANcXI4QAAAAAmhEIAQAAADTjKmMAAAAAzbjKGAAAAEAzpowBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZiYNhKpqq6p25/P5lGUAAAAAtDJpIDTG2BtjbM9msynLAAAAAGjFlDEAAACAZgRCAAAAAM2cmboAAABYxcbOhbW3een8ubW3CQAniUAIAAAus+6QScAEwEljyhgAAABAMwIhAAAAgGZMGQMA4EWZPgXAOvlcORkmHSFUVVtVtTufz6csAwAAAKCVSQOhMcbeGGN7NptNWQYAAABAK9YQAgAAAGhGIAQAAADQjEWl4QgsggYAAMBpZoQQAAAAQDMCIQAAAIBmTBkDAOCaWvfUawDg6gmEuO750gkAAAB/nEAIAAAAOJA/sF+/rCEEAAAA0IxACAAAAKCZSQOhqtqqqt35fD5lGQAAAACtTBoIjTH2xhjbs9lsyjIAAAAAWjFlDAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNnJm6AAAAAICj2ti5sNb2Lp0/t9b2TiojhAAAAACaEQgBAAAANCMQAgAAAGhGIAQAAADQjEWlAQCuM+teXBMAuP4YIQQAAADQjEAIAAAAoBmBEAAAAEAz1hCCE8BaDwAAAFxLRggBAAAANDNpIFRVW1W1O5/PpywDAAAAoJVJA6Exxt4YY3s2m01ZBgAAAEArpowBAAAANCMQAgAAAGhGIAQAAADQjEAIAAAAoBmBEAAAAEAzAiEAAACAZgRCAAAAAM2cmbqA683GzoW1tnfp/Lm1tgcAAABghBAAAABAMwIhAAAAgGYEQgAAAADNCIQAAAAAmhEIAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNnJm6AAAAAI7fxs6Ftbd56fy5tbcJXBtGCAEAAAA0IxACAAAAaGbtgVBVvaqqfqiq3r3utgEAAABY3aECoap6Z1U9W1WPXbb/rqp6sqouVtVOkowxnhpjvOk4igUAAABgdYcdIXR/kruWd1TVDUnenuTuJHckua+q7lhrdQAAAACs3aECoTHGe5N84rLddya5uBgR9KkkDya5Z831AQAAALBmq6whdHOSjy7dfjrJzVX1RVX1jiRfUVXffaUHV9V2VT1aVY8+99xzK5QBAAAAwNU4s+4GxxgfT/KWQxy3m2Q3STY3N8e66wAAAADgYKuMEHomySuXbt+y2AcAAADACbZKIPT+JLdX1W1VdWOSe5M8tJ6yAAAAADguh73s/ANJ3pfk1VX1dFW9aYzxfJK3JnkkyRNJ3jXGePz4SgUAAABgHQ61htAY474r7H84ycNHffKq2kqydfbs2aM2AQAANLCxc2HtbV46f27tbXL9W3df1A+ZyipTxlY2xtgbY2zPZrMpywAAAABoZdJACAAAAIBrTyAEAAAA0IxACAAAAKCZQy0qfVwsKg0AQAcWoYXD8V5Z3XEswM71yaLSAAAAAM2YMgYAAADQjEAIAAAAoBmBEAAAAEAzFpUGAABasoDx6ixgDKeXRaUBAAAAmjFlDAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNuMoYAADgiltr4P8hcJq4yhgAAABAM6aMAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGbOTPnkVbWVZOvs2bNTlnGibexcWGt7l86fW2t7x2HdrxkA4Hrj+1IPfs89+D0zlUlHCI0x9sYY27PZbMoyAAAAAFoxZQwAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaObMlE9eVVtJts6ePTtlGaxoY+fC1CUAAAAAV2HSEUJjjL0xxvZsNpuyDAAAAIBWTBkDAAAAaEYgBAAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzQiEAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANDMmSmfvKq2kmydPXt2yjJa2di5MHUJAAA04HsnJ4F+CFc26QihMcbeGGN7NptNWQYAAABAK6aMAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNCIQAAAAAmhEIAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKAZgRAAAABAMwIhAAAAgGYEQgAAAADNTBoIVdVWVe3O5/MpywAAAABoZdJAaIyxN8bYns1mU5YBAAAA0IopYwAAAADNCIQAAAAAmhEIAQAAADQjEAIAAABoRiAEAAAA0IxACAAAAKCZGmNMXUOq6rkkvz11HazdK5L87tRFMDn9gEQ/4DP0BRL9gH36AYl+wGfoC8fjz48xbjrojhMRCHF9qqpHxxibU9fBtPQDEv2Az9AXSPQD9ukHJPoBn6EvXHumjAEAAAA0IxACAAAAaEYgxHHanboATgT9gEQ/4DP0BRL9gH36AYl+wGfoC9eYNYQAAAAAmjFCCAAAAKAZgRBXraq+sKp+tqo+vPjvy69w3BsXx3y4qt64tP89VfVkVX1osX3xYv/LqurHq+piVf23qtq4Ri+JI1ilH1TV51TVhar6H1X1eFWdXzr+O6rquaX+8Z3X6jVxeFV11+J9fLGqdg64/4rv56r67sX+J6vq6w7bJifPUftBVX1tVX2gqn5j8d+vXnrMgZ8RnFwr9IONqvrDpd/1O5Ye85WL/nGxqt5WVXUNXxJHtEJf+LalfvChqvp0Vb1mcZ9zwilziH7wN6rqg1X1fFW94bL7rvTvB+eEU+ao/aCqXlNV71v8G+HXq+pbl+67v6p+a+l88Jpr9HKuX2MMm+2qtiT/KsnO4uedJN9/wDFfmOSpxX9fvvj55Yv73pNk84DH/KMk71j8fG+SH5/6tdqOpx8k+Zwkf3NxzI1JfinJ3Yvb35HkB6Z+fbYX/d3fkOQjSV61+P39WpI7LjvmwPdzkjsWx78syW2Ldm44TJu2k7Wt2A++IsmXLn7+8iTPLD3mwM8I28ncVuwHG0keu0K7v5Lkq5JUkp954TPCdnK3VfrCZcf85SQfWbrtnHCKtkP2g40kfyXJjyZ5w9L+F/v3g3PCKdpW7AdfluT2xc9fmuRjSb5gcfv+5WNtq29GCHEU9yT5kcXPP5LkGw845uuS/OwY4xNjjN9L8rNJ7rqKdt+d5G9J/0+0I/eDMcYfjDF+MUnGGJ9K8sEktxx/yazJnUkujjGeWvz+Hsx+f1h2pffzPUkeHGN8cozxW0kuLto7TJucLEfuB2OMXx1j/M5i/+NJ/nRVveyaVM26rXI+OFBV/bkknz/G+OWx/y+AH83BnzGcLOvqC/ctHsvp9JL9YIxxaYzx60k+fdljD/ze6JxwKh25H4wxfnOM8eHFz7+T5NkkN12bsvsRCHEUXzLG+Nji5/+Z5EsOOObmJB9duv30Yt8LfngxzO+fLn0R+KPHjDGeTzJP8kVrrZx1Wkc/SFV9QZKtJD+/tPubF0NE311Vr1xfyazJS/5ec+X385Uee5g2OVlW6QfLvjnJB8cYn1zad9BnBCfTqv3gtqr61ar6L1X115eOf/ol2uTkWdc54VuTPHDZPueE02OVz/MX+47gnHC6rOV7XVXdmf0RRh9Z2v0vFv9O+Nf+mLS6M1MXwMlUVT+X5M8ecNf3LN8YY4yqutpL1X3bGOOZqvozSf5jkr+b/aSfE+aY+0Gq6kz2v/S9bYzx1GL3XpIHxhifrKp/kP2/JH71ldoATq+q+ktJvj/J65d2+4zo42NJbh1jfLyqvjLJTy36BE1V1WuT/MEY47Gl3c4J0NBiZNiPJXnjGOOFUUTfnf0/RN+Y/UvUf1eS75umwuuDEUIcaIzxNWOMLz9g++kk/2vxBn3hjfrsAU08k2R5ZMcti30ZY7zw3/+T5N9nf0jhH3vMIiiYJfn4+l8dh3Wc/WBhN8mHxxj/Zuk5P740UuAHk3zlGl8S6/FSv9c/dsxl7+crPfYwbXKyrNIPUlW3JPnJJN8+xvijv/y9yGcEJ9OR+8Fi6ujHk2SM8YHs/wX4yxbHL08jdj44HVY6Jyzcm8tGBzknnDqrfJ6/2HcE54TTZaXvdVX1+UkuJPmeMcYvv7B/jPGxse+TSX44zgcrEwhxFA8leWHV/zcm+ekDjnkkyeur6uW1f/Wp1yd5pKrOVNUrkqSq/lSSr0/ywl+Bltt9Q5JfWMwT5mQ6cj9Ikqr659n/IvhPlh/wQsi08A1Jnlhv2azB+5PcXlW3VdWN2f8C/9Blx1zp/fxQkntr/0oztyW5PfsLRR6mTU6WI/eDxVTRC9lfmP6/vnDwS3xGcDKt0g9uqqobkqSqXpX988FTi+nI/7uqvmoxPejbc/BnDCfLKp8NqarPSvItWVo/yDnhVFrl8/zA743OCafSkfvB4vifTPKjY4x3X3bfC3+MruyvI+V8sKp1rU5t67Nlf673zyf5cJKfS/KFi/2bSX5w6bi/n/0FYy8m+XuLfZ+b5ANJfj37C4n+2yQ3LO777CT/YXH8ryR51dSv1XZs/eCWJCP7Yc+HFtt3Lu77l4u+8WtJfjHJX5j6tdoO/P3/7SS/mf2/6H/PYt/3JfmGxc9XfD9nf8rhR5I8maWrhBzUpu1kb0ftB0m+N8nvL73/P5Tki1/sM8J2crcV+sE3L37PH8r+xQW2ltrczP4X/Y8k+YEkNfXrtB1fX1jc97okv3xZe84Jp3A7RD/4q9lfU+b3sz9C7PGlx/6J742L/c4Jp2w7aj9I8neS/L/LviO8ZnHfLyT5jUVf+HdJPm/q13nat1r8jwUAAACgCVPGAAAAAJoRCAEAAAA0IxACAAAAaEYgBAAAANCMQAgAAACgGYEQAAAAQDMCIQAAAIBmBEIAAAAAzfx/5CzYzS2wNDYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1440x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "reload(epi_figs)\n",
    "new_link = os.path.join(SARS_DIR, 'link-test.npy')\n",
    "linked = np.load(os.path.join(SARS_DIR, old_date, 'linked-sites.npy'), allow_pickle=True)\n",
    "linked = [[linked[i][j][:-2] for j in range(len(linked[i]))] for i in range(len(linked))]\n",
    "np.save(new_link, linked)\n",
    "\n",
    "null = epi_figs.find_null_distribution_ind_old(\n",
    "    os.path.join(INF_DIR, 'infer-2021-01-31-tv.npz'), \n",
    "    new_link,\n",
    "    os.path.join(SARS_DIR, 'old-data', '2021-01-31', 'individual-regions', 'inference'),\n",
    "    neutral_tol=0.01,\n",
    "    old=True\n",
    ")\n",
    "\n",
    "f = open(os.path.join(SARS_DIR, 'null-distribution-old-1pct.csv'), mode='w')\n",
    "for i in range(len(null)):\n",
    "    f.write('%.8f\\n' % null[i])\n",
    "f.close()\n",
    "\n",
    "null = epi_figs.find_null_distribution_ind_old(\n",
    "    os.path.join(INF_DIR, 'infer-2021-01-31-tv.npz'), \n",
    "    new_link,\n",
    "    os.path.join(SARS_DIR, 'old-data', '2021-01-31', 'individual-regions', 'inference'),\n",
    "    neutral_tol=0.05,\n",
    "    old=True\n",
    ")\n",
    "\n",
    "f = open(os.path.join(SARS_DIR, 'null-distribution-old-5pct.csv'), mode='w')\n",
    "for i in range(len(null)):\n",
    "    f.write('%.8f\\n' % null[i])\n",
    "f.close()\n",
    "\n",
    "null = epi_figs.find_null_distribution_ind_old(\n",
    "    os.path.join(INF_DIR, 'infer-2021-01-31-tv.npz'), \n",
    "    new_link,\n",
    "    os.path.join(SARS_DIR, 'old-data', '2021-01-31', 'individual-regions', 'inference'),\n",
    "    neutral_tol=0.1,\n",
    "    old=True\n",
    ")\n",
    "\n",
    "f = open(os.path.join(SARS_DIR, 'null-distribution-old-10pct.csv'), mode='w')\n",
    "for i in range(len(null)):\n",
    "    f.write('%.8f\\n' % null[i])\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NSP13-220-0-G', 'NSP2-523-1-A', 'NSP3-587-1-G', 'ORF3a-260-0-G', 'ORF8-1-1-A']\n",
      "['NSP5-29-0-T', 'NC-151-T', 'NC-164-T', 'NC-245-T', 'NSP1-10-0-T', 'NSP1-14-0-T', 'NSP1-24-1-T', 'NSP1-25-0-T', 'NSP1-26-0-T', 'NSP1-28-0-T', 'NSP1-29-1-T', 'NSP1-35-0-T', 'NSP1-43-1-T', 'NSP1-60-0-T', 'NSP1-73-1-T', 'NSP1-77-1-T', 'NSP1-93-0-T', 'NSP1-98-0-T', 'NSP1-99-1-T', 'NSP1-113-0-T', 'NSP1-119-1-T', 'NSP1-124-1-T', 'NSP1-127-0-T', 'NSP1-137-0-T', 'NSP1-138-0-T', 'NSP1-139-0-T', 'NSP1-147-0-T', 'NSP1-148-0-T', 'NSP1-171-1-T', 'NSP1-175-1-T', 'NSP1-179-0-T', 'NSP2-4-1-T', 'NSP2-7-0-T', 'NSP2-27-1-T', 'NSP2-37-0-T', 'NSP2-52-1-T', 'NSP2-62-2-T', 'NSP2-64-1-T', 'NSP2-121-1-T', 'NSP2-127-2-T', 'NSP2-153-2-T', 'NSP2-155-0-T', 'NSP2-162-0-T', 'NSP2-206-0-T', 'NSP2-218-1-T', 'NSP2-222-1-T', 'NSP2-246-1-T', 'NSP2-249-0-T', 'NSP2-264-0-T', 'NSP2-365-1-T', 'NSP2-370-1-T', 'NSP2-380-1-T', 'NSP2-386-0-T', 'NSP2-430-2-T', 'NSP2-465-0-T', 'NSP2-481-0-T', 'NSP2-531-2-T', 'NSP2-607-0-T', 'NSP2-612-0-T', 'NSP2-638-0-T', 'NSP3-47-0-T', 'NSP3-54-0-T', 'NSP3-175-0-T', 'NSP3-289-0-T', 'NSP3-301-0-T', 'NSP3-324-0-T', 'NSP3-352-1-T', 'NSP3-386-0-T', 'NSP3-558-1-T', 'NSP3-568-1-T', 'NSP3-586-1-T', 'NSP3-614-2-T', 'NSP3-622-0-T', 'NSP3-646-1-T', 'NSP3-810-1-T', 'NSP3-932-0-T', 'NSP3-1032-0-T', 'NSP3-1105-0-T', 'NSP3-1108-0-T', 'NSP3-1184-2-T', 'NSP3-1193-1-T', 'NSP3-1206-2-T', 'NSP3-1215-0-T', 'NSP3-1215-2-T', 'NSP3-1224-0-T', 'NSP3-1252-0-T', 'NSP3-1341-1-T', 'NSP3-1345-1-T', 'NSP3-1377-2-T', 'NSP3-1430-0-T', 'NSP3-1585-0-T', 'NSP3-1597-1-T', 'NSP3-1604-0-T', 'NSP3-1673-0-T', 'NSP3-1777-0-T', 'NSP3-1786-0-T', 'NSP3-1820-1-T', 'NSP3-1850-0-T', 'NSP3-1868-1-T', 'NSP3-1878-2-T', 'NSP3-1879-1-T', 'NSP3-1883-2-T', 'NSP3-1897-0-T', 'NSP3-1908-1-T', 'NSP3-1912-1-T', 'NSP4-55-1-T', 'NSP4-78-1-T', 'NSP4-103-0-T', 'NSP4-109-2-T', 'NSP4-112-1-T', 'NSP4-186-1-T', 'NSP4-190-1-T', 'NSP4-224-0-T', 'NSP4-293-0-T', 'NSP4-401-1-T', 'NSP4-431-1-T', 'NSP4-439-2-T', 'NSP5-35-0-T', 'NSP5-60-1-T', 'NSP5-105-1-T', 'NSP5-210-0-T', 'NSP5-222-1-T', 'NSP5-261-0-T', 'NSP5-279-1-T', 'NSP6-93-1-T', 'NSP6-233-1-T', 'NSP8-57-1-T', 'NSP9-10-1-T', 'NSP9-17-0-T', 'NSP9-47-0-T', 'NSP9-111-1-T', 'NSP10-72-2-T', 'NSP10-78-1-T', 'NSP10-119-0-T', 'NSP10-121-0-T', 'NSP10-134-1-T', 'NSP10-135-0-T', 'NSP12-10-1-T', 'NSP12-13-0-T', 'NSP12-18-1-T', 'NSP12-21-2-T', 'NSP12-23-0-T', 'NSP12-31-0-T', 'NSP12-55-1-T', 'NSP12-61-0-T', 'NSP12-116-1-T', 'NSP12-118-1-T', 'NSP12-128-0-T', 'NSP12-173-1-T', 'NSP12-174-0-T', 'NSP12-176-0-T', 'NSP12-181-1-T', 'NSP12-183-1-T', 'NSP12-197-1-T', 'NSP12-220-0-T', 'NSP12-276-2-T', 'NSP12-285-1-T', 'NSP12-382-0-T', 'NSP12-392-1-T', 'NSP12-413-0-T', 'NSP12-452-0-T', 'NSP12-457-1-T', 'NSP12-484-0-T', 'NSP12-495-0-T', 'NSP12-529-0-T', 'NSP12-533-1-T', 'NSP12-555-1-T', 'NSP12-557-0-T', 'NSP12-581-0-T', 'NSP12-640-1-T', 'NSP12-651-1-T', 'NSP12-717-0-T', 'NSP12-721-1-T', 'NSP12-747-0-T', 'NSP12-750-1-T', 'NSP12-761-0-T', 'NSP12-841-0-T', 'NSP12-848-0-T', 'NSP12-858-1-T', 'NSP12-860-0-T', 'NSP12-927-2-T', 'NSP13-21-1-T', 'NSP13-32-0-T', 'NSP13-101-0-T', 'NSP13-178-1-T', 'NSP13-186-1-T', 'NSP13-212-1-T', 'NSP13-303-1-T', 'NSP13-313-0-T', 'NSP13-337-1-T', 'NSP13-339-1-T', 'NSP13-401-0-T', 'NSP13-409-1-T', 'NSP13-439-0-T', 'NSP13-443-1-T', 'NSP13-495-0-T', 'NSP13-502-1-T', 'NSP13-594-1-T', 'NSP14-76-1-T', 'NSP14-77-0-T', 'NSP14-84-1-T', 'NSP14-90-0-T', 'NSP14-92-0-T', 'NSP14-141-2-T', 'NSP14-163-1-T', 'NSP14-205-1-T', 'NSP14-213-1-T', 'NSP14-236-0-T', 'NSP14-284-0-T', 'NSP14-289-1-T', 'NSP14-307-2-T', 'NSP14-331-0-T', 'NSP14-390-0-T', 'NSP14-472-2-T', 'NSP14-476-1-T', 'NSP15-61-1-T', 'NSP15-138-1-T', 'NSP15-224-1-T', 'NSP15-233-0-T', 'NSP15-236-0-T', 'NSP15-257-1-T', 'NSP15-283-2-T', 'NSP16-45-0-T', 'NSP16-97-0-T', 'NSP16-133-0-T', 'NSP16-167-0-T', 'NSP16-217-0-T', 'NC-21557-T', 'S-34-1-T', 'S-208-2-T', 'S-344-0-T', 'S-372-0-T', 'S-433-0-T', 'S-476-0-T', 'S-498-1-T', 'S-577-1-T', 'S-634-1-T', 'S-646-1-T', 'S-672-0-T', 'S-682-1-T', 'S-683-1-T', 'S-685-1-T', 'S-765-1-T', 'S-857-0-T', 'S-961-2-T', 'S-983-1-T', 'S-1091-1-T', 'S-1185-1-T', 'S-1199-0-T', 'NC-25386-T', 'ORF3a-31-0-T', 'ORF3a-44-0-T', 'ORF3a-50-0-T', 'ORF3a-59-0-T', 'ORF3a-97-0-T', 'ORF3a-250-0-T', 'ORF3a-251-0-T', 'ORF3a-267-2-T', 'ORF3a-268-2-T', 'ORF3a-273-0-T', 'NC-26238-T', 'E-5-0-T', 'E-11-2-T', 'E-17-0-T', 'E-24-0-T', 'E-36-2-T', 'E-38-1-T', 'E-41-2-T', 'NC-26474-T', 'M-10-0-T', 'M-78-0-T', 'M-105-1-T', 'M-107-1-T', 'M-122-0-T', 'M-139-0-T', 'M-141-0-T', 'M-146-1-T', 'M-150-1-T', 'M-158-1-T', 'M-174-1-T', 'M-186-1-T', 'M-198-1-T', 'ORF6-5-0-T', 'ORF6-20-1-T', 'ORF7a-13-0-T', 'ORF7a-41-0-T', 'ORF7a-105-2-T', 'ORF8-75-0-T', 'ORF8-103-2-T', 'ORF8-114-0-T', 'NC-28261-T', 'N-10-1-T', 'N-36-1-T', 'N-40-1-T', 'N-50-2-T', 'N-72-0-T', 'N-88-1-T', 'N-93-1-T', 'N-95-1-T', 'N-99-0-T', 'N-129-0-T', 'N-149-1-T', 'N-158-0-T', 'N-179-0-T', 'N-185-1-T', 'N-189-1-T', 'N-191-1-T', 'N-259-1-T', 'N-262-1-T', 'N-275-0-T', 'N-277-1-T', 'N-311-0-T', 'N-316-0-T', 'N-319-1-T', 'ORF10-6-0-T', 'ORF10-10-2-T']\n",
      "['NSP5-29-0-T', 'NC-151-T', 'NC-164-T', 'NC-245-T', 'NSP1-10-0-T', 'NSP1-14-0-T', 'NSP1-24-1-T', 'NSP1-25-0-T', 'NSP1-26-0-T', 'NSP1-28-0-T', 'NSP1-29-1-T', 'NSP1-35-0-T', 'NSP1-43-1-T', 'NSP1-60-0-T', 'NSP1-73-1-T', 'NSP1-77-1-T', 'NSP1-93-0-T', 'NSP1-98-0-T', 'NSP1-99-1-T', 'NSP1-113-0-T', 'NSP1-119-1-T', 'NSP1-124-1-T', 'NSP1-127-0-T', 'NSP1-137-0-T', 'NSP1-138-0-T', 'NSP1-139-0-T', 'NSP1-147-0-T', 'NSP1-148-0-T', 'NSP1-171-1-T', 'NSP1-175-1-T', 'NSP1-179-0-T', 'NSP2-4-1-T', 'NSP2-7-0-T', 'NSP2-27-1-T', 'NSP2-37-0-T', 'NSP2-52-1-T', 'NSP2-62-2-T', 'NSP2-64-1-T', 'NSP2-121-1-T', 'NSP2-127-2-T', 'NSP2-153-2-T', 'NSP2-155-0-T', 'NSP2-162-0-T', 'NSP2-206-0-T', 'NSP2-218-1-T', 'NSP2-222-1-T', 'NSP2-246-1-T', 'NSP2-249-0-T', 'NSP2-264-0-T', 'NSP2-365-1-T', 'NSP2-370-1-T', 'NSP2-380-1-T', 'NSP2-386-0-T', 'NSP2-430-2-T', 'NSP2-465-0-T', 'NSP2-481-0-T', 'NSP2-531-2-T', 'NSP2-607-0-T', 'NSP2-612-0-T', 'NSP2-638-0-T', 'NSP3-47-0-T', 'NSP3-54-0-T', 'NSP3-175-0-T', 'NSP3-289-0-T', 'NSP3-301-0-T', 'NSP3-324-0-T', 'NSP3-352-1-T', 'NSP3-386-0-T', 'NSP3-558-1-T', 'NSP3-568-1-T', 'NSP3-586-1-T', 'NSP3-614-2-T', 'NSP3-622-0-T', 'NSP3-646-1-T', 'NSP3-810-1-T', 'NSP3-932-0-T', 'NSP3-1032-0-T', 'NSP3-1105-0-T', 'NSP3-1108-0-T', 'NSP3-1184-2-T', 'NSP3-1193-1-T', 'NSP3-1206-2-T', 'NSP3-1215-0-T', 'NSP3-1215-2-T', 'NSP3-1224-0-T', 'NSP3-1252-0-T', 'NSP3-1341-1-T', 'NSP3-1345-1-T', 'NSP3-1377-2-T', 'NSP3-1430-0-T', 'NSP3-1585-0-T', 'NSP3-1597-1-T', 'NSP3-1604-0-T', 'NSP3-1673-0-T', 'NSP3-1777-0-T', 'NSP3-1786-0-T', 'NSP3-1820-1-T', 'NSP3-1850-0-T', 'NSP3-1868-1-T', 'NSP3-1878-2-T', 'NSP3-1879-1-T', 'NSP3-1883-2-T', 'NSP3-1897-0-T', 'NSP3-1908-1-T', 'NSP3-1912-1-T', 'NSP4-55-1-T', 'NSP4-78-1-T', 'NSP4-103-0-T', 'NSP4-109-2-T', 'NSP4-112-1-T', 'NSP4-186-1-T', 'NSP4-190-1-T', 'NSP4-224-0-T', 'NSP4-293-0-T', 'NSP4-401-1-T', 'NSP4-431-1-T', 'NSP4-439-2-T', 'NSP5-35-0-T', 'NSP5-60-1-T', 'NSP5-105-1-T', 'NSP5-210-0-T', 'NSP5-222-1-T', 'NSP5-261-0-T', 'NSP5-279-1-T', 'NSP6-93-1-T', 'NSP6-233-1-T', 'NSP8-57-1-T', 'NSP9-10-1-T', 'NSP9-17-0-T', 'NSP9-47-0-T', 'NSP9-111-1-T', 'NSP10-72-2-T', 'NSP10-78-1-T', 'NSP10-119-0-T', 'NSP10-121-0-T', 'NSP10-134-1-T', 'NSP10-135-0-T', 'NSP12-10-1-T', 'NSP12-13-0-T', 'NSP12-18-1-T', 'NSP12-21-2-T', 'NSP12-23-0-T', 'NSP12-31-0-T', 'NSP12-55-1-T', 'NSP12-61-0-T', 'NSP12-116-1-T', 'NSP12-118-1-T', 'NSP12-128-0-T', 'NSP12-173-1-T', 'NSP12-174-0-T', 'NSP12-176-0-T', 'NSP12-181-1-T', 'NSP12-183-1-T', 'NSP12-197-1-T', 'NSP12-220-0-T', 'NSP12-276-2-T', 'NSP12-285-1-T', 'NSP12-382-0-T', 'NSP12-392-1-T', 'NSP12-413-0-T', 'NSP12-452-0-T', 'NSP12-457-1-T', 'NSP12-484-0-T', 'NSP12-495-0-T', 'NSP12-529-0-T', 'NSP12-533-1-T', 'NSP12-555-1-T', 'NSP12-557-0-T', 'NSP12-581-0-T', 'NSP12-640-1-T', 'NSP12-651-1-T', 'NSP12-717-0-T', 'NSP12-721-1-T', 'NSP12-747-0-T', 'NSP12-750-1-T', 'NSP12-761-0-T', 'NSP12-841-0-T', 'NSP12-848-0-T', 'NSP12-858-1-T', 'NSP12-860-0-T', 'NSP12-927-2-T', 'NSP13-21-1-T', 'NSP13-32-0-T', 'NSP13-101-0-T', 'NSP13-178-1-T', 'NSP13-186-1-T', 'NSP13-212-1-T', 'NSP13-303-1-T', 'NSP13-313-0-T', 'NSP13-337-1-T', 'NSP13-339-1-T', 'NSP13-401-0-T', 'NSP13-409-1-T', 'NSP13-439-0-T', 'NSP13-443-1-T', 'NSP13-495-0-T', 'NSP13-502-1-T', 'NSP13-594-1-T', 'NSP14-76-1-T', 'NSP14-77-0-T', 'NSP14-84-1-T', 'NSP14-90-0-T', 'NSP14-92-0-T', 'NSP14-141-2-T', 'NSP14-163-1-T', 'NSP14-205-1-T', 'NSP14-213-1-T', 'NSP14-236-0-T', 'NSP14-284-0-T', 'NSP14-289-1-T', 'NSP14-307-2-T', 'NSP14-331-0-T', 'NSP14-390-0-T', 'NSP14-472-2-T', 'NSP14-476-1-T', 'NSP15-61-1-T', 'NSP15-138-1-T', 'NSP15-224-1-T', 'NSP15-233-0-T', 'NSP15-236-0-T', 'NSP15-257-1-T', 'NSP15-283-2-T', 'NSP16-45-0-T', 'NSP16-97-0-T', 'NSP16-133-0-T', 'NSP16-167-0-T', 'NSP16-217-0-T', 'NC-21557-T', 'S-34-1-T', 'S-208-2-T', 'S-344-0-T', 'S-372-0-T', 'S-433-0-T', 'S-476-0-T', 'S-498-1-T', 'S-577-1-T', 'S-634-1-T', 'S-646-1-T', 'S-672-0-T', 'S-682-1-T', 'S-683-1-T', 'S-685-1-T', 'S-765-1-T', 'S-857-0-T', 'S-961-2-T', 'S-983-1-T', 'S-1091-1-T', 'S-1185-1-T', 'S-1199-0-T', 'NC-25386-T', 'ORF3a-31-0-T', 'ORF3a-44-0-T', 'ORF3a-50-0-T', 'ORF3a-59-0-T', 'ORF3a-97-0-T', 'ORF3a-250-0-T', 'ORF3a-251-0-T', 'ORF3a-267-2-T', 'ORF3a-268-2-T', 'ORF3a-273-0-T', 'NC-26238-T', 'E-5-0-T', 'E-11-2-T', 'E-17-0-T', 'E-24-0-T', 'E-36-2-T', 'E-38-1-T', 'E-41-2-T', 'NC-26474-T', 'M-10-0-T', 'M-78-0-T', 'M-105-1-T', 'M-107-1-T', 'M-122-0-T', 'M-139-0-T', 'M-141-0-T', 'M-146-1-T', 'M-150-1-T', 'M-158-1-T', 'M-174-1-T', 'M-186-1-T', 'M-198-1-T', 'ORF6-5-0-T', 'ORF6-20-1-T', 'ORF7a-13-0-T', 'ORF7a-41-0-T', 'ORF7a-105-2-T', 'ORF8-75-0-T', 'ORF8-103-2-T', 'ORF8-114-0-T', 'NC-28261-T', 'N-10-1-T', 'N-36-1-T', 'N-40-1-T', 'N-50-2-T', 'N-72-0-T', 'N-88-1-T', 'N-93-1-T', 'N-95-1-T', 'N-99-0-T', 'N-129-0-T', 'N-149-1-T', 'N-158-0-T', 'N-179-0-T', 'N-185-1-T', 'N-189-1-T', 'N-191-1-T', 'N-259-1-T', 'N-262-1-T', 'N-275-0-T', 'N-277-1-T', 'N-311-0-T', 'N-316-0-T', 'N-319-1-T', 'ORF10-6-0-T', 'ORF10-10-2-T']\n",
      "['S-212-0-G', 'S-212-2-G', 'S-213-0-C', 'S-213-2-A', 'S-214-0-G', 'S-214-2-A']\n",
      "S-214q-2-C not in synonymous sites file\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variants</th>\n",
       "      <th>Number of detections</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Selection</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Omicron</td>\n",
       "      <td>100</td>\n",
       "      <td>NSP5:P132H NSP14:I42V S:G339D S:S371F S:S373P ...</td>\n",
       "      <td>1.365548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta</td>\n",
       "      <td>92</td>\n",
       "      <td>NSP6:T77A NC:209 NSP3:A488S NSP3:P1228L NSP3:P...</td>\n",
       "      <td>0.455085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.1</td>\n",
       "      <td>91</td>\n",
       "      <td>NSP6:I189V NSP3:K38R NSP3:A889A NSP3:A1892T NS...</td>\n",
       "      <td>-0.263008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BA.1</td>\n",
       "      <td>88</td>\n",
       "      <td>NSP6:S104- NSP3:S1265- NSP3:L1266- NSP6:L105- ...</td>\n",
       "      <td>0.085201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BA.2 BA.4 BA.5 BA.2.12.1</td>\n",
       "      <td>83</td>\n",
       "      <td>NSP5:D48D NSP1:S135R NSP3:T24I NSP3:G489S NSP3...</td>\n",
       "      <td>0.311589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Omicron</td>\n",
       "      <td>83</td>\n",
       "      <td>N:G30- N:E31- N:R32- N:S33-</td>\n",
       "      <td>0.163238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Delta</td>\n",
       "      <td>81</td>\n",
       "      <td>S:E156- S:F157- S:R158G ORF8:D119- ORF8:F120-</td>\n",
       "      <td>0.093914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BA.2 BA.4 BA.5 BA.2.12.1</td>\n",
       "      <td>57</td>\n",
       "      <td>S:L24- S:P25- S:P26- S:A27S</td>\n",
       "      <td>0.019667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Alpha</td>\n",
       "      <td>44</td>\n",
       "      <td>NSP12:P412P NSP2:S36S NSP3:T183I NSP3:A890D NS...</td>\n",
       "      <td>0.006829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Gamma</td>\n",
       "      <td>36</td>\n",
       "      <td>NSP9:Y31Y NSP1:D156D NSP3:D10D NSP3:S370L NSP3...</td>\n",
       "      <td>0.031859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BA.2.12.1</td>\n",
       "      <td>30</td>\n",
       "      <td>NSP6:Y234Y NSP12:D523D S:D53D S:L452Q S:S704L</td>\n",
       "      <td>0.276532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Alpha Eta BA.1, BA.4 BA.5</td>\n",
       "      <td>24</td>\n",
       "      <td>S:I68- S:H69- S:V70- S:V143- S:Y144-</td>\n",
       "      <td>-0.040236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta ...</td>\n",
       "      <td>9</td>\n",
       "      <td>NSP6:G107- NSP6:F108-</td>\n",
       "      <td>-0.036588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5...</td>\n",
       "      <td>8</td>\n",
       "      <td>NSP6:S106- NSP6:G107-</td>\n",
       "      <td>0.226321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Mu</td>\n",
       "      <td>7</td>\n",
       "      <td>NSP10:S11S NSP3:T237A NSP3:T720I NSP3:S1106S N...</td>\n",
       "      <td>-0.135542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Alpha Omicron</td>\n",
       "      <td>6</td>\n",
       "      <td>S:N501Y S:P681H N:R203K N:G204R</td>\n",
       "      <td>0.177787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>B.1</td>\n",
       "      <td>4</td>\n",
       "      <td>NSP12:P323L NC:240 NSP3:F106F S:D614G</td>\n",
       "      <td>0.076271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>S:C166S</td>\n",
       "      <td>-0.015890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Beta</td>\n",
       "      <td>2</td>\n",
       "      <td>S:D80A NSP3:K837N S:D215G S:T240- S:L241- S:L2...</td>\n",
       "      <td>-0.119233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP15:V320M NSP3:I414I NSP4:T173I S:A575A S:T1...</td>\n",
       "      <td>-0.070192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>B.1.2</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP5:L89F NSP14:N129D NSP16:R216C ORF3a:G172V ...</td>\n",
       "      <td>0.026101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>2</td>\n",
       "      <td>S:G142D</td>\n",
       "      <td>0.063977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AY.9</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP6:V149A NSP3:P822L NSP4:A446V NSP6:T181I NS...</td>\n",
       "      <td>-0.013722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AY.27</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP7:C32C NSP2:H532Q NSP12:G345G NSP13:V34V M:...</td>\n",
       "      <td>0.048689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AY.68</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:L308L NSP3:P192L S:D253A S:D979E ORF3a:C...</td>\n",
       "      <td>0.004075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>NC:28270</td>\n",
       "      <td>0.064697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AY.4.5</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP10:M137K NSP3:R1868P NSP4:S358Y NSP10:L138R...</td>\n",
       "      <td>0.050494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AY.4</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:F694Y</td>\n",
       "      <td>0.014535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>B.1.279</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP2:R370C NSP2:E575K ORF3a:Y109C</td>\n",
       "      <td>0.182182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Delta Omicron</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP4:T492I S:T478K</td>\n",
       "      <td>0.256211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>B.1.258.22</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP9:M101I NSP3:I1683T NSP12:V720I NSP13:A598S...</td>\n",
       "      <td>0.086190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>L.1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP8:N43H NSP3:P1403S S:M1237T</td>\n",
       "      <td>0.360561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>AY.4.5</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:D481A S:D1259Y</td>\n",
       "      <td>0.017742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>B.1.438.1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP2:M117V NSP3:V1811V NSP8:P10P ORF3a:L95F NC...</td>\n",
       "      <td>0.067477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>EU1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP16:A199A NSP1:V60V NSP3:T1189T M:L93L N:A22...</td>\n",
       "      <td>0.050795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>1</td>\n",
       "      <td>S:G75V NSP3:F1569V S:T76I S:P251- N:G214C S:R2...</td>\n",
       "      <td>-0.190919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>synonymous</td>\n",
       "      <td>1</td>\n",
       "      <td>ORF3a:D265D N:D216D</td>\n",
       "      <td>0.078878</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Variants  Number of detections  \\\n",
       "0                                             Omicron                   100   \n",
       "1                                               Delta                    92   \n",
       "2                                                BA.1                    91   \n",
       "3                                                BA.1                    88   \n",
       "4                            BA.2 BA.4 BA.5 BA.2.12.1                    83   \n",
       "5                                             Omicron                    83   \n",
       "6                                               Delta                    81   \n",
       "7                            BA.2 BA.4 BA.5 BA.2.12.1                    57   \n",
       "8                                               Alpha                    44   \n",
       "9                                               Gamma                    36   \n",
       "10                                          BA.2.12.1                    30   \n",
       "11                          Alpha Eta BA.1, BA.4 BA.5                    24   \n",
       "12  Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta ...                     9   \n",
       "13  Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5...                     8   \n",
       "14                                                 Mu                     7   \n",
       "15                                      Alpha Omicron                     6   \n",
       "16                                                B.1                     4   \n",
       "17                                            unknown                     2   \n",
       "18                                               Beta                     2   \n",
       "19                                          B.1.1.318                     2   \n",
       "20                                              B.1.2                     2   \n",
       "21                                               BA.2                     2   \n",
       "22                                               AY.9                     2   \n",
       "23                                              AY.27                     2   \n",
       "24                                              AY.68                     1   \n",
       "25                                              Delta                     1   \n",
       "26                                             AY.4.5                     1   \n",
       "27                                               AY.4                     1   \n",
       "28                                            B.1.279                     1   \n",
       "29                                      Delta Omicron                     1   \n",
       "30                                         B.1.258.22                     1   \n",
       "31                                                L.1                     1   \n",
       "32                                             AY.4.5                     1   \n",
       "33                                          B.1.438.1                     1   \n",
       "34                                                EU1                     1   \n",
       "35                                             Lambda                     1   \n",
       "36                                         synonymous                     1   \n",
       "\n",
       "                                            Mutations  Selection  \n",
       "0   NSP5:P132H NSP14:I42V S:G339D S:S371F S:S373P ...   1.365548  \n",
       "1   NSP6:T77A NC:209 NSP3:A488S NSP3:P1228L NSP3:P...   0.455085  \n",
       "2   NSP6:I189V NSP3:K38R NSP3:A889A NSP3:A1892T NS...  -0.263008  \n",
       "3   NSP6:S104- NSP3:S1265- NSP3:L1266- NSP6:L105- ...   0.085201  \n",
       "4   NSP5:D48D NSP1:S135R NSP3:T24I NSP3:G489S NSP3...   0.311589  \n",
       "5                         N:G30- N:E31- N:R32- N:S33-   0.163238  \n",
       "6       S:E156- S:F157- S:R158G ORF8:D119- ORF8:F120-   0.093914  \n",
       "7                         S:L24- S:P25- S:P26- S:A27S   0.019667  \n",
       "8   NSP12:P412P NSP2:S36S NSP3:T183I NSP3:A890D NS...   0.006829  \n",
       "9   NSP9:Y31Y NSP1:D156D NSP3:D10D NSP3:S370L NSP3...   0.031859  \n",
       "10      NSP6:Y234Y NSP12:D523D S:D53D S:L452Q S:S704L   0.276532  \n",
       "11               S:I68- S:H69- S:V70- S:V143- S:Y144-  -0.040236  \n",
       "12                              NSP6:G107- NSP6:F108-  -0.036588  \n",
       "13                              NSP6:S106- NSP6:G107-   0.226321  \n",
       "14  NSP10:S11S NSP3:T237A NSP3:T720I NSP3:S1106S N...  -0.135542  \n",
       "15                    S:N501Y S:P681H N:R203K N:G204R   0.177787  \n",
       "16              NSP12:P323L NC:240 NSP3:F106F S:D614G   0.076271  \n",
       "17                                            S:C166S  -0.015890  \n",
       "18  S:D80A NSP3:K837N S:D215G S:T240- S:L241- S:L2...  -0.119233  \n",
       "19  NSP15:V320M NSP3:I414I NSP4:T173I S:A575A S:T1...  -0.070192  \n",
       "20  NSP5:L89F NSP14:N129D NSP16:R216C ORF3a:G172V ...   0.026101  \n",
       "21                                            S:G142D   0.063977  \n",
       "22  NSP6:V149A NSP3:P822L NSP4:A446V NSP6:T181I NS...  -0.013722  \n",
       "23  NSP7:C32C NSP2:H532Q NSP12:G345G NSP13:V34V M:...   0.048689  \n",
       "24  NSP12:L308L NSP3:P192L S:D253A S:D979E ORF3a:C...   0.004075  \n",
       "25                                           NC:28270   0.064697  \n",
       "26  NSP10:M137K NSP3:R1868P NSP4:S358Y NSP10:L138R...   0.050494  \n",
       "27                                        NSP12:F694Y   0.014535  \n",
       "28                  NSP2:R370C NSP2:E575K ORF3a:Y109C   0.182182  \n",
       "29                                 NSP4:T492I S:T478K   0.256211  \n",
       "30  NSP9:M101I NSP3:I1683T NSP12:V720I NSP13:A598S...   0.086190  \n",
       "31                     NSP8:N43H NSP3:P1403S S:M1237T   0.360561  \n",
       "32                               NSP12:D481A S:D1259Y   0.017742  \n",
       "33  NSP2:M117V NSP3:V1811V NSP8:P10P ORF3a:L95F NC...   0.067477  \n",
       "34  NSP16:A199A NSP1:V60V NSP3:T1189T M:L93L N:A22...   0.050795  \n",
       "35  S:G75V NSP3:F1569V S:T76I S:P251- N:G214C S:R2...  -0.190919  \n",
       "36                                ORF3a:D265D N:D216D   0.078878  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Finding the groups of mutations that are inferred to pass the maximum value in the null distribution after 2021-02-1\n",
    "cutoff     = np.amax(np.genfromtxt(os.path.join(SARS_DIR, 'null-distribution-old-10pct.csv')))\n",
    "link_file  = os.path.join(SARS_DIR, DATA_DATE, 'linked-sites.npy')\n",
    "inf_dir    = os.path.join(SARS_DIR, DATA_DATE, f'infer-tv')\n",
    "start_date = (dt.date(2021, 2, 1) - dt.date(2020, 1, 1)).days\n",
    "for file in os.listdir(inf_dir):\n",
    "    path = os.path.join(inf_dir, file)\n",
    "    if '---' in file:\n",
    "        os.remove(path)\n",
    "    elif os.path.isdir(path):\n",
    "        os.rmdir(path)\n",
    "        \n",
    "locs, sites, s = dp.calculate_selection(inf_dir, link_file, start_date=start_date, cutoff=cutoff)\n",
    "\n",
    "counts = {\n",
    "    'NSP6-106-0--'  : ['Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5 Eta Iota Lambda', 0],\n",
    "    'NSP6-107-1--'  : ['Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta Iota Lambda', 0],\n",
    "    'S-69-2--'      : ['Alpha Eta BA.1, BA.4 BA.5', 0],\n",
    "    'S-213-1-G'     : ['BA.2 BA.4 BA.5 BA.2.12.1',  0],\n",
    "    'S-26-1--'      : ['BA.2 BA.4 BA.5 BA.2.12.1',  0],\n",
    "    'S-142-1-A'     : ['BA.2 BA.4 BA.5 BA.2.12.1',  0],\n",
    "    'S-478-1-A'     : ['Delta Omicron',   0],    # this is in Delta and Omicron\n",
    "    'S-681-1-A'     : ['Alpha Omicron',   0],\n",
    "    'S-704-1-T'     : ['BA.2.12.1',       0],\n",
    "    'S-614-1-G'     : ['B.1',             0],\n",
    "    'S-681-1-G'     : ['Delta',           0],\n",
    "    'S-796-0-T'     : ['Omicron',         0],\n",
    "    'S-211-1--'     : ['BA.1',            0],\n",
    "    'S-496-0-A'     : ['BA.1',            0],\n",
    "    'N-31-1--'      : ['Omicron',         0],\n",
    "    'S-138-0-T'     : ['Gamma',           0],\n",
    "    'S-76-1-T'      : ['Lambda',          0],\n",
    "    'S-570-1-A'     : ['Alpha',           0],\n",
    "    'S-157-2--'     : ['Delta',           0],\n",
    "    'S-144-1-C'     : ['Mu',              0],\n",
    "    'S-80-1-C'      : ['Beta',            0],\n",
    "    'S-575-2-C'     : ['B.1.1.318',       0],\n",
    "    'N-220-1-T'     : ['EU1',             0],\n",
    "    'N-67-0-T'      : ['B.1.2',           0],\n",
    "    'NSP3-822-1-T'  : ['AY.9',            0],\n",
    "    'N-401-0-T'     : ['B.1.438.1',       0],\n",
    "    'S-1259-0-T'    : ['AY.4.5',          0],\n",
    "    'M-81-0-T'      : ['AY.27',           0],\n",
    "    'S-1237-1-C'    : ['L.1',             0],\n",
    "    'S-1116-2-C'    : ['B.1.258.22',      0],\n",
    "    'S-253-1-C'     : ['AY.68',           0],\n",
    "    'ORF3a-109-1-G' : ['B.1.279',         0],\n",
    "    'NSP12-694-1-A' : ['AY.4',            0],\n",
    "    'NSP4-358-1-A'  : ['AY.4.5',          0],    # sublineage of AY.4.5\n",
    "    'NC-28270--'    : ['Delta',           0],    # NC:A28270T is an Omicron mutation\n",
    "    'S-166-1-C'     : ['unknown',         0],    # IS THIS A SEQUENCING ISSUE OR SOMETHING?\n",
    "    'N-216-2-C'     : ['synonymous',      0]\n",
    "}\n",
    "\n",
    "# SNVs to ignore because they are an alternate nucleotide in a site that belongs to one of the main variants'S-142-1-C'\n",
    "ignore = ['NC-28270-G', 'NC-28270-A', 'NC-28270-C', \n",
    "          'S-142-1-C', 'S-142-1-G', 'S-142-1-T', \n",
    "          'S-166-1-G', 'S-166-1-A', 'S-166-1-T', 'S-166-1-C', 'S-166-1--',\n",
    "          'ORF8-120-2-A', 'ORF8-120-2-T',\n",
    "          'NSP6-77-0--', 'NSP6-77-0-C', 'NSP6-77-0-T',\n",
    "          'S-214n-2-G',    # omicron insertion probably aligned or sequenced incorrectly (in Idaho), but only inferred large in India subregions\n",
    "         ]\n",
    "\n",
    "keys = [i for i in counts]\n",
    "mut_groups = [[] for i in range(len(keys))]\n",
    "\n",
    "for group in sites:\n",
    "    if isinstance(group, list):\n",
    "        test_site = group[0]\n",
    "    else:\n",
    "        test_site = group\n",
    "    test_split = test_site.split('-')\n",
    "    if test_split[0]=='NC':\n",
    "        if int(test_split[1])<150 or int(test_split[1])>29690:\n",
    "            continue\n",
    "    listed = False\n",
    "    for key in keys:\n",
    "        if key in group:\n",
    "            counts[key][-1] += 1\n",
    "            mut_groups[keys.index(key)] = group\n",
    "            listed = True\n",
    "            break\n",
    "    if not listed:\n",
    "        if isinstance(group, list):\n",
    "            if not any(np.isin(group, ignore)):\n",
    "                print(group)\n",
    "        else:\n",
    "            if not group in ignore:\n",
    "                print(group)\n",
    "        \n",
    "values     = [counts[i] for i in keys]\n",
    "labels     = [i[0] for i in values]\n",
    "detections = [i[1] for i in values]\n",
    "mutations  = []\n",
    "\n",
    "syn_file = np.load(os.path.join(DATA_DIR, 'synonymous-prot.npz'), allow_pickle=True)\n",
    "syn_locs = syn_file['locations']\n",
    "aas      = syn_file['aa_changes']\n",
    "initial_aa = [i[0] for i in aas]\n",
    "final_aa   = [i[-1] for i in aas]\n",
    "new_muts   = []\n",
    "for group in mut_groups:\n",
    "    if isinstance(group, list):\n",
    "        new_group = []\n",
    "        for mut in group: \n",
    "            if mut not in syn_locs:\n",
    "                print(f'{mut} not in synonymous sites file')\n",
    "                continue\n",
    "            idx = list(syn_locs).index(mut)\n",
    "            new_label = mut.split('-')\n",
    "            if new_label[0] == 'NC':\n",
    "                new_label = f'NC:{new_label[1]}'\n",
    "            else:\n",
    "                new_label = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "            if new_label not in new_group:\n",
    "                new_group.append(new_label)\n",
    "    else:\n",
    "        idx = list(syn_locs).index(group)\n",
    "        new_label = group.split('-')\n",
    "        if new_label[0] == 'NC':\n",
    "            new_group = f'NC:{new_label[1]}'\n",
    "        else:\n",
    "            new_group = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "        #new_group = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "    new_muts.append(new_group)\n",
    "\n",
    "for group in new_muts:\n",
    "    if isinstance(group, list):\n",
    "        mutations.append(' '.join(group))\n",
    "    else:\n",
    "        mutations.append(group)\n",
    "        \n",
    "# Find the ultimate selection coefficient inferred for each group\n",
    "inf_data  = np.load(os.path.join(DATA_DIR, f'infer-{DATA_DATE}', f'infer-g-10.npz'), allow_pickle=True)\n",
    "alleles   = [dp.get_label_new(i) for i in inf_data['allele_number']]\n",
    "selection = inf_data['selection']\n",
    "\n",
    "s_groups = np.zeros(len(mut_groups))\n",
    "for i in range(len(mut_groups)):\n",
    "    if not isinstance(mut_groups[i], list):\n",
    "        s_groups[i] = selection[alleles.index(mut_groups[i])]\n",
    "        continue\n",
    "    for j in range(len(mut_groups[i])):\n",
    "        if mut_groups[i][j] not in alleles:\n",
    "            print(mut_groups[i][j])\n",
    "            continue\n",
    "        s_groups[i] += selection[alleles.index(mut_groups[i][j])]        \n",
    "\n",
    "detect_dic = {'Variants' : labels, 'Number of detections' : detections, 'Mutations' : mutations, 'Selection' : s_groups}\n",
    "df         = pd.DataFrame.from_dict(detect_dic)\n",
    "df         = df.sort_values(by=['Number of detections'], ascending=False, ignore_index=True)\n",
    "#df['Rank'] = np.arange(1, len(df)+1)\n",
    "df.to_csv(os.path.join(DATA_DIR, 'early-detection-sel.csv'), index=False)\n",
    "display(df)\n",
    "del df['Selection']\n",
    "df.to_csv(os.path.join(DATA_DIR, 'early-detection.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total detections of major variants is 762\n",
      "\n",
      "major variant are 0.9707006369426752 percentage of early detections\n",
      "likely true predictions are 0.9796178343949045 percentage of early detections\n"
     ]
    }
   ],
   "source": [
    "major_var = np.sum([100, 92, 91, 88, 83, 81, 57, 44, 36, 30, 24, 9, 8, 7, 6, 2, 2, 1, 1])\n",
    "minor     = np.sum([4, 2, 1])    # B.1, EU1, B.1.1.318\n",
    "unknown   = np.sum([2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1])\n",
    "total     = major_var + minor + unknown\n",
    "print(f'Total detections of major variants is {major_var}')\n",
    "print('')\n",
    "print(f'major variant are {major_var / total} percentage of early detections')\n",
    "print(f'likely true predictions are {(major_var + minor) / total} percentage of early detections')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-166-1-T\n",
      "europe-estonia\n",
      "ORF8-120-2-A\n",
      "asia-india-andhra pradesh\n",
      "['S-214n-2-G', 'S-214o-2-A', 'S-214p-2-G', 'S-214r-2-C', 'S-214t-2-G']\n",
      "asia-india-tamil nadu\n",
      "['NSP13-220-0-G', 'NSP2-523-1-A', 'NSP3-587-1-G', 'ORF3a-260-0-G', 'ORF8-1-1-A']\n",
      "asia-indonesia\n",
      "['S-214n-2-G', 'S-214o-2-A', 'S-214p-2-G', 'S-214r-2-C', 'S-214t-2-G']\n",
      "asia-india-tamil nadu\n",
      "['NSP5-29-0-T', 'NC-151-T', 'NC-164-T', 'NC-245-T', 'NSP1-10-0-T', 'NSP1-14-0-T', 'NSP1-24-1-T', 'NSP1-25-0-T', 'NSP1-26-0-T', 'NSP1-28-0-T', 'NSP1-29-1-T', 'NSP1-35-0-T', 'NSP1-43-1-T', 'NSP1-60-0-T', 'NSP1-73-1-T', 'NSP1-77-1-T', 'NSP1-93-0-T', 'NSP1-98-0-T', 'NSP1-99-1-T', 'NSP1-113-0-T', 'NSP1-119-1-T', 'NSP1-124-1-T', 'NSP1-127-0-T', 'NSP1-137-0-T', 'NSP1-138-0-T', 'NSP1-139-0-T', 'NSP1-147-0-T', 'NSP1-148-0-T', 'NSP1-171-1-T', 'NSP1-175-1-T', 'NSP1-179-0-T', 'NSP2-4-1-T', 'NSP2-7-0-T', 'NSP2-27-1-T', 'NSP2-37-0-T', 'NSP2-52-1-T', 'NSP2-62-2-T', 'NSP2-64-1-T', 'NSP2-121-1-T', 'NSP2-127-2-T', 'NSP2-153-2-T', 'NSP2-155-0-T', 'NSP2-162-0-T', 'NSP2-206-0-T', 'NSP2-218-1-T', 'NSP2-222-1-T', 'NSP2-246-1-T', 'NSP2-249-0-T', 'NSP2-264-0-T', 'NSP2-365-1-T', 'NSP2-370-1-T', 'NSP2-380-1-T', 'NSP2-386-0-T', 'NSP2-430-2-T', 'NSP2-465-0-T', 'NSP2-481-0-T', 'NSP2-531-2-T', 'NSP2-607-0-T', 'NSP2-612-0-T', 'NSP2-638-0-T', 'NSP3-47-0-T', 'NSP3-54-0-T', 'NSP3-175-0-T', 'NSP3-289-0-T', 'NSP3-301-0-T', 'NSP3-324-0-T', 'NSP3-352-1-T', 'NSP3-386-0-T', 'NSP3-558-1-T', 'NSP3-568-1-T', 'NSP3-586-1-T', 'NSP3-614-2-T', 'NSP3-622-0-T', 'NSP3-646-1-T', 'NSP3-810-1-T', 'NSP3-932-0-T', 'NSP3-1032-0-T', 'NSP3-1105-0-T', 'NSP3-1108-0-T', 'NSP3-1184-2-T', 'NSP3-1193-1-T', 'NSP3-1206-2-T', 'NSP3-1215-0-T', 'NSP3-1215-2-T', 'NSP3-1224-0-T', 'NSP3-1252-0-T', 'NSP3-1341-1-T', 'NSP3-1345-1-T', 'NSP3-1377-2-T', 'NSP3-1430-0-T', 'NSP3-1585-0-T', 'NSP3-1597-1-T', 'NSP3-1604-0-T', 'NSP3-1673-0-T', 'NSP3-1777-0-T', 'NSP3-1786-0-T', 'NSP3-1820-1-T', 'NSP3-1850-0-T', 'NSP3-1868-1-T', 'NSP3-1878-2-T', 'NSP3-1879-1-T', 'NSP3-1883-2-T', 'NSP3-1897-0-T', 'NSP3-1908-1-T', 'NSP3-1912-1-T', 'NSP4-55-1-T', 'NSP4-78-1-T', 'NSP4-103-0-T', 'NSP4-109-2-T', 'NSP4-112-1-T', 'NSP4-186-1-T', 'NSP4-190-1-T', 'NSP4-224-0-T', 'NSP4-293-0-T', 'NSP4-401-1-T', 'NSP4-431-1-T', 'NSP4-439-2-T', 'NSP5-35-0-T', 'NSP5-60-1-T', 'NSP5-105-1-T', 'NSP5-210-0-T', 'NSP5-222-1-T', 'NSP5-261-0-T', 'NSP5-279-1-T', 'NSP6-93-1-T', 'NSP6-233-1-T', 'NSP8-57-1-T', 'NSP9-10-1-T', 'NSP9-17-0-T', 'NSP9-47-0-T', 'NSP9-111-1-T', 'NSP10-72-2-T', 'NSP10-78-1-T', 'NSP10-119-0-T', 'NSP10-121-0-T', 'NSP10-134-1-T', 'NSP10-135-0-T', 'NSP12-10-1-T', 'NSP12-13-0-T', 'NSP12-18-1-T', 'NSP12-21-2-T', 'NSP12-23-0-T', 'NSP12-31-0-T', 'NSP12-55-1-T', 'NSP12-61-0-T', 'NSP12-116-1-T', 'NSP12-118-1-T', 'NSP12-128-0-T', 'NSP12-173-1-T', 'NSP12-174-0-T', 'NSP12-176-0-T', 'NSP12-181-1-T', 'NSP12-183-1-T', 'NSP12-197-1-T', 'NSP12-220-0-T', 'NSP12-276-2-T', 'NSP12-285-1-T', 'NSP12-382-0-T', 'NSP12-392-1-T', 'NSP12-413-0-T', 'NSP12-452-0-T', 'NSP12-457-1-T', 'NSP12-484-0-T', 'NSP12-495-0-T', 'NSP12-529-0-T', 'NSP12-533-1-T', 'NSP12-555-1-T', 'NSP12-557-0-T', 'NSP12-581-0-T', 'NSP12-640-1-T', 'NSP12-651-1-T', 'NSP12-717-0-T', 'NSP12-721-1-T', 'NSP12-747-0-T', 'NSP12-750-1-T', 'NSP12-761-0-T', 'NSP12-841-0-T', 'NSP12-848-0-T', 'NSP12-858-1-T', 'NSP12-860-0-T', 'NSP12-927-2-T', 'NSP13-21-1-T', 'NSP13-32-0-T', 'NSP13-101-0-T', 'NSP13-178-1-T', 'NSP13-186-1-T', 'NSP13-212-1-T', 'NSP13-303-1-T', 'NSP13-313-0-T', 'NSP13-337-1-T', 'NSP13-339-1-T', 'NSP13-401-0-T', 'NSP13-409-1-T', 'NSP13-439-0-T', 'NSP13-443-1-T', 'NSP13-495-0-T', 'NSP13-502-1-T', 'NSP13-594-1-T', 'NSP14-76-1-T', 'NSP14-77-0-T', 'NSP14-84-1-T', 'NSP14-90-0-T', 'NSP14-92-0-T', 'NSP14-141-2-T', 'NSP14-163-1-T', 'NSP14-205-1-T', 'NSP14-213-1-T', 'NSP14-236-0-T', 'NSP14-284-0-T', 'NSP14-289-1-T', 'NSP14-307-2-T', 'NSP14-331-0-T', 'NSP14-390-0-T', 'NSP14-472-2-T', 'NSP14-476-1-T', 'NSP15-61-1-T', 'NSP15-138-1-T', 'NSP15-224-1-T', 'NSP15-233-0-T', 'NSP15-236-0-T', 'NSP15-257-1-T', 'NSP15-283-2-T', 'NSP16-45-0-T', 'NSP16-97-0-T', 'NSP16-133-0-T', 'NSP16-167-0-T', 'NSP16-217-0-T', 'NC-21557-T', 'S-34-1-T', 'S-208-2-T', 'S-344-0-T', 'S-372-0-T', 'S-433-0-T', 'S-476-0-T', 'S-498-1-T', 'S-577-1-T', 'S-634-1-T', 'S-646-1-T', 'S-672-0-T', 'S-682-1-T', 'S-683-1-T', 'S-685-1-T', 'S-765-1-T', 'S-857-0-T', 'S-961-2-T', 'S-983-1-T', 'S-1091-1-T', 'S-1185-1-T', 'S-1199-0-T', 'NC-25386-T', 'ORF3a-31-0-T', 'ORF3a-44-0-T', 'ORF3a-50-0-T', 'ORF3a-59-0-T', 'ORF3a-97-0-T', 'ORF3a-250-0-T', 'ORF3a-251-0-T', 'ORF3a-267-2-T', 'ORF3a-268-2-T', 'ORF3a-273-0-T', 'NC-26238-T', 'E-5-0-T', 'E-11-2-T', 'E-17-0-T', 'E-24-0-T', 'E-36-2-T', 'E-38-1-T', 'E-41-2-T', 'NC-26474-T', 'M-10-0-T', 'M-78-0-T', 'M-105-1-T', 'M-107-1-T', 'M-122-0-T', 'M-139-0-T', 'M-141-0-T', 'M-146-1-T', 'M-150-1-T', 'M-158-1-T', 'M-174-1-T', 'M-186-1-T', 'M-198-1-T', 'ORF6-5-0-T', 'ORF6-20-1-T', 'ORF7a-13-0-T', 'ORF7a-41-0-T', 'ORF7a-105-2-T', 'ORF8-75-0-T', 'ORF8-103-2-T', 'ORF8-114-0-T', 'NC-28261-T', 'N-10-1-T', 'N-36-1-T', 'N-40-1-T', 'N-50-2-T', 'N-72-0-T', 'N-88-1-T', 'N-93-1-T', 'N-95-1-T', 'N-99-0-T', 'N-129-0-T', 'N-149-1-T', 'N-158-0-T', 'N-179-0-T', 'N-185-1-T', 'N-189-1-T', 'N-191-1-T', 'N-259-1-T', 'N-262-1-T', 'N-275-0-T', 'N-277-1-T', 'N-311-0-T', 'N-316-0-T', 'N-319-1-T', 'ORF10-6-0-T', 'ORF10-10-2-T']\n",
      "asia-india-gujarat\n",
      "['NSP5-29-0-T', 'NC-151-T', 'NC-164-T', 'NC-245-T', 'NSP1-10-0-T', 'NSP1-14-0-T', 'NSP1-24-1-T', 'NSP1-25-0-T', 'NSP1-26-0-T', 'NSP1-28-0-T', 'NSP1-29-1-T', 'NSP1-35-0-T', 'NSP1-43-1-T', 'NSP1-60-0-T', 'NSP1-73-1-T', 'NSP1-77-1-T', 'NSP1-93-0-T', 'NSP1-98-0-T', 'NSP1-99-1-T', 'NSP1-113-0-T', 'NSP1-119-1-T', 'NSP1-124-1-T', 'NSP1-127-0-T', 'NSP1-137-0-T', 'NSP1-138-0-T', 'NSP1-139-0-T', 'NSP1-147-0-T', 'NSP1-148-0-T', 'NSP1-171-1-T', 'NSP1-175-1-T', 'NSP1-179-0-T', 'NSP2-4-1-T', 'NSP2-7-0-T', 'NSP2-27-1-T', 'NSP2-37-0-T', 'NSP2-52-1-T', 'NSP2-62-2-T', 'NSP2-64-1-T', 'NSP2-121-1-T', 'NSP2-127-2-T', 'NSP2-153-2-T', 'NSP2-155-0-T', 'NSP2-162-0-T', 'NSP2-206-0-T', 'NSP2-218-1-T', 'NSP2-222-1-T', 'NSP2-246-1-T', 'NSP2-249-0-T', 'NSP2-264-0-T', 'NSP2-365-1-T', 'NSP2-370-1-T', 'NSP2-380-1-T', 'NSP2-386-0-T', 'NSP2-430-2-T', 'NSP2-465-0-T', 'NSP2-481-0-T', 'NSP2-531-2-T', 'NSP2-607-0-T', 'NSP2-612-0-T', 'NSP2-638-0-T', 'NSP3-47-0-T', 'NSP3-54-0-T', 'NSP3-175-0-T', 'NSP3-289-0-T', 'NSP3-301-0-T', 'NSP3-324-0-T', 'NSP3-352-1-T', 'NSP3-386-0-T', 'NSP3-558-1-T', 'NSP3-568-1-T', 'NSP3-586-1-T', 'NSP3-614-2-T', 'NSP3-622-0-T', 'NSP3-646-1-T', 'NSP3-810-1-T', 'NSP3-932-0-T', 'NSP3-1032-0-T', 'NSP3-1105-0-T', 'NSP3-1108-0-T', 'NSP3-1184-2-T', 'NSP3-1193-1-T', 'NSP3-1206-2-T', 'NSP3-1215-0-T', 'NSP3-1215-2-T', 'NSP3-1224-0-T', 'NSP3-1252-0-T', 'NSP3-1341-1-T', 'NSP3-1345-1-T', 'NSP3-1377-2-T', 'NSP3-1430-0-T', 'NSP3-1585-0-T', 'NSP3-1597-1-T', 'NSP3-1604-0-T', 'NSP3-1673-0-T', 'NSP3-1777-0-T', 'NSP3-1786-0-T', 'NSP3-1820-1-T', 'NSP3-1850-0-T', 'NSP3-1868-1-T', 'NSP3-1878-2-T', 'NSP3-1879-1-T', 'NSP3-1883-2-T', 'NSP3-1897-0-T', 'NSP3-1908-1-T', 'NSP3-1912-1-T', 'NSP4-55-1-T', 'NSP4-78-1-T', 'NSP4-103-0-T', 'NSP4-109-2-T', 'NSP4-112-1-T', 'NSP4-186-1-T', 'NSP4-190-1-T', 'NSP4-224-0-T', 'NSP4-293-0-T', 'NSP4-401-1-T', 'NSP4-431-1-T', 'NSP4-439-2-T', 'NSP5-35-0-T', 'NSP5-60-1-T', 'NSP5-105-1-T', 'NSP5-210-0-T', 'NSP5-222-1-T', 'NSP5-261-0-T', 'NSP5-279-1-T', 'NSP6-93-1-T', 'NSP6-233-1-T', 'NSP8-57-1-T', 'NSP9-10-1-T', 'NSP9-17-0-T', 'NSP9-47-0-T', 'NSP9-111-1-T', 'NSP10-72-2-T', 'NSP10-78-1-T', 'NSP10-119-0-T', 'NSP10-121-0-T', 'NSP10-134-1-T', 'NSP10-135-0-T', 'NSP12-10-1-T', 'NSP12-13-0-T', 'NSP12-18-1-T', 'NSP12-21-2-T', 'NSP12-23-0-T', 'NSP12-31-0-T', 'NSP12-55-1-T', 'NSP12-61-0-T', 'NSP12-116-1-T', 'NSP12-118-1-T', 'NSP12-128-0-T', 'NSP12-173-1-T', 'NSP12-174-0-T', 'NSP12-176-0-T', 'NSP12-181-1-T', 'NSP12-183-1-T', 'NSP12-197-1-T', 'NSP12-220-0-T', 'NSP12-276-2-T', 'NSP12-285-1-T', 'NSP12-382-0-T', 'NSP12-392-1-T', 'NSP12-413-0-T', 'NSP12-452-0-T', 'NSP12-457-1-T', 'NSP12-484-0-T', 'NSP12-495-0-T', 'NSP12-529-0-T', 'NSP12-533-1-T', 'NSP12-555-1-T', 'NSP12-557-0-T', 'NSP12-581-0-T', 'NSP12-640-1-T', 'NSP12-651-1-T', 'NSP12-717-0-T', 'NSP12-721-1-T', 'NSP12-747-0-T', 'NSP12-750-1-T', 'NSP12-761-0-T', 'NSP12-841-0-T', 'NSP12-848-0-T', 'NSP12-858-1-T', 'NSP12-860-0-T', 'NSP12-927-2-T', 'NSP13-21-1-T', 'NSP13-32-0-T', 'NSP13-101-0-T', 'NSP13-178-1-T', 'NSP13-186-1-T', 'NSP13-212-1-T', 'NSP13-303-1-T', 'NSP13-313-0-T', 'NSP13-337-1-T', 'NSP13-339-1-T', 'NSP13-401-0-T', 'NSP13-409-1-T', 'NSP13-439-0-T', 'NSP13-443-1-T', 'NSP13-495-0-T', 'NSP13-502-1-T', 'NSP13-594-1-T', 'NSP14-76-1-T', 'NSP14-77-0-T', 'NSP14-84-1-T', 'NSP14-90-0-T', 'NSP14-92-0-T', 'NSP14-141-2-T', 'NSP14-163-1-T', 'NSP14-205-1-T', 'NSP14-213-1-T', 'NSP14-236-0-T', 'NSP14-284-0-T', 'NSP14-289-1-T', 'NSP14-307-2-T', 'NSP14-331-0-T', 'NSP14-390-0-T', 'NSP14-472-2-T', 'NSP14-476-1-T', 'NSP15-61-1-T', 'NSP15-138-1-T', 'NSP15-224-1-T', 'NSP15-233-0-T', 'NSP15-236-0-T', 'NSP15-257-1-T', 'NSP15-283-2-T', 'NSP16-45-0-T', 'NSP16-97-0-T', 'NSP16-133-0-T', 'NSP16-167-0-T', 'NSP16-217-0-T', 'NC-21557-T', 'S-34-1-T', 'S-208-2-T', 'S-344-0-T', 'S-372-0-T', 'S-433-0-T', 'S-476-0-T', 'S-498-1-T', 'S-577-1-T', 'S-634-1-T', 'S-646-1-T', 'S-672-0-T', 'S-682-1-T', 'S-683-1-T', 'S-685-1-T', 'S-765-1-T', 'S-857-0-T', 'S-961-2-T', 'S-983-1-T', 'S-1091-1-T', 'S-1185-1-T', 'S-1199-0-T', 'NC-25386-T', 'ORF3a-31-0-T', 'ORF3a-44-0-T', 'ORF3a-50-0-T', 'ORF3a-59-0-T', 'ORF3a-97-0-T', 'ORF3a-250-0-T', 'ORF3a-251-0-T', 'ORF3a-267-2-T', 'ORF3a-268-2-T', 'ORF3a-273-0-T', 'NC-26238-T', 'E-5-0-T', 'E-11-2-T', 'E-17-0-T', 'E-24-0-T', 'E-36-2-T', 'E-38-1-T', 'E-41-2-T', 'NC-26474-T', 'M-10-0-T', 'M-78-0-T', 'M-105-1-T', 'M-107-1-T', 'M-122-0-T', 'M-139-0-T', 'M-141-0-T', 'M-146-1-T', 'M-150-1-T', 'M-158-1-T', 'M-174-1-T', 'M-186-1-T', 'M-198-1-T', 'ORF6-5-0-T', 'ORF6-20-1-T', 'ORF7a-13-0-T', 'ORF7a-41-0-T', 'ORF7a-105-2-T', 'ORF8-75-0-T', 'ORF8-103-2-T', 'ORF8-114-0-T', 'NC-28261-T', 'N-10-1-T', 'N-36-1-T', 'N-40-1-T', 'N-50-2-T', 'N-72-0-T', 'N-88-1-T', 'N-93-1-T', 'N-95-1-T', 'N-99-0-T', 'N-129-0-T', 'N-149-1-T', 'N-158-0-T', 'N-179-0-T', 'N-185-1-T', 'N-189-1-T', 'N-191-1-T', 'N-259-1-T', 'N-262-1-T', 'N-275-0-T', 'N-277-1-T', 'N-311-0-T', 'N-316-0-T', 'N-319-1-T', 'ORF10-6-0-T', 'ORF10-10-2-T']\n",
      "asia-india-gujarat\n",
      "S-936-0-T\n",
      "europe-finland\n",
      "S-142-1-C\n",
      "asia-india-kerala\n",
      "S-142-1-T\n",
      "asia-india-kerala\n",
      "NC-28270-C\n",
      "asia-india-kerala\n",
      "NC-28270-G\n",
      "asia-india-kerala\n",
      "['S-212-0-G', 'S-212-2-G', 'S-213-0-C', 'S-213-2-A', 'S-214-0-G', 'S-214-2-A']\n",
      "europe-ireland\n",
      "['S-214n-2-G', 'S-214o-2-A', 'S-214p-2-G', 'S-214r-2-C', 'S-214t-2-G']\n",
      "asia-india-tamil nadu\n",
      "['S-214n-2-G', 'S-214o-2-A', 'S-214p-2-G', 'S-214r-2-C', 'S-214t-2-G']\n",
      "asia-india-tamil nadu\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Regions Detected, 5% threshold</th>\n",
       "      <th>Regions Detected, 10% threshold</th>\n",
       "      <th>Regions Detected, 12% threshold</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Final inferred coefficient (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Omicron</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>NSP5:P132H NSP14:I42V S:G339D S:S371F S:S373P ...</td>\n",
       "      <td>1.365548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Delta</td>\n",
       "      <td>97</td>\n",
       "      <td>92</td>\n",
       "      <td>90</td>\n",
       "      <td>NSP6:T77A NC:209 NSP3:A488S NSP3:P1228L NSP3:P...</td>\n",
       "      <td>0.455085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.1</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>90</td>\n",
       "      <td>NSP6:I189V NSP3:K38R NSP3:A889A NSP3:A1892T NS...</td>\n",
       "      <td>-0.263008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BA.1</td>\n",
       "      <td>89</td>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>NSP6:S104- NSP3:S1265- NSP3:L1266- NSP6:L105- ...</td>\n",
       "      <td>0.085201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BA.2 BA.4 BA.5 BA.2.12.1</td>\n",
       "      <td>83</td>\n",
       "      <td>83</td>\n",
       "      <td>81</td>\n",
       "      <td>NSP5:D48D NSP1:S135R NSP3:T24I NSP3:G489S NSP3...</td>\n",
       "      <td>0.311589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Omicron</td>\n",
       "      <td>85</td>\n",
       "      <td>83</td>\n",
       "      <td>81</td>\n",
       "      <td>N:G30- N:E31- N:R32- N:S33-</td>\n",
       "      <td>0.163238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Delta</td>\n",
       "      <td>83</td>\n",
       "      <td>81</td>\n",
       "      <td>79</td>\n",
       "      <td>S:E156- S:F157- S:R158G ORF8:D119- ORF8:F120-</td>\n",
       "      <td>0.093914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BA.2 BA.4 BA.5 BA.2.12.1</td>\n",
       "      <td>64</td>\n",
       "      <td>57</td>\n",
       "      <td>55</td>\n",
       "      <td>S:L24- S:P25- S:P26- S:A27S</td>\n",
       "      <td>0.019667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Alpha</td>\n",
       "      <td>57</td>\n",
       "      <td>44</td>\n",
       "      <td>39</td>\n",
       "      <td>NSP12:P412P NSP2:S36S NSP3:T183I NSP3:A890D NS...</td>\n",
       "      <td>0.006829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Gamma</td>\n",
       "      <td>42</td>\n",
       "      <td>36</td>\n",
       "      <td>34</td>\n",
       "      <td>NSP9:Y31Y NSP1:D156D NSP3:D10D NSP3:S370L NSP3...</td>\n",
       "      <td>0.031859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BA.2.12.1</td>\n",
       "      <td>33</td>\n",
       "      <td>30</td>\n",
       "      <td>29</td>\n",
       "      <td>NSP6:Y234Y NSP12:D523D S:D53D S:L452Q S:S704L</td>\n",
       "      <td>0.276532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Alpha Eta BA.1, BA.4 BA.5</td>\n",
       "      <td>32</td>\n",
       "      <td>24</td>\n",
       "      <td>21</td>\n",
       "      <td>S:I68- S:H69- S:V70- S:V143- S:Y144-</td>\n",
       "      <td>-0.040236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta ...</td>\n",
       "      <td>28</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>NSP6:G107- NSP6:F108-</td>\n",
       "      <td>-0.036588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5...</td>\n",
       "      <td>23</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>NSP6:S106- NSP6:G107-</td>\n",
       "      <td>0.226321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Mu</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>NSP10:S11S NSP3:T237A NSP3:T720I NSP3:S1106S N...</td>\n",
       "      <td>-0.135542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Alpha Omicron</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>S:N501Y S:P681H N:R203K N:G204R</td>\n",
       "      <td>0.177787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>B.1</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NSP12:P323L NC:240 NSP3:F106F S:D614G</td>\n",
       "      <td>0.076271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP15:V320M NSP3:I414I NSP4:T173I S:A575A S:T1...</td>\n",
       "      <td>-0.070192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>S:G142D</td>\n",
       "      <td>0.063977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>B.1.2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP5:L89F NSP14:N129D NSP16:R216C ORF3a:G172V ...</td>\n",
       "      <td>0.026101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>unknown</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>S:C166S</td>\n",
       "      <td>-0.015890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Beta</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>S:D80A NSP3:K837N S:D215G S:T240- S:L241- S:L2...</td>\n",
       "      <td>-0.119233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>AY.27</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NSP7:C32C NSP2:H532Q NSP12:G345G NSP13:V34V M:...</td>\n",
       "      <td>0.048689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>AY.9</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP6:V149A NSP3:P822L NSP4:A446V NSP6:T181I NS...</td>\n",
       "      <td>-0.013722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>AY.4.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP10:M137K NSP3:R1868P NSP4:S358Y NSP10:L138R...</td>\n",
       "      <td>0.050494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>B.1.279</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP2:R370C NSP2:E575K ORF3a:Y109C</td>\n",
       "      <td>0.182182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AY.4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:F694Y</td>\n",
       "      <td>0.014535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Delta Omicron</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP4:T492I S:T478K</td>\n",
       "      <td>0.256211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Delta</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NC:28270</td>\n",
       "      <td>0.064697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>B.1.258.22</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP9:M101I NSP3:I1683T NSP12:V720I NSP13:A598S...</td>\n",
       "      <td>0.086190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>synonymous</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>ORF3a:D265D N:D216D</td>\n",
       "      <td>0.078878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>AY.68</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:L308L NSP3:P192L S:D253A S:D979E ORF3a:C...</td>\n",
       "      <td>0.004075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>EU1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP16:A199A NSP1:V60V NSP3:T1189T M:L93L N:A22...</td>\n",
       "      <td>0.050795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>L.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP8:N43H NSP3:P1403S S:M1237T</td>\n",
       "      <td>0.360561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>AY.4.5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP12:D481A S:D1259Y</td>\n",
       "      <td>0.017742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>B.1.438.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NSP2:M117V NSP3:V1811V NSP8:P10P ORF3a:L95F NC...</td>\n",
       "      <td>0.067477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>S:G75V NSP3:F1569V S:T76I S:P251- N:G214C S:R2...</td>\n",
       "      <td>-0.190919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>AY.98.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>AY.63</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP1:K141- NSP1:S142- NSP1:F143-</td>\n",
       "      <td>0.010882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>AY.23.1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>S:P217P S:L1234I</td>\n",
       "      <td>0.040989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Mu</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>S:R346K NSP2:A555A</td>\n",
       "      <td>0.128451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>B.1.618</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP13:Y277C NSP3:S371G NSP3:S697F NSP3:S848S S...</td>\n",
       "      <td>0.055307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>B.1.279</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP15:D132Y</td>\n",
       "      <td>0.058855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Alpha sublineage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NSP5:G215G S:A292S S:Q787K M:A68S</td>\n",
       "      <td>0.023629</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Variant  \\\n",
       "0                                             Omicron   \n",
       "1                                               Delta   \n",
       "2                                                BA.1   \n",
       "3                                                BA.1   \n",
       "4                            BA.2 BA.4 BA.5 BA.2.12.1   \n",
       "5                                             Omicron   \n",
       "6                                               Delta   \n",
       "7                            BA.2 BA.4 BA.5 BA.2.12.1   \n",
       "8                                               Alpha   \n",
       "9                                               Gamma   \n",
       "10                                          BA.2.12.1   \n",
       "11                          Alpha Eta BA.1, BA.4 BA.5   \n",
       "12  Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta ...   \n",
       "13  Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5...   \n",
       "14                                                 Mu   \n",
       "15                                      Alpha Omicron   \n",
       "16                                                B.1   \n",
       "17                                          B.1.1.318   \n",
       "18                                               BA.2   \n",
       "19                                              B.1.2   \n",
       "20                                            unknown   \n",
       "21                                               Beta   \n",
       "22                                              AY.27   \n",
       "23                                               AY.9   \n",
       "24                                             AY.4.5   \n",
       "25                                            B.1.279   \n",
       "26                                               AY.4   \n",
       "27                                      Delta Omicron   \n",
       "28                                              Delta   \n",
       "29                                         B.1.258.22   \n",
       "30                                         synonymous   \n",
       "31                                              AY.68   \n",
       "32                                                EU1   \n",
       "33                                                L.1   \n",
       "34                                             AY.4.5   \n",
       "35                                          B.1.438.1   \n",
       "36                                             Lambda   \n",
       "37                                            AY.98.1   \n",
       "38                                              AY.63   \n",
       "39                                            AY.23.1   \n",
       "40                                                 Mu   \n",
       "41                                            B.1.618   \n",
       "42                                            B.1.279   \n",
       "43                                   Alpha sublineage   \n",
       "\n",
       "    Regions Detected, 5% threshold  Regions Detected, 10% threshold  \\\n",
       "0                              100                              100   \n",
       "1                               97                               92   \n",
       "2                               91                               91   \n",
       "3                               89                               88   \n",
       "4                               83                               83   \n",
       "5                               85                               83   \n",
       "6                               83                               81   \n",
       "7                               64                               57   \n",
       "8                               57                               44   \n",
       "9                               42                               36   \n",
       "10                              33                               30   \n",
       "11                              32                               24   \n",
       "12                              28                                9   \n",
       "13                              23                                8   \n",
       "14                              11                                7   \n",
       "15                              18                                6   \n",
       "16                               5                                4   \n",
       "17                               2                                2   \n",
       "18                               5                                2   \n",
       "19                               2                                2   \n",
       "20                               2                                2   \n",
       "21                               3                                2   \n",
       "22                               3                                2   \n",
       "23                               4                                2   \n",
       "24                               1                                1   \n",
       "25                               1                                1   \n",
       "26                               1                                1   \n",
       "27                               2                                1   \n",
       "28                               1                                1   \n",
       "29                               1                                1   \n",
       "30                               1                                1   \n",
       "31                               1                                1   \n",
       "32                               1                                1   \n",
       "33                               1                                1   \n",
       "34                               1                                1   \n",
       "35                               1                                1   \n",
       "36                               1                                1   \n",
       "37                               0                                0   \n",
       "38                               1                                0   \n",
       "39                               1                                0   \n",
       "40                               2                                0   \n",
       "41                               1                                0   \n",
       "42                               1                                0   \n",
       "43                               1                                0   \n",
       "\n",
       "    Regions Detected, 12% threshold  \\\n",
       "0                               100   \n",
       "1                                90   \n",
       "2                                90   \n",
       "3                                88   \n",
       "4                                81   \n",
       "5                                81   \n",
       "6                                79   \n",
       "7                                55   \n",
       "8                                39   \n",
       "9                                34   \n",
       "10                               29   \n",
       "11                               21   \n",
       "12                                6   \n",
       "13                                7   \n",
       "14                                5   \n",
       "15                                5   \n",
       "16                                4   \n",
       "17                                2   \n",
       "18                                2   \n",
       "19                                1   \n",
       "20                                1   \n",
       "21                                2   \n",
       "22                                2   \n",
       "23                                0   \n",
       "24                                1   \n",
       "25                                1   \n",
       "26                                1   \n",
       "27                                0   \n",
       "28                                0   \n",
       "29                                1   \n",
       "30                                1   \n",
       "31                                1   \n",
       "32                                1   \n",
       "33                                1   \n",
       "34                                1   \n",
       "35                                1   \n",
       "36                                1   \n",
       "37                                0   \n",
       "38                                0   \n",
       "39                                0   \n",
       "40                                0   \n",
       "41                                0   \n",
       "42                                0   \n",
       "43                                0   \n",
       "\n",
       "                                            Mutations  \\\n",
       "0   NSP5:P132H NSP14:I42V S:G339D S:S371F S:S373P ...   \n",
       "1   NSP6:T77A NC:209 NSP3:A488S NSP3:P1228L NSP3:P...   \n",
       "2   NSP6:I189V NSP3:K38R NSP3:A889A NSP3:A1892T NS...   \n",
       "3   NSP6:S104- NSP3:S1265- NSP3:L1266- NSP6:L105- ...   \n",
       "4   NSP5:D48D NSP1:S135R NSP3:T24I NSP3:G489S NSP3...   \n",
       "5                         N:G30- N:E31- N:R32- N:S33-   \n",
       "6       S:E156- S:F157- S:R158G ORF8:D119- ORF8:F120-   \n",
       "7                         S:L24- S:P25- S:P26- S:A27S   \n",
       "8   NSP12:P412P NSP2:S36S NSP3:T183I NSP3:A890D NS...   \n",
       "9   NSP9:Y31Y NSP1:D156D NSP3:D10D NSP3:S370L NSP3...   \n",
       "10      NSP6:Y234Y NSP12:D523D S:D53D S:L452Q S:S704L   \n",
       "11               S:I68- S:H69- S:V70- S:V143- S:Y144-   \n",
       "12                              NSP6:G107- NSP6:F108-   \n",
       "13                              NSP6:S106- NSP6:G107-   \n",
       "14  NSP10:S11S NSP3:T237A NSP3:T720I NSP3:S1106S N...   \n",
       "15                    S:N501Y S:P681H N:R203K N:G204R   \n",
       "16              NSP12:P323L NC:240 NSP3:F106F S:D614G   \n",
       "17  NSP15:V320M NSP3:I414I NSP4:T173I S:A575A S:T1...   \n",
       "18                                            S:G142D   \n",
       "19  NSP5:L89F NSP14:N129D NSP16:R216C ORF3a:G172V ...   \n",
       "20                                            S:C166S   \n",
       "21  S:D80A NSP3:K837N S:D215G S:T240- S:L241- S:L2...   \n",
       "22  NSP7:C32C NSP2:H532Q NSP12:G345G NSP13:V34V M:...   \n",
       "23  NSP6:V149A NSP3:P822L NSP4:A446V NSP6:T181I NS...   \n",
       "24  NSP10:M137K NSP3:R1868P NSP4:S358Y NSP10:L138R...   \n",
       "25                  NSP2:R370C NSP2:E575K ORF3a:Y109C   \n",
       "26                                        NSP12:F694Y   \n",
       "27                                 NSP4:T492I S:T478K   \n",
       "28                                           NC:28270   \n",
       "29  NSP9:M101I NSP3:I1683T NSP12:V720I NSP13:A598S...   \n",
       "30                                ORF3a:D265D N:D216D   \n",
       "31  NSP12:L308L NSP3:P192L S:D253A S:D979E ORF3a:C...   \n",
       "32  NSP16:A199A NSP1:V60V NSP3:T1189T M:L93L N:A22...   \n",
       "33                     NSP8:N43H NSP3:P1403S S:M1237T   \n",
       "34                               NSP12:D481A S:D1259Y   \n",
       "35  NSP2:M117V NSP3:V1811V NSP8:P10P ORF3a:L95F NC...   \n",
       "36  S:G75V NSP3:F1569V S:T76I S:P251- N:G214C S:R2...   \n",
       "37                                                      \n",
       "38                   NSP1:K141- NSP1:S142- NSP1:F143-   \n",
       "39                                   S:P217P S:L1234I   \n",
       "40                                 S:R346K NSP2:A555A   \n",
       "41  NSP13:Y277C NSP3:S371G NSP3:S697F NSP3:S848S S...   \n",
       "42                                        NSP15:D132Y   \n",
       "43                  NSP5:G215G S:A292S S:Q787K M:A68S   \n",
       "\n",
       "    Final inferred coefficient (%)  \n",
       "0                         1.365548  \n",
       "1                         0.455085  \n",
       "2                        -0.263008  \n",
       "3                         0.085201  \n",
       "4                         0.311589  \n",
       "5                         0.163238  \n",
       "6                         0.093914  \n",
       "7                         0.019667  \n",
       "8                         0.006829  \n",
       "9                         0.031859  \n",
       "10                        0.276532  \n",
       "11                       -0.040236  \n",
       "12                       -0.036588  \n",
       "13                        0.226321  \n",
       "14                       -0.135542  \n",
       "15                        0.177787  \n",
       "16                        0.076271  \n",
       "17                       -0.070192  \n",
       "18                        0.063977  \n",
       "19                        0.026101  \n",
       "20                       -0.015890  \n",
       "21                       -0.119233  \n",
       "22                        0.048689  \n",
       "23                       -0.013722  \n",
       "24                        0.050494  \n",
       "25                        0.182182  \n",
       "26                        0.014535  \n",
       "27                        0.256211  \n",
       "28                        0.064697  \n",
       "29                        0.086190  \n",
       "30                        0.078878  \n",
       "31                        0.004075  \n",
       "32                        0.050795  \n",
       "33                        0.360561  \n",
       "34                        0.017742  \n",
       "35                        0.067477  \n",
       "36                       -0.190919  \n",
       "37                        0.000000  \n",
       "38                        0.010882  \n",
       "39                        0.040989  \n",
       "40                        0.128451  \n",
       "41                        0.055307  \n",
       "42                        0.058855  \n",
       "43                        0.023629  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# with various different cutoffs for the final inferred coefficient\n",
    "\n",
    "start_date = (dt.date(2021, 2, 1) - dt.date(2020, 1, 1)).days\n",
    "inf_dir    = os.path.join(SARS_DIR, DATA_DATE, f'infer-tv2')\n",
    "link_file  = os.path.join(SARS_DIR, DATA_DATE, 'linked-sites.npy')\n",
    "cutoff     = np.amax(np.genfromtxt(os.path.join(SARS_DIR, 'null-distribution-old-5pct.csv')))\n",
    "locs, sites, s = dp.calculate_selection2(\n",
    "    inf_dir,\n",
    "    link_file,\n",
    "    cutoff=cutoff,\n",
    "    start_date=start_date\n",
    ")\n",
    "\n",
    "counts = {\n",
    "    'NSP6-106-0--'  : ['Alpha Beta Gamma BA.1 BA.2 BA.2.12.1 BA.4 BA.5 Eta Iota Lambda', 0, 0, 0],\n",
    "    'NSP6-107-1--'  : ['Alpha Beta Gamma BA.2 BA.2.12.1 BA.4 BA.5 Eta Iota Lambda', 0, 0, 0],\n",
    "    'S-478-1-A'     : ['Delta Omicron',             0, 0, 0],    # this is in Delta and Omicron\n",
    "    'S-681-1-A'     : ['Alpha Omicron',             0, 0, 0],\n",
    "    'S-704-1-T'     : ['BA.2.12.1',                 0, 0, 0],\n",
    "    'S-614-1-G'     : ['B.1',                       0, 0, 0],\n",
    "    'S-681-1-G'     : ['Delta',                     0, 0, 0],\n",
    "    'S-796-0-T'     : ['Omicron',                   0, 0, 0],\n",
    "    'S-211-1--'     : ['BA.1',                      0, 0, 0],\n",
    "    'S-496-0-A'     : ['BA.1',                      0, 0, 0],\n",
    "    'S-213-1-G'     : ['BA.2 BA.4 BA.5 BA.2.12.1',  0, 0, 0],\n",
    "    'S-26-1--'      : ['BA.2 BA.4 BA.5 BA.2.12.1',  0, 0, 0],\n",
    "    'N-31-1--'      : ['Omicron',                   0, 0, 0],\n",
    "    'S-138-0-T'     : ['Gamma',                     0, 0, 0],\n",
    "    'S-76-1-T'      : ['Lambda',                    0, 0, 0],\n",
    "    'S-570-1-A'     : ['Alpha',                     0, 0, 0],\n",
    "    'S-69-2--'      : ['Alpha Eta BA.1, BA.4 BA.5', 0, 0, 0],\n",
    "    'S-157-2--'     : ['Delta',                     0, 0, 0],\n",
    "    'S-144-1-C'     : ['Mu',                        0, 0, 0],\n",
    "    'S-346-1-A'     : ['Mu',                        0, 0, 0],\n",
    "    'S-80-1-C'      : ['Beta',                      0, 0, 0],\n",
    "    'S-575-2-C'     : ['B.1.1.318',                 0, 0, 0],\n",
    "    'N-220-1-T'     : ['EU1',                       0, 0, 0],\n",
    "    'N-67-0-T'      : ['B.1.2',                     0, 0, 0],\n",
    "    'S-142-1-A'     : ['BA.2',                      0, 0, 0],\n",
    "    'S-217-2-C'     : ['AY.23.1',                   0, 0, 0],\n",
    "    'NSP3-822-1-T'  : ['AY.9',                      0, 0, 0],\n",
    "    'N-401-0-T'     : ['B.1.438.1',                 0, 0, 0],\n",
    "    'S-1259-0-T'    : ['AY.4.5',                    0, 0, 0],\n",
    "    'M-81-0-T'      : ['AY.27',                     0, 0, 0],\n",
    "    'NSP1-141-0--'  : ['AY.63',                     0, 0, 0],    # this is also BA.4\n",
    "    'S-1237-1-C'    : ['L.1',                       0, 0, 0],\n",
    "    'S-251-1-T'     : ['AY.98.1',                   0, 0, 0],\n",
    "    'S-1116-2-C'    : ['B.1.258.22',                0, 0, 0],\n",
    "    'S-253-1-C'     : ['AY.68',                     0, 0, 0],\n",
    "    'ORF3a-109-1-G' : ['B.1.279',                   0, 0, 0],\n",
    "    'NSP12-694-1-A' : ['AY.4',                      0, 0, 0],\n",
    "    'NSP4-358-1-A'  : ['AY.4.5',                    0, 0, 0],    # sublineage of AY.4.5\n",
    "    'NC-28270--'    : ['Delta',                     0, 0, 0],    # NC:A28270T is an Omicron mutation\n",
    "    'S-166-1-C'     : ['unknown',                   0, 0, 0],\n",
    "    'N-216-2-C'     : ['synonymous',                0, 0, 0],\n",
    "    'N-18-0-A'      : ['B.1.618',                   0, 0, 0],\n",
    "    'NSP15-132-0-T' : ['B.1.279',                   0, 0, 0],\n",
    "    'S-292-0-T'     : ['Alpha sublineage',          0, 0, 0]\n",
    "}\n",
    "\n",
    "keys = [i for i in counts]\n",
    "mut_groups = [[] for i in range(len(keys))]\n",
    "\n",
    "for group in sites:\n",
    "    if isinstance(group, list):\n",
    "        test_site = group[0]\n",
    "    else:\n",
    "        test_site = group\n",
    "    test_split = test_site.split('-')\n",
    "    if test_split[0]=='NC':\n",
    "        if int(test_split[1])<150 or int(test_split[1])>29690:\n",
    "            continue\n",
    "    listed = False\n",
    "    for key in keys:\n",
    "        if key in group:\n",
    "            counts[key][1] += 1\n",
    "            mut_groups[keys.index(key)] = group\n",
    "            listed = True\n",
    "            break\n",
    "    #if not listed:\n",
    "    #    print(group)\n",
    "    #    print(locs[list(sites).index(group)])\n",
    "\n",
    "cutoff = np.amax(np.genfromtxt(os.path.join(SARS_DIR, 'null-distribution-old-10pct.csv')))\n",
    "locs, sites, s = dp.calculate_selection2(\n",
    "    inf_dir,\n",
    "    link_file,\n",
    "    cutoff=cutoff,\n",
    "    start_date=start_date\n",
    ")\n",
    "\n",
    "for group in sites:\n",
    "    if isinstance(group, list):\n",
    "        test_site = group[0]\n",
    "    else:\n",
    "        test_site = group\n",
    "    test_split = test_site.split('-')\n",
    "    if test_split[0]=='NC':\n",
    "        if int(test_split[1])<150 or int(test_split[1])>29690:\n",
    "            continue\n",
    "    listed = False\n",
    "    for key in keys:\n",
    "        if key in group:\n",
    "            counts[key][2] += 1\n",
    "            mut_groups[keys.index(key)] = group\n",
    "            listed = True\n",
    "            break\n",
    "    #if not listed:\n",
    "    #    print(group)\n",
    "    #    print(locs[list(sites).index(group)])\n",
    "        \n",
    "cutoff = np.amax(np.genfromtxt(os.path.join(SARS_DIR, 'null-distribution-old-12pct.csv')))\n",
    "locs, sites, s = dp.calculate_selection2(\n",
    "    inf_dir,\n",
    "    link_file,\n",
    "    cutoff=cutoff,\n",
    "    start_date=start_date\n",
    ")\n",
    "\n",
    "for group in sites:\n",
    "    if isinstance(group, list):\n",
    "        test_site = group[0]\n",
    "    else:\n",
    "        test_site = group\n",
    "    test_split = test_site.split('-')\n",
    "    if test_split[0]=='NC':\n",
    "        if int(test_split[1])<150 or int(test_split[1])>29690:\n",
    "            continue\n",
    "    listed = False\n",
    "    for key in keys:\n",
    "        if key in group:\n",
    "            counts[key][3] += 1\n",
    "            mut_groups[keys.index(key)] = group\n",
    "            listed = True\n",
    "            break\n",
    "    #if not listed:\n",
    "    #    print(group)\n",
    "    #    print(locs[list(sites).index(group)])\n",
    "        \n",
    "values     = [counts[i] for i in keys]\n",
    "labels     = [i[0] for i in values]\n",
    "detections5  = [i[1] for i in values]\n",
    "detections10 = [i[2] for i in values]\n",
    "detections12 = [i[3] for i in values]\n",
    "mutations    = []\n",
    "\n",
    "syn_file = np.load(os.path.join(DATA_DIR, 'synonymous-prot.npz'), allow_pickle=True)\n",
    "syn_locs = syn_file['locations']\n",
    "aas      = syn_file['aa_changes']\n",
    "initial_aa = [i[0] for i in aas]\n",
    "final_aa   = [i[-1] for i in aas]\n",
    "new_muts   = []\n",
    "for group in mut_groups:\n",
    "    if isinstance(group, list):\n",
    "        new_group = []\n",
    "        for mut in group:\n",
    "            if mut not in syn_locs:\n",
    "                continue\n",
    "            idx = list(syn_locs).index(mut)\n",
    "            new_label = mut.split('-')\n",
    "            if new_label[0] == 'NC':\n",
    "                new_label = f'NC:{new_label[1]}'\n",
    "            else:\n",
    "                new_label = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "            if new_label not in new_group:\n",
    "                new_group.append(new_label)\n",
    "    else:\n",
    "        idx = list(syn_locs).index(group)\n",
    "        new_label = group.split('-')\n",
    "        if new_label[0] == 'NC':\n",
    "            new_group = f'NC:{new_label[1]}'\n",
    "        else:\n",
    "            new_group = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "        #new_group = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "    new_muts.append(new_group)\n",
    "\n",
    "for group in new_muts:\n",
    "    if isinstance(group, list):\n",
    "        mutations.append(' '.join(group))\n",
    "    else:\n",
    "        mutations.append(group)\n",
    "        \n",
    "# Find the ultimate selection coefficient inferred for each group\n",
    "inf_data  = np.load(os.path.join(DATA_DIR, f'infer-{DATA_DATE}', f'infer-g-10.npz'), allow_pickle=True)\n",
    "alleles   = [dp.get_label_new(i) for i in inf_data['allele_number']]\n",
    "selection = inf_data['selection']\n",
    "\n",
    "s_groups = np.zeros(len(mut_groups))\n",
    "for i in range(len(mut_groups)):\n",
    "    if not isinstance(mut_groups[i], list):\n",
    "        s_groups[i] = selection[alleles.index(mut_groups[i])]\n",
    "        continue\n",
    "    for j in range(len(mut_groups[i])):\n",
    "        if mut_groups[i][j] not in alleles:\n",
    "            #print(mut_groups[i][j])\n",
    "            continue\n",
    "        s_groups[i] += selection[alleles.index(mut_groups[i][j])]\n",
    "\n",
    "# Make dataframe\n",
    "detect_dic = {'Variant' : labels, \n",
    "              'Regions Detected, 5% threshold' : detections5, \n",
    "              'Regions Detected, 10% threshold' : detections10, \n",
    "              'Regions Detected, 12% threshold' : detections12, \n",
    "              'Mutations' : mutations}\n",
    "df         = pd.DataFrame.from_dict(detect_dic)\n",
    "df_temp    = df.sort_values(by=['Regions Detected, 10% threshold'], ascending=False, ignore_index=True)\n",
    "df_temp.to_csv(os.path.join(DATA_DIR, 'early-detection-full.csv'), index=False)\n",
    "df['Final inferred coefficient (%)'] = s_groups\n",
    "df = df.sort_values(by=['Regions Detected, 10% threshold'], ascending=False, ignore_index=True)\n",
    "df.to_csv(os.path.join(DATA_DIR, 'early-detection-full-sel.csv'), index=False)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=variant_detection_table></a>\n",
    "## Variant early detection table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff     = np.amax(np.genfromtxt(os.path.join(SARS_DIR, 'null-distribution-old-10pct.csv')))\n",
    "link_file  = os.path.join(SARS_DIR, 'variants-exclusive.npy')\n",
    "inf_dir    = os.path.join(SARS_DIR, DATA_DATE, f'infer-tv')\n",
    "start_date = (dt.date(2021, 2, 1) - dt.date(2020, 1, 1)).days\n",
    "locs, sites, s = dp.calculate_selection(inf_dir, link_file, start_date=start_date, cutoff=cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Number of detections</th>\n",
       "      <th>Mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BA.1</td>\n",
       "      <td>105</td>\n",
       "      <td>E:T9I M:Q19E M:D3G M:A63T N:P13L N:G30- N:E31-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alpha</td>\n",
       "      <td>93</td>\n",
       "      <td>N:R203K N:G204R N:S235F N:D3L NSP12:P412P NSP1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BA.2</td>\n",
       "      <td>89</td>\n",
       "      <td>M:F112F N:S413R NSP1:S135R NSP12:L758L NSP13:R...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Delta</td>\n",
       "      <td>87</td>\n",
       "      <td>N:D63G NSP12:G671S NSP13:P77L ORF7a:T120I ORF8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Beta</td>\n",
       "      <td>83</td>\n",
       "      <td>E:P71L N:T205I NC:173 NSP2:T85I NSP3:K837N NSP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Kappa</td>\n",
       "      <td>72</td>\n",
       "      <td>M:F53F M:I82S N:R203M N:D377Y NC:209 NC:28270 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gamma</td>\n",
       "      <td>43</td>\n",
       "      <td>N:S202S N:P80R NSP1:D156D NSP12:D140D NSP13:E3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Iota</td>\n",
       "      <td>24</td>\n",
       "      <td>N:P199L N:S232R N:M234I NC:28270 NSP13:Q88H NS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Mu</td>\n",
       "      <td>19</td>\n",
       "      <td>NC:26491 NSP10:S11S NSP13:P419S NSP14:L280L NS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Epsilon</td>\n",
       "      <td>5</td>\n",
       "      <td>N:T205I NC:28271 NSP13:D260Y NSP9:I65V ORF3a:Q...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>B.1</td>\n",
       "      <td>4</td>\n",
       "      <td>NC:240 NSP3:F106F NSP12:P323L S:D614G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Lambda</td>\n",
       "      <td>3</td>\n",
       "      <td>N:G214C NSP3:P1469S NSP3:F1569V NSP3:T428I NSP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BA.4</td>\n",
       "      <td>3</td>\n",
       "      <td>N:P151S NSP1:K141- NSP1:S142- NSP1:F143- NSP8:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Eta</td>\n",
       "      <td>1</td>\n",
       "      <td>E:L21F M:I82T N:A12G N:P142P N:S2- N:D3Y NC:29...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BA.2.12.1</td>\n",
       "      <td>0</td>\n",
       "      <td>S:S704L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BA.5</td>\n",
       "      <td>0</td>\n",
       "      <td>M:D3N NSP8:E23E S:F486V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Variant  Number of detections  \\\n",
       "0        BA.1                   105   \n",
       "1       Alpha                    93   \n",
       "2        BA.2                    89   \n",
       "3       Delta                    87   \n",
       "4        Beta                    83   \n",
       "5       Kappa                    72   \n",
       "6       Gamma                    43   \n",
       "7        Iota                    24   \n",
       "8          Mu                    19   \n",
       "9     Epsilon                     5   \n",
       "10        B.1                     4   \n",
       "11     Lambda                     3   \n",
       "12       BA.4                     3   \n",
       "13        Eta                     1   \n",
       "14  BA.2.12.1                     0   \n",
       "15       BA.5                     0   \n",
       "\n",
       "                                            Mutations  \n",
       "0   E:T9I M:Q19E M:D3G M:A63T N:P13L N:G30- N:E31-...  \n",
       "1   N:R203K N:G204R N:S235F N:D3L NSP12:P412P NSP1...  \n",
       "2   M:F112F N:S413R NSP1:S135R NSP12:L758L NSP13:R...  \n",
       "3   N:D63G NSP12:G671S NSP13:P77L ORF7a:T120I ORF8...  \n",
       "4   E:P71L N:T205I NC:173 NSP2:T85I NSP3:K837N NSP...  \n",
       "5   M:F53F M:I82S N:R203M N:D377Y NC:209 NC:28270 ...  \n",
       "6   N:S202S N:P80R NSP1:D156D NSP12:D140D NSP13:E3...  \n",
       "7   N:P199L N:S232R N:M234I NC:28270 NSP13:Q88H NS...  \n",
       "8   NC:26491 NSP10:S11S NSP13:P419S NSP14:L280L NS...  \n",
       "9   N:T205I NC:28271 NSP13:D260Y NSP9:I65V ORF3a:Q...  \n",
       "10              NC:240 NSP3:F106F NSP12:P323L S:D614G  \n",
       "11  N:G214C NSP3:P1469S NSP3:F1569V NSP3:T428I NSP...  \n",
       "12  N:P151S NSP1:K141- NSP1:S142- NSP1:F143- NSP8:...  \n",
       "13  E:L21F M:I82T N:A12G N:P142P N:S2- N:D3Y NC:29...  \n",
       "14                                            S:S704L  \n",
       "15                            M:D3N NSP8:E23E S:F486V  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts = {\n",
    "    'S-614-1-G'     : ['B.1',             0],\n",
    "    'S-716-1-T'     : ['Alpha',           0],\n",
    "    'S-80-1-C'      : ['Beta',            0],\n",
    "    'S-13-1-T'      : ['Epsilon',         0],\n",
    "    'S-20-1-A'      : ['Gamma',           0],\n",
    "    'S-75-1-T'      : ['Lambda',          0],\n",
    "    'S-52-1-G'      : ['Eta',             0],\n",
    "    'S-5-0-T'       : ['Iota',            0],\n",
    "    'S-154-0-A'     : ['Kappa',           0],\n",
    "    'S-346-2-C'     : ['Mu',              0],\n",
    "    'S-158-0--'     : ['Delta',           0],\n",
    "    'S-981-0-T'     : ['BA.1',            0],\n",
    "    'S-405-0-A'     : ['BA.2',            0],\n",
    "    'N-151-0-T'     : ['BA.4',            0],\n",
    "    'M-3-0-A'       : ['BA.5',            0],\n",
    "    'S-704-1-T'     : ['BA.2.12.1',       0],\n",
    "}\n",
    "# NOTE: VERY IMPORTANT THAT VARIANTS BELOW ARE LISTED IN THE ORDER IN WHICH THEY APPEARED IN TIME\n",
    "\n",
    "### USE THE VARIANT NAMES IN THE ABOVE CELL AND THE ACTUAL GROUPS OF MUTATIONS TO IDENTIFY DETECTIONS\n",
    "\n",
    "keys = ['S-614-1-G', 'S-716-1-T', 'S-80-1-C', 'S-13-1-T', 'S-20-1-A', 'S-75-1-T',\n",
    "        'S-52-1-G', 'S-5-0-T', 'S-154-0-A', 'S-346-2-C', 'S-158-0--', 'S-981-0-T', \n",
    "        'S-405-0-A', 'N-151-0-T', 'M-3-0-A', 'S-704-1-T']\n",
    "\n",
    "count_dict = {name : 0 for name in group_order}\n",
    "\n",
    "for group in sites:\n",
    "    for i in range(len(new_groups)):\n",
    "        if list(group)==new_groups[i]:\n",
    "            count_dict[group_order[i]] += 1\n",
    "            break\n",
    "\n",
    "labels     = [i for i in count_dict]\n",
    "detections = [count_dict[i] for i in labels]\n",
    "mutations  = []\n",
    "mut_groups = new_groups\n",
    "\n",
    "syn_file = np.load(os.path.join(DATA_DIR, 'synonymous-prot.npz'), allow_pickle=True)\n",
    "syn_locs = syn_file['locations']\n",
    "aas      = syn_file['aa_changes']\n",
    "initial_aa = [i[0] for i in aas]\n",
    "final_aa   = [i[-1] for i in aas]\n",
    "new_muts   = []\n",
    "for group in mut_groups:\n",
    "    if isinstance(group, list):\n",
    "        new_group = []\n",
    "        for mut in group: \n",
    "            if mut not in syn_locs:\n",
    "                print(f'{mut} not in synonymous sites file')\n",
    "                continue\n",
    "            idx = list(syn_locs).index(mut)\n",
    "            new_label = mut.split('-')\n",
    "            if new_label[0] == 'NC':\n",
    "                new_label = f'NC:{new_label[1]}'\n",
    "            else:\n",
    "                new_label = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "            if new_label not in new_group:\n",
    "                new_group.append(new_label)\n",
    "    else:\n",
    "        idx = list(syn_locs).index(group)\n",
    "        new_label = group.split('-')\n",
    "        if new_label[0] == 'NC':\n",
    "            new_group = f'NC:{new_label[1]}'\n",
    "        else:\n",
    "            new_group = new_label[0] + ':' + initial_aa[idx] + new_label[1] + final_aa[idx]\n",
    "    new_muts.append(new_group)\n",
    "\n",
    "for group in new_muts:\n",
    "    if isinstance(group, list):\n",
    "        mutations.append(' '.join(group))\n",
    "    else:\n",
    "        mutations.append(group)\n",
    "        \n",
    "detect_dic = {'Variant' : labels, 'Number of detections' : detections, 'Mutations' : mutations}\n",
    "df         = pd.DataFrame.from_dict(detect_dic)\n",
    "df         = df.sort_values(by=['Number of detections'], ascending=False, ignore_index=True)\n",
    "df.to_csv(os.path.join(DATA_DIR, 'variant-early-detection.csv'), index=False)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=selection_on_date></a>\n",
    "## Variant selection before appearance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[263, 348, 328, 670]\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel-tv-allregs.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-tv-allregs.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-infer-tv-allregs.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/infer-tv-allregs /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5% in the \n",
    "inf_script    = 'epi-inf-parallel-tv-allregs.py'\n",
    "freqs_dir     = ppath.join(SSH_DATA, 'freqs-regional2')\n",
    "out_file      = ppath.join(SSH_DATA, f'infer-tv-allregs')\n",
    "dates         = ['2020-09-20', '2020-12-14', '2020-11-24', '2021-11-01']\n",
    "new_dates     = [(dt.date.fromisoformat(i) - dt.date(2020,1,1)).days for i in dates]\n",
    "print(new_dates)\n",
    "\n",
    "job_file2 = 'job-infer-tv-allregs.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=20:00:00\n",
    "#SBATCH -e ./MPL/out/inf-tv-all-error-%a\n",
    "#SBATCH -o ./MPL/out/inf-tv-all-out-%a\n",
    "#SBATCH --array=0-{len(dates)-1}\n",
    "\n",
    "dates=(263 348 328 670)\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}/inf-${{dates[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {inf_script} --data {freqs_dir} --timed 1 -o \\\"$outfile\\\" -q 5 --g1 40 --delta_t 15 --infDate ${{dates[$SLURM_ARRAY_TASK_ID]}} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print(f'scp -r {SSH_HOME}{out_file} {INF_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferred selection for Alpha before it appeared = 0.18474387457654953\n",
      "\n",
      "inferred selection for Gamma before it appeared = 0.3855689390579591\n",
      "\n",
      "inferred selection for Delta before it appeared = 0.12185897184153888\n",
      "\n",
      "inferred selection for omicron before it appeared = 0.6612678503524573\n"
     ]
    }
   ],
   "source": [
    "inf_dir = os.path.join(DATA_DIR, 'infer-tv-allregs')\n",
    "\n",
    "alpha      = np.load(os.path.join(DATA_DIR, 'alpha-mutations.npy'), allow_pickle=True)\n",
    "data_alpha = np.load(os.path.join(inf_dir, f'inf-{new_dates[0]}.npz'), allow_pickle=True)\n",
    "muts = [dp.get_label_new(i) for i in data_alpha['allele_number']]\n",
    "idxs = [muts.index(i) for i in alpha]\n",
    "s    = data_alpha['selection']\n",
    "s_alpha = np.sum(s[np.array(idxs)])\n",
    "print(f'inferred selection for Alpha before it appeared = {s_alpha}')\n",
    "print('')\n",
    "\n",
    "gamma      = np.load(os.path.join(DATA_DIR, 'gamma-mutations.npy'), allow_pickle=True)\n",
    "data_gamma = np.load(os.path.join(inf_dir, f'inf-{new_dates[1]}.npz'), allow_pickle=True)\n",
    "muts = [dp.get_label_new(i) for i in data_gamma['allele_number']]\n",
    "idxs = [muts.index(i) for i in gamma]\n",
    "s    = data_gamma['selection']\n",
    "s_gamma = np.sum(s[np.array(idxs)])\n",
    "print(f'inferred selection for Gamma before it appeared = {s_gamma}')\n",
    "print('')\n",
    "\n",
    "delta      = np.load(os.path.join(DATA_DIR, 'delta-mutations.npy'), allow_pickle=True)\n",
    "data_delta = np.load(os.path.join(inf_dir, f'inf-{new_dates[2]}.npz'), allow_pickle=True)\n",
    "muts = [dp.get_label_new(i) for i in data_delta['allele_number']]\n",
    "idxs = [muts.index(i) for i in delta]\n",
    "s    = data_delta['selection']\n",
    "s_delta = np.sum(s[np.array(idxs)])\n",
    "print(f'inferred selection for Delta before it appeared = {s_delta}')\n",
    "print('')\n",
    "\n",
    "omicron      = np.load(os.path.join(DATA_DIR, 'omicron-mutations.npy'), allow_pickle=True)\n",
    "data_omicron = np.load(os.path.join(inf_dir, f'inf-{new_dates[3]}.npz'), allow_pickle=True)\n",
    "muts = [dp.get_label_new(i) for i in data_omicron['allele_number']]\n",
    "idxs = [muts.index(i) for i in omicron]\n",
    "s    = data_omicron['selection']\n",
    "s_omicron = np.sum(s[np.array(idxs)])\n",
    "print(f'inferred selection for omicron before it appeared = {s_omicron}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=variant_detection></a>\n",
    "## Variant Detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5% in the UK\n",
    "freq_script   = 'epi-covar-parallel.py'\n",
    "move_script   = 'make-dirs-tv-inf.py'\n",
    "inf_script    = 'epi-inf-parallel-tv.py'\n",
    "temp_dir      = ppath.join(SSH_DATA, 'freqs-tv-variants')\n",
    "freqs_dir     = ppath.join(SSH_DATA, 'freqs-regional-variants')\n",
    "out_file      = ppath.join(SSH_DATA, f'infer-tv-variants')\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-filtered-variants')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print(f'chdir Archive-tv-int')\n",
    "print('g++ src/main-test.cpp src/inf-test.cpp src/io-test.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11')\n",
    "print('')\n",
    "\n",
    "num_files     = 3    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-freqs-tv-uk.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=150G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-tv-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-tv-out-%a\n",
    "#SBATCH -p highmem\n",
    "#SBATCH --array=0-{num_files - 1}\n",
    "\n",
    "files=({input_dir}/*)\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-tv-int\n",
    "g++ src/main-test.cpp src/inf-test.cpp src/io-test.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python ../{freq_script} --data \\\"$file\\\" -o {temp_dir} -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --tv_inference --delta_t 15\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file3 = 'job-move-data-uk.sh'\n",
    "job_str3  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH -e ./MPL/out/move-uk-error\n",
    "#SBATCH -o ./MPL/out/move-uk-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python ./{move_script} --input {temp_dir} -o {freqs_dir} \n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file3), 'w')\n",
    "f.write('%s\\n' % job_str3)\n",
    "f.close()\n",
    "\n",
    "\n",
    "job_file2 = 'job-infer-tv-uk.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-tv-uk-error\n",
    "#SBATCH -o ./MPL/out/inf-tv-uk-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "dirs=({freqs_dir}/*)\n",
    "dir=${{dirs[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "tempout1=$(basename \"$dir\")\n",
    "tempout2={out_file}/\n",
    "outfile=\"$tempout2$tempout1\"\n",
    "\n",
    "python {inf_script} --data \\\"$dir\\\" --timed 1 -o \\\"$outfile\\\" -q 5 --g1 10 --delta_t 5 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "job_files = [job_file1, job_file2, job_file3]\n",
    "for file in job_files:\n",
    "    job_path = os.path.join(SCRIPT_DIR, file)\n",
    "    print(f'scp {job_path} {SSH_HOME} &&')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "for file in job_files:\n",
    "    print(f'sbatch {file}')\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s' % (SSH_HOME, out_file + '.npz', INF_DIR))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Great Britain (Delta)__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# trajectories\n",
    "input_dir    = ppath.join(SSH_DATA, 'genome-filtered-uk')\n",
    "traj_script  = 'epi-traj-parallel.py'\n",
    "temp_dir     = ppath.join(SSH_DATA, 'trajectories-uk')\n",
    "\n",
    "# run on cluster to find number of files\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 518   # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1     = 'job-find-traj-uk.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/traj-uk-error-%a\n",
    "#SBATCH -o ./MPL/out/traj-uk-out-%a\n",
    "#SBATCH --array=1-{num_files}\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {traj_script} --data \\\"$file\\\" -o {temp_dir} -q 5 --timed 1 --window 5\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file1)\n",
    "print('')\n",
    "\n",
    "print(f'scp -r {SSH_HOME}{temp_dir} {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "twentyE_EU1 = ['NSP16-199-2-C', 'NSP1-60-2-C', 'NSP3-1189-2-T', 'M-93-2-G', 'N-220-1-T', 'ORF10-30-0-T', 'S-222-1-T']    # Done\n",
    "\n",
    "delta       = ['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', \n",
    "               'S-158-0--', 'S-478-1-A', 'S-950-0-A', 'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', \n",
    "               'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'N-63-1-G', \n",
    "               'N-377-0-T', 'S-452-1-G', 'S-681-1-G']  \n",
    "\n",
    "site     = 'S-950-0-A'\n",
    "traj     = []\n",
    "times    = []\n",
    "traj_dir = os.path.join(DATA_DIR, 'trajectories-uk')\n",
    "for file in os.listdir(traj_dir):\n",
    "    if 'england' in file:\n",
    "        df     = pd.read_csv(os.path.join(traj_dir, file))\n",
    "        muts   = list(df.columns)[1:]\n",
    "        sites  = [dp.get_label_new(str(i)) for i in muts]\n",
    "        t_temp = list(df['time'])\n",
    "        if site in sites:\n",
    "            L = len(df)\n",
    "            traj_temp = list(df[muts[sites.index(site)]])\n",
    "            for i in range(len(t_temp)):\n",
    "                times.append(t_temp[i])\n",
    "                traj.append(traj_temp[i])\n",
    "        else:\n",
    "            for i in range(len(t_temp)):\n",
    "                times.append(t_temp[i])\n",
    "                traj.append(0)\n",
    "sorter = np.argsort(times)\n",
    "times  = np.array(times)[sorter]\n",
    "traj   = np.array(traj)[sorter]\n",
    "\n",
    "inf_data  = np.load(os.path.join(DATA_DIR, f'infer-tv-variants', 'europe-united kingdom-england_wales_scotland.npz'), allow_pickle=True)\n",
    "selection = inf_data['selection']\n",
    "alleles   = [dp.get_label_new(i) for i in inf_data['allele_number']]\n",
    "t_infer   = inf_data['time_infer']\n",
    "idxs      = np.array([alleles.index(i) for i in delta])\n",
    "s_delta = np.sum(selection[:, idxs], axis=1)\n",
    "t_common  = [i for i in t_infer if i in times]\n",
    "s_delta = np.array(s_delta)[np.isin(t_infer, t_common)]\n",
    "traj      = traj[np.isin(times, t_common)]\n",
    "\n",
    "sel_traj_file = os.path.join(DATA_DIR, 'delta-tv-s.csv')\n",
    "f = open(sel_traj_file, 'w')\n",
    "f.write('time,frequency,selection\\n')\n",
    "for i in range(len(t_common)):\n",
    "    f.write(f'{t_common[i]},{traj[i]},{s_delta[i]}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Early Detection in London (Alpha)__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-find-traj-lond.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "\n",
      "sbatch job-find-traj-lond.sh\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu://rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09-lond/trajectories-london /Users/brianlee/SARS-CoV-2-Data/2022-06-09\n"
     ]
    }
   ],
   "source": [
    "# trajectories\n",
    "input_dir    = ppath.join(SSH_DATA, 'genome-filtered-london')\n",
    "traj_script  = 'epi-traj-parallel.py'\n",
    "temp_dir     = ppath.join(SSH_DATA, 'trajectories-london')\n",
    "\n",
    "job_file1     = 'job-find-traj-lond.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/traj-lond-error-%a\n",
    "#SBATCH -o ./MPL/out/traj-lond-out-%a\n",
    "#SBATCH --array=1\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "mkdir {temp_dir}\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {traj_script} --data \\\"$file\\\" -o {temp_dir} -q 5 --timed 1 --window 15\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file1)\n",
    "print('')\n",
    "\n",
    "# transfer to local \n",
    "print(f'scp -r {SSH_HOME}/{temp_dir} {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha    = ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "            'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "            'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "            'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "            'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T']\n",
    "\n",
    "site     = 'S-570-1-A'\n",
    "traj     = []\n",
    "times    = []\n",
    "traj_dir = os.path.join(DATA_DIR, 'trajectories-london')\n",
    "for file in os.listdir(traj_dir):\n",
    "    df     = pd.read_csv(os.path.join(traj_dir, file))\n",
    "    muts   = list(df.columns)[1:]\n",
    "    sites  = [dp.get_label_new(str(i)) for i in muts]\n",
    "    t_temp = list(df['time'])\n",
    "    if site in sites:\n",
    "        L = len(df)\n",
    "        traj_temp = list(df[muts[sites.index(site)]])\n",
    "        for i in range(len(t_temp)):\n",
    "            times.append(t_temp[i])\n",
    "            traj.append(traj_temp[i])\n",
    "    else:\n",
    "        for i in range(len(t_temp)):\n",
    "            times.append(t_temp[i])\n",
    "            traj.append(0)\n",
    "sorter = np.argsort(times)\n",
    "times  = np.array(times)[sorter]\n",
    "traj   = np.array(traj)[sorter]\n",
    "\n",
    "inf_data  = np.load(os.path.join(DATA_DIR, f'infer-tv-variants', f'europe-united kingdom-england-london.npz'), allow_pickle=True)\n",
    "selection = inf_data['selection']\n",
    "alleles   = [dp.get_label_new(i) for i in inf_data['allele_number']]\n",
    "t_infer   = inf_data['time_infer']\n",
    "idxs      = np.array([alleles.index(i) for i in alpha])\n",
    "s_alpha   = np.sum(selection[:, idxs], axis=1)\n",
    "t_common  = [i for i in t_infer if i in times]\n",
    "s_alpha   = np.array(s_alpha)[np.isin(t_infer, t_common)]\n",
    "traj      = traj[np.isin(times, t_common)]\n",
    "\n",
    "sel_traj_file = os.path.join(DATA_DIR, 'alpha-tv-s-new.csv')\n",
    "f = open(sel_traj_file, 'w')\n",
    "f.write('time,frequency,selection\\n')\n",
    "for i in range(len(t_common)):\n",
    "    f.write(f'{t_common[i]},{traj[i]},{s_alpha[i]}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Early detection in South Africa (Omicron)__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# trajectories\n",
    "input_dir    = ppath.join(SSH_DATA, 'genome-filtered-south-africa')\n",
    "traj_script  = 'epi-traj-parallel.py'\n",
    "temp_dir     = ppath.join(SSH_DATA, 'trajectories-south-africa')\n",
    "\n",
    "job_file1     = 'job-find-traj-lond.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH -e ./MPL/out/traj-lond-error-%a\n",
    "#SBATCH -o ./MPL/out/traj-lond-out-%a\n",
    "#SBATCH --array=1\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "mkdir {temp_dir}\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {traj_script} --data \\\"$file\\\" -o {temp_dir} -q 5 --timed 1 --window 15\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s' % job_file1)\n",
    "print('')\n",
    "\n",
    "# transfer to local \n",
    "print(f'scp -r {SSH_HOME}/{temp_dir} {DATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# making a file containing the trajectories and the inferred selection over time\n",
    "omicron = ['NSP3-38-1-G', 'NSP3-889-2-G', 'NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--', \n",
    "           'NSP3-1892-0-A', 'NSP5-132-1-A', 'NSP6-189-0-G', 'NSP10-57-2-C', 'NSP14-42-0-G', \n",
    "           'S-211-1--', 'S-211-2--', 'S-212-0--', 'S-446-0-A', 'S-547-1-A', \n",
    "           'S-856-2-A', 'S-954-2-T', 'S-969-2-A', 'S-981-0-T', 'ORF3a-64-2-T', \n",
    "           'E-9-1-T', 'M-19-0-G', 'M-63-0-A', 'ORF6-20-0-C', 'ORF7b-18-0-T', \n",
    "           'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "           'N-32-1--', 'N-32-2--', 'N-33-0--', 'S-484-1-C', 'S-375-1-T', 'S-493-1-G', \n",
    "           'S-496-0-A', 'S-498-1-G', 'S-505-0-C', 'S-547-1-A']\n",
    "# removed: 'S-142-2--', 'S-143-0--', 'S-143-1--', 'S-144-2--', 'S-145-0--'\n",
    "#site = 'S-446-0-A'\n",
    "#site = 'S-67-1-T'\n",
    "site = 'NSP3-38-1-G'\n",
    "\n",
    "traj     = []\n",
    "times    = []\n",
    "traj_dir = os.path.join(DATA_DIR, 'trajectories-south-africa')\n",
    "for file in os.listdir(traj_dir):\n",
    "    if 'south africa' in file:\n",
    "        df     = pd.read_csv(os.path.join(traj_dir, file))\n",
    "        muts   = list(df.columns)[1:]\n",
    "        sites  = [dp.get_label_new(str(i)) for i in muts]\n",
    "        t_temp = list(df['time'])\n",
    "        if site in sites:\n",
    "            L = len(df)\n",
    "            traj_temp = list(df[muts[sites.index(site)]])\n",
    "            for i in range(len(t_temp)):\n",
    "                times.append(t_temp[i])\n",
    "                traj.append(traj_temp[i])\n",
    "        else:\n",
    "            for i in range(len(t_temp)):\n",
    "                times.append(t_temp[i])\n",
    "                traj.append(0)\n",
    "sorter = np.argsort(times)\n",
    "times  = np.array(times)[sorter]\n",
    "traj   = np.array(traj)[sorter]\n",
    "\n",
    "inf_data  = np.load(os.path.join(DATA_DIR, f'infer-tv-variants', 'africa-south-africa.npz'), allow_pickle=True)\n",
    "selection = inf_data['selection']\n",
    "alleles   = [dp.get_label_new(i) for i in inf_data['allele_number']]\n",
    "t_infer   = inf_data['time_infer']\n",
    "idxs      = np.array([alleles.index(i) for i in omicron])\n",
    "s_omicron = np.sum(selection[:, idxs], axis=1)\n",
    "t_common  = [i for i in t_infer if i in times]\n",
    "s_omicron = np.array(s_omicron)[np.isin(t_infer, t_common)]\n",
    "traj      = traj[np.isin(times, t_common)]\n",
    "\n",
    "sel_traj_file = os.path.join(DATA_DIR, 'omicron-tv-s-south-africa.csv')\n",
    "f = open(sel_traj_file, 'w')\n",
    "f.write('time,frequency,selection\\n')\n",
    "for i in range(len(t_common)):\n",
    "    f.write(f'{t_common[i]},{traj[i]},{s_omicron[i]}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Test india early detection (Omicron)__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n",
      "2\n",
      "4\n",
      "2\n",
      "4\n",
      "2\n",
      "5\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "regions = [\n",
    "    'asia-india-west bengal',\n",
    "    'asia-india-maharashtra',\n",
    "    'asia-india-kerala',\n",
    "    'asia-india-karnataka',\n",
    "    'asia-india-telangana', \n",
    "    'asia-india-maharashtra',\n",
    "    'asia-india-gujarat',\n",
    "    'asia-india-andhra pradesh',\n",
    "    'asia-india-haryana',\n",
    "]\n",
    "\n",
    "locs     = regions\n",
    "inf_dir  = os.path.join(DATA_DIR, f'infer-tv2')\n",
    "traj_dir = os.path.join(DATA_DIR, 'trajectories')\n",
    "\n",
    "if not os.path.exists(os.path.join(DATA_DIR, 'infer-tv-india')):\n",
    "    os.mkdir(os.path.join(DATA_DIR, 'infer-tv-india'))\n",
    "    \n",
    "for loc in locs:\n",
    "    inf_data      = np.load(os.path.join(inf_dir, loc + '.npz'), allow_pickle=True)\n",
    "    t_infer       = inf_data['time_infer']\n",
    "    selection     = inf_data['selection']\n",
    "    allele_number = inf_data['allele_number']\n",
    "    mutant_sites  = []\n",
    "    traj          = []\n",
    "    times         = []\n",
    "    if t_infer[-1] < 670:\n",
    "        continue\n",
    "    for file in os.listdir(traj_dir):\n",
    "        if loc in file:\n",
    "            df = pd.read_csv(os.path.join(traj_dir, file))\n",
    "            traj.append(df.to_numpy()[:, 1:])\n",
    "            mutant_sites.append(list(df.columns)[1:])\n",
    "            times.append(list(df['time']))\n",
    "    np.savez_compressed(\n",
    "        os.path.join(DATA_DIR, 'infer-tv-india', f'{loc}'),\n",
    "        times=times,\n",
    "        mutant_sites=mutant_sites,\n",
    "        allele_number=allele_number,\n",
    "        selection=selection,\n",
    "        times_inf=t_infer,\n",
    "        traj=traj\n",
    "    )\n",
    "    \n",
    "    \n",
    "locs = [\n",
    "    #'asia-india-west bengal.npz',\n",
    "    'asia-india-maharashtra',\n",
    "    'asia-india-kerala',\n",
    "    'asia-india-karnataka',\n",
    "    'asia-india-telangana', \n",
    "    'asia-india-maharashtra',\n",
    "    'asia-india-gujarat',\n",
    "    #'asia-india-andhra pradesh',\n",
    "    #'asia-india-haryana',\n",
    "]\n",
    "\n",
    "delta = ['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', \n",
    "         'S-158-0--', 'S-478-1-A', 'S-950-0-A', 'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', \n",
    "         'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'N-63-1-G', \n",
    "         'N-377-0-T', 'S-452-1-G', 'S-681-1-G'] \n",
    "for loc in locs:\n",
    "    infile  = os.path.join(DATA_DIR, 'infer-tv-india', loc + '.npz')\n",
    "    outfile = os.path.join(DATA_DIR, loc + '-traj-selection')\n",
    "    dp.save_traj_selection(infile, group=delta, traj_site='S-19-1-G', out_file=outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=submission_date></a>\n",
    "# Trimming by Submission Date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__London__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-02-20-lond/genome-filtered-alpha | wc -l\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar-sd.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel-sd.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-vector blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-freqs-sd-lond.sh blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-sd-lond.sh blee098@secure.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-freqs-sd-lond.sh &&\n",
      "sbatch job-infer-sd-lond.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-02-20-lond &&\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-02-20-lond-* /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-02-20-lond\n",
      "\n",
      "scp -r blee098@secure.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-02-20-lond /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered-london')\n",
    "freq_script    = 'epi-covar-sd.py'\n",
    "inf_script     = 'epi-inf-parallel-sd.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs-london-sd')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer-london-sd', 'inf')\n",
    "out_dir        = ppath.join(SSH_DATA, f'infer-london-sd')\n",
    "\n",
    "# run on cluster\n",
    "print(f'ls {input_dir} | wc -l')    # in order to find the number of files in the input directory\n",
    "print('')\n",
    "\n",
    "num_files = 5    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1 = 'job-freqs-sd-lond.sh'\n",
    "job_str1  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-sd-lond-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-sd-lond-out-%a\n",
    "#SBATCH --array=0-{num_files - 1}\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-vector\n",
    "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "pwd\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer-sd-lond.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-lond-error\n",
    "#SBATCH -o ./MPL/out/inf-lond-out\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}\n",
    "python {inf_script} --data {temp_dir} --timed 1 -o $outfile -q 5 --g1 10 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s' % (SSH_HOME, out_dir, INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2020-11-30\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEGCAYAAABo25JHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAo3UlEQVR4nO3deZxcdZnv8c9T1Vs6OyQkQCABDZsKiCGiDpoR0OACXrcJiNtVMjpyxzuMd4C5c3FgZq46MzADI74UwRlUFBEVI0bBBRSVJWEn0UAEAklI0iSku7N0p7vqmT/Oqe5K01XnVOqcWlLf9+uVV58651dVT1dO99O/3dwdERFpXZl6ByAiIvWlRCAi0uKUCEREWpwSgYhIi1MiEBFpcW31DqBSM2bM8Hnz5tU7jIaxZk3w9eij6xuHiDS2Bx544AV3nznetaZLBPPmzWPlypX1DqNhLFoUfL3rrnpGISKNzszWlbqmpiERkRanRCAi0uKUCEREWpwSgYhIi1MiEBFpcUoEIiItTolARKTFNd08Amk8Dz37Inf+YUtkueMOmcriV86uQUQiUgklAqnalT97gruffAGz0mXc4cCJHUoEIg1IiUCq1rt7iEVHz+S/PrqwZJnLfrSKWx5YX8OoRCQu9RFI1foHhpnc1V62TFvGGM5pNzyRRqREIFXrHxhiclf5ymVbNsNwPl+jiESkEkoEUrW+geHoRJAxhvOqEYg0IiUCqcrgcI49w3mmRDYNZXCHnJKBSMNRIpCq9A8MA8RoGgqGFKl5SKTxKBFIVWIngkyYCNRhLNJwlAikKv0DQwBM7izfNJQtJAI1DYk0HCUCqUrcGkF7NrjVhnNqGhJpNEoEUpWRGkFEZ3GhRqDOYpHGk2oiMLPFZrbGzNaa2cUlyrzfzFab2Soz+1aa8Ujy+mLXCIJEMKREINJwUltiwsyywDXAGcB6YIWZLXP31UVl5gOXAG9w9xfN7KC04pF0FJqGooaPZjPB3xw5dRaLNJw0awQLgbXu/pS77wFuAs4eU+Z84Bp3fxHA3aOXsJSG0rc7aBqaFLtGoD4CkUaTZiI4FHiu6PH68Fyxo4CjzOy3ZnavmS0e74XMbKmZrTSzlT09PSmFK/uif2CYiR3ZkT6AUtRHINK46t1Z3AbMBxYB5wBfNbNpYwu5+7XuvsDdF8ycObO2EUpZwTpD5ZuFIJhZDDCkUUMiDSfNRLABOKzo8ZzwXLH1wDJ3H3L3p4EnCBKDNIn+GOsMweiEMtUIRBpPmolgBTDfzI4wsw5gCbBsTJlbCWoDmNkMgqaip1KMSRLWPxi98iiMLjExpM5ikYaTWiJw92HgAuB24PfAze6+yswuN7OzwmK3A1vNbDVwJ/B/3H1rWjFJ8uLsRQCjTUOqEYg0nlR3KHP35cDyMecuLTp24MLwnzSh/oFhDj+gO7LcyKJz6iMQaTj17iyWJhe3s7g9q7WGRBqVEoFUpW9gmCkx+ggKE8q0DLVI41EikH1W2JSmklFDWoZapPEoEcg+G115NEZnsZqGRBqWEoHss5F1hibEqREUmoaUCEQaTaqjhqQ5Xf+bp3lw3YuR5fpibkoDxU1D6iMQaTRKBLIXd+eKO9bQ2ZbhwEmdkeWPnzOV4w6ZEllOTUMijUuJQPbSs2OQXXtyXLT4GD78+nmJve5I05A6i0UajvoIZC/Pbt0FwNwDoyeJVaJQI8hp+KhIw1EikL08M5IIJib6uoU+Aq01JNJ4lAhkL+u27iSbMQ6dNiHR123Laq0hkUalRCB7Wbd1F4dM66KjLdlbY6RGoKYhkYajRCB7Wbd1J/MSbhaCov0I1DQk0nCUCGQv67btSryjGEa3qhxS05BIw1EikBG9u4bYvmuIuQckXyMwM9oypgllIg1IiUBGrNu2E0h+6GhBNmPqLBZpQJpQ1iLuf3obX/nVH8l76V/EW3fuAZIfOlrQns1o+KhIA1IiaBG3PryBXz/Zw7EHl18O4vRjZ3HkzHQSQVvWUp1QNjCU459/uoZ7n9p7t9OOtgz/+r7jeflBk1N7b5FmpkTQIjb3DjD/oMksu+BP6hZDW8ZidRZfd/dTPLq+t+LX/8OmPp7YvINT58+gsy0LBHsm3P3kCzy4brsSgUgJSgQtYlPfALOndtU1hrZMJtbw0X//+ZNkjFiL3hXras9y3YcWcPpxs0bO9e4e4oTL7hhZKVVEXirVRGBmi4GrgCxwnbt/fsz1jwD/AmwIT33R3a9LM6ZWtblvgOPnTKtrDNmMxZpQNjic4/xTj+RvFh9T9XtO6gxu8R2Dw1W/lsj+KrVEYGZZ4BrgDGA9sMLMlrn76jFFv+PuF6QVhwS/WF/YsYfZU+pbI2jPRo8ayuWdoZyPNO1UK5sxJnZkRzbREZGXSnP46EJgrbs/5e57gJuAs1N8PylhS98gALOnVtbUkrRsxiKXoR4czgHQ2Z7crTmpq40dSgQiJaWZCA4Fnit6vD48N9Z7zOxRM7vFzA4b74XMbKmZrTSzlT09PWnEul/b3DcAwKy61wgyDEc0DQ0OBdc7E1zraHJXO/2D6iMQKaXeE8p+BMxz9+OBnwE3jFfI3a919wXuvmDmzJk1DXB/sClMBPXuLI5XIwgSQVd7Mk1DEPQTqGlIpLQ0E8EGoPgv/DmMdgoD4O5b3X0wfHgd8JoU42lZm3rDRFDnGkFbNhO5VeVI01CiNYI2dRaLlJHmqKEVwHwzO4IgASwBzi0uYGYHu/vz4cOzgN+nGE/L2tw3QGdbhqkTojeZT1NbxiKbhgZGmoaSqxFM7mrj+TAZSrSBoRw7IxKnmTG9ux0zq1FUkqbIRGBmrwPOA04FDgZ2A48DPwa+6e7jzvxx92EzuwC4nWD46NfcfZWZXQ6sdPdlwF+a2VnAMLAN+Ej135KMtalvkNlTu+r+Q9tWSWdxgjWCSZ37d2fxLQ+sZ+Uz20pe72zL8MHXzY01oe5XT/TwqRsfjFWD+qvTj+LTp8+vKFZpTGUTgZn9BNgI/BD4J2AL0AUcBfwp8EMzuzL8pf4S7r4cWD7m3KVFx5cAl1TzDUi0zb0Dde8ohmCJicJf/KWk0Ucwuaud/v10QtmW/gH+9vuP0dmeobtj/M+sd/cQN973LG86aubI3tHjyTvctWYL8w+azJKF447bGHH9b57m/me2AkoE+4OoGsEH3f2FMed2AA+G/64wsxmpRCaJ2dQ3wAmHTat3GLRlMgznc2XLjIwaSnL4aGcbO/fkyOV9ZF+EpOXzzq+f7Bk30eXd2bpjkB2DwfduBjby1ShU1Mys6Hz4ODxmr2tGxuANL5/Bd1c+x1A+zx0XvJF5M8ZfI2rrjkH+9Y4neOjZFyO/j7e8Yjafe/ermNJVvhnx0fW93LVGI/j2F2UTwThJADM7DegGfuruQ+OVkdrZOTjMcN75zZOl1+bZ3DfA7Cn1nUMAxNqPIK3OYghmF6fVT3Lrwxu48OZHUnntUiZ3tmEGpx0zq2QSgGCpjs+9+1WJvvcxsydzywPr2bpjsOKlQKTxVNRZbGZXAL1AHvgk8LY0gpJ4nurZwWMbgkFX511/X9my5X5R1EpbjJnFaXUWQ7qJ4Jv3ruPIGRP54rknveRaJgMHTOxgcmfw3o7jDg64e/gV8NLXPLg48nj77j185ruP8PiGPj5+6hGpfE/lHD076G9Ys7mf1ysRNL2oPoIrgH9w9+3hqcOB94fHj6UYl8RQGBt/+AHd3PiJ15Usl80Yxx86tVZhldSWyTAUs0bQlWDT0OSwmSOtDuPVG/t48Nnt/N3bj+W4Q8ov852U2VO7uOUTr+cPm/o5sQ7NfkfPChPBpn5e/zK1Dje7qBrB94GbzGw5wbpBXwfuJOgw/mrKsUmEwiYz3R1ZTp53QJ2jiRanRlDoLE6yRlBYeC5uh/ENv3uGJzb3x379xzf20dmW4b2vmbNP8e2rrvZsXZIAwMzJnUzvbq/oc5LGFdVH8FtgsZmdRzAM9Gp3X1SLwCTayG5jTTKUO5uxyB3KBodSGD4aNg31xxgSOTCU47PLVjGxI8uEEqNwxvOxPzmCad0d+xxjszEzjp49mT9sSj4RLHtkIy/0D5Ytk80Y7zj+YPVPJCSqaagNeCvBsNF3AX9lZh8H/p+717ZnTF6i0MpiTZIJ2jOZ+DWCBJuGphQSQYymoWe37QLgc+85nrNOOCSxGPZHR88KOox/8ND6vc6P3Q21u6ONtxw3i0yMEVurNvbyl99+KNb7L3/seW5aekpd58ds6R+g1BzJcmGVutQ/OMzm3gFyYz7EjmyG1x554L4FGUNU09CtwD0Eo4Q+4O4fNrNDgMvNzN39/NQik0iFGkFzpIGgaageM4sndcbvI3jmhZ0AzDuwO7H331+dNHc6N9yzjr/6TvTfhF/6wEm87VUHR5a7Y9VmzOCuzyxi2oTSNazvP7Sey360mttXbWLxK6NfN2l7hvN85ruPsOyRjTV5v4OndnHPJael9vpRiWCuu7/DzDqAewHcfSPwcTM7MbWoJJZ83gFrmkwQLDERPbO4PWuJjvef3BW/j2Dd1qBGMPeA+o+yanRnnXAIr5k7vexscQc+8NV7+fb9z8ZKBD9bvZnXHD6duQeW//w/eMpcbrr/OS7+/mNcd/fTo3Mxwp+HsfM0iq+Xmrcx9hzYXnM+ip+/YftuHl3fy/mnHsGRMyeN/72X+FiCsWDj6+7IMmtKFx3ZvWvE7dl01weNSgTXmtk94fGVxRfc/eFUIpLYgupjszQMhYvOxVh9NMnaAAQ/XBmLt0vZM1t3Mr27nand9V2XqRmYGXOmR9ec3rfgMK7+5ZM8t20Xhx1Quvz6F3ex+vk+Ljkzeme6tmyGK95/AlfcsYbB4fzIENvCL18HPA9OfmQ4LowdkhscjF4rGr5bNISXca5nzPjCe17Fn518eGSszSCqs/g/gP+oUSxSoRj7wDeUOIvODQ7nEu0ohuAXVtylqNdt3RX516hU5v0nB4ng0h8+XnaG+5qw4/mMoj2ny3nloVP5z48uTCLElhfVWfx3wDXuPu7cdDN7M9Dt7relEZyUlw8zQbMsANmWjV50bmAon3gigMJ6Q/FqBAvmTk/8/VvZodMm8M7jD2HZIxu5M2JZipPnTS/Z1CLpiWoaegy4zcwGCNYW6iGYQzAfOBH4OfD/0wxQSiuMwGmWxqFsJtiPwN1LjvQYHM4nuuBcwaTONnZE7FI2OJxj4/bdzD2ptvMBWsFVS07kqiUn1jsMKSGqaeiHBCuMzgfeQLAMdR/wTWCpu+9OP0QpJV+qN6pBtYcdwLm8l1wFc3AoR0cqNYLopqH1L+4m7zBvhkYMJa3eS6BLebHWGnL3J4EnU45FKjQyfLRJfsay4S//4bxTqj94cDhPZxo1gq42tu3cU7bMuq3B0FH1EUirSXOHMklZxLI9Dac9E/ylX24IaRqdxRD0Edzzx628+0u/LVlma5go5ikRSItRImhizTahrDA3IFemw3hgKM+UFFYIfdeJh7B9V/kawcTONt501Eyma+iotBglgiY2utZQc6SC9rBpaKjMENJgHkHyNYLTjp3FacfGG5Yo0mpiJQIzmwmcD8wrfo67/890wpI4RkcNNYdsoWmoTI0graYhESktbo3gh8DdBMNFy+81KDVTaGpvlkTQNtJZXKZGMJT8zGIRKS9uIuh294sqfXEzWwxcBWSB69z98yXKvQe4BTjZ3VdW+j6tKt9kmaAt7CMoXyPIJ7opjYhEi/sTd5uZVbQtpZllCTazORM4DjjHzI4bp9xk4NNA+b0W5SWarbO4LRtj1NBQTjUCkRqLmwg+TTjD2Mz6w399Ec9ZCKx196fcfQ9wE3D2OOX+AfgCMBA7agEYXbO8STqLR2oEUZ3FqhGI1FSsnzh3n+zuGXfvCo8nu3vU5qyHAs8VPV4fnhthZicBh7n7j8u9kJktNbOVZrayp6f8WiWtJN9kncVRTUP5vLMnl86oIREpLfbwUTM7C3hj+PCuaheaM7MMwdLWH4kq6+7XAtcCLFiwoLnWVUhRk3URFHUWj/9fuCeX/KY0IhIt1p9eZvZ5guah1eG/T5vZ5yKetgE4rOjxnPBcwWTglcBdZvYMcAqwzMwWxAtdck2WCdrC4aO5Ek1DA+F+xeosFqmtuDWCtwEnunsewMxuAB4CLinznBXAfDM7giABLAHOLVx0915gRuGxmd0FfEajhuJrvs7icEJZiaahkf2KVSMQqalK/vSaVnQ8Naqwuw8DFwC3A78Hbnb3VWZ2edjMJFUaXX20OVLBaI2gRCIY2a9YNQKRWopbI/gc8JCZ3UnwW+eNwMVRT3L35cDyMecuLVF2UcxYJFRYdK5JBg0V1QjGbxoaHA6ahjRqSKS24i5D/e2w6ebk8NRF7r4ptagklmbbj6CtaD+C8RSahrrUNCRSU2X/9DKzY8KvJxFsSrM+/HdIeE7qqPmGjwa3W6k+gkJnsWoEIrUVVSO4EFgKXDHONQfenHhEEluuyTamKTQNRdUI1FksUltRW1UuDQ/PdPe9Zv6aWVdqUUksI2sNNUmdIGpm8UgfgTqLRWoq7k/c72KekxrKe/PUBmC0aajUzOLCqKE0Nq8XkdLK1gjMbDbBshATzOzVjP7pOQXQDt91lmu2zuKIZahHm4ZUIxCppag+grcSLAExh6CfoJAI+oC/TS8siSOf9yZpFAqMNg2ps1ikkUT1EdwA3GBm73H379UoJokp7441UdtQYRnq1Rv7uPMPW15y/fGNvYA6i0VqLe6EsteY2S/cfTuAmU0H/trd/y61yCRSiXlZDau7I0tHW4Yb73uWG+97dtwynW0ZujuUCERqKW4iONPdR5qC3P3FcKMaJYI6yntzNQ11tWf55V+/iRd27ClZZsakDnUWi9RY3ESQNbNOdx8EMLMJQGd6YUkcefdmGTk6Ys70buZM1zgDkUYSNxHcCPzCzP4zfPxR4IZ0QpK4ck3WWSwijSnuWkNfMLNHgNPDU//g7renF5bEEaw1pFQgItWJvUMZwVLSw+7+czPrNrPJ7t6fVmASLZ9vrgllItKY4u5Qdj5wC/CV8NShwK0pxSQx5Zqss1hEGlPcmTufAt5AMJEMd38SOCitoCSefImJWSIilYibCAbdfWTMn5m1Eaw+KnUUTCirdxQi0uziJoJfmdnfEqw5dAbwXeBH6YUlcQRrtykTiEh14iaCi4Ee4DHgzwm2n9RksjprtrWGRKQxxUoE7p5396+6+/vc/b3hcWTTkJktNrM1ZrbWzF6yx7GZfcLMHjOzh83sN2Z23L58E61KTUMikoSoZagfo0xfgLsfX+a5WeAa4AyC7S1XmNkyd19dVOxb7v7lsPxZwJXA4vjht7ZSO32JiFQiah7BO6p47YXAWnd/CsDMbgLOBkYSgbv3FZWfiDqgK9Jsaw2JSGOKWoZ6XeHYzOYC88MJZROinksw1+C5osfrgdeOLWRmnyLYG7kD7YFckbyjGWUiUrV9nVA2h4QmlLn7Ne7+MuAiSnRAm9lSM1tpZit7enqSeNv9gtYaEpEkpDmhbANwWNHjOeG5Um4C3jXeBXe/1t0XuPuCmTNnxgx5/9eMq4+KSONJc0LZCmC+mR1hZh3AEmBZcQEzm1/08O3AkzHjEdRHICLJiLvo3NgJZX9BxIQydx82swuA24Es8DV3X2VmlwMr3X0ZcIGZnQ4MAS8CH97Xb6QVqWlIRJIQNxFcDHyMvSeUXRf1JHdfHpYtPndp0fGnY0cqL5HPo6YhEala3P0I8sBXzewG4BXAhjgTyiRdQdOQMoGIVKdsH4GZfdnMXhEeTwUeBr4OPGRm56QfnpSTUy4WkQREdRaf6u6rwuOPAk+4+6uA1wB/k2pkEimf1xITIlK9qESwp+j4DMK5A+6+Ka2AJL68Ro+KSAKiEsF2M3uHmb2aYB7BT2Fk+OiEtIOT8rTWkIgkIaqz+M+Bq4HZwP8uqgmcBvw4zcAkWrD6qOoEIlKdqLWGnmCc1UDd/XaC+QFSR3l1FotIAuLOLJYGpAllIpIEJYImFqw+Wu8oRKTZKRE0Ma01JCJJiNqh7MJy1939ymTDkUoETUNKBSJSnahRQ5PDr0cDJzO6eug7gfvTCkriyWsigYgkIGrU0GUAZvZr4CR37w8f/z0aPlp3ygMikoS4fQSz2HuW8Z7wnNRRThvTiEgC4i5D/XXgfjP7Qfj4XcANqUQkseU1fFREEhB3Gep/MrOfAKeGpz7q7g+lF5bEoWWoRSQJlQwf7Qb63P0qYL2ZHZFSTBJTTp0EIpKAWInAzD4LXARcEp5qB76ZVlASj/KAiCQhbo3gfwBnATsB3H0jo0NLpU601pCIJCFuItgTbk3pAGY2Mb2QJK6cNqYRkQTETQQ3m9lXgGlmdj7wc+CrUU8ys8VmtsbM1prZxeNcv9DMVpvZo2b2CzObW1n4rS2oESgTiEh1IkcNWbDg/XeAY4A+glnGl7r7zyKelwWuIdjZbD2wwsyWufvqomIPAQvcfZeZfRL4Z+DP9uk7aUHqIxCRJEQmAnd3M1se7lVc9pf/GAuBte7+FICZ3QScDYwkAne/s6j8vcB5Fbx+y8vlnTZlAhGpUtymoQfN7OQKX/tQ4Lmix+vDc6V8DPjJeBfMbKmZrTSzlT09PRWGsX/Ka5tKEUlI3JnFrwU+YGbrCEYOGUFl4fgkgjCz84AFwJvGu+7u1wLXAixYsEC/ARkdMaQKgYhUK24fwVJgXYWvvQE4rOjxnPDc2Nc/Hfi/wJvcfbDC92hZucLQUQ0bEpEqxe0juCbsI6jECmB+OAN5A7AEOLe4gJm9GvgKsNjdt1T4+i0tnw++Kg2ISLVS6yNw92HgAoJN7n8P3Ozuq8zscjM7Kyz2L8Ak4Ltm9rCZLSvxcjLGSNOQMoGIVKmSPoLzzOwZKugjcPflwPIx5y4tOj69omhlRE6zikUkIXETwVtTjUIqVhg1pAqBiFQrVtOQu68j6Ph9c3i8K+5zJR0jo0fVNiQiVdLqo00qpxqBiCREq482Ka08KiJJ0eqjTUqjhkQkKamuPirpUdOQiCSl7KghM+t090F3/1czO4MKVh+VdBUmlCkViEi1ooaP3gOcZGbfcPcPUtnqo5IiNQ2JSFKiEkGHmZ0LvN7M3j32ort/P52wJIomlIlIUqISwSeADwDTgHeOueaAEkGdaEKZiCSlbCJw998AvzGzle5+fY1ikhjyWnxURBISa4kJd7/ezF4PzCt+jrt/PaW4JEJOG9OISEJiJQIz+wbwMuBhIBeedkCJoE5GN6ZRlUBEqhN30bkFwHHhpDJpACMzi5UHRKRKcSeUPQ7MTjMQqYwmlIlIUuLWCGYAq83sfmBkO0l3P6v0UyRNqhGISFLiJoK/TzMIqdzIqCFlAhGpUtxRQ79KOxCpjJqGRCQpUWsN9ROuODr2EsFWlVNSiUoi5fNqGhKRZERNKNOeAw1KeUBEkhK3j2CfmNli4CogC1zn7p8fc/2NwL8DxwNL3P2WNOOpxJa+AS763qPs2pMrWWbR0QfxyUUvq2FUo7TWkIgkJbV9h80sC1wDnAkcB5xjZseNKfYs8BHgW2nFsa8eem47d67pKZkInnphJ9+6f12Noxo1staQqgQiUqU0awQLgbXu/hSAmd0EnA2sLhRw92fCa/nxXqCedu0ZBuDqc17NETNeuiHbZT9axS0r19c6rBGaWSwiSUmtRgAcCjxX9Hh9eK5iZrbUzFaa2cqenp5EgouyYzCoCUzsyI57/YDuDvoHh9kzXJ8cprWGRCQpaSaCxLj7te6+wN0XzJw5sybvuWswqBFM7By/0jRtYgcA23fvqUk8Y2ljGhFJSpqJYANwWNHjOeG5prBzcBgzmNBeukYA8OLOoVqGNUIVAhFJSpp9BCuA+WZ2BEECWAKcm+L7JWrHYI7u9iyZzPh/ck/vbgfgxV3xagS/+P1mLr9tdWSTzoETO/j20lPo7ij/X6MJZSKSlNQSgbsPm9kFwO0Ew0e/5u6rzOxyYKW7LzOzk4EfANOBd5rZZe7+irRiqsSuPcMlm4UApk8s1AjiJYJfP9HDpt4B3n78wSXLrH9xN/c/vY2N23fz8oPKT+EYXWtIqUBEqpPqPAJ3Xw4sH3Pu0qLjFQRNRg1nx2BEIig0De2K1zT0fO8Acw/s5sr3n1iyzM9Xb+b+p7eVnbtQMDpqSESkOk3RWVwPOweHmdg5fv8AwLQKm4Y29Q0wa0pX2TITwhFKcRJBLhyspEQgItVSIihh554cE8u003e1Z+nuyLItZtPQpt4BDp4aLxHsHopfI1AmEJFqKRGUsDOiaQiC5qE4NYKhXJ6eHYPMnjqhbLnuQiKI0zSkzmIRSYgSQQlxEsEBEztidRZv6R/EncgaQXd78H6xmobUWSwiCVEiKGHnnhyTyvQRQNBPEKezeFPvbgBmx+wj2B0ub1GOVh8VkaQoEZSwc3A4ciz/ARPjNQ1t6g1295wds48g1qghzSgTkYQoEYwjn3d27cnF6yOI0TT0fFgjiOwsbo/fWZzT6qMikhAlgnHsGiq/4FzB9O4O+gaGGc6VX3huU+8AXe0Zpk5oL1sumzE62zLxOos1j0BEEqJEMI6dEQvOFUyfGPxi3767fD/B830DzJ7ShcX48727I1vRhDKlAhGplhLBOAqJYFKMpiGIXmZic+9AZP9AQXdHW2UTypQHRKRKSgTj2BnuRdAdo2kIiJxU9nzvAAdHzCEo6GrPsHsozqghdRaLSDJSXWuoWe2IWyMIm4YuvPmRsstRbOzdHbm8REHcGoEmlIlIUpQIxlHYpjKqj2D+QZM5Z+HhbI8YQjp/1mTOPvGQWO89oSMbq7M4p41pRCQhSgTj2DHSWVy+aaijLcPn3v2qRN877vpFo9MIlAlEpDrqIxhHoWkmqkaQhtijhvJOiT1zREQqokQwjrjDR9Mwob0tdtNQVplARBKgRDCOQtNQd4n9itM0oSMz0kdRTt6djDoIRCQBSgTj2LUnR1d7hrZs7T+e7o62ePsR5JUIRCQZSgTj2DE4XHZTmjRNaM8yMJSPXFQul0dNQyKSCCWCccTZiyAt3TF3KQuahmoRkYjs71JNBGa22MzWmNlaM7t4nOudZvad8Pp9ZjYvzXji2jkYvfJoWrpjLkWddyejTCAiCUgtEZhZFrgGOBM4DjjHzI4bU+xjwIvu/nLg34AvpBVPJXYODkeuPJqWrvZ421Xm8k5WfQQikoA0/+xdCKx196cAzOwm4GxgdVGZs4G/D49vAb5oZuae/EI6N963ji/d+cdYZXv6B3ndyw5MOoRYCpvh7IpYbyjvqEYgIolIMxEcCjxX9Hg98NpSZdx92Mx6gQOBF4oLmdlSYCnA4Ycfvk/BHDJ1AqccGf+X+1kxl4RIWtwN7DWhTESS0hRLTLj7tcC1AAsWLNin2sKfHnMQf3rMQYnGlYYJMRNBztU0JCLJSLOzeANwWNHjOeG5ccuYWRswFdiaYkwNT53FIlJraSaCFcB8MzvCzDqAJcCyMWWWAR8Oj98L/DKN/oFmMpIIooaPakKZiCQktaahsM3/AuB2IAt8zd1XmdnlwEp3XwZcD3zDzNYC2wiSRUsbHTVUvrM458GEspbOmiKSiFT7CNx9ObB8zLlLi44HgPelGUOzKYwaiuwsDieURS9GISJSnmYWNxg1DYlIrSkRNJjOtgxmMSeUqbNYRBKgRNBgzIzu9ujNafKOagQikgglggY0IcYuZcHw0RoFJCL7taaYUNZqJnRkue3Rjax8ZlvJMhu272b+QZNqGJWI7K+UCBrQn7/xZfzujy+ULTN/1iROP3YWV99So6BEZL+lRNCAzjtlLuedMjdW2atTjkVE9n9qZRYRaXFKBCIiLU6JQESkxSkRiIi0OCUCEZEWp0QgItLilAhERFqcEoGISIuzZtsQzMx6gHVVvswMoPzU3fpQXJVRXJVRXJXZ3+Ka6+4zx7vQdIkgCWa20t0X1DuOsRRXZRRXZRRXZVopLjUNiYi0OCUCEZEW16qJ4Np6B1CC4qqM4qqM4qpMy8TVkn0EIiIyqlVrBCIiElIiEBFpcftdIjCzLjO738weMbNVZnZZeP4IM7vPzNaa2XfMrCM83xk+Xhten1fjuG40szVm9riZfc3M2sPzi8ys18weDv9dWuO4/svMni56/xPD82ZmV4ef16NmdlKN47q7KKaNZnZreL4mn1dRfFkze8jMbgsf1/X+KhNXXe+vMnHV9f4qE1ej3F/PmNlj4XutDM8dYGY/M7Mnw6/Tw/PVf2buvl/9AwyYFB63A/cBpwA3A0vC818GPhke/wXw5fB4CfCdGsf1tvCaAd8uimsRcFsdP6//At47Tvm3AT8Jn3cKcF8t4xpT5nvAh2r5eRW994XAtwrvWe/7q0xcdb2/ysRV1/urVFwNdH89A8wYc+6fgYvD44uBLyT1me13NQIP7Agftof/HHgzUNjh9wbgXeHx2eFjwuunmZnVKi53Xx5ec+B+YE7S770vcZV5ytnA18Pn3QtMM7ODax2XmU0h+D+9Nen3jmJmc4C3A9eFj40631/jxQVQ7/urVFxl1OT+ioqrnvdXGcX30th7rKrPbL9LBDBS3XsY2AL8DPgjsN3dh8Mi64FDw+NDgecAwuu9wIG1iMvd7yu61g58EPhp0VNeFzaN/MTMXpFGTBFx/VNY1fw3M+sMz418XqHiz7JWcUHwQ/ALd+8rOleTzwv4d+BvgHz4+EAa4P4aJ64R9by/ysRV1/urTFxQ3/sLgj967jCzB8xsaXhulrs/Hx5vAmaFx1V/ZvtlInD3nLufSPDXz0LgmPpGFBgbl5m9sujyl4Bfu/vd4eMHCdYGOQH4D1L8y6REXJcQfG4nAwcAF6X1/hXGVXAOQVNHQU0+LzN7B7DF3R9I4/X3VYy46nJ/lYmrrvdXjM+rLvdXkT9x95OAM4FPmdkbiy+GNbzExv7vl4mgwN23A3cCryOoLrWFl+YAG8LjDcBhAOH1qcDWGsW1OHzfzwIzCdorC2X6Ck0j7r4caDezGbWKy92fD6uag8B/EiRUKPq8QsWfZepxAYSfw0Lgx0VlavV5vQE4y8yeAW4iaD64ivrfXy+Jy8y+Gb5vPe+vceNqgPur3OdVz/ur8H4bwq9bgB+E8WwuNPmEX7eExav+zPa7RGBmM81sWng8ATgD+D3BL5L3hsU+DPwwPF4WPia8/ssw29Yirj+Y2ceBtwLnuHu+qPzsQluymS0k+L9K/BdImbgKN5wRVJMfD5+yDPhQOFLhFKC3qLqaelzh5fcSdNwNFJWvyefl7pe4+xx3n0fQ+ftLd/8Adb6/SsR1Xr3vrzJx1fX+KhVXeLlu91f4+hPNbHLhGHgLwedTfC+Nvceq+szaoos0nYOBG8wsS/CfdbO732Zmq4GbzOwfgYeA68Py1wPfMLO1wDaCm6KWcQ0TLKt9T3iffd/dLye4GT8ZXt9NMCIljWngpeL6pZnNJBiJ8DDwibD8coJRCmuBXcBHU4ipZFzhtSXA58eUr9XnVcpF1Pf+KuXL1Pf+KuXGOt9f5dT7/poF/CD8/2oDvuXuPzWzFcDNZvYxgv/T94flq/7MtMSEiEiL2++ahkREpDJKBCIiLU6JQESkxSkRiIi0OCUCEZEWp0QgEsHMchasArkqXGLgr82s7M+Omc0zs3NrFaNINZQIRKLtdvcT3f0VBBPbzgQ+G/GceYASgTQFzSMQiWBmO9x9UtHjI4EVwAxgLvANYGJ4+QJ3/52Z3QscCzxNsFLk1QSTlBYBncA17v6Vmn0TImUoEYhEGJsIwnPbgaOBfiDv7gNmNh/4trsvMLNFwGfc/R1h+aXAQe7+jxastPlb4H3u/nQNvxWRce2PS0yI1FI78EULdtjKAUeVKPcW4HgzK6xHNBWYT1BjEKkrJQKRCoVNQzmC1R8/C2wGTiDocxso9TTgf7n77TUJUqQC6iwWqUC4UNqXgS+Gi45NBZ4PV/b8IJANi/YDk4ueejvBomWFPYOPCleWFKk71QhEok2wYKe0dmCYoHP4yvDal4DvmdmHCHb/2hmefxTImdkjBPvzXkUwkujBcDnjHka3GhSpK3UWi4i0ODUNiYi0OCUCEZEWp0QgItLilAhERFqcEoGISItTIhARaXFKBCIiLe6/AVILYVHQW+JlAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data      = np.load(os.path.join(INF_DIR, f'infer-london-sd', 'inf.npz'), allow_pickle=True)    \n",
    "sites     = data['sites']\n",
    "selection = data['selection']\n",
    "times     = data['times']\n",
    "alpha = ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "         'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "         'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "         'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "         'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T']\n",
    "\n",
    "f = open(os.path.join(DATA_DIR, 'alpha-sites-exclusive.npy'), mode='wb')\n",
    "np.save(f, alpha)\n",
    "f.close()\n",
    "\n",
    "alleles    = [dp.get_label_new(i) for i in sites]\n",
    "idxs_alpha = [alleles.index(i) for i in alpha]\n",
    "s_alpha    = np.sum(np.array(selection)[:, idxs_alpha], axis=1)\n",
    "x_detect   = np.array(times)[s_alpha>0.13183][0]\n",
    "\n",
    "out_file = os.path.join(DATA_DIR, 'alpha-subdate-early-detection.csv')\n",
    "f = open(out_file, mode='w')\n",
    "f.write('time,selection\\n')\n",
    "for i in range(len(times)):\n",
    "    f.write(f'{times[i]},{s_alpha[i]:.5f}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Great Britain__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-02-20-uk/genome-filtered | wc -l\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar-sd.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel-sd.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-vector blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-freqs-sd-uk.sh blee098@secure.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-sd-uk.sh blee098@secure.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-freqs-sd-uk.sh &&\n",
      "sbatch job-infer-sd-uk.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-02-20-uk &&\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-02-20-uk-* /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-02-20-lond\n",
      "\n",
      "scp -r blee098@secure.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-02-20-uk /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered-uk')\n",
    "freq_script    = 'epi-covar-sd.py'\n",
    "inf_script     = 'epi-inf-parallel-sd.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs-uk-sd')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer-uk-sd', 'inf')\n",
    "out_dir        = ppath.join(SSH_DATA, f'infer-uk-sd')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files = 6    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1 = 'job-freqs-sd-uk.sh'\n",
    "job_str1  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-sd-uk-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-sd-uk-out-%a\n",
    "#SBATCH --array=1-{num_files - 1}\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-vector\n",
    "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "pwd\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer-sd-uk.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=12-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-uk-error\n",
    "#SBATCH -o ./MPL/out/inf-uk-out\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}\n",
    "python {inf_script} --data {temp_dir} --timed 1 -o $outfile -q 5 --g1 10 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s' % (SSH_HOME, out_dir, INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-05-03\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEGCAYAAABo25JHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAkTklEQVR4nO3deXxV9Z3/8dcnNwtJgLBFQAOCilpc0Yh1tEqtWrAW26m2aqerLd38TTtOp9Vpx1rb6d4+altm6tIF7eLW1jJKS12rWBdQEQVEI4sEZYcESUK2z++PcxJuYpJ7gZxzT3Lfz4fXnPM9y/1c7k0+9/v9nvP9mrsjIiL5qyDXAYiISG4pEYiI5DklAhGRPKdEICKS55QIRETyXGGuA9hXY8aM8UmTJuU6DJEDsmpV8POoo3Ibh+SPp59+equ7V/a0bcAlgkmTJrFkyZJchyFyQGbMCH4+/HAuo5B8YmbretumpiERkTynRCAikueUCERE8pwSgYhInlMiEBHJc0oEIiJ5TolARCTPKRGIDDL/qNnKXU/X5joMGUAG3A1lItK3y25+EoC3TRnD2OFDchyNDARKBCIDzIMvbmLFa/UcNW44D6zcxLf/+TjMDID0iaa2vrFHiUCyokQgMsB8/NfBECsjy4rY0dDCB06ZwLSJIwGob2rt3G9nQ0tO4pOBR30EIgnX2NzGD/+2ilnXP8rjr2zrLC8vCb7HPblmOxDUBr55z4rO7Vt27WHx2u3xBisDkhKBSMItqtnKTx+sYeXr9Tzz6o7O8todjQDMfbCGtnZn/fZG7kzrJP7C7Uu5+OeP8/KmXbHHLAOLEoFIwjU0723uaWxue9P2XXta+cPTtZz5/YcA+OVHqykssM7tm3ftiT5IGdDURyCScHta2zuXt3T7o/6htx7KrU+sY1HN1s6yyWOGcsOHTubf73yOnQ0t/O6pV2lobmPymHKOOGhobHFL4PnaOubcuoTmtPdxf10162gurp7QD1F1FWkiMLOZwPVACrjZ3b/Twz7vB64FHHjO3S+LMiaRgaYjERQXFnD7kvUA3Hr5dN42pZKNdU3c+sQ6Vm0Mmn/GVwyhamQpk8eUs+jLZ3Ps1xZy77LXuXfZ6wC8v7qKUeUlFKWMwoICClPGmVMqOa6qostzbqxr4u8vbabAjKJUsF9hgTG0pIjjJ1SQCq9SMoM39rTy6VufZvvu5rDM6KyPGJ3L6eVm0LFmeysvb97HOk5jact7d+h6vq7nKDAjVWAUpLV79PSc6c+dfv4ucfWyT9f9eng9wJqtu9nR0MxFJ1e9+Un30aGjyw/4HD2JLBGYWQqYC5wL1AKLzWy+u69I22cKcDVwurvvMLODoopHZKDa0xI0B40pL+a1uiZKCgs444gxQdnQYgBWhf0Aj37p7RSmgr98Q8PO5NMOG81nZhzON+9dwT3LXqe13Wlta6c9vNL0Zw/W8KmzDmPp+p188NRDOXfqWH74t1Vd+huyccHx4ynouIyVvZeyeuf/wHE6rnB1D9b3Lu8tp8dy77JPejldyoP1tnantb0db+t47q777z2KvTHx5kLvtk/H6+hyXJdtXZ8nVWB8+qzD+cI5R5JUUdYIpgM17r4awMxuAy4EVqTt80lgrrvvAHD3zRHGIzIgddQIRg8t4bW6JsZVDOn8ZlqYKuCECSN4bv3OzvV0K657JyWFKVIFxt+OPKvLtvZ2Z/XWN5j540f58f0vA0EfxLlTx/Lixl2cOnkUP7j4BFra2sPk4azdtpvaHQ1A1z9+E0eVMeu48VG8fIlBlIngEGB92notcGq3fY4EMLPHCJqPrnX3v3Y/kZnNAeYATJw4MZJgRZJqT0sbZjCyPPj23/0msbs/+09MvnpBj8eWFff+K15QYBxx0DB+/bHpvLazkWfX7+C2xes55pq/sru5jcvPmMyEUWVdjpl68PADfDWSRLnuLC4EpgAzgCrgETM7zt13pu/k7jcCNwJUV1c7InlkT2s7JYUFTBxVCsCRY7t2+JoZ33rvcexoaN6v858xJWhmmj55FEOKUqTMGFKU4gOn9H+npCRTlIlgA5D+SaoKy9LVAk+6ewuwxsxeIkgMiyOMS2RACRJBiq/PPpZPnXk4B48ofdM+l5164DXlSWPK+dq7jzng88jAE+V9BIuBKWY22cyKgUuA+d32uZugNoCZjSFoKlodYUwiA05TSxslhQWkCowJo8pIFfRwyYvIAYgsEbh7K3AFsBBYCdzh7svN7Dozmx3uthDYZmYrgIeA/3D3bT2fUSQ/7Wltp6RI935KdCLtI3D3BcCCbmXXpC07cGX4EJEe7Glto6QwleswZBDT1wyRBFv08lYWPL+R4pR+VSU6+nSJJNgjL28BYPaJB+c4EhnMlAhEEq60KMWnzzo812HIIKZEIJJg6cMniERFiUAkwdx7HiBNpD8pEYgkmNN1pEuRKCgRiCRYUCNQKpBoKRGIJJjjqhFI5JQIRBLM1TYkMVAiEEk45QGJmhKBSIK5u/oIJHJKBCIJ5ujyUYmeEoFIgrmraUiip0QgkmCOmoYkekoEIgmmGoHEQYlAJMHURyBxUCIQSbBgzDllAomWEoFIorlqBBI5JQKRBFMfgcRBiUAkwTQMtcRBiUAkwYJB55QJJFpKBCIJphqBxCHSRGBmM81slZnVmNlVPWz/qJltMbOl4eMTUcYjMtBo8FGJQ2FUJzazFDAXOBeoBRab2Xx3X9Ft19vd/Yqo4hAZyDQxjcQhyhrBdKDG3Ve7ezNwG3BhhM8nMug4mrxeohdlIjgEWJ+2XhuWdfc+M1tmZneZ2YSeTmRmc8xsiZkt2bJlSxSxiiST+ggkBrnuLP4/YJK7Hw/cB8zraSd3v9Hdq929urKyMtYARXJJQ0xIHKJMBBuA9G/4VWFZJ3ff5u57wtWbgZMjjEdkwHHX5aMSvSgTwWJgiplNNrNi4BJgfvoOZjY+bXU2sDLCeEQGHNUIJA6RXTXk7q1mdgWwEEgBv3T35WZ2HbDE3ecD/2pms4FWYDvw0ajiERmolAckapElAgB3XwAs6FZ2Tdry1cDVUcYgMpDp8lGJQ8ZEYGanAf8CvA0YDzQCLwD3Ar9x97pIIxTJY7qhTOLQZx+Bmf0F+ARB885MgkQwFfgqMAT4c9i0IyIRcNd9BBK9TDWCD7n71m5lbwDPhI8fmtmYSCITETQvjcShzxpBD0kAM3uHmb3bzIp620dE+onmI5AY7NPlo2b2Q+B04ATgz5FEJCKdHFdnsUSuz6ah8A//N9x9Z1g0EXh/uPx8hHGJCJqhTOKRqUbwR+A2M/vXcDTRW4CHgMeBm6IOTiTfaT4CiUOmPoLH3H0mwc1eCwFz9xnu/lZ3vz6WCEXymGYokzhkuny00MzeBWwG3gOcYGbzzeyEOIITyXeqEUgcMl0+ejdBM1AZ8EF3/4iZHQxcZ2bu7p+MOkCRfKa7CCQOmRLBoe5+QTho3BMA7v4a8AkzOzHq4ETynYaYkDhkSgQ3mtnj4fKP0je4+9JIIhKRNK4eAolcn4nA3X8K/DSmWESkG/URSBwydRZ/1cxG9rH9bDO7oP/DEhHQfAQSj0xNQ88D95hZE8HYQlsIBpubApwI3A98K8oARfKZZiiTOGRqGvozwQijUwiGlhgP1AO/Aea4e2P0IYrkL9UIJA5ZTUzj7i8DL0cci4h0oyEmJA5RzlksIgcoGIZaqUCipUQgkmBBH4FItJQIRBJOFQKJWlZ9BGZWCXwSmJR+jLt/PJqwRATURyDxyCoREExC8yjB5aJt0YUjIuk0MY3EIdtEUObuX97Xk5vZTOB6IAXc7O7f6WW/9wF3Aae4+5J9fR6RwUo1AolDtn0E95jZ+fty4nAim7nALGAqcKmZTe1hv2HA54En9+X8IvlAQ0xIHLJNBJ8nvMPYzHaFj/oMx0wHatx9tbs3A7cBF/aw3zeA7wJNWUctkic0MY3EIatE4O7D3L3A3YeEy8PcfXiGww4B1qet14ZlnczsJGCCu9+7T1GL5Al31DYkkcu2jwAzmw2cGa4+7O73HMgTm1kBwdDWH81i3znAHICJEyceyNOKDDjKAxK1rGoEZvYdguahFeHj82b27QyHbQAmpK1XhWUdhgHHAg+b2VrgrcB8M6vufiJ3v9Hdq929urKyMpuQRQYFjTUkcci2RnA+cKK7twOY2TzgWeDqPo5ZDEwxs8kECeAS4LKOje5eB4zpWDezh4Ev6qohkTSaoUxisC93Fo9IW67ItLO7twJXAAuBlcAd7r7czK4Lm5lEJIPgPoJcRyGDXbY1gm8Dz5rZQwRNlmcCV2U6yN0XAAu6lV3Ty74zsoxFJG/o8lGJQ7bDUP8+bLo5JSz6srtvjCwqEQHC0UdFIpZpqsqjw58nEUxKUxs+Dg7LRCRCmqFM4pCpRnAlwWWbP+xhmwNn93tEItJJVw1JHDJNVTknXJzl7l3u/DWzIZFFJSJAeEOZSMSyvWroH1mWiUg/CmoEqhJItPqsEZjZOIJhIUrNbBp7b3IcDpRFHJuIaIYyiUGmPoJ3EgwBUUXQT9DxmawH/jO6sEQE1Ecg8cjURzAPmGdm73P3P8QUk4iENB+BxCHbPoKTzWxEx4qZjTSzb0YTkoh00AxlEodsE8Esd9/ZseLuOwjGHxKRCKlGIHHINhGkzKykY8XMSoGSPvYXkX6gISYkDtmONfRb4AEz+1W4/jFgXjQhiUiH4DYCZQKJVrZjDX3XzJ4DzgmLvuHuC6MLS0QgHGJCeUAilvUMZQRDSbe6+/1mVmZmw9x9V1SBiUhAeUCilu0MZZ8E7gJuCIsOAe6OKCYRCamPQOKQbWfx54DTCW4kw91fBg6KKigRCTgafVSil20i2OPuzR0rZlaIhkoXiZxqBBKHbBPB383sPwnGHDoXuBP4v+jCEhHQEBMSj2wTwVXAFuB54FME009+NaqgRCSgiWkkDtlePtoO3BQ+RCQmDrpsSCKXaRjq5+mjL8Ddj+/3iERkLw0xITHIVCO4IJYoRKRHmphG4pBpGOp1HctmdigwJbyhrDTTsSJy4FwT00gM9veGsiqyuKHMzGaa2SozqzGzq3rY/mkze97MlprZIjObug+xi+QFVQgkapHdUGZmKWAuMAuYClzawx/637n7ce5+IvA94EfZhy4y+DnqI5DoRXlD2XSgxt1Xh8feBlyYvoO716etlmdxTpG8EtxQplQg0cq2nb/7DWWfJfMNZYcA69PWa4FTu+9kZp8DrgSKgbN7OpGZzQHmAEycODHLkEUGvmCICZFo5fyGMnef6+6HA1/u7ZzufqO7V7t7dWVlZX88rciA4Gobkhjs0w1lZjYPOAbY4O6ZmnE2ABPS1qvCst7cBvxvNvGI5ItgqkplAolWnzUCM/u5mR0TLlcAS4FbgGfN7NIM514MTDGzyWZWDFwCzO92/ilpq+8CXt638EUGP3URSNQyNQ29zd2Xh8sfA15y9+OAk4Ev9XWgu7cCVwALCSa1ucPdl5vZdWY2O9ztCjNbbmZLCfoJPrKfr0NkUMpc8RY5cJmahprTljtGHcXdN2ZzJYO7LyDoT0gvuyZt+fNZRyqSh9RFIHHIVCPYaWYXmNk0gvsI/gqdl4+WRh2cSL7TfAQSh0w1gk8BPwHGAV9w941h+TuAe6MMTEQ0Q5nEI9NYQy8BM3soX0jQ9i8iEVKNQOKQ7X0EIpIDmqFM4qBEIJJgrplpJAZKBCKJ5qoRSOQyzVB2ZV/b3V2jhYpEyDVDmcQg01VDw8KfRwGnsPfO4HcDT0UVlIgE1Ecgcch01dDXAczsEeAkd98Vrl+LLh8ViVwwQ5kygUQr2z6CsXS9y7g5LBORCKlGIHHIdj6CW4CnzOxP4fp7gHmRRCQindRHIHHIdhjq/zazvwBvC4s+5u7PRheWiEDYNKQqgURsXy4fLQPq3f16oNbMJkcUk4iENPaoxCGrRGBmXyOYQezqsKgI+E1UQYlISENMSAyy7SN4LzANeAbA3V8zs2F9HyIi2Vq/vYGN9U3s3tNKxxQEFWVFtOmqIYlBtomg2d3dzIIb3s3KI4xJJK88sHITl89b0uv2kiINACDRyjYR3GFmNwAjzOyTwMeBm6ILSyR/fOKWIAnc+KGTGT20mFRBAe7OxromGlvaOOvIyhxHKINdxkRgwSULtwNHA/UEdxlf4+73RRybyKDX2NyGO1x44sGcd8y4XIcjeSpjIgibhBaEcxXrj79IP9pY3wTAmVP0rV9yJ9vGx2fM7JRIIxHJM82t7fz2iXUAjKsYkuNoJJ9l20dwKvBBM1sH7Ca42dHd/fjIIhMZ5F6va+TmRWsoLizg8MqhuQ5H8li2fQRzgHXRhyOSP9rdGV1ezGNfPpvS4lSuw5E8lrFpyN0dmOvu67o/Mh1rZjPNbJWZ1ZjZVT1sv9LMVpjZMjN7wMwO3c/XITLguENJYYGSgORcZH0EZpYC5gKzgKnApWY2tdtuzwLVYRPTXcD39uU5RAaydneKC3WPgORetp/CU4EnzOyV8Nv782a2LMMx04Ead1/t7s3AbcCF6Tu4+0Pu3hCuPgFU7UvwIgOZOxSllAgk97LtLH7nfpz7EGB92notQULpzeXAX3raYGZzCPopmDhx4n6EIpI87SgRSDJk9SkM+wMmAGeHyw3ZHpsNM/sXoBr4fi/Pf6O7V7t7dWWlrreWwcHdKU5pHCHJvaxqBOHoo9UEdxX/ir2jj57ex2EbCJJHh6qwrPu5zwG+Apzl7nuyC1tk4HNHfQSSCNl+Ct8LzCa4hwB3f429E9v3ZjEwxcwmm1kxcAkwP30HM5sG3ADMdvfN+xK4yEDX7q6mIUmEbD+FzeFlpFmPPururcAVwEJgJXCHuy83s+vMbHa42/eBocCdZrbUzOb3cjqRQUedxZIUkY4+6u4LgAXdyq5JWz5nH2IVGVR0+agkRZ+JwMxK3H2Pu//AzM5Fo4+K9BsHilUjkATIVCN4HDjJzG519w+h0UdF+o27U6SrhiQBMiWCYjO7DPgnM/vn7hvd/Y/RhCUy+LWrj0ASIlMi+DTwQWAE8O5u2xxQIhDZT64+AkmIPhOBuy8CFpnZEnf/RUwxieQFXTUkSZHVVUPu/gsz+ydgUvox7n5LRHGJDGrrtzfQ2q4agSRDtncW3wocDiwF2sJiB5QIRPbDjoZmAM55y9gcRyKS/X0E1cDU8KYyEekHo8qLmT450w36ItHLtl76AjAuykBE8okTzPcqkgTZ1gjGACvM7Cmgc2A4d5/d+yEi0itlAkmQbBPBtVEGIZJvHMeUCSQhsr1q6O9RByKST9zBlAckITKNNbSLcMTR7psI5rUfHklUIoOcu1qGJDky3VCmSxpEIuC4agSSGLqbRSQHghqBMoEkgxKBSA446iOQ5FAiEMkBdzUNSXIoEYjkQHAbgTKBJIMSgUgO6PJRSRIlApGYtbdryC5JFiUCkZi1honAVCWQhFAiEIlZW0ciyHEcIh0iTQRmNtPMVplZjZld1cP2M83sGTNrNbOLooxFJCla29sB9RFIckSWCMwsBcwFZgFTgUvNbGq33V4FPgr8Lqo4RJJGNQJJmmxHH90f04Ead18NYGa3ARcCKzp2cPe14bb2COMQSRT1EUjSRNk0dAiwPm29NiwTyWuqEUjSDIjOYjObY2ZLzGzJli1bch2OyAHZWyPIcSAioSgTwQZgQtp6VVi2z9z9RnevdvfqysrKfglOJG6/eWId1/z5BdraOu4jUCaQZIiyj2AxMMXMJhMkgEuAyyJ8PpFE++rdLwAwfEgRMEY1AkmMyGoE7t4KXAEsBFYCd7j7cjO7zsxmA5jZKWZWC1wM3GBmy6OKRyQpfvZQDaCmIUmOKGsEuPsCYEG3smvSlhcTNBmJDHojy4ooKUyxsb4J0KBzkhwDorNYZKD75aI17Gho4fzjxneWFSgPSEJEWiMQkUBHc9Cs48ZxyqSRXPm3coaXFuU4KpGAEoFIDJpb2/n46ZM5ZdIoAL47PMcBiaRR05BIDBpb2igt1q+bJJM+mSIRa2lrp63dKS1K5ToUkR4pEYhErLGlDYAhSgSSUEoEIhFralYikGRTIhCJWFNLMLiuEoEklRKBSMQ6mobURyBJpUQgErGmjkSgq4YkofTJFIlYZ2dxoWoEkkxKBCIR66gRDClWIpBkUiIQiZC781jNVgAOrijNcTQiPVMiEInQz/++mpseXcNhleWMqxiS63BEeqREIBKhFzbUAfA/Hzwpx5GI9E6JQKSfuQdTUTa1tPHU2u2ceWQlR4/TKHOSXBp9VKSfffKWJSyq2UphQQFv7Gnl/GPLch2SSJ9UI5BBb3N9E/92+1LqGlsiOX9jcxsX/mwR184PZlq9f+VmmlraOf2I0VxzwVQ+d/YRkTyvSH9RjUAGvXmPr+VPz27g8Mpyrjh7Sr+f/5lXd/BcbR3P1dbx1Xe9hQKDK95+BFeed1S/P5dIFJQIZND65aI13PDIK2yq3wPAD/72Es/V1vGV89/C2OFDmPf4WopTBXz8jMmdxzy8ajN3Pl0LQElhAV9659GdV/tsrGtiSFEBQ0sKKUztrUwvXru9c3nttt20Oxw0XFcIycChRCCJtKm+ifrGFg4aNoSKsq5TOu7Y3Ux5SSGvbHmDV7a8wdqtu5k+eTTjwz/Y7rByYz3X3bOi85hhJYVMGTuU+1Zs4r4Vmzh36ljuW7EJgDHDSvjpAy8zYVQZy2rr2NPaRuWwEtZu3c3qLbs5etwwaja/wZJ1O4L9hxbz8H+8nevvf4ltu5v54zMbOp/nnB89AsBYJQIZQJQIJKf2tLbxyEtbGTu8hOOrRvDazkY+/uvFvLhxFwCFBcY7jx3H8CGF3L9yM0UFxmt1TVmf/9n/OpfmtnZKi1MMLS7kz89t4Et3LetMAgBf+dPz7GpqJVVgjCwr4r8uOIEzj6zkv+9dwT3LXue1nY0AXHxyFSPKirjp0TV8768vcsvj6zrPMevYcWx7o5mTJ41kaEkhZxwxpp/+hUSiZx2Xug0U1dXVvmTJklyHIfth/fYGXtvZyFNrtvPAi5tZt203u5paaW0PPoMVpUWUF6d4ra6Ji06u4tTJo/jFojVs2NFIUWEBo8qLGVlWxLbdzZx++BiOq6qgamQpk0aX8/gr22h3x8wAMOCwynKmTRz5pjgamlt5o6mVksIUl89bzI6GZj4z4wguOrkq42tobG7j1G/dT31TK0Up46EvzqC1zTlkZClFqeyvvZgxI/j58MNZHyJyQMzsaXev7nFblInAzGYC1wMp4GZ3/0637SXALcDJwDbgA+6+tq9zKhH0j6aWti7j4z+8ajP3LnsdgKLCAqYcNJSiVAFjhhZz2mFjWL31DSpKiygsKKC1vZ2SohQHh00xy2rr2LCzkcbmNh57ZSvrtzewq6mVXU2tbN/dzKjyYsZXDOlsWgEYXzGEaRNHMKKsmFMmjWRZbR31ja3samph2sSRfGbG4fH+g+yD+qYW6hpaGFpSyMjy4v06hxKBxK2vRBBZ05CZpYC5wLlALbDYzOa7+4q03S4Hdrj7EWZ2CfBd4ANRxdSTFzbUYQbDhxT1uk/4JRMIJhcZM7Skz3Pu3tNKU0sbTtBe7Tjhf7S2Ozc9spr3TjuEEyaM6JfXsK9e3FjPzB8/yulHjGZZbR3Vh47kudo69rS0UVFaRF1jC7vDWbX6UlhgmEFL294vE0UpY9rEkUwYVcawkkIqyopYvWU3r25v4APVE5h13DjGV5Ry1LhhXc713mmZv40nxfAhRX1+XkQGmij7CKYDNe6+GsDMbgMuBNITwYXAteHyXcDPzMw8gmrKHYvXc9Ojq7uUNba0UbujcZ/PdciIUopSxo6Gls5r0wsMDh5RihnU7mgk0yv49T/WMnFUzzcapSeeN23rrbyXg3oq3RnG/FjNNgBe3d5A5dASrrvwGE49bDQtbe3UN7bQ1u78dflGanc0cuKEEWzZtYfClDG0pJAdu5vZtCu4GmfS6DJOmDCCksIUo8qLqSjVH0mRgSTKRHAIsD5tvRY4tbd93L3VzOqA0cDW9J3MbA4wB2DixIn7FcyIsiKmjB36pvKZx4xj6sHDaU/7w52eh7r/PX91WwOvbm8Aghmnxg4vATP2tLSxqT7oxDz/2PGdScGCF4AFPzCM2h0NbKxretO5uz93l/I+XltvSaevY+oaWzjukOG8d1oVRxzU9d+lKFXA6LDW8+HTJvVxFhEZDAbEVUPufiNwIwR9BPtzjvOOGcd5x4zr17hERAaDKIeY2ABMSFuvCst63MfMCoEKgk5jERGJSZSJYDEwxcwmm1kxcAkwv9s+84GPhMsXAQ9G0T8gIiK9i6xpKGzzvwJYSHD56C/dfbmZXQcscff5wC+AW82sBthOkCxERCRGkfYRuPsCYEG3smvSlpuAi6OMQURE+qZhqEVE8pwSgYhInlMiEBHJc0oEIiJ5bsCNPmpmW4B1GXfM3hi63cmcQIqxfyjG/qEY+0fcMR7q7pU9bRhwiaC/mdmS3kbkSwrF2D8UY/9QjP0jSTGqaUhEJM8pEYiI5DklgnAwu4RTjP1DMfYPxdg/EhNj3vcRiIjkO9UIRETynBKBiEieG9SJwMx+aWabzeyFtLJRZnafmb0c/hwZlpuZ/cTMasxsmZmdFFOME8zsITNbYWbLzezzSYvTzIaY2VNm9lwY49fD8slm9mQYy+3hcOOYWUm4XhNunxR1jGmxpszsWTO7J4kxmtlaM3vezJaa2ZKwLDHvdfi8I8zsLjN70cxWmtlpSYrRzI4K//06HvVm9oUkxRg+77+Fvy8vmNnvw9+jRH0eO7n7oH0AZwInAS+klX0PuCpcvgr4brh8PvAXgtkl3wo8GVOM44GTwuVhwEvA1CTFGT7X0HC5CHgyfO47gEvC8p8DnwmXPwv8PFy+BLg9xvf8SuB3wD3heqJiBNYCY7qVJea9Dp93HvCJcLkYGJG0GNNiTQEbgUOTFCPBNLxrgNK0z+FHk/Z57Iw3zifLxQOYRNdEsAoYHy6PB1aFyzcAl/a0X8zx/hk4N6lxAmXAMwTzT28FCsPy04CF4fJC4LRwuTDcz2KIrQp4ADgbuCf8xU9ajGt5cyJIzHtNMEvgmu7/FkmKsVtc5wGPJS1G9s7HPir8fN0DvDNpn8eOx6BuGurFWHd/PVzeCIwNlzveuA61YVlswurgNIJv3ImKM2xyWQpsBu4DXgF2untrD3F0xhhurwNGRx0j8GPgS0B7uD46gTE68Dcze9rM5oRlSXqvJwNbgF+FTWw3m1l5wmJMdwnw+3A5MTG6+wbgB8CrwOsEn6+nSd7nERjkfQSZeJB+E3H9rJkNBf4AfMHd69O3JSFOd29z9xMJvnVPB47OZTzdmdkFwGZ3fzrXsWRwhrufBMwCPmdmZ6ZvTMB7XUjQnPq/7j4N2E3QzNIpATECELavzwbu7L4t1zGG/RMXEiTWg4FyYGau4skkHxPBJjMbDxD+3ByWbwAmpO1XFZZFzsyKCJLAb939j0mNE8DddwIPEVRrR5hZxyx36XF0xhhurwC2RRza6cBsM1sL3EbQPHR9wmLs+KaIu28G/kSQVJP0XtcCte7+ZLh+F0FiSFKMHWYBz7j7pnA9STGeA6xx9y3u3gL8keAzmqjPY4d8TATzgY+Eyx8haJPvKP9weIXBW4G6tGpmZMzMCOZuXunuP0pinGZWaWYjwuVSgj6MlQQJ4aJeYuyI/SLgwfAbWmTc/Wp3r3L3SQTNBQ+6+weTFKOZlZvZsI5lgvbtF0jQe+3uG4H1ZnZUWPQOYEWSYkxzKXubhTpiSUqMrwJvNbOy8He8498xMZ/HLuLqjMjFg+BD8jrQQvBN53KCdrcHgJeB+4FR4b4GzCVo+34eqI4pxjMIqrDLgKXh4/wkxQkcDzwbxvgCcE1YfhjwFFBDUD0vCcuHhOs14fbDYn7fZ7D3qqHExBjG8lz4WA58JSxPzHsdPu+JwJLw/b4bGJnAGMsJvjFXpJUlLcavAy+GvzO3AiVJ+jymPzTEhIhInsvHpiEREUmjRCAikueUCERE8pwSgYhInlMiEBHJc0oEIhmYWVs4yuVyC0Zg/Xcz6/N3x8wmmdllccUociCUCEQya3T3E939GIKb6WYBX8twzCRAiUAGBN1HIJKBmb3h7kPT1g8DFgNjCIY/vpXgBieAK9z9H2b2BPAWgpE85wE/Ab5DcLNbCTDX3W+I7UWI9EGJQCSD7okgLNsJHAXsAtrdvcnMpgC/d/dqM5sBfNHdLwj3nwMc5O7fNLMS4DHgYndfE+NLEelRYeZdRKQPRcDPzOxEoA04spf9zgOON7OOcWYqgCkENQaRnFIiENlHYdNQG8Holl8DNgEnEPS5NfV2GPD/3H1hLEGK7AN1FovsAzOrJJhi8GcetKtWAK+7ezvwIYKpEyFoMhqWduhC4DPhkOOY2ZHhCKQiOacagUhmpeHsbEVAK0HncMeQ4f8D/MHMPgz8lWAiFwhG7mwzs+eAXxPMjTAJeCYclngL8J54whfpmzqLRUTynJqGRETynBKBiEieUyIQEclzSgQiInlOiUBEJM8pEYiI5DklAhGRPPf/AVKDOKCkjzscAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = np.load(os.path.join(DATA_DIR, 'infer-uk-sd', 'inf.npz'), allow_pickle=True)\n",
    "selection = data['selection']\n",
    "sites     = data['allele_number']\n",
    "times     = data['times']\n",
    "    \n",
    "f = open(os.path.join(INF_DIR, f'infer-2022-05-10-uk-sd.npz'), mode='wb')\n",
    "np.savez_compressed(f, selection=selection, sites=sites, times=times)\n",
    "f.close()\n",
    "    \n",
    "delta = ['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', \n",
    "         'S-158-0--', 'S-478-1-A', 'S-950-0-A', 'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', \n",
    "         'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'N-63-1-G', \n",
    "         'N-377-0-T', 'S-452-1-G', 'S-681-1-G']  \n",
    "\n",
    "f = open(os.path.join(DATA_DIR, 'delta-sites-exclusive.npy'), mode='wb')\n",
    "np.save(f, delta)\n",
    "f.close()\n",
    "\n",
    "alleles    = [dp.get_label_new(i) for i in sites]\n",
    "idxs_delta = [alleles.index(i) for i in delta]\n",
    "s_delta    = np.sum(np.array(selection)[:, idxs_delta], axis=1)\n",
    "x_detect   = np.array(times)[s_delta>0.13183][0]\n",
    "\n",
    "out_file = os.path.join(DATA_DIR, 'delta-subdate-early-detection.csv')\n",
    "f = open(out_file, mode='w')\n",
    "f.write('time,selection\\n')\n",
    "for i in range(len(times)):\n",
    "    f.write(f'{times[i]},{s_delta[i]:.5f}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__South Africa__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered-sa\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered/africa-south* /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered-sa\n",
      "readarray -d '' files < <(find /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/genome-filtered-sa ! -name \"*sites*\" -print0)\n",
      "echo ${#files[*]}\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/2022-06-09/freqs-south-africa-sd\n",
      "\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/epi-covar-sd.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/Python/MPL/epi-inf-parallel-sd.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp -r /Users/brianlee/Python/MPL/6-29-20-epidemiological/Archive-vector blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/data_processing.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-freqs-sd-sa.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-infer-sd-sa.sh blee098@cluster.hpcc.ucr.edu:.\n",
      "\n",
      "sbatch job-freqs-sd-sa.sh &&\n",
      "sbatch job-infer-sd-sa.sh &&\n",
      "mkdir /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-06-09-uk &&\n",
      "mv /rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-uk-* /rhome/blee098/bigdata/SARS-CoV-2-Data/inf-2022-06-09-uk\n",
      "\n",
      "scp -r blee098@cluster.hpcc.ucr.edu:/rhome/blee098/bigdata/SARS-CoV-2-Data/infer-2022-06-09-uk /Users/brianlee/Python/MPL/6-29-20-epidemiological\n"
     ]
    }
   ],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "input_dir      = ppath.join(SSH_DATA, 'genome-filtered-south-africa')\n",
    "freq_script    = 'epi-covar-sd.py'\n",
    "inf_script     = 'epi-inf-parallel-sd.py'\n",
    "temp_dir       = ppath.join(SSH_DATA, 'freqs-south-africa-sd')\n",
    "out_file       = ppath.join(SSH_DATA, f'infer-south-africa-sd', 'inf')\n",
    "out_dir        = ppath.join(SSH_DATA, f'infer-south-africa-sd')\n",
    "\n",
    "# run on cluster\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files = 3    # the number of data files that must be read from the directory\n",
    "\n",
    "job_file1 = 'job-freqs-sd-sa.sh'\n",
    "job_str1  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH -e ./MPL/out/freq-sd-sa-error-%a\n",
    "#SBATCH -o ./MPL/out/freq-sd-sa-out-%a\n",
    "#SBATCH --array=1-{num_files - 1}\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "cd ./Archive-vector\n",
    "g++ src/main.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "pwd\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "echo \"$file\"\n",
    "python ../{freq_script} --data \\\"$file\\\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {SCRATCH} --timed 1 --delta_t 15\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_file2 = 'job-infer-sd-sa.sh'\n",
    "job_str2  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH -e ./MPL/out/inf-sa-error\n",
    "#SBATCH -o ./MPL/out/inf-sa-out\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}\n",
    "python {inf_script} --data {temp_dir} --timed 1 -o $outfile -q 5 --g1 10 --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file2), 'w')\n",
    "f.write('%s\\n' % job_str2)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "script_path1 = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {script_path1} {SSH_HOME}')\n",
    "script_path2 = os.path.join(SCRIPT_DIR, job_file2)\n",
    "print(f'scp {script_path2} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file1)\n",
    "print('sbatch %s &&' % job_file2)\n",
    "print('')\n",
    "\n",
    "# transfer files from cluster\n",
    "print('scp -r %s%s %s' % (SSH_HOME, out_dir, INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-12-07\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAEGCAYAAACQO2mwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAApqElEQVR4nO3deZxcdZnv8c+3qpcknRCyG5KQEGQR2WnZRGVUELwIeMV7wZUZNXccnes2dwTHEcVxm9GZ6yhXjMKIOoLIIpEBEQVcUJaEnbBF1gRIQkL29FJdz/3jnGqKTldXddLVdbr7+369+tXn/M45VU9XKvXUbz2KCMzMzPqTa3QAZmaWXU4SZmZWkZOEmZlV5CRhZmYVOUmYmVlFTY0OYChNnz49FixY0OgwzGwQHnkk+b3ffo2NY6xatmzZCxExo9LxuiUJSRcDpwBrIuLAfo7/H+DdZXG8CpgREeslPQlsBnqAQkS01/KcCxYsYOnSpUMRvpkNk+OPT37fcksjoxi7JD010PF6Njf9ADip0sGI+JeIODQiDgXOBX4bEevLTvmL9HhNCcLMzIZe3ZJERPwOWF/1xMRZwKX1isXMzHZOwzuuJU0gqXFcWVYcwK8kLZO0qMr1iyQtlbR07dq19QzVzGzMaXiSAN4G3Nqnqem4iDgcOBn4iKTXV7o4IhZHRHtEtM+YUbHvxczMdkIWksSZ9GlqiohV6e81wNXAkQ2Iy8xszGtokpA0GXgDcE1ZWZukSaVt4ETggcZEaGY2ttVzCOylwPHAdEkrgfOAZoCIuDA97e3AryJia9mls4CrJZXi+0lE/LJecZqZWWV1SxIRcVYN5/yAZKhsednjwCH1icrMsqarUGRLZwFoaXQo1o9RNePazEaeFWs2s6mjwItbJzGlzYkia7LQcW1mY1jptmf3rtzQyDCsAicJM2uoCS15AFas2dLgSKw/ThJm1lC5ZJAK3T2+lXIWOUmYWSYUw0kii5wkzKyh0ooEPUUniSxykjCzTCg4SWSSk4SZZULRSSKTnCTMrKFKXRE97pPIJCcJM8sE90lkk5OEmWWCk0Q2OUmYWUOVUoOTRDY5SZhZJjhJZJOThJk1ljuuM81JwswywUNgs8lJwswaKtKqhCfTZZOThJllgmsS2eQkYWaZ4D6JbHKSMLOGKqUGNzdlk5OEmTVWmhvc3JRNdUsSki6WtEbSAxWOHy9po6R70p/PlR07SdIjklZIOqdeMZpZdnieRDbVsybxA+CkKuf8PiIOTX/OB5CUBy4ATgYOAM6SdEAd4zSzBvKM62yrW5KIiN8B63fi0iOBFRHxeER0AZcBpw1pcGaWOe64zqZG90kcI+leSddLenVaNgd4puyclWlZvyQtkrRU0tK1a9fWM1YzqyPXJLKpkUniLmB+RBwCfAv4+c48SEQsjoj2iGifMWPGUMZnZsOhtCyHk0QmNSxJRMSmiNiSbl8HNEuaDqwC5pWdOjctM7NRyH0S2dawJCHpFVJyC3RJR6axrAPuBPaRtJekFuBMYEmj4jSzekuSQ9F9EpnUVK8HlnQpcDwwXdJK4DygGSAiLgTOAD4sqQBsB86MiAAKkj4K3ADkgYsj4sF6xWlm2eDJdNlUtyQREWdVOf5t4NsVjl0HXFePuMwsW0qpwZPpsqnRo5vMzAAPgc0qJwkza6w0NxR6nCSyyEnCzBqqt7nJNYlMcpIws0zwENhscpIws0xwksgmJwkza6jeyXRubsokJwkza6ze+0k0Ngzrn5OEmTVYkiW6e5wlsshJwswyYWtnodEhWD+cJMysoUo9EVu7eugs9DQ0FtuRk4SZZcaGbd2NDsH6cJIws8xYv7Wr0SFYH04SZtZQ5SNfX9jS2bhArF9OEmbWcE05AfDFa5c3OBLry0nCzBpufEuelqYc+Zw/krLG/yJmlglv3G8mPZ5RlzlOEmbWUEEgoCkv350ug5wkzKyx0rzQlJPvKZFBVW9fKukY4D3A64DZJPejfgD4L+DHEbGxrhGa2egnaMrnvBJsBg1Yk5B0PfBB4AbgJJIkcQDwWWAccI2kU+sdpJmNXklaEE05ef2mDKpWk3hvRLzQp2wLcFf68w1J0/u7UNLFwCnAmog4sJ/j7wY+DQjYDHw4Iu5Njz2ZlvUAhYhor/kvMrMRp9Qn4ZpE9gxYk+gnQSDpTZLeJqm50jmpH5DUPip5AnhDRBwEfBFY3Of4X0TEoU4QZmNDUy7nmkQGDarjWtI3gNcChwDXDHRuRPwOWD/A8T9GxIvp7m3A3MHEYmajQ6nu0JRzTSKLqvVJfEPS7mVFe5J86/9Suj1UPgBcX7YfwK8kLZO0qEqMiyQtlbR07dq1QxiSmQ2LAAnyedHtJJE51fokrgIuk3QdcAHwQ+Bmkk7r7w1FAJL+giRJHFdWfFxErJI0E7hR0sNpzWQHEbGYtKmqvb3d7zCzEao559FNWVStT+LWiDiJpNnoBkARcXxEHB0R39zVJ5d0MPB94LSIWFf2vKvS32uAq4Ejd/W5zCzb8mlzU/he15lSrbmpSdJ/A9YApwOHSFoi6ZBdfWJJe5LUVN4bEY+WlbdJmlTaBk4kmZdhZqNQkI5uShf586zrbKnW3PRz4E/ABODdEfF+SXsA50uKiPhQpQslXQocD0yXtBI4DyiNiLoQ+BwwDfh/kuCloa6zgKvTsibgJxHxy53+C80s45I00ZRPvrMWeoLmfGMjspdUSxLzI+IUSS0kI5CIiGeBD0o6dKALI+KsKsc/SDJRr2/54ySjp8xsLChblgOgUCwCzhJZUS1JLJb0p3T7X8sPRMQ9dYnIzMaUAJBoyidJwp3X2TJgkoiIbwHfGqZYzGyMKu+T6PYif5lSreP6s5KmDHD8jZJOGfqwzGysKfVJuCaRLdWam+4HrpXUQbJW01qSORL7AIcCvwa+XM8AzWxsyPfWJLw0R5ZUa266hmSl131IluOYDWwCfgwsiojt9Q/RzEaz0hDYZvdJZFLV+0kARMRjwGN1jsXMxqI0S5Tub13wLUwzxXemM7NMaPZkukyqqSZhZlYvpZRQ6pMY7C1MN27r5uJbn6C7p8iCaW2844i5vY8FsGLNFs648I+89+j5fOrE/YYq7DHDScLMGizSPolSc9PgksT3fv843755Re/+FctWcvlfHwNAR3cP37rpMTZs6+ba+55zktgJNSUJSTOADwELyq+JiL+qT1hmNtaUvv0/+OxGDp23e83X/fbRtRy9cCo/+eDRnHrBH1j29It0FYq0NOX4wi+Wc809zwKwratQj7BHvVr7JK4BJpMMef2vsh8zs11Susf11LYWAP7h6gdYtaH2gZMvbuti9uTx5HLiQ69bSE8xuOaeVXz/949z6R1Ps3B6G+88Yi5rNnfy0HObWL+1qy5/x2hVa3PThIj4dF0jMbOxKb3p0IFzJvPltx/EZ66+n6de2Mqc3cfXdPnG7d1MHt8MwOF7TkGC/3PFfQBMbWvhn04/kNWbO/jZspWc/M3fk8+Jv3rtAv7hvx1Qtz9pNKk1SVwr6a0RcV1dozGzMe24V04HqLkm0VMMNncU2C1NEvOmTuD2c9/Ef97+NJPHN/P+YxeQz4nOQg/jm5voLPTwscvu4YYHVztJ1KjWJPEx4DOSuoDutCwiYrf6hGVmY9Gsya0APLuho6bzN3ckH0elmgTAzN3G8YkT9n3Zea1NeU468BUA3Pb4em5cvnoowh0Tap1MN6negZjZ2FSacQ3Jh/kek8fx6OrNNV27cfuOSaKatpa8O7EHoebJdJJOlfT19MeL+plZXbxhvxn81/3P8cFL7qx67s4kiQmtTWzr6qE4yKG2Hd09gzp/tKgpSUj6KkmT0/L052OSvlLPwMxsjCivSgCLXr83AL95eE3VS1dv6gRg+sSWmp+urSW5odH2Gj70n1m/jWVPreeiPzzB/v/4S25/fF3NzzNa1Non8Vbg0IgoAki6BLgbOLdegZnZ2ND3+/xe09v4uxP35eu/epTOQg+tTZXvUvf0+m0AzJ/WVvPzTWhNPva2dhVoa638EXjnk+s5c/FtL1tw8KHnNnHUwmk1P9doMJgZ17sD69PtyUMfipmNTVFekQBgYumDvLP/JHHrihe48Ld/JiI5d8qEwfVJAGzr7IGy3tauQpF7V26gu1Dk0dWb+febVtCUE999zxE8t3E7/3jNg3Sly5hv6yqwpaPAlLaW3pnio1WtSeIrwN2SbiapGL4eOKduUZnZmFb6hr+lo9A7ya7cjctX8/vHXgCgff4UpL5ppvpjb+0qEBF86vJ7ueruVTuclxO875gFvPmAWUQE51+7nHVbu9jaWeDor/yGzR0FjnvldH78waN25k8cMWod3XSppFuA16RFn46I56tdJ+li4BRgTUQc2M9xAd8kac7aBpwdEXelx94PfDY99Z8i4pJaYjWzkaU047rcpHFpkujsfxRSa1Py7f3GT7yePWqcdFfS1pI89u8fe4GHntvMVXev4o37z+SgOZPZZ9ZEpk9sZeakVvaa3tabfKRkRviLW7t4+PnNbO4osPeMNv70+DrOvep+7l+1gXVbujjt0Dmcc/L+g4on6wZMEpL2j4iHJR2eFq1Mf+8haY/SB/oAfgB8G/hhheMnk9zlbh/gKOA7wFGSpgLnAe0k76FlkpZExIvV/iAzG3n61gN6axIVkkRHdw+Txzezz6zBj86fuVsyF+Or1z/cW/bF0w+sOsN7yoQWbn9ife8cjo+/eV/+9tK7ufa+Zzl47mQeWLWJ6x94bmwlCeCTwCLgG/0cC+CNA10cEb+TtGCAU04DfhgRAdwmaXdJs4HjgRsjYj2ApBuBk4BLq8RrZiNNn9FNUN4nUSlJFBnXvHN9AfvOmsRNn3oDL27r5ud3r2Jcc66mJUCO2XsaP1u6ki0dBY5ZOI1TDp7N6/eZwaRxTeRy4jNX38+vHqzawDLiVLt96aJ08+SIeNkUSEnjhuD55wDPlO2vTMsqlZvZKNPfbIVSkrh1xQv8xf4zdzjeUehhXHPlUU/VLJwxEYAj5k+p+Zrz3vZqznvbq19WNrmsw3xiaxNbO0ffXIpaU/EfaywbdpIWSVoqaenatWsbHY6Z7YS+zU0zd0u+g1525zP0FIO/v+JePvvz+3uPd3T39PZLZMWEljzbu3tG3T26B3yVJb1C0hHAeEmHSTo8/TkemDAEz78KmFe2Pzctq1S+g4hYHBHtEdE+Y8aMIQjJzBpt8vhmznvbAWzpLLD3Z67j8qUr+fFtT1NIh6AmzU07X5Ooh4llo6ZGk2qp+C3A10k+pL9R9vMJ4DND8PxLgPcpcTSwMSKeA24ATpQ0RdIU4MS0zMzGiNMPncOJB8zirCPnsXvarLOpI/kA7ujuYdwAk+waoa2ffpSO7h4u+kNya9WRqlqfxCXAJZLeERFXDvbBJV1K0gk9XdJKkhFLzeljXwhcRzL8dQXJENi/TI+tl/RFoLR4y/mlTmwzG12iQuvMlLYWFr+vHYDXLJjKJy+/l03bu5na1kJHochu47J19+X+ksR/3PokX/vlw4xrzvHuo+Y3KrRdUuurfISk30TEBoD02/2nIuKzA10UEWdVOR7ARyocuxi4uMb4zGzECqrNhSst4Fda0K+zu4dxk1rrHdigTGxNajZbyjqvSwnjrqc2jNgkUWvPz8mlBAGQzld4a10iMjPrY4ckUchen0Rpkt4Vy14amFmK98q7VrJ6U233yMiaWpNEXlJv2pY0HshWGjezUat057lN6U2Gkj6JbI1uOnDOZCa2NrH0yWTO75pNHfzotqd6j9/+xMhsMa/1Vf5P4DeSPiDpA8CNgJfJMLMhMnB7U6kmcd41D/KGf7mZ1Zs6aN3JyXT10tbaxFlHzuOJF7bSUwzuenoDAOecvD8TW5u4ctlKbn54zYgbIlvr2k1fk3Qv8Oa06IsR4dFGZrbLavnInDmplQ8fvzfPpfe+PnzPKZxxxLwqVw2/vWdMpLNQ5Liv3cRBc5LFst979Hw2bu/mO7f8md8+upYTDpjFMQuncfaxC8jlal+YsFEGMzzgIaAQEb+WNEHSpIio7R6DZmYDqPZRKYlPn5T9NZFOOGAWy5/bxBXLVvKr5avZc+oE2lqb+LsT9+Mdh8/hP259kv+8/WluXL6aY/aexqtm79bokKuq9c50HwKuAL6bFs0Bfl6nmMzMRqRpE1s5/7QDOfvYBQC85+g9AcjnxCtnTuJLbz+Iq/7mWACe3zgyOrJrrUl8BDgSuB0gIh6TtOOCKmZmgzWymuhr8skT9uWd7fOYP3XHhSlmT06WHHluhCSJWnt+OiOiq7QjqYlR+U9rZsMtoHp70wjTlM+x1/S2fvscZkxMBoZedufTFZdCz5Jak8RvJX2GZA2nE4CfAb+oX1hmNpaMshwxoKZ8joXT27hv5UYW//bPjQ6nqlqTxDnAWuB+4H+RLKcx4GxrMzPr35UfPpbdxjVx36qNjQ6lqpqSREQUI+J7EfHOiDgj3XZzk5ntkrH6MTKlrYU3HzCLPzz2Aif862/589otjQ6pomq3L72fAfoeIuLgIY/IzMaMMZojAPjAcXvRks9x+dJn+PZNKzjryD2ZM2U8v3t0LY88v5mHn9/EgXtM5pi9pzF9Yiv7vWJSQ5YiqTa66ZRhicLMxqQxnCN49R6T+eo7Dubp9du4+u5VXH33KnYb18SmjgL5nJjQnOe2x9fz/T88AcCb9p/JRWe/ZtjjrLZUeO/CI5LmA/ukk+nGV7vWzKyaUnPTWOq47ut772vn4ec38Y7v/IlNHQX+/qT9+KvX7sW45jzPrN/G2i2dfOEXy7nl0bWc8Z0/8uK2LiJgQmueQk/w12/Ym9MPq9/dnWv6oE8n0y0CpgJ7k9yE6ELgTXWLzMxGvbFckyhpa23iiPlTufjsdiLgjfvPROna6fOmTmDe1Al8+6zD+MdrHqC7p8jCGRNpbcqxvauH3zy8hqvuXtX4JIEn05lZHfT2SYzlqkTqjfvPqnhs3tQJ/OAvj9yh/Nyr7uPSO57hrMW3cemio+sSV61JojMiukrZzZPpzGwoBM4Su+JdR85nw7bu3rvi1UOtj9x3Mt3f4Ml0ZraLxvLopqFw0NzJfOc9R9T1OTyZzswazvWI7Kr1fhJF4HuSLgFeDazyZDoz21X+FMm+AWsSki6U9Op0ezJwD/BD4G5JZ1V7cEknSXpE0gpJ5/Rz/N8k3ZP+PCppQ9mxnrJjSwb5d5nZCBDu2sy8ajWJ10XEX6fbfwk8GhGnS3oFcD1waaULJeWBC4ATgJXAnZKWRMTy0jkR8Ymy8/8WOKzsIbZHxKGD+WPMbGRxTSL7qvVJdJVtn0B6o6GIeL6Gxz4SWBERj6fLjF8GnDbA+WcxQNIxs9Gnd2yTOyUyq1qS2CDpFEmHAa8Ffgm9Q2DHV7l2DvBM2f7KtGwH6WzuvYCbyorHSVoq6TZJp1d5LjMbgdy1mX3Vmpv+F/DvwCuAj5fVIN4E/NcQxnEmcEVE9JSVzY+IVZIWAjdJuj8idlh8XdIiktng7LnnnkMYkpnVm1NE9lVbu+lR4KR+ym8Abqjy2KuAeWX7c9Oy/pxJMqu7/DlWpb8fl3QLSX/FDkkiIhYDiwHa29v9njMbQVyRyL5a50nsjDuBfSTtJamFJBHsMEpJ0v7AFOBPZWVTJLWm29NJmrqW973WzEa4NEm4SyK76jaXOyIKkj5KUuPIAxdHxIOSzgeWRkQpYZwJXNZn3sWrgO9KKpIksq+Wj4oys9HBQ2Czr67LfUfEdSSzs8vLPtdn//P9XPdH4KB6xmZmjVd0jsi8anem++RAxyPiX4c2HDMbS3obEDwGNrOq1SQmpb/3A17DS30KbwPuqFdQZjY2eA3Y7Ks2uukLAJJ+BxweEZvT/c8ztENgzWwM8uim7Kt1dNMsXj77uistMzPbae64zr5aO65/CNwh6ep0/3TgkrpEZGZjh3NE5tW6VPiXJF0PvC4t+suIuLt+YZnZWOAckX2DmUw3AdgUEd8EVkraq04xmdkYEZ5Ml3k1JQlJ5wGfBs5Ni5qBH9crKDMbG8JTrjOv1prE24FTga0AEfEsLw2PNTPbKR7dlH21JomudNmMAJDUVr+QzGyscI7IvlqTxOWSvgvsLulDwK+B79UvLDMbC0ozrt3alF1VRzdJEvBTYH9gE8ns689FxI11js3MRjk3N2Vf1SQRESHpuog4CHBiMDMbQ2ptbrpL0mvqGomZjTkv1STc4JRVtc64Pgp4t6SnSEY4iaSScXDdIjOzUc/LcmRfrX0Si4Cn6h+OmY0lXik8+2rtk7gg7ZMwMxsyrkdkn/skzKxhwsObMm8wfRLvkfQk7pMwsyHiFJF9tSaJt9Q1CjMbk1yRyL6ampsi4ilgHvDGdHtbrdeamVXmGddZV9dVYCWdJOkRSSskndPP8bMlrZV0T/rzwbJj75f0WPrz/tr+HDMbSVyTyL5am5veDhwG3AXJKrCSBlwFVlIeuAA4AVgJ3ClpSUQs73PqTyPio32unQqcB7STfNVYll77Yo3xmtkI4Ll02VfPVWCPBFZExOMR0QVcBpxW4/O9BbgxItanieFG4KQarzWzEcI1ieyr5yqwc4BnyvZXpmV9vUPSfZKukDRvkNciaZGkpZKWrl27tpa/xcwyItwnkXkDJglJrQAR8XXgCuBKXloF9ltD8Py/ABakQ2lvBC4Z7ANExOKIaI+I9hkzZgxBSGY2XFyTyL5qfRJ/Ag6X9KOIeC+DWwV2FcmIqJK5aVmviFhXtvt94J/Lrj2+z7W3DOK5zWwEcJLIvmpJokXSu4BjJf33vgcj4qoBrr0T2EfSXiQf+mcC7yo/QdLsiHgu3T0VeCjdvgH4sqQp6f6JvDSyysxGiZcW+HODU1ZVSxJ/Dbwb2B14W59jAVRMEhFRkPRRkg/8PHBxRDwo6XxgaUQsAf63pFOBArAeODu9dr2kL5IkGoDzI2L9YP4wM8s+1ySyb8AkERF/AP4gaWlEXDTYB4+I64Dr+pR9rmz7XCrUECLiYuDiwT6nmY08rkdkV03zJCLiIknHAgvKr4mIH9YpLjMbA8KtTZlXU5KQ9CNgb+AeoCctDsBJwsx2mm86lH21zrhuBw4Ir+trZkPInyjZV+tkugeAV9QzEDMbe9zalH211iSmA8sl3QF0lgoj4tS6RGVmY4IbJ7Kv1iTx+XoGYWZjk1NE9tU6uum39Q7EzMYeVySyb8AkIWkz/Sf70u1Ld6tLVGY2RqQfL+6UyKxqk+kGvGeEmdmuCOeIzPMtSM2sYdzalH1OEmbWMOFb02Wek4SZNYyHwGafk4SZNYwn02Wfk4SZNYwrEtnnJGFmDeMF/rLPScLMGsftTZnnJGFmDeN6RPY5SZhZw3gyXfY5SZhZw7hPIvucJMysYTy6KfvqmiQknSTpEUkrJJ3Tz/FPSlou6T5Jv5E0v+xYj6R70p8l9YzTzBrDOSL7ar2fxKBJygMXACcAK4E7JS2JiOVlp90NtEfENkkfBv4Z+J/pse0RcWi94jOzxvOM6+yrZ03iSGBFRDweEV3AZcBp5SdExM0RsS3dvQ2YW8d4zCxjekfAuuc6s+qZJOYAz5Ttr0zLKvkAcH3Z/jhJSyXdJun0ShdJWpSet3Tt2rW7FLCZDTNXJDKvbs1NgyHpPUA78Iay4vkRsUrSQuAmSfdHxJ/7XhsRi4HFAO3t7X7LmY0g4dl0mVfPmsQqYF7Z/ty07GUkvRn4B+DUiOgslUfEqvT348AtwGF1jNXMGsDzJLKvnkniTmAfSXtJagHOBF42SknSYcB3SRLEmrLyKZJa0+3pwGuB8g5vMxsF3G+dfXVrboqIgqSPAjcAeeDiiHhQ0vnA0ohYAvwLMBH4mZKeq6cj4lTgVcB3JRVJEtlX+4yKMrNRwDki++raJxER1wHX9Sn7XNn2mytc90fgoHrGZmaN5yGw2ecZ12bWML57afY5SZhZw7jjOvucJMysgdzclHVOEmbWMO6SyD4nCTNrGOeI7HOSMLOGcZ9E9jlJmFnD+KZD2eckYWYNE166KfOcJMysYV6qRzhLZJWThJk1jGdcZ5+ThJk1jDuus89Jwswaxh3X2eckYWYN49am7HOSMLOGcZLIPicJM2sYj4DNPicJM2uYcM915jlJmFnDuLUp+5wkzKxxnCUyz0nCrE4uuHkFR3/5N54wNgAPgc0+JwkbNW5cvpqjvvxrPvTDpY0OBYB/ueERnt/UwcoXtzc6lMxyl0T21TVJSDpJ0iOSVkg6p5/jrZJ+mh6/XdKCsmPnpuWPSHpLPeO00eGmh1ezelMnNy5fzYZtXY0Oh9am5L/XLY+saWgc3T1FunuKFIsDf2vv7ilS6Cnu8vNt2NbFV657iH+6djmX3/kM67dW/rdwPSL7mur1wJLywAXACcBK4E5JSyJiedlpHwBejIhXSjoT+BrwPyUdAJwJvBrYA/i1pH0joqde8Y5WHd09rN7U8bKyNZs7eWrdNrZ399DR1cPmjm76fn7sP3sSB8/ZveLjFiNYtWE7XYXkQyUIIpJvhqWHigham/PMnjxuWL4pPvjspt7t9150BxNa8kksURZfGtd+r5jE2cfuRT4ncoKcRE5CglxOKH28+1ZuoCmXoykvWvI5Wppy7Da+icnjm5nY2gwkH4r7zppEPvfyv3LSuGY6t3Ty+V8s54d/eopxzXkmtOSZ2NpEa3MOIRAsmDaB1+49PY0NeiJ4at1WOruLvLC1k/ue2UhbaxP5HORzIp/L0ZwTE1rzTGxtpjkvmvM5nlq3jXwOWpqS7e6eItu7i9y/cgPFSK49du9pTB7f3Bvjms2dPPHCVsY151i3pYumnFg4YyITWpJYW5vzbO0sML45D0ChGBSLwZbOAps7CjQ35RCg9E8X8NzGDtZs7qQ5Lzq6i3Al7DauiYUzJnLqIXtwxPwpva/1sxu2A224LpFdqld7qaRjgM9HxFvS/XMBIuIrZefckJ7zJ0lNwPPADOCc8nPLzxvoOdvb22Pp0sE3Nbzl335HR+Hl+afvW1bSgMdfVhDJf5SeYpB+DvT+p1D6WEqv6d1PP6he+g9Xfj4o3c/nkg+rWq3Z3MmWzsKA55Seu6SnyjfOLPsf7XPZsK2bDdu6y17fl14/CdZt6eLh5zcPSzzvOmpPAF7c2kVHdw9bu3rY0lGgq6dIRNDRXWTVhoGbo+ZPm8D45jwRSXIuFINCsciWjgJbO3voLhaJgGltLTTlxeaOAvOntTGptYlcDvadNYmZk1p5/IWt3PPMhpc9dnMux0FzJ7O9u4e8koS5bmsX27t62NbVQ0ehh7aWJrZ39yBKSUq0NuWY2tZCV0/yXin/HMnnxKLXLeTohdO47fF13PLoWpY+uZ4Hnt3U+6Wi3PM/OZr2BVP4w+/c+t0IkpZFRHul43WrSQBzgGfK9lcCR1U6JyIKkjYC09Ly2/pcO6e/J5G0CFgEsOeee+5UoIfMm/yyN2/fj8i+eXTH47HD8dftM53xLU1Jx1z6DbZYjN5vi6VvtqXrS+XFsm1e9u08KEbyGF09xR2SViWH5XO0L5jS2/QByTfNA/eYzITWPOOb87S1NJEr+xbcVSjy64dWs71r4IrbzN1amdja1BtL3w9kgI3bu3lxmJp+RPJNeUpby4DnRQS3P7GeF7Z0UoxkvxhBsZi+/um/Q0tTjpMPnE1LUy5piikGXYUim7Z3s2F7N1s6ChSKRTZu7+73wy+fE2961ayXfXPvz/JnN7Gpozv5MpHWYmZOGsfUiS1EBJPGDXw9wKaObiaV/VtkxbGvnM6xr5wOQKGnyB1PrGd7d8/L3uufuWU3mnJOEFlVzyQxLCJiMbAYkprEzjzGP59xyJDGNNK1NOV460GzGx1G3Uji6IXTBnVNPpc0t9AKU6skocE6YI/ddvkxdqshkTRaUz7XmzDKfWVcA4KxmtUzfa8C5pXtz03L+j0nbW6aDKyr8VozM6uzeiaJO4F9JO0lqYWkI3pJn3OWAO9Pt88Aboqk7WYJcGY6+mkvYB/gjjrGamZm/ahbc1Pax/BR4AYgD1wcEQ9KOh9YGhFLgIuAH0laAawnSSSk510OLAcKwEc8ssnMbPjVtU8iIq4DrutT9rmy7Q7gnRWu/RLwpXrGZ2ZmA/OQAjMzq8hJwszMKnKSMDOzipwkzMysoroty9EIktYCTw3ikunAC3UKp15GWswjLV5wzMPFMQ+PajHPj4gZlQ6OqiQxWJKWDrRmSRaNtJhHWrzgmIeLYx4euxqzm5vMzKwiJwkzM6torCeJxY0OYCeMtJhHWrzgmIeLYx4euxTzmO6TMDOzgY31moSZmQ3AScLMzCoatUlC0jxJN0taLulBSR9Ly6dKulHSY+nvKWm5JP27pBWS7pN0eANiHifpDkn3pjF/IS3fS9LtaWw/TZdeJ11K/adp+e2SFgx3zGkceUl3S7p2JMSbxvKkpPsl3SNpaVqW2fdGGsfukq6Q9LCkhyQdk9WYJe2Xvraln02SPp7VeMvi/kT6f+8BSZem/ycz/X6W9LE03gclfTwtG7rXOSJG5Q8wGzg83Z4EPAocAPwzcE5afg7wtXT7rcD1JHfhPBq4vQExC5iYbjcDt6exXA6cmZZfCHw43f4b4MJ0+0zgpw16rT8J/AS4Nt3PdLzp8z8JTO9Tltn3RhrHJcAH0+0WYPesx5zGkie5f/38LMdLcovkJ4Dx6f7lwNlZfj8DBwIPABNIVvX+NfDKoXydG/KmacQPcA1wAvAIMDstmw08km5/Fzir7Pze8xoU7wTgLpL7gr8ANKXlxwA3pNs3AMek203peRrmOOcCvwHeCFybvvkyG29Z3E+yY5LI7HuD5K6NT/R9vbIcc9lznwjcmvV4SZLEM8DU9P15LfCWLL+fSW61cFHZ/j8Cfz+Ur/OobW4ql1YDDyP5Zj4rIp5LDz0PzEq3S2+QkpVp2bBKm27uAdYANwJ/BjZERKGfuHpjTo9vBAZ38+Zd939J3pTFdH8a2Y63JIBfSVomaVFaluX3xl7AWuA/0qa970tqI9sxl5wJXJpuZzbeiFgFfB14GniO5P25jGy/nx8AXidpmqQJJDWFeQzh6zzqk4SkicCVwMcjYlP5sUhSaabGAEdET0QcSvIN/Uhg/8ZGVJmkU4A1EbGs0bHshOMi4nDgZOAjkl5ffjCD740m4HDgOxFxGLCVpBmhVwZjJm2/PxX4Wd9jWYs3bbc/jSQh7wG0ASc1NKgqIuIh4GvAr4BfAvcAPX3O2aXXeVQnCUnNJAniPyPiqrR4taTZ6fHZJN/YAVaRZOCSuWlZQ0TEBuBmkurt7pJKdxEsj6s35vT4ZGDdMIb5WuBUSU8Cl5E0OX0zw/H2Sr81EhFrgKtJEnKW3xsrgZURcXu6fwVJ0shyzJAk4bsiYnW6n+V43ww8ERFrI6IbuIrkPZ7p93NEXBQRR0TE64EXSfpfh+x1HrVJQpJI7qH9UET8a9mhJcD70+33k/RVlMrfl/b+Hw1sLKuuDQtJMyTtnm6PJ+lDeYgkWZxRIebS33IGcFP6rWFYRMS5ETE3IhaQNCncFBHvzmq8JZLaJE0qbZO0mT9Aht8bEfE88Iyk/dKiN5HcAz6zMafO4qWmplJcWY33aeBoSRPSz4/Sa5z19/PM9PeewH8nGUQydK/zcHayDHOHznEkVaz7SKpg95C0100j6Wh9jGQkwNT0fAEXkPQB3A+0NyDmg4G705gfAD6Xli8E7gBWkFTbW9Pycen+ivT4wga+3sfz0uimTMebxndv+vMg8A9peWbfG2kchwJL0/fHz4EpWY6ZpLlmHTC5rCyz8aZxfAF4OP3/9yOgdQS8n39PkszuBd401K+zl+UwM7OKRm1zk5mZ7TonCTMzq8hJwszMKnKSMDOzipwkzMysIicJs50kqSdd4fRBJSv3fkrSgP+nJC2Q9K7hitFsVzlJmO287RFxaES8mmTi48nAeVWuWQA4SdiI4XkSZjtJ0paImFi2vxC4E5hOsiz2j0gmlAF8NCL+KOk24FUkK7peAvw78FWSyYitwAUR8d1h+yPMqnCSMNtJfZNEWrYB2A/YDBQjokPSPsClEdEu6Xjg7yLilPT8RcDMiPgnSa3ArcA7I+KJYfxTzCpqqn6Kme2EZuDbkg4lWZVz3wrnnQgcLKm0NtBkYB+SmoZZwzlJmA2RtLmph2TFzfOA1cAhJH1/HZUuA/42Im4YliDNBskd12ZDQNIMkltbfjuSNtzJwHMRUQTeS3ILT0iaoSaVXXoD8OF0WXsk7ZuuTGuWCa5JmO288eldBJuBAklHdWlZ+v8HXCnpfSQ3g9malt8H9Ei6F/gByf03FgB3pctTrwVOH57wzapzx7WZmVXk5iYzM6vIScLMzCpykjAzs4qcJMzMrCInCTMzq8hJwszMKnKSMDOziv4/8+pN5Goa9PAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = np.load(os.path.join(SARS_DIR, DATA_DATE, f'infer-south-africa-sd', 'inf.npz'), allow_pickle=True)\n",
    "selection = data['selection']\n",
    "sites     = data['allele_number']\n",
    "times     = data['times']\n",
    "    \n",
    "f = open(os.path.join(INF_DIR, f'infer-{DATA_DATE}-tv-sd-south-africa.npz'), mode='wb')\n",
    "np.savez_compressed(f, selection=selection, sites=sites, times=times)\n",
    "f.close()\n",
    "    \n",
    "omicron = ['NSP3-38-1-G', 'NSP3-889-2-G', 'NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--', \n",
    "           'NSP3-1892-0-A', 'NSP5-132-1-A', 'NSP6-189-0-G', 'NSP10-57-2-C', 'NSP14-42-0-G', \n",
    "           'S-211-1--', 'S-211-2--', 'S-212-0--', 'S-446-0-A', 'S-547-1-A', \n",
    "           'S-856-2-A', 'S-954-2-T', 'S-969-2-A', 'S-981-0-T', 'ORF3a-64-2-T', \n",
    "           'E-9-1-T', 'M-19-0-G', 'M-63-0-A', 'ORF6-20-0-C', 'ORF7b-18-0-T', \n",
    "           'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "           'N-32-1--', 'N-32-2--', 'N-33-0--', 'S-484-1-C', 'S-375-1-T', 'S-493-1-G', \n",
    "           'S-496-0-A', 'S-498-1-G', 'S-505-0-C', 'S-547-1-A']\n",
    "\n",
    "f = open(os.path.join(DATA_DIR, 'omicron-sites-exclusive.npy'), mode='wb')\n",
    "np.save(f, omicron)\n",
    "f.close()\n",
    "\n",
    "alleles    = [dp.get_label_new(i) for i in sites]\n",
    "idxs_omicron = [alleles.index(i) for i in omicron]\n",
    "s_omicron    = np.sum(np.array(selection)[:, idxs_omicron], axis=1)\n",
    "x_detect   = np.array(times)[s_omicron>0.13183][0]\n",
    "\n",
    "out_file = os.path.join(DATA_DIR, 'omicron-subdate-early-detection.csv')\n",
    "f = open(out_file, mode='w')\n",
    "f.write('time,selection\\n')\n",
    "for i in range(len(times)):\n",
    "    f.write(f'{times[i]},{s_omicron[i]:.5f}\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='statistics'></a>\n",
    "# Calculating Statistics  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Error in inferred coefficients from bootstrapping sequences and from subsampling regions\n",
    "\n",
    "data = pd.read_csv(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-subsample-corrected.csv'))\n",
    "mean_s  = data['selection coefficient']\n",
    "std_dev = data['standard deviation']\n",
    "print('mean error from subsampling 80% of the regions analyzed:', np.mean(np.absolute(std_dev)))\n",
    "\n",
    "data = pd.read_csv(os.path.join(SARS_DIR, DATA_DATE, f'selection-nonsyn-bootstrap.csv'))\n",
    "mean_s  = data['selection coefficient']\n",
    "std_dev = data['standard deviation']\n",
    "print('error from subsampling 20 sequences per time point in every region, with replacement:', np.mean(np.absolute(std_dev)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding the number of sites present in any region after filtering\n",
    "counts = 0\n",
    "for file in os.listdir(os.path.join(DATA_DIR, 'genome-filtered')):\n",
    "    if 'sites' in file:\n",
    "        continue\n",
    "    df = pd.read_csv(os.path.join(DATA_DIR, 'genome-filtered', file))\n",
    "    counts += len(df)\n",
    "print(f'total number of sequences used = {counts}')\n",
    "\n",
    "lengths = []\n",
    "for file in os.listdir(os.path.join(DATA_DIR, 'genome-filtered')):\n",
    "    if 'sites' not in file:\n",
    "        continue\n",
    "    lengths.append(len(pd.read_csv(os.path.join(DATA_DIR, 'genome-filtered', file))))\n",
    "L = np.amax(lengths)\n",
    "print(f'longest sequence in any region is {L}, which is approximately {L / 29900} percent of sites')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/count-ambiguous.py blee098@secure.hpcc.ucr.edu: &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-count-ambiguous.sh blee098@secure.hpcc.ucr.edu:\n"
     ]
    }
   ],
   "source": [
    "# Counting the number of ambiguous nucleotides in the sequences used for analysis\n",
    "regions_dir   = ppath.join(SSH_DATA, f'regions-times-{DATA_DATE}')\n",
    "script        = 'count-ambiguous.py'\n",
    "out_file      = ppath.join(SSH_DATA, 'number-ambiguous')\n",
    "msa_file      = ppath.join(SSH_DATA, MERGED_FINAL)\n",
    "\n",
    "job_file1     = 'job-count-ambiguous.sh'\n",
    "job_str1      = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=700G\n",
    "#SBATCH --time=18:00:00\n",
    "#SBATCH -e ./MPL/out/count-ambiguous-error\n",
    "#SBATCH -o ./MPL/out/count-ambiguous-out\n",
    "#SBATCH -p highmem\n",
    "\n",
    "module unload miniconda2 \n",
    "module load miniconda3\n",
    "\n",
    "conda activate sars-env\n",
    "\n",
    "python {script} --regions {regions_dir} -o {out_file} --input_file {msa_file}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file1), 'w')\n",
    "f.write('%s\\n' % job_str1)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file1)\n",
    "print(f'scp {job_path} {SSH_HOME}')\n",
    "\n",
    "print(f'sbatch {job_file1}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['number_ambiguous', 'number_ns', 'number_sequences', 'ns_sequence']\n",
      "There are 1346553 sequences in the analysis\n",
      "There are 200899091 ambiguous nucleotides in the used sequences\n",
      "There are 200476507 Ns in the used sequences\n",
      "There are a total of 39789294597 sites in all of the used data\n",
      "Percentage of sites that are ambiguous is 0.005049073954056658\n",
      "\n",
      "number of sequences with more than 1000 Ns is 48097\n",
      "number of sequences with more than 300 Ns is 215780\n",
      "number of sequences with more than 100 Ns is 398532\n",
      "number of sequences with more than 0 Ns is 600715\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlYAAAI/CAYAAAC1XpeNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYJUlEQVR4nO3df6zd933X8dd7MUnXDk5aEsaIE5ySKMJ/scpKi4ZQtEHrzHWDpgklVKKFgEVRJGBIzKMSaOKfDBC/IyJrDR1oSxZCgLgJCgU2ZRJdSTJolzTLalJvcdSSlHSXH0OkoR/+uN9k15adXPe+z/0e3/N4SFc+53uOnc/95Ot7nv5+vud7aowRAAB27jvmHgAAwF4hrAAAmggrAIAmwgoAoImwAgBoIqwAAJrsm3sASXLVVVeNAwcOzD0MAIC39fTTT399jHH1+R5bibA6cOBAnnrqqbmHAQDwtqrq1y70mKVAAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbtV16vqu9I8jeS/I4kT40xfqr7vwEAsIq2dcSqqu6rqper6plzth+uquer6lRVHZ8235Zkf5JvJjnTO1wAgNW13aXATyc5vHVDVV2W5J4ktyY5mOSOqjqY5KYk/3GM8SNJPtE3VACA1batsBpjPJHk1XM235zk1BjjhTHGa0keyObRqjNJvjE95/91DRQAYNXt5OT1a5K8uOX+mWnbw0k+VFX/MMkTF/rNVXWsqp6qqqdeeeWVHQwDAGA1tJ+8Psb4zSR3buN5J5KcSJJDhw6N7nEAAOy2nRyxeinJtVvu75+2AQCspZ2E1ZNJbqyq66vq8iS3J3mkZ1gAAJee7V5u4f4kn0tyU1Wdqao7xxivJ7kryeNJnkvy4Bjj2eUNFQBgtW3rHKsxxh0X2P5YksdaRwQAcInykTYAAE2EFQBAk1nDqqqOVtWJjY2NOYcBANBi1rAaY5wcYxxbLBZzDgMAoEX7BUJX1YHjj277uafvPrLEkQAAe5VzrAAAmggrAIAmwgoAoImwAgBoIqwAAJq4jhUAQBPXsQIAaGIpEACgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJq68DgDQxJXXAQCaWAoEAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaOIjbQAAmvhIGwCAJpYCAQCaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJD2EGAGjiQ5gBAJpYCgQAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCazhlVVHa2qExsbG3MOAwCgxaxhNcY4OcY4tlgs5hwGAEALS4EAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQJNZw6qqjlbViY2NjTmHAQDQYtawGmOcHGMcWywWcw4DAKCFpUAAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmswaVlV1tKpObGxszDkMAIAWs4bVGOPkGOPYYrGYcxgAAC0sBQIANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0KQ9rKrqlqr6haq6t6pu6f7zAQBW1bbCqqruq6qXq+qZc7Yfrqrnq+pUVR2fNo8k/yvJO5Kc6R0uAMDq2u4Rq08nObx1Q1VdluSeJLcmOZjkjqo6mOQXxhi3JvnRJD/eN1QAgNW2rbAaYzyR5NVzNt+c5NQY44UxxmtJHkhy2xjjW9Pj30hyRdtIAQBW3L4d/N5rkry45f6ZJO+vqh9K8qEkVyb5Rxf6zVV1LMmxJLnuuut2MAwAgNWwk7A6rzHGw0ke3sbzTiQ5kSSHDh0a3eMAANhtO3lX4EtJrt1yf/+0DQBgLe0krJ5McmNVXV9Vlye5PckjPcMCALj0bPdyC/cn+VySm6rqTFXdOcZ4PcldSR5P8lySB8cYzy5vqAAAq21b51iNMe64wPbHkjzWOiIAgEuUj7QBAGgya1hV1dGqOrGxsTHnMAAAWswaVmOMk2OMY4vFYs5hAAC0sBQIANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATVwgFACgiQuEAgA0sRQIANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATVwgFACgiQuEAgA0sRQIANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA08VmBAABNfFYgAEATS4EAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0MSHMAMANPEhzAAATSwFAgA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE1mDauqOlpVJzY2NuYcBgBAi1nDaoxxcoxxbLFYzDkMAIAWlgIBAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCb75h4Ae9OB449u63mn7z6y5JEAwO4RVjsgHgCArSwFAgA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNXG6Bi7LdS0wAwDpyxAoAoImwAgBoIqwAAJo4x2oPcz4UAOyuWY9YVdXRqjqxsbEx5zAAAFrMGlZjjJNjjGOLxWLOYQAAtHCOFQBAE2EFANDEyevn4aRvAODb4YgVAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE32zT0ALt6B44/OPQQA4DwcsQIAaCKsAACaCCsAgCbCCgCgibACAGjiXYG7YLvv4jt995EljwQAWCZHrAAAmswaVlV1tKpObGxszDkMAIAWs4bVGOPkGOPYYrGYcxgAAC0sBQIANBFWAABNhBUAQBNhBQDQxHWsVsh2r3cFAKwmR6wAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJvvmHgDr7cDxR7f1vNN3H1nySABg5xyxAgBoIqwAAJpYCoQ1YdkVYPkcsQIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBospSwqqp3VdVTVfXhZfz5AACraFthVVX3VdXLVfXMOdsPV9XzVXWqqo5veehHkzzYOVAAgFW33SNWn05yeOuGqrosyT1Jbk1yMMkdVXWwqv5oki8leblxnAAAK29bH8I8xniiqg6cs/nmJKfGGC8kSVU9kOS2JN+V5F3ZjK3/U1WPjTG+1TdkAIDVtK2wuoBrkry45f6ZJO8fY9yVJFX18SRfv1BUVdWxJMeS5LrrrtvBMFgHB44/uq3nnb77yJJHAgAXtrR3BY4xPj3G+MxbPH5ijHFojHHo6quvXtYwAAB2zU7C6qUk1265v3/aBgCwlnayFPhkkhur6vpsBtXtSf5Ey6hgyeZcWrSsCbB3bSusqur+JLckuaqqziT562OMT1XVXUkeT3JZkvvGGM8ubaQwg+1GUNIfQgIM4NKz3XcF3nGB7Y8leax1RAAAl6idLAXCyrmYI0wA0E1YQRNRB8CsH8JcVUer6sTGxsacwwAAaDFrWI0xTo4xji0WizmHAQDQYtawAgDYS4QVAEATYQUA0ERYAQA0cbkFuMS5zAPA6nDECgCgietYAQA0cR0rAIAmlgIBAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGjiyusAAE1m/RDmMcbJJCcPHTr0Z+ccBwC742I+NPz03UeWOBJYDkuBAABNhBUAQBNhBQDQRFgBADSZ9eR1gHW33ZO5V/1E7os5KR32MkesAACaCCsAgCbCCgCgibACAGjiI20AAJr4SBvgLHvlXWoAc7AUCADQxHWsgG/LOn6YrqN5wNtxxAoAoImwAgBoIqwAAJoIKwCAJk5eB/YsJ5sDu01YAUsncIB1YSkQAKCJI1bA2ruYa3IBvBVHrAAAmsx6xKqqjiY5esMNN8w5DGBFOBcLuNT5EGbgkrPqS3erPj5geSwFAgA0EVYAAE2EFQBAE2EFANDEdawAuCAn4sPFccQKAKCJsAIAaCKsAACaOMcKYA9x9XqYlyNWAABNhBUAQBNLgQBryGUUYDkcsQIAaCKsAACazBpWVXW0qk5sbGzMOQwAgBazhtUY4+QY49hisZhzGAAALZy8DnAJcLI5XBqcYwUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANHG5BQBW0nYvMXH67iNLHglsn7AC4JImwFgllgIBAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAms36kTVUdTXL0hhtumHMYAKwBH33Dbpj1iNUY4+QY49hisZhzGAAALSwFAgA0EVYAAE2EFQBAE2EFANBEWAEANJn1cgsAsGpcloGdcMQKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoIkrrwPAmnO1+T6OWAEANBFWAABNLAUCwCVmu0t37D5hBQDfBuclcT7CCgCW6GKOLq16hInJt+ccKwCAJsIKAKCJsAIAaCKsAACaOHkdAFaEyyhc+hyxAgBoMusRq6o6muToDTfcMOcwAIBG63xZhlmPWI0xTo4xji0WizmHAQDQwlIgAEATYQUA0ERYAQA0EVYAAE2EFQBAE2EFANBEWAEANBFWAABNhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0GTf3AMAANbTgeOPbvu5p+8+ssSR9HHECgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoImwAgBoIqwAAJoIKwCAJsIKAKCJsAIAaCKsAACaCCsAgCbCCgCgibACAGgirAAAmggrAIAmwgoAoEl7WFXV76+qe6vqoar6RPefDwCwqrYVVlV1X1W9XFXPnLP9cFU9X1Wnqup4kowxnhtj/LkkfzzJ9/UPGQBgNW33iNWnkxzeuqGqLktyT5JbkxxMckdVHZwe+0iSR5M81jZSAIAVt62wGmM8keTVczbfnOTUGOOFMcZrSR5Ictv0/EfGGLcm+WjnYAEAVtm+Hfzea5K8uOX+mSTvr6pbkvxQkivyFkesqupYkmNJct111+1gGAAAq2EnYXVeY4yfT/Lz23jeiSQnkuTQoUOjexwAALttJ+8KfCnJtVvu75+2AQCspZ2E1ZNJbqyq66vq8iS3J3mkZ1gAAJee7V5u4f4kn0tyU1Wdqao7xxivJ7kryeNJnkvy4Bjj2eUNFQBgtW3rHKsxxh0X2P5YXFIBAFiyA8cf3dbzTt99ZMkjeWs+0gYAoImwAgBoMmtYVdXRqjqxsbEx5zAAAFrMGlZjjJNjjGOLxWLOYQAAtLAUCADQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE1cIBQAoIkLhAIANLEUCADQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0EVYAAE1cIBQAoIkLhAIANLEUCADQRFgBADSpMcbcY0hVvZLk15b8n7kqydeX/N+4lJiPs5mPs5mPs5mPs5mPs5mPs63DfPzeMcbV53tgJcJqN1TVU2OMQ3OPY1WYj7OZj7OZj7OZj7OZj7OZj7Ot+3xYCgQAaCKsAACarFNYnZh7ACvGfJzNfJzNfJzNfJzNfJzNfJxtredjbc6xAgBYtnU6YgUAsFRrEVZVdbiqnq+qU1V1fO7x7Iaquraqfq6qvlRVz1bVX5i2v6eqPltVX55+ffe0varqH0xz9MWqet+830G/qrqsqv5zVX1mun99VX1++p5/tqoun7ZfMd0/NT1+YNaBL0lVXVlVD1XVr1TVc1X1B9d8//hL09+VZ6rq/qp6xzrtI1V1X1W9XFXPbNl20ftDVX1sev6Xq+pjc3wvHS4wH39r+vvyxar6l1V15ZbHfmyaj+er6kNbtu+J15/zzceWx/5yVY2qumq6v+f3j7c0xtjTX0kuS/Jfk7w3yeVJvpDk4Nzj2oXv+3uSvG+6/duT/GqSg0n+ZpLj0/bjSX5iuv2DSf5NkkrygSSfn/t7WMKc/EiSn0nymen+g0lun27fm+QT0+0/n+Te6fbtSX527rEvaT5+KsmfmW5fnuTKdd0/klyT5CtJvnPLvvHxddpHkvzhJO9L8syWbRe1PyR5T5IXpl/fPd1+99zfW+N8fDDJvun2T2yZj4PTa8sVSa6fXnMu20uvP+ebj2n7tUkez+a1KK9al/3jrb7W4YjVzUlOjTFeGGO8luSBJLfNPKalG2N8dYzxS9Pt/5nkuWy+eNyWzRfUTL/+sen2bUn+6dj0i0murKrv2d1RL09V7U9yJMlPTvcryfcneWh6yrlz8cYcPZTkB6bn7xlVtcjmD8pPJckY47Uxxm9kTfePyb4k31lV+5K8M8lXs0b7yBjjiSSvnrP5YveHDyX57Bjj1THGN5J8NsnhpQ9+Cc43H2OMfzvGeH26+4tJ9k+3b0vywBjj/44xvpLkVDZfe/bM688F9o8k+btJ/kqSrSds7/n9462sQ1hdk+TFLffPTNvWxrRM8b1JPp/ku8cYX50e+lqS755u7/V5+nvZ/Mv/ren+70zyG1t+SG79ft+ci+nxjen5e8n1SV5J8k+m5dGfrKp3ZU33jzHGS0n+dpJfz2ZQbSR5Ouu9jyQXvz/s6f3kHH86m0dlkjWdj6q6LclLY4wvnPPQWs7HG9YhrNZaVX1Xkn+R5C+OMf7H1sfG5rHZPf+20Kr6cJKXxxhPzz2WFbIvm4f1//EY43uT/O9sLvW8aV32jySZzh26LZvB+XuSvCt78F/SO7FO+8PbqapPJnk9yU/PPZa5VNU7k/zVJH9t7rGsmnUIq5eyuQb8hv3Ttj2vqn5bNqPqp8cYD0+b/9sbSzjTry9P2/fyPH1fko9U1elsHor//iR/P5uHp/dNz9n6/b45F9PjiyT/fTcHvAvOJDkzxvj8dP+hbIbWOu4fSfJHknxljPHKGOObSR7O5n6zzvtIcvH7w17fT1JVH0/y4SQfnWIzWc/5+H3Z/IfIF6afrfuT/FJV/e6s53y8aR3C6skkN07v7rk8myeaPjLzmJZuOt/jU0meG2P8nS0PPZLkjXdifCzJv96y/U9O7+b4QJKNLUsAl7Qxxo+NMfaPMQ5k8///fxhjfDTJzyX54elp587FG3P0w9Pz99S/1McYX0vyYlXdNG36gSRfyhruH5NfT/KBqnrn9HfnjflY231kcrH7w+NJPlhV756OAn5w2rYnVNXhbJ5S8JExxm9ueeiRJLdP7xa9PsmNSf5T9vDrzxjjl8cYv2uMcWD62Xomm2+Y+lrWdP9409xnz+/GVzbfofCr2Xx3xifnHs8ufc9/KJuH7b+Y5L9MXz+YzfNA/n2SLyf5d0neMz2/ktwzzdEvJzk09/ewpHm5Jb/1rsD3ZvOH36kk/zzJFdP2d0z3T02Pv3fucS9pLv5AkqemfeRfZfNdOmu7fyT58SS/kuSZJP8sm+/wWpt9JMn92Ty/7JvZfJG889vZH7J57tGp6etPzf19Nc/HqWyeI/TGz9R7tzz/k9N8PJ/k1i3b98Trz/nm45zHT+e33hW45/ePt/py5XUAgCbrsBQIALArhBUAQBNhBQDQRFgBADQRVgAATYQVAEATYQUA0ERYAQA0+f8SDcqC/o5W5gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = np.load(os.path.join(SARS_DIR, old_date, 'number-ambiguous.npz'), allow_pickle=True)\n",
    "print([i for i in data])\n",
    "num_ambig = data['number_ambiguous']\n",
    "num_ns    = data['number_ns']\n",
    "num_seqs  = data['number_sequences']\n",
    "genome_length = 29903 - (29903 - 29690) - 150 + 9\n",
    "print(f'There are {num_seqs} sequences in the analysis')\n",
    "print(f'There are {num_ambig} ambiguous nucleotides in the used sequences')\n",
    "print(f'There are {num_ns} Ns in the used sequences')\n",
    "print(f'There are a total of {num_seqs * genome_length} sites in all of the used data')\n",
    "print(f'Percentage of sites that are ambiguous is {num_ambig / (num_seqs * genome_length)}')\n",
    "print('')\n",
    "ns_sequence = data['ns_sequence']\n",
    "\n",
    "#fig, ax = plt.subplots(1,1, figsize=[10,10])\n",
    "#_ = ax.hist(ns_sequence, log=True, bins=50)\n",
    "print(f'number of sequences with more than 1000 Ns is {np.count_nonzero(ns_sequence>1000)}')\n",
    "print(f'number of sequences with more than 300 Ns is {np.count_nonzero(ns_sequence>300)}')\n",
    "print(f'number of sequences with more than 100 Ns is {np.count_nonzero(ns_sequence>100)}')\n",
    "print(f'number of sequences with more than 0 Ns is {np.count_nonzero(ns_sequence>0)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "alpha covariance entry: 405.97825335960823\n",
      "delta covariance entry: 287.5670967294807\n"
     ]
    }
   ],
   "source": [
    "# Looking at integrated correlations for sites in variants\n",
    "data_date = DATA_DATE\n",
    "\n",
    "def normalizer(m, g):\n",
    "    \"\"\"given a covariance matrix m and a regularization g, \n",
    "    return the maximum value that the correlation matrix should have if sites were fully linked\"\"\"\n",
    "    return 1 / (1 - (g / np.mean(m)) - len(m))\n",
    "\n",
    "alpha = ['S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "         'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C',\n",
    "         'S-570-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "         'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T'] \n",
    "\n",
    "delta = ['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-478-1-A', 'S-681-1-G', 'S-950-0-A', \n",
    "         'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', 'N-63-1-G', 'N-203-1-T', \n",
    "         'N-377-0-T', 'S-452-1-G']  \n",
    "\n",
    "omicron = ['NSP3-38-1-G', 'NSP3-889-2-G', \n",
    "           'NSP3-1892-0-A', 'NSP4-492-1-T', 'NSP5-132-1-A', 'NSP6-189-0-G', 'NSP10-57-2-C', \n",
    "           'NSP12-600-2-T', 'S-67-1-T', 'S-339-1-A', 'S-371-0-C', \n",
    "           'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-446-0-A', 'S-547-1-A', 'S-655-0-T', \n",
    "           'S-679-2-G', 'S-764-2-A', 'S-796-0-T', 'S-856-2-A', 'S-954-2-T', 'S-969-2-A', 'S-981-0-T', \n",
    "           'S-1146-2-T', 'ORF3a-64-2-T', 'E-9-1-T', 'M-3-1-G', 'M-19-0-G', 'M-63-0-A', 'ORF6-20-0-C', \n",
    "           'N-13-1-T', 'S-375-1-T', 'S-477-1-A', 'S-478-1-A', \n",
    "           'S-484-1-C', 'S-493-1-G', 'S-496-0-A', 'S-498-1-G', 'S-505-0-C', 'NSP14-42-0-G'] \n",
    "\n",
    "n = 10000\n",
    "k = 0.1\n",
    "R = 2\n",
    "g = 40\n",
    "c = (n * k * R) / (k + R)\n",
    "#g *= c\n",
    "data  = np.load(os.path.join(INF_DIR, f'infer-{data_date}.npz'), allow_pickle=True)\n",
    "muts  = [dp.get_label_new(i) for i in data['allele_number']]\n",
    "covar = (data['covar_int'] / c)\n",
    "alpha = sorted([i for i in alpha if i in muts])\n",
    "delta = sorted([i for i in delta if i in muts])\n",
    "print(f'alpha covariance entry: {np.mean([covar[muts.index(i), muts.index(i)] for i in alpha])}')\n",
    "print(f'delta covariance entry: {np.mean([covar[muts.index(i), muts.index(i)] for i in delta])}')\n",
    "covar = covar + np.diag(np.ones(len(covar))) * g\n",
    "covar_inv  = linalg.inv(covar)\n",
    "bare_covar = covar - np.diag(np.ones(len(covar))) * g\n",
    "\n",
    "alpha_idxs  = np.array([muts.index(i) for i in alpha])\n",
    "alpha_covar = covar_inv[alpha_idxs][:, alpha_idxs]    # the inverse covariance matrix for only alpha entries\n",
    "L           = len(alpha_covar)\n",
    "new_covar   = np.ones((L, L))\n",
    "for i in range(L):\n",
    "    for j in range(L):\n",
    "        new_covar[i, j] = alpha_covar[i, j] / np.sqrt(alpha_covar[i, i] * alpha_covar[j, j])\n",
    "alpha_df = pd.DataFrame(data=new_covar, columns=alpha, index=alpha)\n",
    "alpha_df.to_csv(os.path.join(SARS_DIR, data_date, f'alpha-correlation-{DATA_DATE}.csv'))\n",
    "alpha_sub_covar = bare_covar[alpha_idxs][:, alpha_idxs]\n",
    "new_covar       = new_covar / normalizer(alpha_sub_covar, g)\n",
    "alpha_df = pd.DataFrame(data=new_covar, columns=alpha, index=alpha)\n",
    "alpha_df.to_csv(os.path.join(SARS_DIR, data_date, f'alpha-correlation-normed-{DATA_DATE}.csv'))\n",
    "    \n",
    "delta       = sorted([i for i in delta if i in muts])\n",
    "delta_idxs  = np.array([muts.index(i) for i in delta])\n",
    "delta_covar = covar_inv[delta_idxs][:, delta_idxs]\n",
    "L           = len(delta_covar)\n",
    "new_covar   = np.ones((L, L))\n",
    "for i in range(L):\n",
    "    for j in range(L):\n",
    "        new_covar[i, j] = delta_covar[i, j] / np.sqrt(delta_covar[i, i] * delta_covar[j, j])\n",
    "delta_df = pd.DataFrame(data=new_covar, columns=delta, index=delta)\n",
    "delta_df.to_csv(os.path.join(SARS_DIR, data_date, f'delta-correlation-{DATA_DATE}.csv'))\n",
    "delta_sub_covar = bare_covar[delta_idxs][:, delta_idxs]\n",
    "new_covar       = new_covar / normalizer(delta_sub_covar, g)\n",
    "delta_df = pd.DataFrame(data=new_covar, columns=delta, index=delta)\n",
    "delta_df.to_csv(os.path.join(SARS_DIR, data_date, f'delta-correlation-normed-{DATA_DATE}.csv'))\n",
    "\n",
    "omicron       = sorted([i for i in omicron if i in muts])\n",
    "omicron_idxs  = np.array([muts.index(i) for i in omicron])\n",
    "omicron_covar = covar_inv[omicron_idxs][:, omicron_idxs]\n",
    "L           = len(omicron_covar)\n",
    "new_covar   = np.ones((L, L))\n",
    "for i in range(L):\n",
    "    for j in range(L):\n",
    "        new_covar[i, j] = omicron_covar[i, j] / np.sqrt(omicron_covar[i, i] * omicron_covar[j, j])\n",
    "omicron_df = pd.DataFrame(data=new_covar, columns=omicron, index=omicron)\n",
    "omicron_df.to_csv(os.path.join(SARS_DIR, data_date, f'omicron-correlation-{DATA_DATE}.csv'))\n",
    "omicron_sub_covar = bare_covar[omicron_idxs][:, omicron_idxs]\n",
    "new_covar       = new_covar / normalizer(omicron_sub_covar, g)\n",
    "omicron_df = pd.DataFrame(data=new_covar, columns=omicron, index=omicron)\n",
    "omicron_df.to_csv(os.path.join(SARS_DIR, data_date, f'omicron-correlation-normed-{DATA_DATE}.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Manually fixing gaps in the map from nucleotide mutations to amino acid mutations__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--'], [], ['S-240-2--', 'S-241-0--', 'S-242-1--', 'S-242-2--', 'S-243-0--', 'S-240-1--', 'S-241-1--', 'S-241-2--', 'S-242-0--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--'], ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--'], ['S-246-2--', 'S-247-0--', 'S-247-1--', 'S-247-2--', 'S-248-0--', 'S-248-1--', 'S-248-2--', 'S-249-0--', 'S-249-1--', 'S-249-2--', 'S-250-0--', 'S-250-1--', 'S-250-2--', 'S-251-0--', 'S-251-1--', 'S-251-2--', 'S-252-0--', 'S-252-1--', 'S-252-2--', 'S-253-0--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--'], ['S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', 'S-158-0--', 'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'NC-28270--'], [], [], ['NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--', 'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'S-142-1--', 'S-142-2--', 'S-143-0--', 'S-143-1--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-144-2--', 'S-145-0--', 'S-211-1--', 'S-211-2--', 'S-212-0--', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', 'N-32-2--', 'N-33-0--'], ['N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', 'N-32-2--', 'N-33-0--', 'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--'], ['ORF6-2-0--', 'ORF6-2-1--', 'ORF6-2-2--', 'N-2-1--', 'N-2-2--', 'N-3-0--', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--'], ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--', 'NC-28270--'], ['ORF7b-44-1--', 'NC-27887--', 'NC-27888--', 'NC-27889--', 'NC-27890--', 'NC-27891--', 'NC-27892--', 'ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'ORF8-2-0--', 'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--', 'N-208-2--', 'N-209-0--', 'N-208-1--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--', 'S-143-2--', 'S-144-0--', 'S-144-1--']]\n"
     ]
    }
   ],
   "source": [
    "gap_muts = []\n",
    "for i in var_dic:\n",
    "    gaps_temp = [j for j in var_dic[i] if j[-1]=='-']\n",
    "    gap_muts.append(gaps_temp)\n",
    "print(gap_muts)\n",
    "gaps [#['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--',\n",
    "      # 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--'], \n",
    "      ['NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--']\n",
    "      ['S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--'],\n",
    "      ['S-240-1--', 'S-240-2--', 'S-241-0--', 'S-241-1--', 'S-241-2--', 'S-242-0--' 'S-242-1--', 'S-242-2--', 'S-243-0--'],\n",
    "      ['S-246-1--', 'S-246-2--', 'S-247-0--', 'S-247-1--', 'S-247-2--', 'S-248-0--', 'S-248-1--', \n",
    "       'S-248-2--', 'S-249-0--', 'S-249-1--', 'S-249-2--', 'S-250-0--', 'S-250-1--', 'S-250-2--', \n",
    "       'S-251-0--', 'S-251-1--', 'S-251-2--', 'S-252-0--', 'S-252-1--', 'S-252-2--', 'S-253-0--'],\n",
    "      ['S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', 'S-158-0--'],\n",
    "      #['ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--'],\n",
    "      ['NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--'],\n",
    "      ['S-211-1--', 'S-211-2--', 'S-212-0--'],\n",
    "      ['N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', 'N-32-2--', 'N-33-0--'],\n",
    "      ['S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--'],\n",
    "      #['ORF6-2-0--', 'ORF6-2-1--', 'ORF6-2-2--'],\n",
    "      ['N-2-1--', 'N-2-2--', 'N-3-0--']]\n",
    "      \n",
    "      \n",
    "additional = [['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--',\n",
    "               'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--'],\n",
    "              ['NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--'],\n",
    "              ['ORF7b-44-1--'],\n",
    "              ['S-142-1--', 'S-142-2--', 'S-143-0--', 'S-143-1--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-144-2--', 'S-145-0--'],\n",
    "              ['S-143-2--', 'S-144-0--', 'S-144-1--'],\n",
    "              ['ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'ORF8-2-0--', 'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--']]\n",
    "# How to deal with it when there are different overlapping groups of deletions\n",
    "\n",
    "# orf7b-44-1 and the orf8 mutations should cause a frameshift, correct? (both are in B.1.1.318)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 482,
   "metadata": {},
   "outputs": [],
   "source": [
    "syn_file  = np.load(os.path.join(DATA_DIR, 'synonymous-prot-test.npz'), allow_pickle=True)\n",
    "nuc_index  = syn_file['nuc_index']\n",
    "locs       = list(syn_file['locations'])\n",
    "aa_changes = syn_file['aa_changes']\n",
    "types      = syn_file['types']\n",
    "\n",
    "nuc_pos    = [i[:-2] for i in nuc_index]\n",
    "nucs       = [i[-1]  for i in nuc_index]\n",
    "\n",
    "ref_df     = pd.read_csv(os.path.join(DATA_DIR, 'ref-index.csv'))\n",
    "ref_index  = list(ref_df['ref_index'])\n",
    "ref_nucs   = list(ref_df['nucleotide'])\n",
    "ref_nucs   = [ref_nucs[i] for i in range(len(ref_nucs)) if ref_index[i][0]!='-']\n",
    "ref_index  = [ref_index[i] for i in range(len(ref_index)) if ref_index[i][0]!='-']\n",
    "ref_labels = [dp.get_label_new(ref_index[i] + '-' + ref_nucs[i]) for i in range(len(ref_index))]\n",
    "ref_short  = [i[:-2] for i in ref_labels]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 484,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-142-0-G\n",
      "S-142-1-G\n",
      "S-142-2-T\n",
      "S-145-0-T\n",
      "S-145-1-A\n",
      "S-145-2-C\n",
      "D G Y\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-142'\n",
    "end_site   = 'S-145'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'GAC'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 485,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'Y>D'\n",
    "#aa_changes[list(locs).index(start_site + '-2--')] = 'S>S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 473,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-2-0-T\n",
      "N-2-1-C\n",
      "N-2-2-T\n",
      "N-3-0-G\n",
      "N-3-1-A\n",
      "N-3-2-T\n",
      "S S D\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-24'\n",
    "end_site   = 'S-27'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'TCA'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 472,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'A>S'\n",
    "#aa_changes[list(locs).index(start_site + '-2--')] = 'S>S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 475,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-2-0-T\n",
      "N-2-1-C\n",
      "N-2-2-T\n",
      "N-3-0-G\n",
      "N-3-1-A\n",
      "N-3-2-T\n",
      "Y S D\n"
     ]
    }
   ],
   "source": [
    "start_site = 'N-2'\n",
    "end_site   = 'N-3'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'TAT'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 476,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'D>Y'\n",
    "#aa_changes[list(locs).index(start_site + '-2--')] = 'S>S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-30-0-G\n",
      "N-30-1-G\n",
      "N-30-2-A\n",
      "N-33-0-A\n",
      "N-33-1-G\n",
      "N-33-2-T\n",
      "G G S\n"
     ]
    }
   ],
   "source": [
    "start_site = 'N-30'\n",
    "end_site   = 'N-33'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'GGT'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(start_site + '-1--')] = 'G>G'\n",
    "aa_changes[list(locs).index(start_site + '-2--')] = 'G>G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-211-0-A\n",
      "S-211-1-A\n",
      "S-211-2-T\n",
      "S-212-0-T\n",
      "S-212-1-T\n",
      "S-212-2-A\n",
      "I N L\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-211'\n",
    "end_site   = 'S-212'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'ATA'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 466,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'L>I'\n",
    "#aa_changes[list(locs).index(start_site + '-2--')] = 'S>S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NSP3-1265-0-A\n",
      "NSP3-1265-1-G\n",
      "NSP3-1265-2-T\n",
      "NSP3-1266-0-T\n",
      "NSP3-1266-1-T\n",
      "NSP3-1266-2-A\n",
      "S S L\n"
     ]
    }
   ],
   "source": [
    "start_site = 'NSP3-1265'\n",
    "end_site   = 'NSP3-1266'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "#print(ref_labels[30000:])\n",
    "new_codon = 'TCA'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(start_site + '-1--')] = 'S>S'\n",
    "aa_changes[list(locs).index(start_site + '-2--')] = 'S>S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-156-0-G\n",
      "S-156-1-A\n",
      "S-156-2-G\n",
      "S-158-0-A\n",
      "S-158-1-G\n",
      "S-158-2-A\n",
      "G E R\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-156'\n",
    "end_site   = 'S-158'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "new_codon = 'GGA'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'R>G'\n",
    "#aa_changes[list(locs).index(end_site + '-2--')] = 'R>G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-246-0-A\n",
      "S-246-1-G\n",
      "S-246-2-A\n",
      "S-253-0-G\n",
      "S-253-1-A\n",
      "S-253-2-T\n",
      "N R D\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-246'\n",
    "end_site   = 'S-253'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "new_codon = 'AAT'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_changes[list(locs).index(end_site + '-0--')] = 'D>N'\n",
    "#aa_changes[list(locs).index(end_site + '-2--')] = 'D>N'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-240-0-A\n",
      "S-240-1-C\n",
      "S-240-2-T\n",
      "S-243-0-G\n",
      "S-243-1-C\n",
      "S-243-2-T\n",
      "T T A\n"
     ]
    }
   ],
   "source": [
    "start_site = 'S-240'\n",
    "end_site   = 'S-243'\n",
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in [start_site, end_site]:\n",
    "        print(ref_labels[i])\n",
    "new_codon = 'ACT'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index(start_site + '-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index(end_site + '-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)\n",
    "\n",
    "aa_changes[list(locs).index(start_site + '-1--')] = 'T>T'\n",
    "aa_changes[list(locs).index(start_site + '-2--')] = 'T>T'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S-68-0-A\n",
      "S-68-1-T\n",
      "S-68-2-A\n",
      "S-70-0-G\n",
      "S-70-1-T\n",
      "S-70-2-C\n",
      "I I V\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in ['S-68', 'S-70']:\n",
    "        print(ref_labels[i])\n",
    "new_codon = 'ATC'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index('S-68-1-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index('S-70-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)\n",
    "\n",
    "aa_changes[list(locs).index('S-68-1--')] = 'I>I'\n",
    "aa_changes[list(locs).index('S-68-2--')] = 'I>I'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NSP6-104-0-A\n",
      "NSP6-104-1-G\n",
      "NSP6-104-2-T\n",
      "NSP6-107-0-G\n",
      "NSP6-107-1-G\n",
      "NSP6-107-2-T\n",
      "S G S\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(ref_short)):\n",
    "    if ref_short[i][:-2] in ['NSP6-104', 'NSP6-107']:\n",
    "        print(ref_labels[i])\n",
    "new_codon = 'AGT'\n",
    "new_aa    = dp.codon2aa(new_codon)\n",
    "old_aa1   = aa_changes[list(locs).index('NSP6-107-0-A')][0]\n",
    "old_aa2   = aa_changes[list(locs).index('NSP6-104-1-A')][0]\n",
    "print(new_aa, old_aa1, old_aa2)\n",
    "\n",
    "aa_changes[list(locs).index('NSP6-104-1--')] = 'S>S'\n",
    "aa_changes[list(locs).index('NSP6-104-2--')] = 'S>S'\n",
    "# NOTE: LEAVING NONSYNONYMOUS BECUASE THESE MUTATIONS ARE IMPORTANT FOR INFERENCE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_file = open(os.path.join(DATA_DIR, 'synonymous-prot-fixed.npz'), mode='wb')\n",
    "np.savez_compressed(new_file, nuc_index=nuc_index, locations=locs, aa_changes=aa_changes, types=types)\n",
    "new_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=nextstrain_version></a>\n",
    "# Nextstrain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "name\tn dupes\ttotal s\n",
      "NSP1-K141-\t3\t0.001\n",
      "NSP1-K141-\t3\t0.001\n",
      "NSP1-K141-\t3\t0.001\n",
      "NSP1-S142-\t3\t-0.004\n",
      "NSP1-S142-\t3\t0.001\n",
      "NSP1-S142-\t3\t0.001\n",
      "NSP1-F143-\t3\t0.005\n",
      "NSP1-F143-\t3\t0.004\n",
      "NSP3-S1265-\t2\t-0.004\n",
      "NSP6-S104-\t2\t-0.013\n",
      "NSP6-S104-\t2\t-0.012\n",
      "NSP6-L105-\t3\t-0.013\n",
      "NSP6-L105-\t3\t-0.044\n",
      "NSP6-L105-\t3\t-0.040\n",
      "NSP6-S106-\t3\t0.060\n",
      "NSP6-S106-\t3\t0.069\n",
      "NSP6-S106-\t3\t0.021\n",
      "NSP6-G107-\t3\t0.039\n",
      "NSP6-G107-\t3\t-0.007\n",
      "NSP6-G107-\t3\t-0.008\n",
      "NSP6-F108-\t3\t-0.005\n",
      "NSP6-F108-\t3\t-0.007\n",
      "NSP6-F108-\t3\t0.028\n",
      "NSP14-E204D\t2\t0.013\n",
      "NSP14-E204D\t2\t0.013\n",
      "S-L24-\t2\t0.013\n",
      "S-L24-\t2\t0.013\n",
      "S-P25-\t3\t0.045\n",
      "S-P25-\t3\t-0.002\n",
      "S-P25-\t3\t-0.003\n",
      "S-P26-\t3\t-0.002\n",
      "S-P26-\t3\t-0.001\n",
      "S-P26-\t3\t-0.012\n",
      "S-A27S\t2\t-0.013\n",
      "S-A27S\t2\t-0.013\n",
      "S-I68-\t2\t0.003\n",
      "S-I68-\t2\t0.005\n",
      "S-H69-\t3\t-0.010\n",
      "S-H69-\t3\t-0.012\n",
      "S-G142-\t2\t0.111\n",
      "S-G142D\t2\t0.064\n",
      "S-G142-\t2\t0.024\n",
      "S-G142D\t2\t0.064\n",
      "S-V143-\t3\t0.032\n",
      "S-V143-\t3\t0.050\n",
      "S-V143-\t3\t0.005\n",
      "S-Y144-\t3\t0.002\n",
      "S-Y144-\t3\t-0.010\n",
      "S-Y144-\t3\t-0.044\n",
      "S-E156-\t2\t0.004\n",
      "S-E156-\t2\t0.003\n",
      "S-F157-\t3\t-0.005\n",
      "S-F157-\t3\t0.019\n",
      "S-F157-\t3\t0.023\n",
      "S-R158G\t2\t-0.010\n",
      "S-R158G\t2\t-0.010\n",
      "S-N211-\t2\t-0.017\n",
      "S-N211-\t2\t-0.009\n",
      "S-L212I\t2\t0.017\n",
      "S-L212I\t2\t0.017\n",
      "S-T240-\t2\t-0.015\n",
      "S-T240-\t2\t-0.008\n",
      "S-L241-\t3\t-0.009\n",
      "S-L241-\t3\t-0.008\n",
      "S-L241-\t3\t-0.006\n",
      "S-L242-\t3\t-0.006\n",
      "S-L242-\t3\t-0.007\n",
      "S-L242-\t3\t-0.006\n",
      "S-R246-\t2\t-0.003\n",
      "S-S247-\t3\t-0.003\n",
      "S-S247-\t3\t-0.003\n",
      "S-S247-\t3\t-0.006\n",
      "S-Y248-\t3\t-0.007\n",
      "S-Y248-\t3\t-0.004\n",
      "S-Y248-\t3\t-0.006\n",
      "S-L249-\t3\t-0.006\n",
      "S-L249-\t3\t-0.006\n",
      "S-L249-\t3\t-0.006\n",
      "S-T250-\t3\t-0.006\n",
      "S-T250-\t3\t-0.001\n",
      "S-T250-\t3\t-0.006\n",
      "S-P251-\t3\t-0.004\n",
      "S-P251-\t3\t-0.007\n",
      "S-P251-\t3\t-0.006\n",
      "S-G252-\t3\t-0.006\n",
      "S-G252-\t3\t-0.006\n",
      "S-G252-\t3\t-0.006\n",
      "S-N440K\t2\t0.078\n",
      "S-N440K\t2\t0.078\n",
      "S-Q675H\t2\t0.015\n",
      "S-Q675H\t2\t0.015\n",
      "S-Q677H\t2\t0.025\n",
      "S-Q677H\t2\t0.025\n",
      "S-N679K\t2\t0.049\n",
      "S-N679K\t2\t0.049\n",
      "S-S943P\t2\t0.012\n",
      "S-S943P\t2\t0.013\n",
      "S-V1264L\t2\t0.007\n",
      "S-V1264L\t2\t0.007\n",
      "ORF3a-V256-\t3\t0.003\n",
      "ORF3a-V256-\t3\t0.008\n",
      "ORF3a-V256-\t3\t0.006\n",
      "ORF6-D61L\t3\t-0.009\n",
      "ORF6-D61L\t3\t-0.055\n",
      "ORF6-D61L\t3\t0.011\n",
      "ORF8-M1-\t3\t-0.002\n",
      "ORF8-M1-\t3\t0.005\n",
      "ORF8-M1-\t3\t-0.001\n",
      "ORF8-K2-\t3\t-0.004\n",
      "ORF8-K2-\t3\t-0.004\n",
      "ORF8-K2-\t3\t-0.004\n",
      "ORF8-F3-\t2\t-0.002\n",
      "ORF8-F3-\t2\t-0.006\n",
      "ORF8-V62L\t2\t0.008\n",
      "ORF8-V62L\t2\t0.008\n",
      "ORF8-D119-\t3\t0.022\n",
      "ORF8-D119-\t3\t-0.010\n",
      "ORF8-D119-\t3\t0.009\n",
      "ORF8-F120-\t3\t0.022\n",
      "ORF8-F120L\t2\t0.023\n",
      "ORF8-F120-\t3\t0.005\n",
      "ORF8-F120-\t3\t0.016\n",
      "ORF8-F120L\t2\t0.023\n",
      "N-D3L\t3\t0.033\n",
      "N-D3L\t3\t0.026\n",
      "N-D3L\t3\t0.025\n",
      "N-G30-\t2\t-0.006\n",
      "N-G30-\t2\t0.025\n",
      "N-E31-\t3\t0.025\n",
      "N-E31-\t3\t0.022\n",
      "N-E31-\t3\t0.026\n",
      "N-R32-\t3\t0.027\n",
      "N-R32-\t3\t0.020\n",
      "N-R32-\t3\t0.026\n",
      "N-R203K\t2\t0.092\n",
      "N-R203K\t2\t-0.045\n",
      "N-A208-\t2\t-0.002\n",
      "N-A208-\t2\t-0.002\n",
      "N-M234I\t3\t0.027\n",
      "N-M234I\t3\t0.027\n",
      "N-M234I\t3\t0.027\n"
     ]
    }
   ],
   "source": [
    "# Instead of summing contributions from linked mutations that result in the same amino acid, leave these alone\n",
    "# But for non-linked mutations resulting in the same amino acid change, use the largest inferred mutation for all of them, \n",
    "# since this is probably the most reliable estimate\n",
    "\n",
    "# Find mutations that are in the same codon and strongly linked to one another and sum their selection coefficients\n",
    "data  = np.load(os.path.join(DATA_DIR, 'linked-sites.npy'), allow_pickle=True)\n",
    "nucs  = [np.array([data[i][j][-1] for j in range(len(data[i]))]) for i in range(len(data))]\n",
    "idxs  = []\n",
    "for i in range(len(data)):\n",
    "    if len(np.array(nucs[i])[nucs[i]=='-']) != len(nucs[i]):\n",
    "        idxs.append(i)\n",
    "data = data[idxs]\n",
    "split = [[data[i][j].split('-') for j in range(len(data[i]))] for i in range(len(data))]\n",
    "cods  = [[split[i][j][1] for j in range(len(split[i]))] for i in range(len(split))]\n",
    "prots = [[split[i][j][0] for j in range(len(split[i]))] for i in range(len(split))]\n",
    "\n",
    "linked_groups     = []\n",
    "for i in range(len(data)):\n",
    "    group   = data[i]\n",
    "    g_cods  = cods[i]\n",
    "    g_prots = prots[i]\n",
    "    prot_unique = np.unique(g_prots)\n",
    "    for j in range(len(prot_unique)):\n",
    "        temp_cods   = np.array(g_cods)[np.array(g_prots)==prot_unique[j]]\n",
    "        temp_labs   = np.array(group)[np.array(g_prots)==prot_unique[j]]\n",
    "        cods_unique = np.unique(temp_cods)\n",
    "        for k in range(len(cods_unique)):\n",
    "            labs_same_cod = temp_labs[temp_cods==cods_unique[k]]\n",
    "            if len(labs_same_cod)>1:\n",
    "                linked_groups.append(labs_same_cod)\n",
    "                \n",
    "linked_groups     = [[linked_groups[i][j] for j in range(len(linked_groups[i])) if linked_groups[i][j][-1]!='-'] for i in range(len(linked_groups))]\n",
    "linked_groups     = [i for i in linked_groups if len(i)!=0]\n",
    "linked_all        = np.unique([linked_groups[i][j] for i in range(len(linked_groups)) for j in range(len(linked_groups[i]))])\n",
    "\n",
    "# replacing coefficients for mutations that lead to the same amino acid\n",
    "g = 40\n",
    "df_sel = pd.read_csv(os.path.join(DATA_DIR, f'selection-nonsyn-subsample.csv'))\n",
    "\n",
    "f = open(os.path.join(DATA_DIR, f'selection_coefficients_nextstrain.csv'), 'w')\n",
    "\n",
    "headers = list(df_sel.columns)\n",
    "f.write('%s\\n' % ','.join(headers))\n",
    "\n",
    "print('name\\tn dupes\\ttotal s')\n",
    "\n",
    "#dup_names = []\n",
    "for iter, entry in df_sel.iterrows():\n",
    "    if entry['protein']=='NC':\n",
    "        continue\n",
    "    loc      = str(entry['amino acid number in protein'])\n",
    "    anc      = entry['amino acid mutation'][0]\n",
    "    mut      = entry['amino acid mutation'][-1]\n",
    "    var_name = str(entry['protein']) + '-' + anc + loc + mut\n",
    "    label    = dp.get_label_new(str(entry['nucleotide number']) + '-' + entry['nucleotide'])\n",
    "    #if var_name in dup_names:\n",
    "    #    continue\n",
    "    \n",
    "    df_same = df_sel[(df_sel['protein']==entry['protein']) \n",
    "                     & (df_sel['amino acid number in protein']==entry['amino acid number in protein'])\n",
    "                     & (df_sel['amino acid mutation']==entry['amino acid mutation'])]\n",
    "    \n",
    "    s = entry['selection coefficient']\n",
    "    if len(df_same)>1:\n",
    "        if var_name[-1]!='-' and label not in linked_all:\n",
    "            s = np.max(df_same['selection coefficient'])\n",
    "        if np.fabs(s)>0.001:\n",
    "            print('%s\\t%d\\t%.3f' % (var_name, len(df_same), s))\n",
    "        \n",
    "    for i in range(len(headers)):\n",
    "        if headers[i]=='selection coefficient':\n",
    "            f.write('%s' % str(s))\n",
    "        elif headers[i]=='linked sites':\n",
    "            f.write('\"%s\"' % str(entry[headers[i]]))\n",
    "        elif headers[i]=='nucleotide number':\n",
    "            num, letters = dp.separate_label_idx(str(entry[headers[i]]))\n",
    "            new_nuc = str(int(num) + 1) + letters\n",
    "            f.write('%s' % new_nuc)\n",
    "        else:\n",
    "            f.write('%s' % str(entry[headers[i]]))\n",
    "        if i!=len(headers)-1:\n",
    "            f.write(',')\n",
    "    f.write('\\n')\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of different mutations in regular inference and in only nonsynonymous inference\n",
    "df = pd.read_csv(os.path.join(DATA_DIR, f'selection-observed.csv'))\n",
    "df = df[df['selection coefficient']>0.01]\n",
    "v  = np.array(df['nucleotide'].value_counts()) / np.sum(df['nucleotide'].value_counts())\n",
    "print(df['nucleotide'].value_counts())\n",
    "print(v)\n",
    "print('')\n",
    "df = pd.read_csv(os.path.join(DATA_DIR, f'selection-nonsyn-observed.csv'))\n",
    "df = df[df['selection coefficient']>0.01]\n",
    "v  = np.array(df['nucleotide'].value_counts()) / np.sum(df['nucleotide'].value_counts())\n",
    "print(df['nucleotide'].value_counts())\n",
    "print(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove error column from nextstrain file\n",
    "df = pd.read_csv(os.path.join(DATA_DIR, f'selection_coefficients_nextstrain.csv'), index_col=False)\n",
    "df = df.drop(columns='error')\n",
    "df.to_csv(os.path.join(DATA_DIR, f'selection_coefficients_nextstrain.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=EU1_travel></a>\n",
    "# Inferring the selection coefficients in the United Kingdom including migration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting an estimate for the time-varying number of infected in the United Kingdom\n",
    "uk_filepath = []\n",
    "uk_filename = []\n",
    "for file in sorted(os.listdir(os.path.join(DATA_DIR, 'freqs'))):\n",
    "    file_split = file.split('-')\n",
    "    if file_split[1] == 'united kingdom' and file_split[2] == 'england_wales_scotland':\n",
    "        year_end  = int(file_split[-3])\n",
    "        month_end = int(file_split[-2])\n",
    "        day_end   = int(file_split[-1][:-4])  \n",
    "        if (dt.date(year_end, month_end, day_end) - dt.date(2021, 5, 5)).days < 0:  # take only data before the middle of may\n",
    "            uk_filepath.append(os.path.join(DATA_DIR, 'freqs', file))\n",
    "            uk_filename.append(file)\n",
    "uk_dir = os.path.join(DATA_DIR, 'genome-uk')\n",
    "if not os.path.exists(uk_dir):\n",
    "    os.mkdir(uk_dir)\n",
    "for path in uk_filepath:\n",
    "    shutil.copy(path, uk_dir)\n",
    "\n",
    "print(f'scp -r {uk_dir} {SSH_HOME}/{SSH_DATA}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding selection coefficients in the uk only (up to may 1st 2021)\n",
    "out_file   = 'infer-2021-08-14--2021-05-01-uk-g-10'\n",
    "data       = os.path.join(SSH_DATA, 'genome-uk')\n",
    "inf_script = 'epi-inf-parallel.py'\n",
    "\n",
    "job_file  = 'job-infer-uk.sh'\n",
    "job_str   = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=20:00:00\n",
    "#SBATCH -e ./MPL/out/inf-error\n",
    "#SBATCH -o ./MPL/out/inf-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file}\n",
    "python {inf_script} --data {data} --timed 1 -o $outfile -q 5 --g1 10\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "print('scp %s/%s %s.'      % (SCRIPT_DIR, job_file, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file)\n",
    "print('')\n",
    "\n",
    "print('scp %s%s %s' % (SSH_HOME, out_file + '.npz', INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finding all uk files and transferring them to their own directory\n",
    "uk_filepath = []\n",
    "uk_filename = []\n",
    "for file in sorted(os.listdir(os.path.join(DATA_DIR, 'freqs'))):\n",
    "    file_split = file.split('-')\n",
    "    if file_split[1] == 'united kingdom' and file_split[2] == 'england_wales_scotland':\n",
    "        year_end  = int(file_split[-3])\n",
    "        month_end = int(file_split[-2])\n",
    "        day_end   = int(file_split[-1][:-4])  \n",
    "        uk_filepath.append(os.path.join(DATA_DIR, 'freqs', file))\n",
    "        uk_filename.append(file)\n",
    "uk_dir = os.path.join(DATA_DIR, 'genome-uk-full')\n",
    "if not os.path.exists(uk_dir):\n",
    "    os.mkdir(uk_dir)\n",
    "for path in uk_filepath:\n",
    "    shutil.copy(path, uk_dir)\n",
    "    \n",
    "print(f'scp -r {uk_dir} {SSH_HOME}{SSH_DATA}/{DATA_DATE}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding selection coefficients in the uk only \n",
    "out_file2  = 'infer-2021-08-14-uk-g-10'\n",
    "data       = os.path.join(SSH_DATA, 'genome-uk-full')\n",
    "inf_script = 'epi-inf-parallel.py'\n",
    "\n",
    "job_file  = 'job-infer-uk2.sh'\n",
    "job_str   = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=200G\n",
    "#SBATCH --time=10:00:00\n",
    "#SBATCH -e ./MPL/out/inf-error\n",
    "#SBATCH -o ./MPL/out/inf-out\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "outfile={out_file2}\n",
    "python {inf_script} --data {data} --timed 1 -o $outfile -q 5 --g1 10\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "print('scp %s/%s %s.'      % (SCRIPT_DIR, job_file, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file)\n",
    "print('')\n",
    "\n",
    "print('scp %s%s %s' % (SSH_HOME, out_file2 + '.npz', INF_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff    = 0\n",
    "\n",
    "data      = np.load(os.path.join(DATA_DIR, 'nucleotide-counts.npz'), allow_pickle=True)\n",
    "muts      = data['allele_number']\n",
    "counts    = data['counts']\n",
    "regs      = [10]\n",
    "for reg in regs:\n",
    "\n",
    "    inf_data  = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-g-{reg}.npz'), allow_pickle=True)\n",
    "    alleles   = inf_data['allele_number']\n",
    "    selection = inf_data['selection']\n",
    "    errors    = inf_data['error_bars']\n",
    "    s_ind     = inf_data['selection_independent']\n",
    "    numerator = inf_data['numerator']\n",
    "    covar_int = inf_data['covar_int']\n",
    "    locations = inf_data['locations']\n",
    "    times     = inf_data['times']\n",
    "    \n",
    "    mask   = np.isin(muts, alleles)\n",
    "    muts   = muts[mask]\n",
    "    counts = counts[mask]\n",
    "\n",
    "    print('data loaded')\n",
    "    clip_start= 150\n",
    "    nucs      = np.array([i[:-2] for i in alleles], dtype=int)\n",
    "\n",
    "    mask      = np.nonzero(np.logical_and(np.logical_and(counts>cutoff, selection!=0), nucs > clip_start))[0]\n",
    "\n",
    "    selection = selection[mask]\n",
    "    alleles   = alleles[mask]\n",
    "    errors    = errors[mask]\n",
    "    s_ind     = s_ind[mask]\n",
    "    numerator = numerator[mask]\n",
    "    covar_int = covar_int[mask][:, mask]\n",
    "\n",
    "    print('selection data filtered')\n",
    "\n",
    "    traj      = inf_data['traj']\n",
    "    mutants   = inf_data['mutant_sites']\n",
    "    for i in range(len(mutants)):\n",
    "        idxs_keep  = np.array([j for j in range(len(mutants[i])) if mutants[i][j] in alleles])\n",
    "        mutants[i] = np.array(mutants[i])[idxs_keep]\n",
    "        traj[i]    = np.array([traj[i][:, j] for j in range(len(traj[i][0])) if j in idxs_keep])\n",
    "        traj[i]    = np.swapaxes(traj[i], 0, 1)\n",
    "        #print(f'trajectory data filtered for  region {i} out of {len(mutants)}')\n",
    "    \n",
    "    np.savez_compressed(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-g-{reg}-observed.npz'), covar_int=covar_int, \n",
    "                       selection=selection, selection_independent=s_ind, error_bars=errors, numerator=numerator,\n",
    "                       allele_number=alleles, traj=traj, mutant_sites=mutants, locations=locations, times=times)\n",
    "    \n",
    "    dp.website_file(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-g-{reg}-observed.npz'),\n",
    "                    os.path.join(DATA_DIR, 'synonymous-test.npz'),\n",
    "                    os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                    os.path.join(DATA_DIR, 'protein-changes'), \n",
    "                    os.path.join(DATA_DIR, f'selection-g-40-observed.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sites linked to S-222 mutation\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "infer_data = np.load(os.path.join(INF_DIR, out_file + '.npz'), allow_pickle=True)\n",
    "allele_num = infer_data['allele_number']\n",
    "labels     = [dp.get_label(i[:-2]) + '-' + i[-1] for i in allele_num]\n",
    "\n",
    "freq_data  = np.load(os.path.join(DATA_DIR, 'freq_0.05', \n",
    "                     'europe-united kingdom-england_wales_scotland-2020-01-01-2021-01-01.n---2020-2-24-2021-1-1.npz'),\n",
    "                     allow_pickle=True)\n",
    "nVec       = freq_data['nVec']\n",
    "sVec       = freq_data['sVec']\n",
    "times      = freq_data['times']\n",
    "mutants    = freq_data['mutant_sites']\n",
    "mask       = np.nonzero(mutants<29700)[0]\n",
    "#mask       = np.nonzero(np.logical_and(mutants<29700), mutants>150)[0]\n",
    "mutants    = mutants[mask]\n",
    "sVec_temp  = []\n",
    "for t in sVec:\n",
    "    sVec_t = []\n",
    "    for i in t:\n",
    "        sVec_t.append(i[mask])\n",
    "    sVec_temp.append(sVec_t)\n",
    "sVec = sVec_temp\n",
    "nuc_nums   = [dp.get_label(i) for i in mutants]\n",
    "loc_sites  = []\n",
    "for site in mutants:\n",
    "    for j in range(5):\n",
    "        loc_sites.append(dp.get_label(site) + '-' + NUC[j])\n",
    "\n",
    "link_sites = ['ORF10-30-0-T', 'N-220-1-T', 'M-93-2-G', 'S-222-1-T', 'NSP16-199-2-C', 'NSP3-1189-2-T', 'NSP1-60-2-C']    # The linked mutations that will inflow\n",
    "sites      = [allele_num[labels.index(i)] for i in link_sites]\n",
    "\n",
    "# Find a sequence containing the desired list of mutations\n",
    "inflow_seq = None\n",
    "seq_found  = False\n",
    "for i in range(len(sVec)):\n",
    "    for j in range(len(sVec[i])):\n",
    "        muts = True\n",
    "        for k in range(len(link_sites)):\n",
    "            if sVec[i][j][nuc_nums.index(link_sites[k][:-2])] != NUC.index(link_sites[k][-1]):\n",
    "                muts = False\n",
    "        if muts:\n",
    "            inflow_seq = sVec[i][j]\n",
    "            seq_found  = True\n",
    "            break\n",
    "    if seq_found:\n",
    "        break\n",
    "print(inflow_seq)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(SCRIPT_DIR)\n",
    "reload(dp)\n",
    "os.chdir(SARS_DIR)\n",
    "\n",
    "# Running the inference with different numbers of infected individuals migrating into the United Kingdom\n",
    "est_dir = os.path.join(SARS_DIR, IHME_DATA)\n",
    "file1   = os.path.join(est_dir, 'reference_hospitalization_all_locs.csv')\n",
    "data1   = pd.read_csv(file1)\n",
    "popsize = list(data1.est_infections_mean)\n",
    "dates   = list(data1.date)\n",
    "locs    = [i.lower() for i in list(data1.location_name)]\n",
    "uk_popsize  = np.array(popsize)[np.array(locs)=='united kingdom']\n",
    "uk_dates    = np.array(dates)[np.array(locs)=='united kingdom']\n",
    "uk_dates    = uk_dates[~np.isnan(uk_popsize)]\n",
    "uk_times    = [(dt.date.fromisoformat(uk_dates[i]) - dt.date(2020,1,1)).days for i in range(len(uk_dates))]\n",
    "uk_popsize  = uk_popsize[~np.isnan(uk_popsize)]\n",
    "\n",
    "inf_data  = np.load(os.path.join(INF_DIR, 'infer-2021-08-14--2021-05-01-uk-g-10' + '.npz'), allow_pickle=True)\n",
    "alleles   = inf_data['allele_number']\n",
    "numerator = inf_data['numerator']\n",
    "covar_int = inf_data['covar_int']\n",
    "traj      = inf_data['traj']\n",
    "mutants   = inf_data['mutant_sites']\n",
    "locs      = inf_data['locations']\n",
    "dates     = inf_data['times']\n",
    "idx       = list(locs).index('europe-united kingdom-england_wales_scotland-2020-01-01-2021-01-01.n---2020-2-24-2021-1-1')\n",
    "traj_uk   = traj[idx]\n",
    "muts_uk   = mutants[idx]\n",
    "time      = dates[idx]\n",
    "\n",
    "q   = 5\n",
    "N   = 10000\n",
    "k   = 0.1\n",
    "R   = 2\n",
    "g   = 10\n",
    "g  *= (N * k) / (1 + (k / R))\n",
    "\n",
    "for i in range(len(covar_int)):\n",
    "    covar_int[i, i] += g\n",
    "    \n",
    "start_t    = 188    # number of days after february 24th 2020 that the individuals infected with 20E-EU1 start migrating\n",
    "end_t      = 288\n",
    "T          = end_t - start_t + 1\n",
    "uk_popsize = uk_popsize[uk_times.index(start_t):uk_times.index(end_t)+1]\n",
    "traj_uk    = traj_uk[list(time).index(start_t):list(time).index(end_t)+1]\n",
    "\n",
    "ref_seq, ref_tag  = dp.get_MSA(REF_TAG +'.fasta')\n",
    "ref_seq       = list(ref_seq[0])\n",
    "allele_number = pd.unique([int(i[:-2]) for i in alleles])\n",
    "mut_number    = pd.unique([int(i[:-2]) for i in muts_uk])\n",
    "ref_poly      = np.array(ref_seq)[allele_number]\n",
    "\n",
    "alleles_sorted = np.argsort(allele_number)\n",
    "positions      = np.searchsorted(allele_number[alleles_sorted], mut_number)\n",
    "\n",
    "counts     = [0, 4, 8, 10, 20, 50, 100, 500, 1000]\n",
    "s_tot      = []\n",
    "for i in range(len(counts)):\n",
    "    count = counts[i]\n",
    "    inflow_seqs   = [[inflow_seq] for i in range(T)]\n",
    "    inflow_counts = [[count] for i in range(T)]\n",
    "    pop_in        = [np.sum(i) for i in inflow_counts]\n",
    "    inflow_term   = dp.allele_counter_in(inflow_seqs, inflow_counts, traj_uk, pop_in, uk_popsize, k, R, T, d=5)\n",
    "    inflow_tot    = np.zeros(len(alleles))\n",
    "    for j in range(len(mut_number)):\n",
    "        inflow_tot[positions[j] * 5 : (positions[j] + 1) * 5] += inflow_term[j * 5 : (j + 1) * 5]\n",
    "    A = covar_int\n",
    "    b = numerator - inflow_tot\n",
    "    \n",
    "    s = linalg.solve(A, b, assume_a='sym')\n",
    "\n",
    "    # normalize reference nucleotide to zero\n",
    "    L = int(len(s) / 5)\n",
    "    selection  = np.reshape(s, (L, 5))\n",
    "\n",
    "    s_new = []\n",
    "    for i in range(L):\n",
    "        idx = NUC.index(ref_poly[i])\n",
    "        temp_s    = selection[i]\n",
    "        temp_s    = temp_s - temp_s[idx]\n",
    "        s_new.append(temp_s)\n",
    "    selection = s_new\n",
    "    selection = np.array(selection).flatten()\n",
    "    s_tot.append(selection)\n",
    "    np.savez_compressed(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-inflow{count}.npz'), \n",
    "                        selection=selection, allele_number=allele_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EU1 = [['NSP16-199-2-C', 'NSP1-60-2-C', 'NSP3-1189-2-T', 'M-93-2-G', 'N-220-1-T', 'ORF10-30-0-T', 'S-222-1-T']]\n",
    "np.save(os.path.join(DATA_DIR, '20E-EU1'), EU1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(INF_DIR, 'inflow-uk')):\n",
    "    os.mkdir(os.path.join(INF_DIR, 'inflow-uk'))\n",
    "for count in counts:\n",
    "    shutil.copy(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-inflow{count}.npz'), \n",
    "                os.path.join(INF_DIR, 'inflow-uk', f'infer-{DATA_DATE}-uk-inflow{count}.npz'))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(INF_DIR, 'inflow-uk')):\n",
    "    os.mkdir(os.path.join(INF_DIR, 'inflow-uk'))\n",
    "inf_data  = np.load(os.path.join(INF_DIR, 'infer-2021-08-14--2021-05-01-uk-g-10' + '.npz'), allow_pickle=True)\n",
    "alleles   = inf_data['allele_number']\n",
    "counts     = [0, 4, 8, 10, 20, 50, 100, 500, 1000]\n",
    "for count in counts:\n",
    "    data = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-uk-inflow{count}.npz.npy'), allow_pickle=True)\n",
    "    np.savez_compressed(os.path.join(INF_DIR, 'inflow-uk', f'infer-{DATA_DATE}-uk-inflow{count}.npz'),\n",
    "                        selection=data, allele_number=alleles)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites    = [dp.get_label(i[:-2]) + '-' + i[-1] for i in alleles]\n",
    "var_idxs = np.array([sites.index(site) for site in link_sites])\n",
    "s_var    = np.zeros(len(s_tot))\n",
    "for i in range(len(var_idxs)):\n",
    "    for j in range(len(s_tot)):\n",
    "        s_var[j] += s_tot[j][var_idxs[i]]\n",
    "slope  = (s_var[4] - s_var[0]) / (counts[4] - counts[0])\n",
    "x_zero = - s_var[0] / slope\n",
    "print('number of migrating individuals per day in order for 20E-EU1 variant to be zero', x_zero)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=alphamask></a>\n",
    "# Inferring the effect of Alpha on selection coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "         'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "         'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "         'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "         'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T',\n",
    "         'N-203-1-A', 'N-203-2-A', 'N-204-0-C']\n",
    "np.save(os.path.join(SARS_DIR, 'b117-muts.npy'), alpha)\n",
    "\n",
    "cutoff     = 0\n",
    "\n",
    "count_data = np.load(os.path.join(DATA_DIR, 'nucleotide-counts.npz'), allow_pickle=True)\n",
    "muts       = count_data['allele_number']\n",
    "counts     = count_data['counts']\n",
    "\n",
    "# Same as above but for a single file\n",
    "syn_data  = np.load(os.path.join(DATA_DIR, 'synonymous-prot.npz'), allow_pickle=True)\n",
    "data      = np.load(os.path.join(INF_DIR, 'inf-2021-08-14', 'infer-2021-08-14-g-40.npz'), allow_pickle=True)\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "q   = 5\n",
    "N   = 10000\n",
    "k   = 0.1\n",
    "R   = 2\n",
    "g   = 40\n",
    "alleles   = data['allele_number']\n",
    "numerator = data['numerator']\n",
    "covar_int = data['covar_int']\n",
    "syn_muts  = syn_data['nuc_index']\n",
    "types     = syn_data['types']\n",
    "types     = types[np.isin(syn_muts, alleles)]\n",
    "g1 = np.array(g, dtype=float) * (N * k) / (1 + (k / R))\n",
    "for i in range(len(covar_int)):\n",
    "    covar_int[i, i] += g1\n",
    "    \n",
    "# Mask out sites that have mutations in alpha\n",
    "alpha_short = np.array([i[:-2] for i in alpha])\n",
    "muts_short  = np.array([dp.get_label2(i)[:-2] for i in alleles])\n",
    "mask        = np.isin(muts_short, alpha_short, invert=True)\n",
    "alleles     = alleles[mask]\n",
    "numerator   = numerator[mask]\n",
    "covar_int   = covar_int[mask][:, mask]\n",
    "types       = types[mask]\n",
    "counts      = counts[mask]\n",
    "\n",
    "# infer selection coefficients\n",
    "s      = linalg.solve(covar_int, numerator, assume_a='sym')\n",
    "s_ind  = numerator / np.diag(covar_int)\n",
    "errors = 1 / np.sqrt(np.absolute(np.diag(covar_int)))\n",
    "\n",
    "print('selection coefficients found')\n",
    "\n",
    "# normalize reference nucleotide to zero\n",
    "L = int(len(s) / 5)\n",
    "selection  = np.reshape(s, (L, q))\n",
    "selection_nocovar = np.reshape(s_ind, (L, q))\n",
    "ref_seq, ref_tag  = dp.get_MSA(REF_TAG +'.fasta')\n",
    "ref_seq  = list(ref_seq[0])\n",
    "allele_number = np.unique([int(i[:-2]) for i in alleles])\n",
    "ref_poly = np.array(ref_seq)[allele_number]\n",
    "\n",
    "s_new = []\n",
    "s_SL  = []\n",
    "for i in range(L):\n",
    "    idx = NUC.index(ref_poly[i])\n",
    "    temp_s    = selection[i]\n",
    "    temp_s    = temp_s - temp_s[idx]\n",
    "    temp_s_SL = selection_nocovar[i]\n",
    "    temp_s_SL = temp_s_SL - temp_s_SL[idx]\n",
    "    s_new.append(temp_s)\n",
    "    s_SL.append(temp_s_SL)\n",
    "selection         = s_new\n",
    "selection_nocovar = s_SL\n",
    "\n",
    "selection         = np.array(selection).flatten()\n",
    "selection_nocovar = np.array(selection_nocovar).flatten()\n",
    "error_bars        = errors\n",
    "\n",
    "# Eliminate reference nucleotides and mutations that aren't observed\n",
    "\n",
    "mask      = np.nonzero(np.logical_and(selection!=0, counts>cutoff))[0]\n",
    "alleles   = alleles[mask]\n",
    "selection = selection[mask]\n",
    "s_ind     = selection_nocovar[mask]\n",
    "errors    = error_bars[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "traj      = data['traj']\n",
    "mutants   = data['mutant_sites']\n",
    "for i in range(len(mutants)):\n",
    "    idxs_keep  = np.array([j for j in range(len(mutants[i])) if mutants[i][j] in alleles])\n",
    "    mutants[i] = np.array(mutants[i])[idxs_keep]\n",
    "    traj[i]    = np.array([traj[i][:, j] for j in range(len(traj[i][0])) if j in idxs_keep])\n",
    "    traj[i]    = np.swapaxes(traj[i], 0, 1)\n",
    "\n",
    "np.savez_compressed(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-{g}-observed-mask-alpha.npz'),\n",
    "                    selection=selection, selection_independent=s_ind, error_bars=errors,\n",
    "                    allele_number=alleles, traj=traj, mutant_sites=mutants)\n",
    "    \n",
    "dp.website_file2(os.path.join(INF_DIR,  f'infer-{DATA_DATE}-g-{g}-observed-mask-alpha.npz'),\n",
    "                 os.path.join(DATA_DIR,  'synonymous-test.npz'),\n",
    "                 os.path.join(DATA_DIR,  'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR,  'protein-changes-test'), \n",
    "                 os.path.join(DATA_DIR, f'selection-g-40-observed-mask-alpha.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmasked    = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-40-observed.npz'), allow_pickle=True)\n",
    "masked      = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-{g}-observed-mask-alpha.npz'), allow_pickle=True)\n",
    "s_unmask    = unmasked['selection']\n",
    "s_mask      = masked['selection']\n",
    "muts_mask   = [dp.get_label2(i) for i in masked['allele_number']]\n",
    "muts_unmask = [dp.get_label2(i) for i in unmasked['allele_number']]\n",
    "link_sites  = ['ORF10-30-0-T', 'N-220-1-T', 'M-93-2-G', 'S-222-1-T', 'NSP16-199-2-C', 'NSP3-1189-2-T', 'NSP1-60-2-C']\n",
    "\n",
    "sel_mask    = np.sum(s_mask[np.isin(muts_mask, link_sites)])\n",
    "sel_unmask  = np.sum(s_unmask[np.isin(muts_unmask, link_sites)])\n",
    "print('selection coefficient for 20E-EU1 when alpha is kept in the data:                               ', sel_unmask)\n",
    "print('selection coefficient for 20E-EU1 when alpha mutations are reverted to the reference nucleotide:', sel_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selection coefficient for 20E-EU1 when alpha is kept in the data:                                0.10225873676301382\n",
      "selection coefficient for 20E-EU1 when alpha mutations are reverted to the reference nucleotide: -0.01673100542328889\n"
     ]
    }
   ],
   "source": [
    "unmasked    = np.load(os.path.join(INF_DIR, f'infer-2021-08-14-g-40-observed.npz'), allow_pickle=True)\n",
    "masked      = np.load(os.path.join(INF_DIR, f'infer-2021-08-14-g-40-observed-mask-alpha.npz'), allow_pickle=True)\n",
    "s_unmask    = unmasked['selection']\n",
    "s_mask      = masked['selection']\n",
    "muts_mask   = [dp.get_label2(i) for i in masked['allele_number']]\n",
    "muts_unmask = [dp.get_label2(i) for i in unmasked['allele_number']]\n",
    "link_sites  = ['ORF10-30-0-T', 'N-220-1-T', 'M-93-2-G', 'S-222-1-T', 'NSP16-199-2-C', 'NSP3-1189-2-T', 'NSP1-60-2-C']\n",
    "\n",
    "sel_mask    = np.sum(s_mask[np.isin(muts_mask, link_sites)])\n",
    "sel_unmask  = np.sum(s_unmask[np.isin(muts_unmask, link_sites)])\n",
    "print('selection coefficient for 20E-EU1 when alpha is kept in the data:                               ', sel_unmask)\n",
    "print('selection coefficient for 20E-EU1 when alpha mutations are reverted to the reference nucleotide:', sel_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finding the opposite: the inferred coefficient for 20E-EU1 when Alpha is masked\n",
    "EU1 = ['ORF10-30-0-T', 'N-220-1-T', 'M-93-2-G', 'S-222-1-T', 'NSP16-199-2-C', 'NSP3-1189-2-T', 'NSP1-60-2-C']\n",
    "np.save(os.path.join(SARS_DIR, 'eu1-muts.npy'), EU1)\n",
    "\n",
    "cutoff     = 0\n",
    "\n",
    "count_data = np.load(os.path.join(DATA_DIR, 'nucleotide-counts.npz'), allow_pickle=True)\n",
    "muts       = count_data['allele_number']\n",
    "counts     = count_data['counts']\n",
    "\n",
    "# Same as above but for a single file\n",
    "syn_data  = np.load(os.path.join(DATA_DIR, 'synonymous-prot-test.npz'), allow_pickle=True)\n",
    "data      = np.load(os.path.join(INF_DIR, f'inf-{DATA_DATE}', f'infer-{DATA_DATE}-g-40.npz'), allow_pickle=True)\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "q   = 5\n",
    "N   = 10000\n",
    "k   = 0.1\n",
    "R   = 2\n",
    "g   = 40\n",
    "alleles   = data['allele_number']\n",
    "numerator = data['numerator']\n",
    "covar_int = data['covar_int']\n",
    "syn_muts  = syn_data['nuc_index']\n",
    "types     = syn_data['types']\n",
    "types     = types[np.isin(syn_muts, alleles)]\n",
    "g1 = np.array(g, dtype=float) * (N * k) / (1 + (k / R))\n",
    "for i in range(len(covar_int)):\n",
    "    covar_int[i, i] += g1\n",
    "    \n",
    "# Mask out sites that have mutations in alpha\n",
    "EU1_short   = np.array([i[:-2] for i in EU1])\n",
    "muts_short  = np.array([dp.get_label2(i)[:-2] for i in alleles])\n",
    "mask        = np.isin(muts_short, EU1_short, invert=True)\n",
    "alleles     = alleles[mask]\n",
    "numerator   = numerator[mask]\n",
    "covar_int   = covar_int[mask][:, mask]\n",
    "types       = types[mask]\n",
    "counts      = counts[mask]\n",
    "\n",
    "# infer selection coefficients\n",
    "s      = linalg.solve(covar_int, numerator, assume_a='sym')\n",
    "s_ind  = numerator / np.diag(covar_int)\n",
    "errors = 1 / np.sqrt(np.absolute(np.diag(covar_int)))\n",
    "\n",
    "print('selection coefficients found')\n",
    "\n",
    "# normalize reference nucleotide to zero\n",
    "L = int(len(s) / 5)\n",
    "selection  = np.reshape(s, (L, q))\n",
    "selection_nocovar = np.reshape(s_ind, (L, q))\n",
    "ref_seq, ref_tag  = dp.get_MSA(REF_TAG +'.fasta')\n",
    "ref_seq  = list(ref_seq[0])\n",
    "allele_number = np.unique([int(i[:-2]) for i in alleles])\n",
    "ref_poly = np.array(ref_seq)[allele_number]\n",
    "\n",
    "s_new = []statisticsne\n",
    "s_SL  = []\n",
    "for i in range(L):\n",
    "    idx = NUC.index(ref_poly[i])\n",
    "    temp_s    = selection[i]\n",
    "    temp_s    = temp_s - temp_s[idx]\n",
    "    temp_s_SL = selection_nocovar[i]\n",
    "    temp_s_SL = temp_s_SL - temp_s_SL[idx]\n",
    "    s_new.append(temp_s)\n",
    "    s_SL.append(temp_s_SL)\n",
    "selection         = s_new\n",
    "selection_nocovar = s_SL\n",
    "\n",
    "selection         = np.array(selection).flatten()\n",
    "selection_nocovar = np.array(selection_nocovar).flatten()\n",
    "error_bars        = errors\n",
    "\n",
    "# Eliminate reference nucleotides and mutations that aren't observed\n",
    "\n",
    "mask      = np.nonzero(np.logical_and(selection!=0, counts>cutoff))[0]\n",
    "alleles   = alleles[mask]\n",
    "selection = selection[mask]\n",
    "s_ind     = selection_nocovar[mask]\n",
    "errors    = error_bars[mask]\n",
    "\n",
    "traj      = data['traj']\n",
    "mutants   = data['mutant_sites']\n",
    "for i in range(len(mutants)):\n",
    "    idxs_keep  = np.array([j for j in range(len(mutants[i])) if mutants[i][j] in alleles])\n",
    "    mutants[i] = np.array(mutants[i])[idxs_keep]\n",
    "    traj[i]    = np.array([traj[i][:, j] for j in range(len(traj[i][0])) if j in idxs_keep])\n",
    "    traj[i]    = np.swapaxes(traj[i], 0, 1)\n",
    "\n",
    "np.savez_compressed(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-{g}-observed-mask-eu1.npz'),\n",
    "                    selection=selection, selection_independent=s_ind, error_bars=errors,\n",
    "                    allele_number=alleles, traj=traj, mutant_sites=mutants)\n",
    "\n",
    "dp.website_file2(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-{g}-observed-mask-eu1.npz'),\n",
    "                 os.path.join(DATA_DIR, 'synonymous-prot-test.npz'),\n",
    "                 os.path.join(DATA_DIR, 'linked-sites.npy'), \n",
    "                 os.path.join(DATA_DIR, f'selection-g-40-observed-mask-eu1.csv'))\n",
    "\n",
    "unmasked    = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-40-observed.npz'), allow_pickle=True)\n",
    "masked      = np.load(os.path.join(INF_DIR, f'infer-{DATA_DATE}-g-{g}-observed-mask-eu1.npz'), allow_pickle=True)\n",
    "s_unmask    = unmasked['selection']\n",
    "s_mask      = masked['selection']\n",
    "muts_mask   = [dp.get_label2(i) for i in masked['allele_number']]\n",
    "muts_unmask = [dp.get_label2(i) for i in unmasked['allele_number']]\n",
    "link_sites  = ['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "               'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "               'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "               'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "               'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T',\n",
    "               'N-203-1-A', 'N-203-2-A', 'N-204-0-C']\n",
    "\n",
    "sel_mask    = np.sum(s_mask[np.isin(muts_mask, link_sites)])\n",
    "sel_unmask  = np.sum(s_unmask[np.isin(muts_unmask, link_sites)])\n",
    "print('selection coefficient for Alpha when EU1 is kept in the data:                               ', sel_unmask)\n",
    "print('selection coefficient for Alpha when EU1 mutations are reverted to the reference nucleotide:', sel_mask)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=remove_seqs></a>\n",
    "# Remove problematic sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/remove-seqs.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-remove-seqs.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "\n",
      "sbatch job-remove-seqs.sh &&\n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_dir = os.path.join(SSH_DATA, 'genome-filtered')\n",
    "out_dir   = os.path.join(SSH_DATA, 'genome-filtered-new')\n",
    "script    = 'remove-seqs.py'\n",
    "\n",
    "file1 = os.path.join(input_dir, 'asia-indonesia-2020-02-12---2020-12-25-2022-4-15.csv')\n",
    "site1 = 'ORF8-1-1-A'\n",
    "\n",
    "file2 = os.path.join(input_dir, 'north america-usa-idaho-2020-05-31---2021-2-22-2022-2-28.csv')\n",
    "site2 = 'S-214n-2-G'\n",
    "\n",
    "file3 = os.path.join(input_dir, 'europe-ireland-2021-12-14-2022-05-31---2021-12-14-2022-3-3.csv')\n",
    "site3 = 'S-214-2-A'\n",
    "\n",
    "file4a = os.path.join(input_dir, 'asia-india-gujarat-2020-02-28---2021-11-27-2022-3-5.csv')\n",
    "file4b = os.path.join(input_dir, 'asia-india-karnataka-2020-02-16---2021-10-7-2022-2-21.csv')\n",
    "site4  = 'NSP8-57-1-T'\n",
    "\n",
    "job_file3 = 'job-remove-seqs.sh'\n",
    "job_str3  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH -e ./MPL/out/remove-error\n",
    "#SBATCH -o ./MPL/out/remove-out\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python ./{script} --input_file {file1}  -o {out_dir} --site {site1}\n",
    "python ./{script} --input_file {file2}  -o {out_dir} --site {site2}\n",
    "python ./{script} --input_file {file3}  -o {out_dir} --site {site3}\n",
    "python ./{script} --input_file {file4a} -o {out_dir} --site {site4}\n",
    "python ./{script} --input_file {file4b} -o {out_dir} --site {site4}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file3), 'w')\n",
    "f.write('%s\\n' % job_str3)\n",
    "f.close()\n",
    "\n",
    "# transfer files to cluster\n",
    "print('scp %s/%s %s. &&'   % (SCRIPT_DIR, job_file3, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file3)\n",
    "print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/brianlee/SARS-CoV-2-Data/processing-new/remove-seqs.py blee098@cluster.hpcc.ucr.edu:. &&\n",
      "scp /Users/brianlee/SARS-CoV-2-Data/Processing-files/job-remove-seqs.sh blee098@cluster.hpcc.ucr.edu:. &&\n",
      "\n",
      "sbatch job-remove-seqs.sh &&\n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_dir = os.path.join(SSH_DATA, 'genome-filtered')\n",
    "out_dir   = os.path.join(SSH_DATA, 'genome-filtered-new')\n",
    "script    = 'remove-seqs.py'\n",
    "\n",
    "file1 = os.path.join(input_dir, 'asia-indonesia-2020-02-12---2020-12-25-2022-4-15.csv')\n",
    "site1 = 'ORF8-1-1-A'\n",
    "\n",
    "file2 = os.path.join(input_dir, 'north america-usa-idaho-2020-05-31---2021-2-22-2022-2-28.csv')\n",
    "site2 = 'S-214n-2-G'\n",
    "\n",
    "file3 = os.path.join(input_dir, 'europe-ireland-2021-12-14-2022-05-31---2021-12-14-2022-3-3.csv')\n",
    "site3 = 'S-214-2-A'\n",
    "\n",
    "file4a = os.path.join(input_dir, 'asia-india-gujarat-2020-02-28---2021-11-27-2022-3-5.csv')\n",
    "file4b = os.path.join(input_dir, 'asia-india-karnataka-2020-02-16---2021-10-7-2022-2-21.csv')\n",
    "site4  = 'NSP8-57-1-T'\n",
    "\n",
    "job_file3 = 'job-remove-seqs.sh'\n",
    "job_str3  = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --time=2:00:00\n",
    "#SBATCH -e ./MPL/out/remove-error\n",
    "#SBATCH -o ./MPL/out/remove-out\n",
    "#SBATCH -p batch\n",
    "\n",
    "module unload miniconda2\n",
    "module load miniconda3\n",
    "conda activate sars-env\n",
    "\n",
    "python ./{script} --input_file {file2}  -o {out_dir} --site {site2}\n",
    "python ./{script} --input_file {file3}  -o {out_dir} --site {site3}\n",
    "python ./{script} --input_file {file4a} -o {out_dir} --site {site4}\n",
    "python ./{script} --input_file {file4b} -o {out_dir} --site {site4}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file3), 'w')\n",
    "f.write('%s\\n' % job_str3)\n",
    "f.close()\n",
    "\n",
    "\n",
    "# transfer files to cluster\n",
    "print('scp %s/%s %s. &&'   % (SCRIPT_DIR, job_file3, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# run on cluster\n",
    "print('sbatch %s &&' % job_file3)\n",
    "print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Streamlined"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Combine separate job steps into single slurm jobs using the srun command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Command to check fairshare score on different partitions__:\n",
    "sshare -U $USER -o partition,account,User,RawShares,NormShares,RawUsage,EffectvUsage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
